<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Complete protocols | DataViz protocols</title>
  <meta name="description" content="This is an introduction to data visualization protocols for wet lab scientists" />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Complete protocols | DataViz protocols" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is an introduction to data visualization protocols for wet lab scientists" />
  <meta name="github-repo" content="JoachimGoedhart/DataViz-protocols" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Complete protocols | DataViz protocols" />
  
  <meta name="twitter:description" content="This is an introduction to data visualization protocols for wet lab scientists" />
  

<meta name="author" content="Joachim Goedhart" />


<meta name="date" content="2024-09-11" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="plotting-the-data.html"/>
<link rel="next" href="questions-and-answers.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>

<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-ZDBZZ5FR0E"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-ZDBZZ5FR0E');
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DataViz Protocols</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#a-toast"><i class="fa fa-check"></i>A toast</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>1</b> Getting started with R</a>
<ul>
<li class="chapter" data-level="1.1" data-path="getting-started.html"><a href="getting-started.html#running-r"><i class="fa fa-check"></i><b>1.1</b> Running R</a></li>
<li class="chapter" data-level="1.2" data-path="getting-started.html"><a href="getting-started.html#using-the-command-line"><i class="fa fa-check"></i><b>1.2</b> Using the command line</a></li>
<li class="chapter" data-level="1.3" data-path="getting-started.html"><a href="getting-started.html#help"><i class="fa fa-check"></i><b>1.3</b> ?Help</a></li>
<li class="chapter" data-level="1.4" data-path="getting-started.html"><a href="getting-started.html#installing-packages"><i class="fa fa-check"></i><b>1.4</b> Installing packages</a></li>
<li class="chapter" data-level="1.5" data-path="getting-started.html"><a href="getting-started.html#multiline-code"><i class="fa fa-check"></i><b>1.5</b> Multiline code</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="read-and-reshape.html"><a href="read-and-reshape.html"><i class="fa fa-check"></i><b>2</b> Reading and reshaping data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="read-and-reshape.html"><a href="read-and-reshape.html#introduction"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="read-and-reshape.html"><a href="read-and-reshape.html#types-of-data"><i class="fa fa-check"></i><b>2.2</b> Types of data</a></li>
<li class="chapter" data-level="2.3" data-path="read-and-reshape.html"><a href="read-and-reshape.html#reading-data"><i class="fa fa-check"></i><b>2.3</b> Reading data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="read-and-reshape.html"><a href="read-and-reshape.html#loading-data-from-a-text-or-csv-file"><i class="fa fa-check"></i><b>2.3.1</b> Loading data from a text or csv file</a></li>
<li class="chapter" data-level="2.3.2" data-path="read-and-reshape.html"><a href="read-and-reshape.html#loading-data-from-a-url"><i class="fa fa-check"></i><b>2.3.2</b> Loading data from a URL</a></li>
<li class="chapter" data-level="2.3.3" data-path="read-and-reshape.html"><a href="read-and-reshape.html#retrieving-data-from-excel"><i class="fa fa-check"></i><b>2.3.3</b> Retrieving data from Excel</a></li>
<li class="chapter" data-level="2.3.4" data-path="read-and-reshape.html"><a href="read-and-reshape.html#retrieving-data-from-multiple-files"><i class="fa fa-check"></i><b>2.3.4</b> Retrieving data from multiple files</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="read-and-reshape.html"><a href="read-and-reshape.html#tidy-data"><i class="fa fa-check"></i><b>2.4</b> Reshaping data</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="read-and-reshape.html"><a href="read-and-reshape.html#quantitative-data-discrete-conditions"><i class="fa fa-check"></i><b>2.4.1</b> Quantitative data, discrete conditions</a></li>
<li class="chapter" data-level="2.4.2" data-path="read-and-reshape.html"><a href="read-and-reshape.html#multiple-discrete-conditions"><i class="fa fa-check"></i><b>2.4.2</b> Multiple discrete conditions</a></li>
<li class="chapter" data-level="2.4.3" data-path="read-and-reshape.html"><a href="read-and-reshape.html#double-quantitative-data"><i class="fa fa-check"></i><b>2.4.3</b> Double quantitative data</a></li>
<li class="chapter" data-level="2.4.4" data-path="read-and-reshape.html"><a href="read-and-reshape.html#data-from-multiple-files"><i class="fa fa-check"></i><b>2.4.4</b> Data from multiple files</a></li>
<li class="chapter" data-level="2.4.5" data-path="read-and-reshape.html"><a href="read-and-reshape.html#data-in-96-wells-format-data-in-96-wells-format"><i class="fa fa-check"></i><b>2.4.5</b> Data in 96-wells format {data-in-96-wells-format}</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="plotting-the-data.html"><a href="plotting-the-data.html"><i class="fa fa-check"></i><b>3</b> Plotting the Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="plotting-the-data.html"><a href="plotting-the-data.html#data-over-time-continuous-vs.-continuous"><i class="fa fa-check"></i><b>3.1</b> Data over time (continuous vs. continuous)</a></li>
<li class="chapter" data-level="3.2" data-path="plotting-the-data.html"><a href="plotting-the-data.html#discrete-conditions"><i class="fa fa-check"></i><b>3.2</b> Discrete conditions</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="plotting-the-data.html"><a href="plotting-the-data.html#x-axis-data-qualitative-versus-quantitative-data"><i class="fa fa-check"></i><b>3.2.1</b> X-axis data: qualitative versus quantitative data</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="plotting-the-data.html"><a href="plotting-the-data.html#statistics"><i class="fa fa-check"></i><b>3.3</b> Statistics</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="plotting-the-data.html"><a href="plotting-the-data.html#introduction-1"><i class="fa fa-check"></i><b>3.3.1</b> Introduction</a></li>
<li class="chapter" data-level="3.3.2" data-path="plotting-the-data.html"><a href="plotting-the-data.html#data-summaries-directly-added-as-a-plot-layer"><i class="fa fa-check"></i><b>3.3.2</b> Data summaries directly added as a plot layer</a></li>
<li class="chapter" data-level="3.3.3" data-path="plotting-the-data.html"><a href="plotting-the-data.html#data-summaries-from-a-dataframe"><i class="fa fa-check"></i><b>3.3.3</b> Data summaries from a dataframe</a></li>
<li class="chapter" data-level="3.3.4" data-path="plotting-the-data.html"><a href="plotting-the-data.html#data-summaries-for-continuous-x-axis-data"><i class="fa fa-check"></i><b>3.3.4</b> Data summaries for continuous x-axis data</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="plotting-the-data.html"><a href="plotting-the-data.html#plot-a-lot---discrete-data"><i class="fa fa-check"></i><b>3.4</b> Plot-a-lot - discrete data</a></li>
<li class="chapter" data-level="3.5" data-path="plotting-the-data.html"><a href="plotting-the-data.html#optimizing-the-data-visualization"><i class="fa fa-check"></i><b>3.5</b> Optimizing the data visualization</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="plotting-the-data.html"><a href="plotting-the-data.html#rotation"><i class="fa fa-check"></i><b>3.5.1</b> Rotation</a></li>
<li class="chapter" data-level="3.5.2" data-path="plotting-the-data.html"><a href="plotting-the-data.html#ordering-conditions"><i class="fa fa-check"></i><b>3.5.2</b> Ordering conditions</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="plotting-the-data.html"><a href="plotting-the-data.html#adjusting-the-layout"><i class="fa fa-check"></i><b>3.6</b> Adjusting the layout</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="plotting-the-data.html"><a href="plotting-the-data.html#themes"><i class="fa fa-check"></i><b>3.6.1</b> Themes</a></li>
<li class="chapter" data-level="3.6.2" data-path="plotting-the-data.html"><a href="plotting-the-data.html#legend"><i class="fa fa-check"></i><b>3.6.2</b> Legend</a></li>
<li class="chapter" data-level="3.6.3" data-path="plotting-the-data.html"><a href="plotting-the-data.html#grids"><i class="fa fa-check"></i><b>3.6.3</b> Grids</a></li>
<li class="chapter" data-level="3.6.4" data-path="plotting-the-data.html"><a href="plotting-the-data.html#labelstitles"><i class="fa fa-check"></i><b>3.6.4</b> Labels/Titles</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="plotting-the-data.html"><a href="plotting-the-data.html#plot-a-lot---continuous-data"><i class="fa fa-check"></i><b>3.7</b> Plot-a-lot - continuous data</a>
<ul>
<li class="chapter" data-level="3.7.1" data-path="plotting-the-data.html"><a href="plotting-the-data.html#small-multiples"><i class="fa fa-check"></i><b>3.7.1</b> Small multiples</a></li>
<li class="chapter" data-level="3.7.2" data-path="plotting-the-data.html"><a href="plotting-the-data.html#heatmaps"><i class="fa fa-check"></i><b>3.7.2</b> Heatmaps</a></li>
<li class="chapter" data-level="3.7.3" data-path="plotting-the-data.html"><a href="plotting-the-data.html#custom-objects"><i class="fa fa-check"></i><b>3.7.3</b> Custom objects</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="complete-protocols.html"><a href="complete-protocols.html"><i class="fa fa-check"></i><b>4</b> Complete protocols</a>
<ul>
<li class="chapter" data-level="4.1" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-1"><i class="fa fa-check"></i><b>4.1</b> Protocol 1 - Spectra of fluorescent proteins</a></li>
<li class="chapter" data-level="4.2" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-2"><i class="fa fa-check"></i><b>4.2</b> Protocol 2 - A superplot of calcium concentrations</a></li>
<li class="chapter" data-level="4.3" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-3"><i class="fa fa-check"></i><b>4.3</b> Protocol 3 - small multiples of time courses</a></li>
<li class="chapter" data-level="4.4" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-4"><i class="fa fa-check"></i><b>4.4</b> Protocol 4 - Plotting data from a 96-wells experiment</a></li>
<li class="chapter" data-level="4.5" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-5"><i class="fa fa-check"></i><b>4.5</b> Protocol 5 - A map of amino acids</a></li>
<li class="chapter" data-level="4.6" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-6"><i class="fa fa-check"></i><b>4.6</b> Protocol 6 - Heatmap style visualization of timelapse data</a></li>
<li class="chapter" data-level="4.7" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-7"><i class="fa fa-check"></i><b>4.7</b> Protocol 7 - Ridgeline plot</a></li>
<li class="chapter" data-level="4.8" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-8"><i class="fa fa-check"></i><b>4.8</b> Protocol 8 - Plotting data in a 96-wells layout</a></li>
<li class="chapter" data-level="4.9" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-9"><i class="fa fa-check"></i><b>4.9</b> Protocol 9 - A dose response curve with fit</a></li>
<li class="chapter" data-level="4.10" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-10"><i class="fa fa-check"></i><b>4.10</b> Protocol 10 - Plotting data that was harvested with a Google form</a></li>
<li class="chapter" data-level="4.11" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-11"><i class="fa fa-check"></i><b>4.11</b> Protocol 11 - Plotting time data from a Google form</a></li>
<li class="chapter" data-level="4.12" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-12"><i class="fa fa-check"></i><b>4.12</b> Protocol 12 - Plotting grouped data</a></li>
<li class="chapter" data-level="4.13" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-13"><i class="fa fa-check"></i><b>4.13</b> Protocol 13 - Plotting multiple plots side-by-side</a></li>
<li class="chapter" data-level="4.14" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-14"><i class="fa fa-check"></i><b>4.14</b> Protocol 14 - Volcano plot</a>
<ul>
<li class="chapter" data-level="4.14.1" data-path="complete-protocols.html"><a href="complete-protocols.html#step-1---define-thresholds"><i class="fa fa-check"></i><b>4.14.1</b> Step 1 - Define thresholds</a></li>
<li class="chapter" data-level="4.14.2" data-path="complete-protocols.html"><a href="complete-protocols.html#step-2---display-thresholds"><i class="fa fa-check"></i><b>4.14.2</b> Step 2 - Display thresholds</a></li>
<li class="chapter" data-level="4.14.3" data-path="complete-protocols.html"><a href="complete-protocols.html#step-3---filtering-with-thresholds"><i class="fa fa-check"></i><b>4.14.3</b> Step 3 - Filtering with thresholds</a></li>
<li class="chapter" data-level="4.14.4" data-path="complete-protocols.html"><a href="complete-protocols.html#step-4---annotations"><i class="fa fa-check"></i><b>4.14.4</b> Step 4 - Annotations</a></li>
</ul></li>
<li class="chapter" data-level="4.15" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-15"><i class="fa fa-check"></i><b>4.15</b> Protocol 15 - Timeline</a>
<ul>
<li class="chapter" data-level="4.15.1" data-path="complete-protocols.html"><a href="complete-protocols.html#improve-the-timeline"><i class="fa fa-check"></i><b>4.15.1</b> Improve the timeline</a></li>
<li class="chapter" data-level="4.15.2" data-path="complete-protocols.html"><a href="complete-protocols.html#add-and-format-text"><i class="fa fa-check"></i><b>4.15.2</b> Add and format text</a></li>
</ul></li>
<li class="chapter" data-level="4.16" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-16"><i class="fa fa-check"></i><b>4.16</b> Protocol 16 - Scatterplots and correlation</a>
<ul>
<li class="chapter" data-level="4.16.1" data-path="complete-protocols.html"><a href="complete-protocols.html#linear-fit"><i class="fa fa-check"></i><b>4.16.1</b> Linear fit</a></li>
<li class="chapter" data-level="4.16.2" data-path="complete-protocols.html"><a href="complete-protocols.html#multiple-scatterplots"><i class="fa fa-check"></i><b>4.16.2</b> Multiple scatterplots</a></li>
<li class="chapter" data-level="4.16.3" data-path="complete-protocols.html"><a href="complete-protocols.html#fit-parameters"><i class="fa fa-check"></i><b>4.16.3</b> Fit parameters</a></li>
</ul></li>
<li class="chapter" data-level="4.17" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-17"><i class="fa fa-check"></i><b>4.17</b> Protocol 17 - Animated plots of data from optogenetics experiments</a></li>
<li class="chapter" data-level="4.18" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-18"><i class="fa fa-check"></i><b>4.18</b> Protocol 18 - Combining animated plots with a movie</a>
<ul>
<li class="chapter" data-level="4.18.1" data-path="complete-protocols.html"><a href="complete-protocols.html#intermezzo"><i class="fa fa-check"></i><b>4.18.1</b> Intermezzo</a></li>
<li class="chapter" data-level="4.18.2" data-path="complete-protocols.html"><a href="complete-protocols.html#time-for-some-magick"><i class="fa fa-check"></i><b>4.18.2</b> Time for some magick</a></li>
</ul></li>
<li class="chapter" data-level="4.19" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-19"><i class="fa fa-check"></i><b>4.19</b> Protocol 19 - Plotting ratiometric FRET data</a></li>
<li class="chapter" data-level="4.20" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-20"><i class="fa fa-check"></i><b>4.20</b> Protocol 20 - On- and off-kinetics</a>
<ul>
<li class="chapter" data-level="4.20.1" data-path="complete-protocols.html"><a href="complete-protocols.html#reading-and-processing-the-data"><i class="fa fa-check"></i><b>4.20.1</b> Reading and processing the data</a></li>
<li class="chapter" data-level="4.20.2" data-path="complete-protocols.html"><a href="complete-protocols.html#trial-fitting-on-a-single-trace"><i class="fa fa-check"></i><b>4.20.2</b> Trial fitting on a single trace</a></li>
<li class="chapter" data-level="4.20.3" data-path="complete-protocols.html"><a href="complete-protocols.html#on-kinetics---alltogether"><i class="fa fa-check"></i><b>4.20.3</b> On kinetics - alltogether</a></li>
<li class="chapter" data-level="4.20.4" data-path="complete-protocols.html"><a href="complete-protocols.html#off-kinetics"><i class="fa fa-check"></i><b>4.20.4</b> Off kinetics</a></li>
</ul></li>
<li class="chapter" data-level="4.21" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-21"><i class="fa fa-check"></i><b>4.21</b> Protocol 21 - A spiral plot</a></li>
<li class="chapter" data-level="4.22" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-22"><i class="fa fa-check"></i><b>4.22</b> Protocol 22 - Frequencies of discrete data</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="questions-and-answers.html"><a href="questions-and-answers.html"><i class="fa fa-check"></i><b>5</b> Questions and Answers</a>
<ul>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#what-is-the-difference-between-a-data.frame-and-a-tibble"><i class="fa fa-check"></i>What is the difference between a data.frame and a tibble?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#whats-up-with-spaces-in-variable-names-and-in-files-names"><i class="fa fa-check"></i>What’s up with spaces in variable names and in files names?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#what-are-the-rules-for-naming-dataframes-or-variables"><i class="fa fa-check"></i>What are the rules for naming dataframes or variables?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#what-is-the-difference-between---and-for-assigning-variables"><i class="fa fa-check"></i>What is the difference between <code>&lt;-</code> and <code>=</code> for assigning variables?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#what-does-na-mean-in-a-dataframe"><i class="fa fa-check"></i>What does ‘NA’ mean in a dataframe?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#what-is-the-beste-way-to-type-a-operator"><i class="fa fa-check"></i>What is the beste way to type a ‘%&gt;%’ operator?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#where-do-i-find-the-example-data"><i class="fa fa-check"></i>Where do I find the example data?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#is-there-a-way-to-re-use-or-adjust-the-protocols"><i class="fa fa-check"></i>Is there a way to re-use or adjust the protocols?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#which-packages-are-included-in-the-tidyverse-package"><i class="fa fa-check"></i>Which packages are included in the <code>{tidyverse}</code> package?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#what-is-the-difference-between-require-and-library-for-loading-a-package"><i class="fa fa-check"></i>What is the difference between <code>require()</code> and <code>library()</code> for loading a package?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#is-it-possible-to-have-interactive-file-selection"><i class="fa fa-check"></i>Is it possible to have interactive file selection?</a></li>
<li><a href="questions-and-answers.html#section" id="toc-section"></a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DataViz protocols</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="complete-protocols" class="section level1 hasAnchor" number="4">
<h1><span class="header-section-number">Chapter 4</span> Complete protocols<a href="complete-protocols.html#complete-protocols" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>This chapter showcases several <em>complete</em> protocols for different kinds of data visualizations. Each protocol starts with the raw data and ends with a publication quality plot. The data is available from Github in the ‘/data’ folder on <a href="https://github.com/JoachimGoedhart/DataViz-protocols/tree/main/Protocols">this page</a>. The github page holds a separate Rmarkdown file for each protocol. To reproduce the data visualizations you can either take the data and follow the instructions in the chapter. Alternatively you can download the Rmarkdown file (and the data) and run it step by step. The Rmarkdown files can also be used as a starting point to apply the same visualization to your own data.</p>
<p>An overview of the data visualizations that are generated by the protocols:</p>
<p><img src="Protocols/overview_of_protocols.png" /></p>
<div id="protocol-1" class="section level2 hasAnchor" number="4.1">
<h2><span class="header-section-number">4.1</span> Protocol 1 - Spectra of fluorescent proteins<a href="complete-protocols.html#protocol-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This protocol describes how you can turn a csv with spectral data that is obtained from FPbase.org into a plot of those spectra. First, we load the required package:</p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="complete-protocols.html#cb206-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>For this data visualization, I selected several spectra from fluorescent proteins at FPbase.org: <a href="https://www.fpbase.org/spectra/?s=1746,6551,101,102,123,124,1604,1606&amp;showY=0&amp;showX=1&amp;showGrid=0&amp;areaFill=1&amp;logScale=0&amp;scaleEC=0&amp;scaleQY=0&amp;shareTooltip=1&amp;palette=wavelength">https://www.fpbase.org/spectra/?s=1746,6551,101,102,123,124,1604,1606&amp;showY=0&amp;showX=1&amp;showGrid=0&amp;areaFill=1&amp;logScale=0&amp;scaleEC=0&amp;scaleQY=0&amp;shareTooltip=1&amp;palette=wavelength</a></p>
<p>The data was downloaded in CSV format (by clicking on the button in the lower right corner of the webpage) and renamed to ‘FPbase_Spectra_4FPs.csv’.
We read the data from the CSV by using the <code>read_csv()</code> function. This function is part of the tidy verse and loads the data as a tibble. It also guesses type of data for each column. To hide that information, we use <code>show_col_types = FALSE</code> here.</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="complete-protocols.html#cb207-1" tabindex="-1"></a>df_raw <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;data/FPbase_Spectra_4FPs.csv&quot;</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Let’s briefly look at what we have loaded:</p>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="complete-protocols.html#cb208-1" tabindex="-1"></a><span class="fu">glimpse</span>(df_raw)</span></code></pre></div>
<pre><code>Rows: 512
Columns: 9
$ Wavelength       &lt;dbl&gt; 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310…
$ `mTurquoise2 EM` &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `mTurquoise2 EX` &lt;dbl&gt; 0.2484, 0.2266, 0.2048, 0.1852, 0.1634, 0.1482, 0.132…
$ `mNeonGreen EM`  &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `mNeonGreen EX`  &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `mScarlet-I EM`  &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `mScarlet-I EX`  &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `miRFP670 EM`    &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `miRFP670 EX`    &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…</code></pre>
<p>The data needs conversion to a tidy format before plotting. Since we have a single continuous data column with Wavelength information that is used for the x-axis, it is excluded from the operation:</p>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb210-1"><a href="complete-protocols.html#cb210-1" tabindex="-1"></a>df_1 <span class="ot">&lt;-</span></span>
<span id="cb210-2"><a href="complete-protocols.html#cb210-2" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb210-3"><a href="complete-protocols.html#cb210-3" tabindex="-1"></a>    df_raw,</span>
<span id="cb210-4"><a href="complete-protocols.html#cb210-4" tabindex="-1"></a>    <span class="at">cols =</span> <span class="sc">-</span>Wavelength,</span>
<span id="cb210-5"><a href="complete-protocols.html#cb210-5" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">&quot;Sample&quot;</span>,</span>
<span id="cb210-6"><a href="complete-protocols.html#cb210-6" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">&quot;Intensity&quot;</span></span>
<span id="cb210-7"><a href="complete-protocols.html#cb210-7" tabindex="-1"></a>  )</span></code></pre></div>
<p>There are several rows that have NA values for Intensity and this is how to get rid of that:</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="complete-protocols.html#cb211-1" tabindex="-1"></a>df_1 <span class="ot">&lt;-</span> df_1 <span class="sc">%&gt;%</span> <span class="fu">drop_na</span>(Intensity)</span></code></pre></div>
<p>The column ‘Sample’ has labels for the fluorescent protein and the type of spectrum. We can separate that column into two different columns that we name ‘Fluorescent Protein’ and ‘Spectrum’:</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="complete-protocols.html#cb212-1" tabindex="-1"></a>df_1 <span class="ot">&lt;-</span> df_1 <span class="sc">%&gt;%</span></span>
<span id="cb212-2"><a href="complete-protocols.html#cb212-2" tabindex="-1"></a>  <span class="fu">separate</span>(Sample, <span class="fu">c</span>(<span class="st">&quot;Fluorescent Protein&quot;</span>, <span class="st">&quot;Spectrum&quot;</span>), <span class="at">sep =</span> <span class="st">&quot; &quot;</span>)</span></code></pre></div>
<p>Let’s do a first attempt and plot the data:</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="complete-protocols.html#cb213-1" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb213-2"><a href="complete-protocols.html#cb213-2" tabindex="-1"></a>  <span class="at">data =</span> df_1,</span>
<span id="cb213-3"><a href="complete-protocols.html#cb213-3" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">x =</span> Wavelength, <span class="at">y =</span> Intensity, <span class="at">color =</span> <span class="st">`</span><span class="at">Fluorescent Protein</span><span class="st">`</span>)</span>
<span id="cb213-4"><a href="complete-protocols.html#cb213-4" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb213-5"><a href="complete-protocols.html#cb213-5" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">linetype =</span> Spectrum), <span class="at">size =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-345-1.png" width="672" /></p>
<p>This looks pretty good already. Now let’s change the order of the fluorescent proteins to their order in the plot:</p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="complete-protocols.html#cb214-1" tabindex="-1"></a>df_1 <span class="ot">&lt;-</span> df_1 <span class="sc">%&gt;%</span></span>
<span id="cb214-2"><a href="complete-protocols.html#cb214-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">`</span><span class="at">Fluorescent Protein</span><span class="st">`</span> <span class="ot">=</span> forcats<span class="sc">::</span><span class="fu">fct_relevel</span>(</span>
<span id="cb214-3"><a href="complete-protocols.html#cb214-3" tabindex="-1"></a>    <span class="st">`</span><span class="at">Fluorescent Protein</span><span class="st">`</span>,</span>
<span id="cb214-4"><a href="complete-protocols.html#cb214-4" tabindex="-1"></a>    <span class="fu">c</span>(<span class="st">&quot;mTurquoise2&quot;</span>, <span class="st">&quot;mNeonGreen&quot;</span>, <span class="st">&quot;mScarlet-I&quot;</span>, <span class="st">&quot;miRFP670&quot;</span>)</span>
<span id="cb214-5"><a href="complete-protocols.html#cb214-5" tabindex="-1"></a>  ))</span></code></pre></div>
<p>The data is in the right shape now, so let’s save it:</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="complete-protocols.html#cb215-1" tabindex="-1"></a>df_1 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_1.csv&quot;</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>We define the plot object and add an extra geometry, <code>geom_area()</code> to fill the area under the curves:</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="complete-protocols.html#cb216-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb216-2"><a href="complete-protocols.html#cb216-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb216-3"><a href="complete-protocols.html#cb216-3" tabindex="-1"></a>    <span class="at">data =</span> df_1,</span>
<span id="cb216-4"><a href="complete-protocols.html#cb216-4" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb216-5"><a href="complete-protocols.html#cb216-5" tabindex="-1"></a>      <span class="at">x =</span> Wavelength, <span class="at">y =</span> Intensity,</span>
<span id="cb216-6"><a href="complete-protocols.html#cb216-6" tabindex="-1"></a>      <span class="at">fill =</span> <span class="st">`</span><span class="at">Fluorescent Protein</span><span class="st">`</span></span>
<span id="cb216-7"><a href="complete-protocols.html#cb216-7" tabindex="-1"></a>    )</span>
<span id="cb216-8"><a href="complete-protocols.html#cb216-8" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb216-9"><a href="complete-protocols.html#cb216-9" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">linetype =</span> Spectrum),</span>
<span id="cb216-10"><a href="complete-protocols.html#cb216-10" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.5</span></span>
<span id="cb216-11"><a href="complete-protocols.html#cb216-11" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb216-12"><a href="complete-protocols.html#cb216-12" tabindex="-1"></a>  <span class="fu">geom_area</span>(</span>
<span id="cb216-13"><a href="complete-protocols.html#cb216-13" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">linetype =</span> Spectrum),</span>
<span id="cb216-14"><a href="complete-protocols.html#cb216-14" tabindex="-1"></a>    <span class="at">color =</span> <span class="cn">NA</span>,</span>
<span id="cb216-15"><a href="complete-protocols.html#cb216-15" tabindex="-1"></a>    <span class="at">position =</span> <span class="st">&quot;identity&quot;</span>,</span>
<span id="cb216-16"><a href="complete-protocols.html#cb216-16" tabindex="-1"></a>    <span class="at">size =</span> <span class="dv">1</span>,</span>
<span id="cb216-17"><a href="complete-protocols.html#cb216-17" tabindex="-1"></a>    <span class="at">alpha =</span> <span class="fl">0.5</span></span>
<span id="cb216-18"><a href="complete-protocols.html#cb216-18" tabindex="-1"></a>  )</span></code></pre></div>
<p>Let’s check the result:</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="complete-protocols.html#cb217-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-349-1.png" width="672" /></p>
<p>Next, we set the limits of the axis and force the y-axis to start at 0</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="complete-protocols.html#cb218-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb218-2"><a href="complete-protocols.html#cb218-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">1.1</span>)) <span class="sc">+</span> </span>
<span id="cb218-3"><a href="complete-protocols.html#cb218-3" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">350</span>, <span class="dv">810</span>))</span></code></pre></div>
<p>Add labels:</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="complete-protocols.html#cb219-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb219-2"><a href="complete-protocols.html#cb219-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb219-3"><a href="complete-protocols.html#cb219-3" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Spectra of Fluorescent Proteins&quot;</span>,</span>
<span id="cb219-4"><a href="complete-protocols.html#cb219-4" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Wavelength [nm]&quot;</span>,</span>
<span id="cb219-5"><a href="complete-protocols.html#cb219-5" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Normalized Intensity [a.u.]&quot;</span>,</span>
<span id="cb219-6"><a href="complete-protocols.html#cb219-6" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">(based on data from FPbase.org)&quot;</span>,</span>
<span id="cb219-7"><a href="complete-protocols.html#cb219-7" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 1&quot;</span></span>
<span id="cb219-8"><a href="complete-protocols.html#cb219-8" tabindex="-1"></a>  )</span></code></pre></div>
<p>Modify the layout by adjusting the theme. Comments are used to explain effect of the individual lines of code:</p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="complete-protocols.html#cb220-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb220-2"><a href="complete-protocols.html#cb220-2" tabindex="-1"></a>  <span class="co">#Set text size</span></span>
<span id="cb220-3"><a href="complete-protocols.html#cb220-3" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span> <span class="fu">theme</span>(</span>
<span id="cb220-4"><a href="complete-protocols.html#cb220-4" tabindex="-1"></a>    <span class="at">plot.caption =</span> <span class="fu">element_text</span>(</span>
<span id="cb220-5"><a href="complete-protocols.html#cb220-5" tabindex="-1"></a>      <span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>,</span>
<span id="cb220-6"><a href="complete-protocols.html#cb220-6" tabindex="-1"></a>      <span class="at">hjust =</span> <span class="dv">1</span></span>
<span id="cb220-7"><a href="complete-protocols.html#cb220-7" tabindex="-1"></a>    ),</span>
<span id="cb220-8"><a href="complete-protocols.html#cb220-8" tabindex="-1"></a>    <span class="co">#Remove grid</span></span>
<span id="cb220-9"><a href="complete-protocols.html#cb220-9" tabindex="-1"></a>    <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb220-10"><a href="complete-protocols.html#cb220-10" tabindex="-1"></a>    <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb220-11"><a href="complete-protocols.html#cb220-11" tabindex="-1"></a>    <span class="co">#Set position of legend</span></span>
<span id="cb220-12"><a href="complete-protocols.html#cb220-12" tabindex="-1"></a>    <span class="at">legend.position =</span> <span class="st">&quot;top&quot;</span>,</span>
<span id="cb220-13"><a href="complete-protocols.html#cb220-13" tabindex="-1"></a>    <span class="at">legend.justification =</span> <span class="st">&quot;left&quot;</span></span>
<span id="cb220-14"><a href="complete-protocols.html#cb220-14" tabindex="-1"></a>    <span class="co">#Define the legend layout</span></span>
<span id="cb220-15"><a href="complete-protocols.html#cb220-15" tabindex="-1"></a>  ) <span class="sc">+</span> <span class="fu">guides</span>(</span>
<span id="cb220-16"><a href="complete-protocols.html#cb220-16" tabindex="-1"></a>    <span class="at">linetype =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb220-17"><a href="complete-protocols.html#cb220-17" tabindex="-1"></a>    <span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="cn">NULL</span>, <span class="at">label.position =</span> <span class="st">&quot;right&quot;</span>)</span>
<span id="cb220-18"><a href="complete-protocols.html#cb220-18" tabindex="-1"></a>  )</span></code></pre></div>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="complete-protocols.html#cb221-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-353-1.png" width="672" /></p>
<p>We are almost there, except that the colors of the plot do not match with the natural colors of the fluorescent proteins. Let’s fix that by defining a custom color palette. The order of the colors matches with the order of the fluorescent proteins that was defined earlier:</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="complete-protocols.html#cb222-1" tabindex="-1"></a>custom_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;green&quot;</span>, <span class="st">&quot;orange&quot;</span>, <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p>To apply the custom colors to the filled area:</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="complete-protocols.html#cb223-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> custom_colors)</span></code></pre></div>
<p>This is the result:</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="complete-protocols.html#cb224-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-356-1.png" width="672" /></p>
<p>To save this plot as a PNG file:</p>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="complete-protocols.html#cb225-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_01.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb225-2"><a href="complete-protocols.html#cb225-2" tabindex="-1"></a>  p</span>
<span id="cb225-3"><a href="complete-protocols.html#cb225-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-2" class="section level2 hasAnchor" number="4.2">
<h2><span class="header-section-number">4.2</span> Protocol 2 - A superplot of calcium concentrations<a href="complete-protocols.html#protocol-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This protocol is used to create a superplot which differentiates between technical and biological replicates. The concept of superplots has been reported by Lord and colleagues <a href="https://doi.org/10.1083/jcb.202001064">(Lord et al., 2021)</a>. We will use the data that was used to create figure 5e in a publication by <a href="https://doi.org/10.1038/s41467-021-27249-w">van der Linden et al. (2021)</a>.</p>
<p>The figure in the publication summarizes the data from all experiments and does not identify the biological replicates. Below, we will differentiate the biological replicates, by treating each batch of neutrophils as a biological replicate.</p>
<p>We start by loading the required tidyverse package:</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="complete-protocols.html#cb227-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>We define the confidence level as 95%:</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="complete-protocols.html#cb228-1" tabindex="-1"></a>Confidence_level <span class="ot">&lt;-</span>  <span class="fl">0.95</span></span></code></pre></div>
<p>The data is stored in an excel sheet and we read it, skipping the first 6 lines which contain comments:</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="complete-protocols.html#cb229-1" tabindex="-1"></a>df_raw <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">&#39;data/figure5.xlsx&#39;</span>, <span class="at">skip=</span><span class="dv">6</span>)</span></code></pre></div>
<p>Let’s look at the data:</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="complete-protocols.html#cb230-1" tabindex="-1"></a><span class="fu">head</span>(df_raw)</span></code></pre></div>
<pre><code># A tibble: 6 × 7
  `Experimental day` `Replicate no.` `Neutrophil no.` `Batch no. neutrophils`
               &lt;dbl&gt;           &lt;dbl&gt;            &lt;dbl&gt;                   &lt;dbl&gt;
1                  1               1                1                       1
2                  1               1                1                       1
3                  1               1                1                       1
4                  1               1                1                       1
5                  1               1                1                       1
6                  1               1                1                       1
# ℹ 3 more variables: Stage &lt;chr&gt;, `dF/F0` &lt;chr&gt;, `Calcium (uM)` &lt;dbl&gt;</code></pre>
<p>The data is already in a tidy format. The column with ‘Stage’ has the four different conditions for which we will compare the data in the column ‘Calcium (uM)’. We change the name of the column ‘Batch no. neutrophils’ to ‘Replicate’ and make sure the different replicates are treated as factors (qualitative data):</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="complete-protocols.html#cb232-1" tabindex="-1"></a>df_raw <span class="ot">&lt;-</span> df_raw <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Replicate =</span> <span class="fu">as.factor</span>(<span class="st">`</span><span class="at">Batch no. neutrophils</span><span class="st">`</span>))</span></code></pre></div>
<p>Let’s look at the data, and identify the biological replicates, as suggested in the original publication on Superplot by <a href="https://doi.org/10.1083/jcb.202001064">(Lord et al., 2021)</a>. In this example a color code is used to label the replicates:</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="complete-protocols.html#cb233-1" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>df_raw, <span class="fu">aes</span>(<span class="at">x=</span>Stage)) <span class="sc">+</span></span>
<span id="cb233-2"><a href="complete-protocols.html#cb233-2" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">data=</span>df_raw, <span class="fu">aes</span>(<span class="at">x=</span>Stage, <span class="at">y=</span><span class="st">`</span><span class="at">Calcium (uM)</span><span class="st">`</span>, <span class="at">color=</span>Replicate))</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-383-1.png" width="672" /></p>
<p>To display the statistics for the individual biological replicates, we define a new dataframe. To this end, we group the data for the different stages and biological replicates:</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="complete-protocols.html#cb234-1" tabindex="-1"></a>df_summary <span class="ot">&lt;-</span> df_raw <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Stage, Replicate) <span class="sc">%&gt;%</span></span>
<span id="cb234-2"><a href="complete-protocols.html#cb234-2" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n=</span><span class="fu">n</span>(), <span class="at">mean=</span><span class="fu">mean</span>(<span class="st">`</span><span class="at">Calcium (uM)</span><span class="st">`</span>))</span></code></pre></div>
<pre><code>`summarise()` has grouped output by &#39;Stage&#39;. You can override using the
`.groups` argument.</code></pre>
<p>Next, we use ‘df_summary’ which holds the averages of each biological replicate, and we calculate the statistics for the different conditions:</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="complete-protocols.html#cb236-1" tabindex="-1"></a>df_summary_replicas <span class="ot">&lt;-</span> df_summary <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Stage) <span class="sc">%&gt;%</span></span>
<span id="cb236-2"><a href="complete-protocols.html#cb236-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">n_rep=</span><span class="fu">n</span>(), <span class="at">mean_rep=</span><span class="fu">mean</span>(mean), <span class="at">sd_rep =</span> <span class="fu">sd</span>(mean))  <span class="sc">%&gt;%</span></span>
<span id="cb236-3"><a href="complete-protocols.html#cb236-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sem =</span> sd_rep <span class="sc">/</span> <span class="fu">sqrt</span>(n_rep <span class="sc">-</span> <span class="dv">1</span>),</span>
<span id="cb236-4"><a href="complete-protocols.html#cb236-4" tabindex="-1"></a>         <span class="st">`</span><span class="at">95%CI_lo</span><span class="st">`</span> <span class="ot">=</span> mean_rep <span class="sc">+</span> <span class="fu">qt</span>((<span class="dv">1</span><span class="sc">-</span>Confidence_level)<span class="sc">/</span><span class="dv">2</span>, n_rep <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">*</span> sem,</span>
<span id="cb236-5"><a href="complete-protocols.html#cb236-5" tabindex="-1"></a>         <span class="st">`</span><span class="at">95%CI_hi</span><span class="st">`</span> <span class="ot">=</span> mean_rep <span class="sc">-</span> <span class="fu">qt</span>((<span class="dv">1</span><span class="sc">-</span>Confidence_level)<span class="sc">/</span><span class="dv">2</span>, n_rep <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">*</span> sem,</span>
<span id="cb236-6"><a href="complete-protocols.html#cb236-6" tabindex="-1"></a>         <span class="cn">NULL</span>)</span></code></pre></div>
<p>The dataframe has the summary of the conditions and note that each condition has a summary of 4 biological replicates:</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="complete-protocols.html#cb237-1" tabindex="-1"></a><span class="fu">head</span>(df_summary_replicas)</span></code></pre></div>
<pre><code># A tibble: 6 × 10
# Groups:   Stage [2]
  Stage    Replicate     n   mean n_rep mean_rep  sd_rep     sem `95%CI_lo`
  &lt;chr&gt;    &lt;fct&gt;     &lt;int&gt;  &lt;dbl&gt; &lt;int&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;
1 before   1            38 0.0305     4   0.0301 0.00414 0.00239     0.0225
2 before   2            67 0.0270     4   0.0301 0.00414 0.00239     0.0225
3 before   3            56 0.0358     4   0.0301 0.00414 0.00239     0.0225
4 before   4            55 0.0270     4   0.0301 0.00414 0.00239     0.0225
5 crawling 1             7 0.0296     4   0.0317 0.00339 0.00196     0.0255
6 crawling 2            29 0.0289     4   0.0317 0.00339 0.00196     0.0255
# ℹ 1 more variable: `95%CI_hi` &lt;dbl&gt;</code></pre>
<p>We can now add or ‘bind’ the data of ‘df_summary_replicas’ to the original dataframe ‘df’ and store this as a dataframe ‘df_2’:</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="complete-protocols.html#cb239-1" tabindex="-1"></a>df_2 <span class="ot">&lt;-</span> df_raw <span class="sc">%&gt;%</span> <span class="fu">left_join</span>(df_summary_replicas, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;Stage&quot;</span>,<span class="st">&quot;Replicate&quot;</span>))</span></code></pre></div>
<p>Let’s save this data:</p>
<div class="sourceCode" id="cb240"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb240-1"><a href="complete-protocols.html#cb240-1" tabindex="-1"></a>df_2 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_2.csv&quot;</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>Let’s first define a basic plot with all of the data for each stage shown as a violinplot:</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="complete-protocols.html#cb241-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>df_2, <span class="fu">aes</span>(<span class="at">x=</span>Stage)) <span class="sc">+</span></span>
<span id="cb241-2"><a href="complete-protocols.html#cb241-2" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">data=</span>df_2, <span class="fu">aes</span>(<span class="at">x=</span>Stage, <span class="at">y=</span><span class="st">`</span><span class="at">Calcium (uM)</span><span class="st">`</span>), <span class="at">color=</span><span class="cn">NA</span>, <span class="at">fill=</span><span class="st">&quot;grey80&quot;</span>)</span></code></pre></div>
<p>This is what it looks like:</p>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="complete-protocols.html#cb242-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-390-1.png" width="672" /></p>
<p>We add the 95% confidence interval from the summary of the biological replicates as a line:</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="complete-protocols.html#cb243-1" tabindex="-1"></a> p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_linerange</span>(<span class="at">data =</span> df_2, <span class="fu">aes</span>(<span class="at">ymin=</span><span class="st">`</span><span class="at">95%CI_lo</span><span class="st">`</span>, <span class="at">ymax=</span><span class="st">`</span><span class="at">95%CI_hi</span><span class="st">`</span>), <span class="at">size=</span><span class="dv">1</span>, <span class="at">alpha=</span><span class="fl">0.8</span>) </span></code></pre></div>
<p>And we add the mean value of each replicate as a dot. Here, the size of the dot is reflecting n:</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="complete-protocols.html#cb244-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">data=</span>df_2, <span class="fu">aes</span>(<span class="at">x=</span>Stage, <span class="at">y=</span>mean, <span class="at">size=</span>n, <span class="at">fill=</span>Replicate),  <span class="at">shape=</span><span class="dv">21</span>, <span class="at">stroke =</span> <span class="dv">1</span>)</span></code></pre></div>
<p>The function <code>scale_size_area()</code> ensures that 0 is represented as an area of 0 and allows to to define that an n of 10,50 and 90 is shown in the legend:</p>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="complete-protocols.html#cb245-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_size_area</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">50</span>, <span class="dv">90</span>), <span class="at">max_size =</span> <span class="dv">6</span>)</span></code></pre></div>
<p>This is what that looks like:</p>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="complete-protocols.html#cb246-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-394-1.png" width="672" /></p>
<p>Next, one of my favorite tweaks for discrete conditions is to rotate the plot 90 degrees. At the same time, the limits are defined.</p>
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="complete-protocols.html#cb247-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">coord_flip</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="fl">0.02</span>,<span class="fl">0.09</span>)) <span class="sc">+</span></span>
<span id="cb247-2"><a href="complete-protocols.html#cb247-2" tabindex="-1"></a>  <span class="co"># This ensures correct order of conditions when plot is rotated 90 degrees</span></span>
<span id="cb247-3"><a href="complete-protocols.html#cb247-3" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">limits =</span> rev)</span></code></pre></div>
<p>Rotation improves readability of the labels for the conditions, even when they are long. It also easier to read the different calcium levels:</p>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="complete-protocols.html#cb248-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-396-1.png" width="672" /></p>
<p>To guide the interpretation, a line is added as a threshold of 0.06 µM (=60 nM):</p>
<div class="sourceCode" id="cb249"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb249-1"><a href="complete-protocols.html#cb249-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">0.060</span>, <span class="at">linetype=</span><span class="st">&#39;dotted&#39;</span>)</span></code></pre></div>
<p>Adjusting the axis labels and adding a title and caption:</p>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="complete-protocols.html#cb250-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb250-2"><a href="complete-protocols.html#cb250-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb250-3"><a href="complete-protocols.html#cb250-3" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Calcium concentrations are less than 60 nM&quot;</span>,</span>
<span id="cb250-4"><a href="complete-protocols.html#cb250-4" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;at different stages of transendothelial migration&quot;</span>,</span>
<span id="cb250-5"><a href="complete-protocols.html#cb250-5" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Stage&quot;</span>,</span>
<span id="cb250-6"><a href="complete-protocols.html#cb250-6" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Calcium [µM]&quot;</span>,</span>
<span id="cb250-7"><a href="complete-protocols.html#cb250-7" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">(based on data from van der Linden, DOI: 10.1101/2021.06.21.449214)&quot;</span>,</span>
<span id="cb250-8"><a href="complete-protocols.html#cb250-8" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 2&quot;</span></span>
<span id="cb250-9"><a href="complete-protocols.html#cb250-9" tabindex="-1"></a>  )</span></code></pre></div>
<p>The layout it further optimized. The most tricky part is positioning of the label for the different conditions. It is placed right above the conditions, which I really like. However, getting this right involves a bit of trial and error and I recommend playing with the parameters to see how it affects the positioning. Something similar applies to the legend which is moved into the lower right corner of the plot, although this is eassier to accomplish. The comments explain the effect of the different lines:</p>
<div class="sourceCode" id="cb251"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb251-1"><a href="complete-protocols.html#cb251-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb251-2"><a href="complete-protocols.html#cb251-2" tabindex="-1"></a>  <span class="co">#Set text size</span></span>
<span id="cb251-3"><a href="complete-protocols.html#cb251-3" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">16</span>) <span class="sc">+</span> <span class="fu">theme</span>(</span>
<span id="cb251-4"><a href="complete-protocols.html#cb251-4" tabindex="-1"></a>    <span class="at">plot.caption =</span> <span class="fu">element_text</span>(</span>
<span id="cb251-5"><a href="complete-protocols.html#cb251-5" tabindex="-1"></a>      <span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>,</span>
<span id="cb251-6"><a href="complete-protocols.html#cb251-6" tabindex="-1"></a>      <span class="at">hjust =</span> <span class="dv">1</span></span>
<span id="cb251-7"><a href="complete-protocols.html#cb251-7" tabindex="-1"></a>    ),</span>
<span id="cb251-8"><a href="complete-protocols.html#cb251-8" tabindex="-1"></a>    <span class="co">#Set position of legend to lower right corner</span></span>
<span id="cb251-9"><a href="complete-protocols.html#cb251-9" tabindex="-1"></a>    <span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.88</span>,<span class="fl">0.15</span>),</span>
<span id="cb251-10"><a href="complete-protocols.html#cb251-10" tabindex="-1"></a>    </span>
<span id="cb251-11"><a href="complete-protocols.html#cb251-11" tabindex="-1"></a>    <span class="co">#This line positions the label (&#39;title&#39;) of the conditions</span></span>
<span id="cb251-12"><a href="complete-protocols.html#cb251-12" tabindex="-1"></a>    <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="fl">0.98</span>, <span class="at">angle =</span> <span class="dv">0</span>, <span class="at">margin=</span><span class="fu">margin</span>(<span class="at">l=</span><span class="dv">70</span>)),</span>
<span id="cb251-13"><a href="complete-protocols.html#cb251-13" tabindex="-1"></a>    </span>
<span id="cb251-14"><a href="complete-protocols.html#cb251-14" tabindex="-1"></a>    <span class="co">#This line positions the names of the conditions</span></span>
<span id="cb251-15"><a href="complete-protocols.html#cb251-15" tabindex="-1"></a>    <span class="co">#A negative margin is needed for aligning the y-axis &#39;title&#39; with the &#39;text&#39;</span></span>
<span id="cb251-16"><a href="complete-protocols.html#cb251-16" tabindex="-1"></a>    <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>, <span class="at">angle =</span> <span class="dv">0</span>, <span class="at">margin=</span><span class="fu">margin</span>(<span class="at">l=</span><span class="sc">-</span><span class="dv">90</span>, <span class="at">r=</span><span class="dv">5</span>)),</span>
<span id="cb251-17"><a href="complete-protocols.html#cb251-17" tabindex="-1"></a>    </span>
<span id="cb251-18"><a href="complete-protocols.html#cb251-18" tabindex="-1"></a>    <span class="co">#Move &#39;tag&#39;, so its position partially overlaps with the conditions</span></span>
<span id="cb251-19"><a href="complete-protocols.html#cb251-19" tabindex="-1"></a>    <span class="at">plot.tag.position =</span> <span class="fu">c</span>(<span class="fl">0.06</span>,<span class="fl">0.99</span>)</span>
<span id="cb251-20"><a href="complete-protocols.html#cb251-20" tabindex="-1"></a>    ) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">fill =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb251-21"><a href="complete-protocols.html#cb251-21" tabindex="-1"></a>               <span class="at">size =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">&#39;n per replicate&#39;</span>, <span class="at">label.position =</span> <span class="st">&quot;left&quot;</span>)</span>
<span id="cb251-22"><a href="complete-protocols.html#cb251-22" tabindex="-1"></a>               )</span></code></pre></div>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="complete-protocols.html#cb252-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-400-1.png" width="672" /></p>
<p>To save the plot as a PNG file:</p>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb253-1"><a href="complete-protocols.html#cb253-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_02.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb253-2"><a href="complete-protocols.html#cb253-2" tabindex="-1"></a>  p</span>
<span id="cb253-3"><a href="complete-protocols.html#cb253-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-3" class="section level2 hasAnchor" number="4.3">
<h2><span class="header-section-number">4.3</span> Protocol 3 - small multiples of time courses<a href="complete-protocols.html#protocol-3" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This protocol displays a number of different timecourses as ‘small multiples’. Small multiples, as the name suggests, displays many small plot separately as a stamp collection. By stressing the data, rather than the labels and grids, this can be a powerful visualization strategy.</p>
<p>The data is taken from a publication by <a href="https://doi.org/10.3389/fimmu.2021.667213">Arts et al. (2021)</a> and we recreate figure panel 1F. The original figure is in small multiple format, but we tweak it a bit more to increase the focus on the data.</p>
<p>Let’s first load the necessary package:</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb255-1"><a href="complete-protocols.html#cb255-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>The data comes from an excel file:</p>
<div class="sourceCode" id="cb256"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb256-1"><a href="complete-protocols.html#cb256-1" tabindex="-1"></a>df_raw <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">&quot;data/Data_Arts_Circularity.xlsx&quot;</span>)</span>
<span id="cb256-2"><a href="complete-protocols.html#cb256-2" tabindex="-1"></a><span class="fu">head</span>(df_raw)</span></code></pre></div>
<pre><code># A tibble: 6 × 13
   time `neutro 1` `neutro 2` `neutro 3` `neutro 4` `neutro 5` `neutro 6`
  &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
1     0       0.53       0.54       0.55       0.59       0.58       0.34
2    10       0.44       0.6        0.5        0.4        0.54       0.4 
3    20       0.33       0.55       0.64       0.3        0.48       0.41
4    30       0.35       0.54       0.69       0.28       0.52       0.32
5    40       0.41       0.53       0.57       0.23       0.44       0.29
6    50       0.32       0.4        0.4        0.26       0.43       0.25
# ℹ 6 more variables: `neutro 7` &lt;dbl&gt;, `neutro 8` &lt;dbl&gt;, `neutro 9` &lt;dbl&gt;,
#   `neutro 10` &lt;dbl&gt;, `neutro 11` &lt;dbl&gt;, `neutro 12` &lt;dbl&gt;</code></pre>
<p>It is in a wide format, so we need to make it tidy. The parameter that was measured over time is the ‘roundness’ of cells:</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="complete-protocols.html#cb258-1" tabindex="-1"></a>df_3 <span class="ot">&lt;-</span></span>
<span id="cb258-2"><a href="complete-protocols.html#cb258-2" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb258-3"><a href="complete-protocols.html#cb258-3" tabindex="-1"></a>    df_raw,</span>
<span id="cb258-4"><a href="complete-protocols.html#cb258-4" tabindex="-1"></a>    <span class="at">cols =</span> <span class="sc">-</span>time,</span>
<span id="cb258-5"><a href="complete-protocols.html#cb258-5" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">&quot;Cell&quot;</span>,</span>
<span id="cb258-6"><a href="complete-protocols.html#cb258-6" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">&quot;Roundness&quot;</span></span>
<span id="cb258-7"><a href="complete-protocols.html#cb258-7" tabindex="-1"></a>  )</span></code></pre></div>
<p>The data is in the right shape now, so let’s save it:</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="complete-protocols.html#cb259-1" tabindex="-1"></a>df_3 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_3.csv&quot;</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>First we create a line plot of all the data:</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="complete-protocols.html#cb260-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_3, <span class="fu">aes</span>(<span class="at">x=</span>time, <span class="at">y=</span>Roundness, <span class="at">group=</span>Cell)) <span class="sc">+</span> <span class="fu">geom_line</span>()</span>
<span id="cb260-2"><a href="complete-protocols.html#cb260-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-431-1.png" width="672" /></p>
<p>With the facet_wrap() function, we turn this into a small multiple:</p>
<div class="sourceCode" id="cb261"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb261-1"><a href="complete-protocols.html#cb261-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>Cell)</span>
<span id="cb261-2"><a href="complete-protocols.html#cb261-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-432-1.png" width="672" /></p>
<p>Set the limits of the axis and force the y-axis to start at 0</p>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="complete-protocols.html#cb262-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb262-2"><a href="complete-protocols.html#cb262-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">1.0</span>)) <span class="sc">+</span> </span>
<span id="cb262-3"><a href="complete-protocols.html#cb262-3" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">300</span>))</span></code></pre></div>
<p>Use a minimal theme and remove the strips and grid to increase focus on the data:</p>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="complete-protocols.html#cb263-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">14</span>)</span>
<span id="cb263-2"><a href="complete-protocols.html#cb263-2" tabindex="-1"></a></span>
<span id="cb263-3"><a href="complete-protocols.html#cb263-3" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">strip.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb263-4"><a href="complete-protocols.html#cb263-4" tabindex="-1"></a>               <span class="at">strip.text =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb263-5"><a href="complete-protocols.html#cb263-5" tabindex="-1"></a>               <span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>),</span>
<span id="cb263-6"><a href="complete-protocols.html#cb263-6" tabindex="-1"></a>               <span class="co">#Remove grid</span></span>
<span id="cb263-7"><a href="complete-protocols.html#cb263-7" tabindex="-1"></a>               <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb263-8"><a href="complete-protocols.html#cb263-8" tabindex="-1"></a>               <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>()</span>
<span id="cb263-9"><a href="complete-protocols.html#cb263-9" tabindex="-1"></a>)</span>
<span id="cb263-10"><a href="complete-protocols.html#cb263-10" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-434-1.png" width="672" /></p>
<p>I do not like the repeated axis for the different plots. We can remove those:</p>
<div class="sourceCode" id="cb264"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb264-1"><a href="complete-protocols.html#cb264-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(        </span>
<span id="cb264-2"><a href="complete-protocols.html#cb264-2" tabindex="-1"></a>        <span class="co">#Remove axis labels</span></span>
<span id="cb264-3"><a href="complete-protocols.html#cb264-3" tabindex="-1"></a>        <span class="at">axis.text =</span> <span class="fu">element_blank</span>()</span>
<span id="cb264-4"><a href="complete-protocols.html#cb264-4" tabindex="-1"></a>        )</span>
<span id="cb264-5"><a href="complete-protocols.html#cb264-5" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-435-1.png" width="672" /></p>
<p>This is a very minimal plot, focusing entirely on the data. It may work well, but it is informative to add some information about the scaling of the x- and y-axis. To achieve this, I add lines to the lower left plot, which correspond to the data of ‘neutro 6’ (you can see this in the small multiple plot where each plot was labeled). I define a new dataframe with the x- and y-scale for ‘neutro 6’ to do just that:</p>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="complete-protocols.html#cb265-1" tabindex="-1"></a>ann_line<span class="ot">&lt;-</span><span class="fu">data.frame</span>(<span class="at">xmin=</span><span class="dv">0</span>,<span class="at">xmax=</span><span class="dv">300</span>,<span class="at">ymin=</span><span class="dv">0</span>,<span class="at">ymax=</span><span class="dv">1</span>,</span>
<span id="cb265-2"><a href="complete-protocols.html#cb265-2" tabindex="-1"></a>                     <span class="at">Cell=</span><span class="fu">factor</span>(<span class="st">&quot;neutro 6&quot;</span>,<span class="at">levels=</span><span class="fu">c</span>(<span class="st">&quot;neutro 6&quot;</span>)))</span>
<span id="cb265-3"><a href="complete-protocols.html#cb265-3" tabindex="-1"></a>ann_line</span></code></pre></div>
<pre><code>  xmin xmax ymin ymax     Cell
1    0  300    0    1 neutro 6</code></pre>
<p>This dataframe can now be used to draw two lines with <code>geom_segment()</code>:</p>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="complete-protocols.html#cb267-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb267-2"><a href="complete-protocols.html#cb267-2" tabindex="-1"></a>  <span class="co">#Vertical line</span></span>
<span id="cb267-3"><a href="complete-protocols.html#cb267-3" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">data=</span>ann_line, <span class="fu">aes</span>(<span class="at">x=</span>xmin,<span class="at">xend=</span>xmin,<span class="at">y=</span>ymin,<span class="at">yend=</span>ymax), <span class="at">size=</span><span class="dv">2</span>, <span class="at">color=</span><span class="st">&#39;grey40&#39;</span>) <span class="sc">+</span></span>
<span id="cb267-4"><a href="complete-protocols.html#cb267-4" tabindex="-1"></a>  <span class="co">#Horizontal line</span></span>
<span id="cb267-5"><a href="complete-protocols.html#cb267-5" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">data=</span>ann_line, <span class="fu">aes</span>(<span class="at">x=</span>xmin,<span class="at">xend=</span>xmax,<span class="at">y=</span>ymin,<span class="at">yend=</span>ymin), <span class="at">size=</span><span class="dv">2</span>, <span class="at">color=</span><span class="st">&#39;grey40&#39;</span>) <span class="sc">+</span></span>
<span id="cb267-6"><a href="complete-protocols.html#cb267-6" tabindex="-1"></a>  <span class="cn">NULL</span></span>
<span id="cb267-7"><a href="complete-protocols.html#cb267-7" tabindex="-1"></a></span>
<span id="cb267-8"><a href="complete-protocols.html#cb267-8" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-437-1.png" width="672" /></p>
<p>The plot is now in black and white which gives it a strong contrast. We can make it a bit more soft and pleasant to look at by changing to shades of grey. Also, the labels of the axes are moved next to the lines:</p>
<div class="sourceCode" id="cb268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb268-1"><a href="complete-protocols.html#cb268-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb268-2"><a href="complete-protocols.html#cb268-2" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="st">&#39;grey98&#39;</span>, <span class="at">color=</span><span class="cn">NA</span>),</span>
<span id="cb268-3"><a href="complete-protocols.html#cb268-3" tabindex="-1"></a>        <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">color=</span><span class="st">&#39;grey90&#39;</span>, <span class="at">fill=</span><span class="cn">NA</span>),</span>
<span id="cb268-4"><a href="complete-protocols.html#cb268-4" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>, <span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">color=</span><span class="st">&#39;grey40&#39;</span>),</span>
<span id="cb268-5"><a href="complete-protocols.html#cb268-5" tabindex="-1"></a>        <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>, <span class="at">vjust =</span> <span class="dv">0</span>, <span class="at">hjust=</span><span class="dv">0</span>, <span class="at">angle =</span> <span class="dv">90</span>, <span class="at">color=</span><span class="st">&#39;grey40&#39;</span>),</span>
<span id="cb268-6"><a href="complete-protocols.html#cb268-6" tabindex="-1"></a>  ) </span></code></pre></div>
<p>Finally, we add a title, caption, and labels (and a scale in brackets):</p>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="complete-protocols.html#cb269-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb269-2"><a href="complete-protocols.html#cb269-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb269-3"><a href="complete-protocols.html#cb269-3" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Changes in the shape of migrating cells over time&quot;</span>,</span>
<span id="cb269-4"><a href="complete-protocols.html#cb269-4" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Time [300s]&quot;</span>,</span>
<span id="cb269-5"><a href="complete-protocols.html#cb269-5" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Circularity [0-1]&quot;</span>,</span>
<span id="cb269-6"><a href="complete-protocols.html#cb269-6" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">(based on data from Arts et al., DOI: 10.3389/fimmu.2021.667213)&quot;</span>,</span>
<span id="cb269-7"><a href="complete-protocols.html#cb269-7" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 3&quot;</span></span>
<span id="cb269-8"><a href="complete-protocols.html#cb269-8" tabindex="-1"></a>  ) </span></code></pre></div>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="complete-protocols.html#cb270-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-440-1.png" width="672" /></p>
<div class="sourceCode" id="cb271"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb271-1"><a href="complete-protocols.html#cb271-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_03.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb271-2"><a href="complete-protocols.html#cb271-2" tabindex="-1"></a> p</span>
<span id="cb271-3"><a href="complete-protocols.html#cb271-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-4" class="section level2 hasAnchor" number="4.4">
<h2><span class="header-section-number">4.4</span> Protocol 4 - Plotting data from a 96-wells experiment<a href="complete-protocols.html#protocol-4" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This protocol showcases some serious data wrangling and tidying. One reason is that the data is acquired with a 96-wells plate reader and the data is stored according to the layout of the plate. This makes total sense from a human perspective, but it is not well suited for data visualization.</p>
<p>In addition, the plate is measured twice. One measurement is the luminescence from a luciferase and the other measurement is the luminescence from Renilla. The latter reading serves as a reference and therefore, the luciferase data is divided by the Renilla intensities. A final step before the data is visualized is a normalization to a control condition.</p>
<p>The code that is shown here is also the basis for the <a href="https://doi.org/10.12688/f1000research.73641.1">plotXpress app</a> that can be used to process and visualize the data. In fact, the data visualization is very close to the standard output of plotXpress and uses the same example data.</p>
<p>We start by loading a package that we need:</p>
<div class="sourceCode" id="cb273"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb273-1"><a href="complete-protocols.html#cb273-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>The measured data is read from an excel sheet. Note that this is the raw data that is stored by the software that operates the plate reader:</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="complete-protocols.html#cb274-1" tabindex="-1"></a>df_raw <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">&quot;data/DualLuc_example_data.xlsx&quot;</span>, <span class="at">sheet =</span> <span class="st">&quot;Results&quot;</span>)</span></code></pre></div>
<pre><code>New names:
• `` -&gt; `...1`
• `` -&gt; `...2`
• `` -&gt; `...3`
• `` -&gt; `...5`
• `` -&gt; `...6`
• `` -&gt; `...7`
• `` -&gt; `...8`
• `` -&gt; `...9`
• `` -&gt; `...10`
• `` -&gt; `...11`
• `` -&gt; `...12`
• `` -&gt; `...13`
• `` -&gt; `...14`
• `` -&gt; `...15`
• `` -&gt; `...16`
• `` -&gt; `...17`</code></pre>
<p>The experimental conditions for each well are stored in a separate CSV file, generated by the experimentalist that did the experiment:</p>
<div class="sourceCode" id="cb276"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb276-1"><a href="complete-protocols.html#cb276-1" tabindex="-1"></a>df_design <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;data/Tidy_design.csv&quot;</span>)</span>
<span id="cb276-2"><a href="complete-protocols.html#cb276-2" tabindex="-1"></a><span class="fu">head</span>(df_design)</span></code></pre></div>
<pre><code>  Wells condition treatment1 treatment2
1   A01         -        HEK          -
2   B01         -        HEK          -
3   C01         -        HEK          -
4   D01         -        HEK          -
5   E01         -     neuron          -
6   F01         -     neuron          -</code></pre>
<p>You can see that the design file is tidy. In contrast the excel file with data is far from tidy. In the excel sheet, two ‘rectangles’ of cells define the data for the firefly &amp; renilla reads.
The data is subset and converted to a vector</p>
<div class="sourceCode" id="cb278"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb278-1"><a href="complete-protocols.html#cb278-1" tabindex="-1"></a>firefly <span class="ot">&lt;-</span> df_raw[<span class="dv">19</span><span class="sc">:</span><span class="dv">26</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">17</span>] <span class="sc">%&gt;%</span> <span class="fu">unlist</span>(<span class="at">use.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb278-2"><a href="complete-protocols.html#cb278-2" tabindex="-1"></a>renilla <span class="ot">&lt;-</span> df_raw[<span class="dv">40</span><span class="sc">:</span><span class="dv">47</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">17</span>] <span class="sc">%&gt;%</span> <span class="fu">unlist</span>(<span class="at">use.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Define a dataframe with wells</p>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="complete-protocols.html#cb279-1" tabindex="-1"></a>column <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="at">each=</span><span class="dv">8</span>)</span>
<span id="cb279-2"><a href="complete-protocols.html#cb279-2" tabindex="-1"></a>row <span class="ot">&lt;-</span> <span class="fu">rep</span>(LETTERS[<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>],<span class="dv">12</span>)</span></code></pre></div>
<p>For convenience, all numbers should consist of 2 digits and so we add a zero that precedes the single digit numbers:</p>
<div class="sourceCode" id="cb280"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb280-1"><a href="complete-protocols.html#cb280-1" tabindex="-1"></a>x0 <span class="ot">&lt;-</span> <span class="fu">str_pad</span>(column, <span class="dv">2</span>, <span class="at">pad =</span> <span class="st">&quot;0&quot;</span>)</span></code></pre></div>
<p>To generate a unique index for each row in the dataframe, we define ‘Wells’, which combines the row and column index:</p>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="complete-protocols.html#cb281-1" tabindex="-1"></a>Wells <span class="ot">&lt;-</span> <span class="fu">paste0</span>(row,x0)</span></code></pre></div>
<p>Next, we create a dataframe that holds the data of individual columns, rows and wells:</p>
<div class="sourceCode" id="cb282"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb282-1"><a href="complete-protocols.html#cb282-1" tabindex="-1"></a>df_plate <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(column,row,Wells)</span>
<span id="cb282-2"><a href="complete-protocols.html#cb282-2" tabindex="-1"></a><span class="fu">head</span>(df_plate)</span></code></pre></div>
<pre><code>  column row Wells
1      1   A   A01
2      1   B   B01
3      1   C   C01
4      1   D   D01
5      1   E   E01
6      1   F   F01</code></pre>
<p>Add to df_plate the vectors with data from firefly and renilla reads</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="complete-protocols.html#cb284-1" tabindex="-1"></a>df_4 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(df_plate,firefly,renilla)</span></code></pre></div>
<p>Merge the design with the data, based on the well ID - left_join() is used to add only data for wells that are listed in the design dataframe</p>
<div class="sourceCode" id="cb285"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb285-1"><a href="complete-protocols.html#cb285-1" tabindex="-1"></a>df_4 <span class="ot">&lt;-</span> <span class="fu">left_join</span>(df_design, df_4, <span class="at">by=</span><span class="st">&#39;Wells&#39;</span>)</span>
<span id="cb285-2"><a href="complete-protocols.html#cb285-2" tabindex="-1"></a><span class="fu">head</span>(df_4)</span></code></pre></div>
<pre><code>  Wells condition treatment1 treatment2 column row firefly renilla
1   A01         -        HEK          -      1   A    2010 2391540
2   B01         -        HEK          -      1   B    3210 2391639
3   C01         -        HEK          -      1   C    1965 2390991
4   D01         -        HEK          -      1   D    2381 2391774
5   E01         -     neuron          -      1   E    1292  269021
6   F01         -     neuron          -      1   F     991  268918</code></pre>
<p>Calculate the relative expression from the firefly/renilla ratio</p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="complete-protocols.html#cb287-1" tabindex="-1"></a>df_4 <span class="ot">&lt;-</span> df_4 <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">expression=</span>firefly<span class="sc">/</span>renilla)</span></code></pre></div>
<p>Take all control conditions and calculate the average value</p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="complete-protocols.html#cb288-1" tabindex="-1"></a>df_norm <span class="ot">&lt;-</span> df_4 <span class="sc">%&gt;%</span> <span class="fu">filter</span>(condition <span class="sc">==</span> <span class="st">&quot;-&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb288-2"><a href="complete-protocols.html#cb288-2" tabindex="-1"></a>  <span class="fu">group_by</span>(treatment1,treatment2) <span class="sc">%&gt;%</span></span>
<span id="cb288-3"><a href="complete-protocols.html#cb288-3" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean=</span><span class="fu">mean</span>(expression)) </span></code></pre></div>
<pre><code>`summarise()` has grouped output by &#39;treatment1&#39;. You can override using the
`.groups` argument.</code></pre>
<p>Combine the mean values (needed for normalization) values with the df_4 dataframe</p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="complete-protocols.html#cb290-1" tabindex="-1"></a>df_4 <span class="ot">&lt;-</span> df_4 <span class="sc">%&gt;%</span></span>
<span id="cb290-2"><a href="complete-protocols.html#cb290-2" tabindex="-1"></a>  <span class="fu">full_join</span>(df_norm, <span class="at">by=</span><span class="fu">c</span>(<span class="st">&quot;treatment1&quot;</span>,<span class="st">&quot;treatment2&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb290-3"><a href="complete-protocols.html#cb290-3" tabindex="-1"></a>  <span class="fu">na.omit</span>(condition)</span></code></pre></div>
<p>Calculate the <code>Fold Change</code> by normalizing all measurements against the control (-)</p>
<div class="sourceCode" id="cb291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb291-1"><a href="complete-protocols.html#cb291-1" tabindex="-1"></a>df_4 <span class="ot">&lt;-</span> df_4 <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="st">`</span><span class="at">Fold Change</span><span class="st">`</span> <span class="ot">=</span> expression<span class="sc">/</span>mean)</span></code></pre></div>
<p>The result is a dataframe that holds all the necessary data:</p>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="complete-protocols.html#cb292-1" tabindex="-1"></a><span class="fu">head</span>(df_4)</span></code></pre></div>
<pre><code>  Wells condition treatment1 treatment2 column row firefly renilla   expression
1   A01         -        HEK          -      1   A    2010 2391540 0.0008404626
2   B01         -        HEK          -      1   B    3210 2391639 0.0013421758
3   C01         -        HEK          -      1   C    1965 2390991 0.0008218350
4   D01         -        HEK          -      1   D    2381 2391774 0.0009954954
5   E01         -     neuron          -      1   E    1292  269021 0.0048025991
6   F01         -     neuron          -      1   F     991  268918 0.0036851382
          mean Fold Change
1 0.0009999922   0.8404692
2 0.0009999922   1.3421863
3 0.0009999922   0.8218414
4 0.0009999922   0.9955032
5 0.0040858977   1.1754085
6 0.0040858977   0.9019164</code></pre>
<p>The data is in the right shape now, so let’s save it:</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="complete-protocols.html#cb294-1" tabindex="-1"></a>df_4 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_4.csv&quot;</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Based on the dataframe, we can create a plot with jittered dots that show the data:</p>
<div class="sourceCode" id="cb295"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb295-1"><a href="complete-protocols.html#cb295-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_4, <span class="fu">aes</span>(<span class="at">x=</span>condition, <span class="at">y=</span><span class="st">`</span><span class="at">Fold Change</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb295-2"><a href="complete-protocols.html#cb295-2" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.2</span>, <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">size=</span><span class="dv">3</span>)</span></code></pre></div>
<p>To splot the graphs based on treatment1 (vertical) and treatment2 (horizontal) we use the <code>facet_grid()</code> function:</p>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="complete-protocols.html#cb296-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">facet_grid</span>(treatment1<span class="sc">~</span>treatment2)</span></code></pre></div>
<p>We add a horizontal line for mean value:</p>
<div class="sourceCode" id="cb297"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb297-1"><a href="complete-protocols.html#cb297-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">stat_summary</span>(<span class="at">fun.min=</span>mean, <span class="at">fun.max=</span>mean, <span class="at">geom=</span><span class="st">&#39;errorbar&#39;</span>, <span class="at">width=</span><span class="fl">0.6</span>, <span class="at">size=</span><span class="fl">0.5</span>)</span></code></pre></div>
<p>Add labels:</p>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="complete-protocols.html#cb298-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb298-2"><a href="complete-protocols.html#cb298-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb298-3"><a href="complete-protocols.html#cb298-3" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Effect of DNA sequences on reporter levels under different conditions&quot;</span>,</span>
<span id="cb298-4"><a href="complete-protocols.html#cb298-4" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;The expression level was determined by a dual luciferase assay</span><span class="sc">\n</span><span class="st"> and the values were normalized to a control with no DNA sequence (-)&quot;</span>,</span>
<span id="cb298-5"><a href="complete-protocols.html#cb298-5" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;DNA Sequence&quot;</span>,</span>
<span id="cb298-6"><a href="complete-protocols.html#cb298-6" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Fold change of the reporter relative to the control (-)&quot;</span>,</span>
<span id="cb298-7"><a href="complete-protocols.html#cb298-7" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">(based on data from Brandorff et al., DOI: 10.1101/2021.07.08.451595)&quot;</span>,</span>
<span id="cb298-8"><a href="complete-protocols.html#cb298-8" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 4&quot;</span></span>
<span id="cb298-9"><a href="complete-protocols.html#cb298-9" tabindex="-1"></a>  ) </span></code></pre></div>
<p>Set the theme and font size:</p>
<div class="sourceCode" id="cb299"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb299-1"><a href="complete-protocols.html#cb299-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">14</span>)</span></code></pre></div>
<p>Format the facet labels (strips) and the caption + subtitle</p>
<div class="sourceCode" id="cb300"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb300-1"><a href="complete-protocols.html#cb300-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="st">&quot;grey90&quot;</span>, <span class="at">color=</span><span class="st">&quot;grey50&quot;</span>),</span>
<span id="cb300-2"><a href="complete-protocols.html#cb300-2" tabindex="-1"></a>               <span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">color=</span><span class="st">&quot;grey50&quot;</span>),</span>
<span id="cb300-3"><a href="complete-protocols.html#cb300-3" tabindex="-1"></a>               <span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>),</span>
<span id="cb300-4"><a href="complete-protocols.html#cb300-4" tabindex="-1"></a>               <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey50&quot;</span>, <span class="at">face =</span> <span class="st">&quot;italic&quot;</span>),</span>
<span id="cb300-5"><a href="complete-protocols.html#cb300-5" tabindex="-1"></a>               <span class="co">#Remove the grid</span></span>
<span id="cb300-6"><a href="complete-protocols.html#cb300-6" tabindex="-1"></a>               <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb300-7"><a href="complete-protocols.html#cb300-7" tabindex="-1"></a>               <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>()</span>
<span id="cb300-8"><a href="complete-protocols.html#cb300-8" tabindex="-1"></a>)</span></code></pre></div>
<p>Let’s look at the result:</p>
<div class="sourceCode" id="cb301"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb301-1"><a href="complete-protocols.html#cb301-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-481-1.png" width="672" /></p>
<p>To save the plot as a png file:</p>
<div class="sourceCode" id="cb302"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb302-1"><a href="complete-protocols.html#cb302-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_04.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb302-2"><a href="complete-protocols.html#cb302-2" tabindex="-1"></a> p</span>
<span id="cb302-3"><a href="complete-protocols.html#cb302-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-5" class="section level2 hasAnchor" number="4.5">
<h2><span class="header-section-number">4.5</span> Protocol 5 - A map of amino acids<a href="complete-protocols.html#protocol-5" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This data visualization plots the position of amino acids in a given protein. It is inspired by figure 2A of the paper by <a href="https://www.sciencedirect.com/science/article/pii/S0092867420304815">Basu et al. (2020)</a></p>
<p>Let’s load the tidyverse package:</p>
<div class="sourceCode" id="cb304"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb304-1"><a href="complete-protocols.html#cb304-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>First, we define a vector with the 20 amino acids and the order in which we plot them. The amino acids are grouped as hydrophobic (G,A,V,C,P,I,L,M,F,W), hydrophilic (S,T,Y,N,Q), acidic (D,E) and basic (R,H,K).</p>
<div class="sourceCode" id="cb305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb305-1"><a href="complete-protocols.html#cb305-1" tabindex="-1"></a>amino_acid_ordered <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(<span class="st">&quot;GAVCPILMFWSTYNQDERHK&quot;</span>,<span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span> <span class="fu">unlist</span>()</span>
<span id="cb305-2"><a href="complete-protocols.html#cb305-2" tabindex="-1"></a>amino_acid_ordered</span></code></pre></div>
<pre><code> [1] &quot;G&quot; &quot;A&quot; &quot;V&quot; &quot;C&quot; &quot;P&quot; &quot;I&quot; &quot;L&quot; &quot;M&quot; &quot;F&quot; &quot;W&quot; &quot;S&quot; &quot;T&quot; &quot;Y&quot; &quot;N&quot; &quot;Q&quot; &quot;D&quot; &quot;E&quot; &quot;R&quot; &quot;H&quot;
[20] &quot;K&quot;</code></pre>
<p>The protein sequence that we will use is the <em>Homo sapiens</em> Homeobox protein Hox-D13:</p>
<div class="sourceCode" id="cb307"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb307-1"><a href="complete-protocols.html#cb307-1" tabindex="-1"></a>protein <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;MSRAGSWDMDGLRADGGGAGGAPASSSSSSVAAAAASGQCRGFLSAPVFAGTHSGRAAAA</span></span>
<span id="cb307-2"><a href="complete-protocols.html#cb307-2" tabindex="-1"></a><span class="st">AAAAAAAAAAASGFAYPGTSERTGSSSSSSSSAVVAARPEAPPAKECPAPTPAAAAAAPP</span></span>
<span id="cb307-3"><a href="complete-protocols.html#cb307-3" tabindex="-1"></a><span class="st">SAPALGYGYHFGNGYYSCRMSHGVGLQQNALKSSPHASLGGFPVEKYMDVSGLASSSVPA</span></span>
<span id="cb307-4"><a href="complete-protocols.html#cb307-4" tabindex="-1"></a><span class="st">NEVPARAKEVSFYQGYTSPYQHVPGYIDMVSTFGSGEPRHEAYISMEGYQSWTLANGWNS</span></span>
<span id="cb307-5"><a href="complete-protocols.html#cb307-5" tabindex="-1"></a><span class="st">QVYCTKDQPQGSHFWKSSFPGDVALNQPDMCVYRRGRKKRVPYTKLQLKELENEYAINKF</span></span>
<span id="cb307-6"><a href="complete-protocols.html#cb307-6" tabindex="-1"></a><span class="st">INKDKRRRISAATNLSERQVTIWFQNRRVKDKKIVSKLKDTVS&quot;</span>)</span></code></pre></div>
<p>The protein sequence may contain end-of-line characters “” after copy pasting and we need to remove these. The <code>gsub()</code> function can be used:</p>
<div class="sourceCode" id="cb308"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb308-1"><a href="complete-protocols.html#cb308-1" tabindex="-1"></a><span class="fu">gsub</span>(<span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>, <span class="st">&quot;&quot;</span>, protein)</span></code></pre></div>
<pre><code>[1] &quot;MSRAGSWDMDGLRADGGGAGGAPASSSSSSVAAAAASGQCRGFLSAPVFAGTHSGRAAAAAAAAAAAAAAASGFAYPGTSERTGSSSSSSSSAVVAARPEAPPAKECPAPTPAAAAAAPPSAPALGYGYHFGNGYYSCRMSHGVGLQQNALKSSPHASLGGFPVEKYMDVSGLASSSVPANEVPARAKEVSFYQGYTSPYQHVPGYIDMVSTFGSGEPRHEAYISMEGYQSWTLANGWNSQVYCTKDQPQGSHFWKSSFPGDVALNQPDMCVYRRGRKKRVPYTKLQLKELENEYAINKFINKDKRRRISAATNLSERQVTIWFQNRRVKDKKIVSKLKDTVS&quot;</code></pre>
<p>But we can also use <code>str_replace_all()</code> from the {tidyverse} package:</p>
<div class="sourceCode" id="cb310"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb310-1"><a href="complete-protocols.html#cb310-1" tabindex="-1"></a>protein <span class="ot">&lt;-</span> protein <span class="sc">%&gt;%</span> <span class="fu">str_replace_all</span>(<span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>, <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Next, the protein sequence is split into single characters and we assign this vector to <code>aa</code>:</p>
<div class="sourceCode" id="cb311"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb311-1"><a href="complete-protocols.html#cb311-1" tabindex="-1"></a>aa <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(protein, <span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span> <span class="fu">unlist</span>()</span></code></pre></div>
<p>We generate a dataframe with a column with the amino acids and a column that defines their position:</p>
<div class="sourceCode" id="cb312"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb312-1"><a href="complete-protocols.html#cb312-1" tabindex="-1"></a>df_5 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">aa=</span>aa, <span class="at">position=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(aa))</span></code></pre></div>
<p>Now we reorder the data frame to the order of the amino acids that we defined earlier in the vector <code>amino_acid_ordered</code>:</p>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="complete-protocols.html#cb313-1" tabindex="-1"></a>df_5 <span class="ot">&lt;-</span> df_5 <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">aa =</span> <span class="fu">fct_relevel</span>(aa, amino_acid_ordered))</span></code></pre></div>
<p>The basic plot shows a black tile for each amino acid. Note that the y-axis order is defined by the vector <code>amino_acid_ordered</code>, but it needs to be reverted to order the amino acids from top to bottom along the y-axis (which is naturally starts at the bottom it corresponds to the origin).</p>
<p>The data is in the right shape now, so let’s save it:</p>
<div class="sourceCode" id="cb314"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb314-1"><a href="complete-protocols.html#cb314-1" tabindex="-1"></a>df_5 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_5.csv&quot;</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb315"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb315-1"><a href="complete-protocols.html#cb315-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span> <span class="fu">geom_tile</span>(<span class="at">data=</span>df_5, <span class="fu">aes</span>(<span class="at">x=</span>position, <span class="at">y=</span>aa)) <span class="sc">+</span></span>
<span id="cb315-2"><a href="complete-protocols.html#cb315-2" tabindex="-1"></a>        <span class="fu">scale_y_discrete</span>(<span class="at">limits =</span> <span class="fu">rev</span>(amino_acid_ordered))</span>
<span id="cb315-3"><a href="complete-protocols.html#cb315-3" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-518-1.png" width="672" /></p>
<p>Set the theme to classic, to get rid off the ‘frame’ around the plot and the grid.</p>
<div class="sourceCode" id="cb316"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb316-1"><a href="complete-protocols.html#cb316-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p<span class="sc">+</span><span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">16</span>)</span></code></pre></div>
<p>For each of the four classes of amino acids we can define a box with a color that indicates the class. For example, there are three basic residues that will have a rectangle filled with blue in the background. The amino acids are factors, but we need numbers to define the coordinates for the rectangle. In a plot with a factors (here on the y-axis) their position is defined by a (non visible) natural number. Therefore we can define a box with the function <code>annotate()</code> for the first residue with y-coordinates <code>ymin=0.5</code> and <code>ymax=1.5</code>:</p>
<div class="sourceCode" id="cb317"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb317-1"><a href="complete-protocols.html#cb317-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="fl">0.5</span>, <span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymax=</span><span class="fl">1.5</span>, <span class="at">fill=</span><span class="st">&#39;blue&#39;</span>, <span class="at">alpha=</span><span class="fl">0.4</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-520-1.png" width="672" />
In this way, we define four colored rectangles that reflect the different amino acids categories; blue=basic, red=acidic, yellow=hydrophilic, grey=hydrophobic:</p>
<div class="sourceCode" id="cb318"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb318-1"><a href="complete-protocols.html#cb318-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="fl">0.5</span>, <span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymax=</span><span class="fl">3.5</span>, <span class="at">fill=</span><span class="st">&#39;blue&#39;</span>, <span class="at">alpha=</span><span class="fl">0.15</span>)</span>
<span id="cb318-2"><a href="complete-protocols.html#cb318-2" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="fl">3.5</span>, <span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymax=</span><span class="fl">5.5</span>, <span class="at">fill=</span><span class="st">&#39;red&#39;</span>, <span class="at">alpha=</span><span class="fl">0.15</span>)</span>
<span id="cb318-3"><a href="complete-protocols.html#cb318-3" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="fl">5.5</span>, <span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymax=</span><span class="fl">10.5</span>, <span class="at">fill=</span><span class="st">&#39;yellow&#39;</span>, <span class="at">alpha=</span><span class="fl">0.15</span>)</span>
<span id="cb318-4"><a href="complete-protocols.html#cb318-4" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="fl">10.5</span>, <span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymax=</span><span class="fl">20.5</span>, <span class="at">fill=</span><span class="st">&#39;black&#39;</span>, <span class="at">alpha=</span><span class="fl">0.15</span>)</span></code></pre></div>
<p>Let’s look at the result:</p>
<div class="sourceCode" id="cb319"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb319-1"><a href="complete-protocols.html#cb319-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-522-1.png" width="672" /></p>
<p>Adjusting the axis labels and adding a title and caption:</p>
<div class="sourceCode" id="cb320"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb320-1"><a href="complete-protocols.html#cb320-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb320-2"><a href="complete-protocols.html#cb320-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb320-3"><a href="complete-protocols.html#cb320-3" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Mapping the amino acid positions of HOXD13&quot;</span>,</span>
<span id="cb320-4"><a href="complete-protocols.html#cb320-4" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;shows a high abundance of alanines in the IDR&quot;</span>,</span>
<span id="cb320-5"><a href="complete-protocols.html#cb320-5" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Amino acid&quot;</span>,</span>
<span id="cb320-6"><a href="complete-protocols.html#cb320-6" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart | based on data from Basu et al, DOI: 10.1016/j.cell.2020.04.018&quot;</span>,</span>
<span id="cb320-7"><a href="complete-protocols.html#cb320-7" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 5&quot;</span></span>
<span id="cb320-8"><a href="complete-protocols.html#cb320-8" tabindex="-1"></a>  )</span></code></pre></div>
<p>And a final tweak of the label style and location:</p>
<div class="sourceCode" id="cb321"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb321-1"><a href="complete-protocols.html#cb321-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(</span>
<span id="cb321-2"><a href="complete-protocols.html#cb321-2" tabindex="-1"></a>    <span class="at">plot.caption =</span> <span class="fu">element_text</span>(</span>
<span id="cb321-3"><a href="complete-protocols.html#cb321-3" tabindex="-1"></a>      <span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>,</span>
<span id="cb321-4"><a href="complete-protocols.html#cb321-4" tabindex="-1"></a>      <span class="at">hjust =</span> <span class="dv">1</span></span>
<span id="cb321-5"><a href="complete-protocols.html#cb321-5" tabindex="-1"></a>    ))</span></code></pre></div>
<p>In the original paper, a region of the protein is annotated as an ‘intrinsically disordered region’ abbreviated as IDR. Here, we we use the <code>annotate()</code> function to add a rectangle and a label to the top of the plot:</p>
<div class="sourceCode" id="cb322"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb322-1"><a href="complete-protocols.html#cb322-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="st">&quot;rect&quot;</span>, <span class="at">xmin=</span><span class="dv">0</span>, <span class="at">xmax=</span><span class="dv">118</span>, <span class="at">ymin=</span><span class="dv">21</span>, <span class="at">ymax=</span><span class="dv">22</span>, <span class="at">fill=</span><span class="st">&#39;darkblue&#39;</span>) <span class="sc">+</span></span>
<span id="cb322-2"><a href="complete-protocols.html#cb322-2" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="at">x=</span><span class="dv">59</span>, <span class="at">y=</span><span class="dv">23</span>, <span class="at">alpha=</span><span class="dv">1</span>, <span class="at">color=</span><span class="st">&#39;darkblue&#39;</span>, <span class="at">size=</span><span class="dv">4</span>,<span class="at">label=</span><span class="st">&#39;IDR&#39;</span>)</span>
<span id="cb322-3"><a href="complete-protocols.html#cb322-3" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-525-1.png" width="672" /></p>
<p>To avoid clipping of the new label by the subtitle of the plot:</p>
<div class="sourceCode" id="cb323"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb323-1"><a href="complete-protocols.html#cb323-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">coord_cartesian</span>(<span class="at">clip =</span> <span class="st">&#39;off&#39;</span>)</span>
<span id="cb323-2"><a href="complete-protocols.html#cb323-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-526-1.png" width="672" /></p>
<p>The subtitle is quite close to the IDR label. Let’s give the subtitle a bit more room, by adding a margin at the bottom of the subtitle. This can be done with the <code>theme()</code> function to style the subtitle:</p>
<div class="sourceCode" id="cb324"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb324-1"><a href="complete-protocols.html#cb324-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">margin =</span> <span class="fu">margin</span>(<span class="at">b=</span><span class="dv">20</span>)))</span>
<span id="cb324-2"><a href="complete-protocols.html#cb324-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-527-1.png" width="672" /></p>
<p>Finally we can save the plot:</p>
<div class="sourceCode" id="cb325"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb325-1"><a href="complete-protocols.html#cb325-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_05.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb325-2"><a href="complete-protocols.html#cb325-2" tabindex="-1"></a> p</span>
<span id="cb325-3"><a href="complete-protocols.html#cb325-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-6" class="section level2 hasAnchor" number="4.6">
<h2><span class="header-section-number">4.6</span> Protocol 6 - Heatmap style visualization of timelapse data<a href="complete-protocols.html#protocol-6" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Lineplots are typically used to plot data from timeseries (also known as longitudinal data). However, in case of many samples/objects, this may result in a cluttered data visualization. To solve this, the same data can be presented as a heatmap, where every row is an object and the response is coded as a color. A downside is that it is less quantitative as it is difficult to ‘read’ the numbers from this kind of visualization. Still, it is a powerful visualization to plot a lot of data and show its dynamics and heterogeneity. The data and visualization is originally published by <a href="https://www.biorxiv.org/content/10.1101/2021.07.27.453948v3">Chavez-Abiega et al. (2021)</a></p>
<p>The multipurpose <code>{tidyverse}</code> package is used for data wrangling and plotting:</p>
<div class="sourceCode" id="cb327"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb327-1"><a href="complete-protocols.html#cb327-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>A CSV file with the data is loaded. Since the file is a couple of Megabytes, we use the faster <code>fread()</code> function from the package <code>data.table</code>:</p>
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb328-1"><a href="complete-protocols.html#cb328-1" tabindex="-1"></a><span class="fu">library</span>(data.table)</span></code></pre></div>
<pre><code>
Attaching package: &#39;data.table&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:lubridate&#39;:

    hour, isoweek, mday, minute, month, quarter, second, wday, week,
    yday, year</code></pre>
<pre><code>The following objects are masked from &#39;package:dplyr&#39;:

    between, first, last</code></pre>
<pre><code>The following object is masked from &#39;package:purrr&#39;:

    transpose</code></pre>
<div class="sourceCode" id="cb333"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb333-1"><a href="complete-protocols.html#cb333-1" tabindex="-1"></a>df_S1P <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="st">&quot;data/ERK_repsonse_to_S1P.csv&quot;</span>) </span>
<span id="cb333-2"><a href="complete-protocols.html#cb333-2" tabindex="-1"></a><span class="fu">head</span>(df_S1P)</span></code></pre></div>
<pre><code>   Time_in_min Ligand Unique_Object CN_ERK  Slide
         &lt;num&gt; &lt;char&gt;         &lt;int&gt;  &lt;num&gt; &lt;char&gt;
1:         2.5    S1P           100 0.3941     P1
2:         3.0    S1P           100 0.3921     P1
3:         3.5    S1P           100 0.3900     P1
4:         4.0    S1P           100 0.3880     P1
5:         4.5    S1P           100 0.3859     P1
6:         5.0    S1P           100 0.3839     P1</code></pre>
<p>The column ‘CN_ERK’ has the data on the activity that we will plot over time. Each number in the ‘Unique_Object’ column reflects an individual cell measurement and so we can use that to group the measurements using <code>group_by(Unique_Object)</code>. We subtract the average baseline activity from each trace by subtracting the data acquired at the first 5 timepoints: <code>CN_ERK[1:5]</code>. The data is stored in a new column with normalized ERK activity data ‘ERKn’:</p>
<div class="sourceCode" id="cb335"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb335-1"><a href="complete-protocols.html#cb335-1" tabindex="-1"></a>df_sub <span class="ot">&lt;-</span> df_S1P <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Unique_Object) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(Time_in_min) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">ERKn=</span>CN_ERK<span class="sc">-</span><span class="fu">mean</span>(CN_ERK[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])) <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>()</span></code></pre></div>
<p>Around Timepoint 23 (minutes), the ligand was added. To set this time point to zero, we subtract a value of 23 (minutes) from each point:</p>
<div class="sourceCode" id="cb336"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb336-1"><a href="complete-protocols.html#cb336-1" tabindex="-1"></a>df_sub<span class="sc">$</span>Time_in_min <span class="ot">&lt;-</span> df_sub<span class="sc">$</span>Time_in_min<span class="dv">-23</span></span></code></pre></div>
<p>The column ‘Unique_Object’ that identifies the individual cells contains natural numbers, but these need to be treated as qualitative data. Therefore, we change the data type in this column to a factor with <code>as.factor()</code>:</p>
<div class="sourceCode" id="cb337"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb337-1"><a href="complete-protocols.html#cb337-1" tabindex="-1"></a>df_sub <span class="ot">&lt;-</span> df_sub <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Unique_Object=</span><span class="fu">as.factor</span>(Unique_Object))</span></code></pre></div>
<p>To order objects, we need to order ‘Unique_Object’ according to something. That something can be the maximum value of the Erk activity:</p>
<div class="sourceCode" id="cb338"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb338-1"><a href="complete-protocols.html#cb338-1" tabindex="-1"></a>df_sub <span class="ot">&lt;-</span> df_sub <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Unique_Object =</span> <span class="fu">fct_reorder</span>(Unique_Object, ERKn, max))</span></code></pre></div>
<p>Plot the data in heatmap style. We use theme_void here to focus only on the data and we will deal with styling of the axes and labels later:</p>
<div class="sourceCode" id="cb339"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb339-1"><a href="complete-protocols.html#cb339-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_sub, <span class="fu">aes</span>(<span class="at">x=</span>Time_in_min, <span class="at">y=</span>Unique_Object,<span class="at">fill=</span>ERKn)) <span class="sc">+</span></span>
<span id="cb339-2"><a href="complete-protocols.html#cb339-2" tabindex="-1"></a>  <span class="fu">geom_tile</span>() <span class="sc">+</span> <span class="fu">theme_void</span>()<span class="sc">+</span></span>
<span id="cb339-3"><a href="complete-protocols.html#cb339-3" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>() </span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-557-1.png" width="672" /></p>
<p>Can we also sort the data based on something else? Definitely, but it requires a bit of understanding of functions. Previously, we used the maximum value. This is defined by the function <code>max</code>, which takes the maxium value from a vector of numbers. Let’s look at an artificial example:</p>
<div class="sourceCode" id="cb340"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb340-1"><a href="complete-protocols.html#cb340-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>,<span class="dv">9</span>)</span>
<span id="cb340-2"><a href="complete-protocols.html#cb340-2" tabindex="-1"></a><span class="fu">max</span>(x)</span></code></pre></div>
<pre><code>[1] 9</code></pre>
<p>Other functions that take a vector as input and return a single value as output can be used. Other existing examples are <code>mean()</code>, <code>sum()</code> and <code>min()</code>. We can also define a function:</p>
<div class="sourceCode" id="cb342"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb342-1"><a href="complete-protocols.html#cb342-1" tabindex="-1"></a>second <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {x[<span class="dv">2</span>]}</span>
<span id="cb342-2"><a href="complete-protocols.html#cb342-2" tabindex="-1"></a><span class="fu">second</span>(x)</span></code></pre></div>
<pre><code>[1] 3</code></pre>
<div class="sourceCode" id="cb344"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb344-1"><a href="complete-protocols.html#cb344-1" tabindex="-1"></a>one_but_last <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {x[<span class="fu">length</span>(x)<span class="sc">-</span><span class="dv">1</span>]}</span>
<span id="cb344-2"><a href="complete-protocols.html#cb344-2" tabindex="-1"></a><span class="fu">one_but_last</span>(x)</span></code></pre></div>
<pre><code>[1] 7</code></pre>
<p>We can use the new function to sort the dataframe:</p>
<div class="sourceCode" id="cb346"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb346-1"><a href="complete-protocols.html#cb346-1" tabindex="-1"></a>df_sub <span class="ot">&lt;-</span> df_sub <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Unique_Object =</span> <span class="fu">fct_reorder</span>(Unique_Object, ERKn, one_but_last)) </span>
<span id="cb346-2"><a href="complete-protocols.html#cb346-2" tabindex="-1"></a></span>
<span id="cb346-3"><a href="complete-protocols.html#cb346-3" tabindex="-1"></a><span class="fu">ggplot</span>(df_sub, <span class="fu">aes</span>(<span class="at">x=</span>Time_in_min, <span class="at">y=</span>Unique_Object,<span class="at">fill=</span>ERKn)) <span class="sc">+</span></span>
<span id="cb346-4"><a href="complete-protocols.html#cb346-4" tabindex="-1"></a>  <span class="fu">geom_tile</span>() <span class="sc">+</span> <span class="fu">theme_void</span>()<span class="sc">+</span></span>
<span id="cb346-5"><a href="complete-protocols.html#cb346-5" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>() </span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-560-1.png" width="672" /></p>
<p>If we want to sort on the sum of the top five values we can define a function:</p>
<div class="sourceCode" id="cb347"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb347-1"><a href="complete-protocols.html#cb347-1" tabindex="-1"></a>top_five <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {<span class="fu">sum</span>(<span class="fu">tail</span>(<span class="fu">sort</span>(x),<span class="dv">5</span>))}</span></code></pre></div>
<p>But we can also directly implement the function in the <code>fct_reorder()</code> function:</p>
<div class="sourceCode" id="cb348"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb348-1"><a href="complete-protocols.html#cb348-1" tabindex="-1"></a>df_6 <span class="ot">&lt;-</span> df_sub <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Unique_Object =</span> <span class="fu">fct_reorder</span>(Unique_Object, ERKn, <span class="cf">function</span>(x) {<span class="fu">sum</span>(<span class="fu">tail</span>(<span class="fu">sort</span>(x),<span class="dv">5</span>))}))</span></code></pre></div>
<p>Let’s save this data:</p>
<div class="sourceCode" id="cb349"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb349-1"><a href="complete-protocols.html#cb349-1" tabindex="-1"></a>df_6 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_6.csv&quot;</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb350"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb350-1"><a href="complete-protocols.html#cb350-1" tabindex="-1"></a>p <span class="ot">&lt;-</span>  <span class="fu">ggplot</span>(df_6, <span class="fu">aes</span>(<span class="at">x=</span>Time_in_min, <span class="at">y=</span>Unique_Object,<span class="at">fill=</span>ERKn)) <span class="sc">+</span> <span class="fu">geom_tile</span>() <span class="sc">+</span></span>
<span id="cb350-2"><a href="complete-protocols.html#cb350-2" tabindex="-1"></a>  <span class="co"># scale_x_continuous(breaks=seq(0,60, by=15), labels=seq(0,60, by=15), limits=c(-8,60))  +</span></span>
<span id="cb350-3"><a href="complete-protocols.html#cb350-3" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>(<span class="at">na.value=</span><span class="st">&quot;black&quot;</span>, <span class="at">limits =</span> <span class="fu">range</span>(<span class="sc">-</span><span class="fl">0.1</span>,<span class="fl">1.5</span>))</span></code></pre></div>
<p>Let’s look at the plot</p>
<div class="sourceCode" id="cb351"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb351-1"><a href="complete-protocols.html#cb351-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-565-1.png" width="672" /></p>
<p>Add labels:</p>
<div class="sourceCode" id="cb352"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb352-1"><a href="complete-protocols.html#cb352-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb352-2"><a href="complete-protocols.html#cb352-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb352-3"><a href="complete-protocols.html#cb352-3" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;ERK activity is increased by S1P&quot;</span>,</span>
<span id="cb352-4"><a href="complete-protocols.html#cb352-4" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;and varies strongly between cells&quot;</span>,</span>
<span id="cb352-5"><a href="complete-protocols.html#cb352-5" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Time after addition of S1P (min)&quot;</span>,</span>
<span id="cb352-6"><a href="complete-protocols.html#cb352-6" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Cells sorted according to response&quot;</span>,</span>
<span id="cb352-7"><a href="complete-protocols.html#cb352-7" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart | data from Chavez-Abiega, DOI: 10.1101/2021.07.27.453948&quot;</span>,</span>
<span id="cb352-8"><a href="complete-protocols.html#cb352-8" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 6&quot;</span>,</span>
<span id="cb352-9"><a href="complete-protocols.html#cb352-9" tabindex="-1"></a>    <span class="at">fill=</span> <span class="st">&quot;ERK C/N ratio&quot;</span></span>
<span id="cb352-10"><a href="complete-protocols.html#cb352-10" tabindex="-1"></a>  )</span></code></pre></div>
<p>The <code>theme_void()</code> would be close to what we want as a theme, but I prefer to start from <code>theme_light</code> and remove the redundant features (grids and y-axis labels):</p>
<div class="sourceCode" id="cb353"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb353-1"><a href="complete-protocols.html#cb353-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb353-2"><a href="complete-protocols.html#cb353-2" tabindex="-1"></a>  </span>
<span id="cb353-3"><a href="complete-protocols.html#cb353-3" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>, <span class="at">hjust =</span> <span class="fl">2.0</span>),</span>
<span id="cb353-4"><a href="complete-protocols.html#cb353-4" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.1</span>, <span class="at">margin =</span> <span class="fu">margin</span>(<span class="at">t=</span><span class="dv">10</span>)),</span>
<span id="cb353-5"><a href="complete-protocols.html#cb353-5" tabindex="-1"></a>        <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.1</span>, <span class="at">margin =</span> <span class="fu">margin</span>(<span class="at">t=</span><span class="dv">2</span>, <span class="at">b=</span><span class="dv">5</span>)),</span>
<span id="cb353-6"><a href="complete-protocols.html#cb353-6" tabindex="-1"></a>        <span class="co"># Remove background</span></span>
<span id="cb353-7"><a href="complete-protocols.html#cb353-7" tabindex="-1"></a>        <span class="at">panel.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb353-8"><a href="complete-protocols.html#cb353-8" tabindex="-1"></a>        <span class="co"># Remove borders</span></span>
<span id="cb353-9"><a href="complete-protocols.html#cb353-9" tabindex="-1"></a>        <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb353-10"><a href="complete-protocols.html#cb353-10" tabindex="-1"></a>        <span class="co"># Remove grid</span></span>
<span id="cb353-11"><a href="complete-protocols.html#cb353-11" tabindex="-1"></a>        <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb353-12"><a href="complete-protocols.html#cb353-12" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb353-13"><a href="complete-protocols.html#cb353-13" tabindex="-1"></a>        <span class="co"># Remove text of the y-axis</span></span>
<span id="cb353-14"><a href="complete-protocols.html#cb353-14" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb353-15"><a href="complete-protocols.html#cb353-15" tabindex="-1"></a>        <span class="co"># Remove ticks on y-axis</span></span>
<span id="cb353-16"><a href="complete-protocols.html#cb353-16" tabindex="-1"></a>        <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb353-17"><a href="complete-protocols.html#cb353-17" tabindex="-1"></a>        <span class="co"># Remove label of y-axis</span></span>
<span id="cb353-18"><a href="complete-protocols.html#cb353-18" tabindex="-1"></a>        <span class="at">axis.title.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb353-19"><a href="complete-protocols.html#cb353-19" tabindex="-1"></a>        <span class="co"># Make x-axis ticks more pronounced</span></span>
<span id="cb353-20"><a href="complete-protocols.html#cb353-20" tabindex="-1"></a>        <span class="at">axis.ticks =</span> <span class="fu">element_line</span>(<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb353-21"><a href="complete-protocols.html#cb353-21" tabindex="-1"></a>  )</span>
<span id="cb353-22"><a href="complete-protocols.html#cb353-22" tabindex="-1"></a></span>
<span id="cb353-23"><a href="complete-protocols.html#cb353-23" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-567-1.png" width="672" /></p>
<p>The caption does not look good in this plot, but it has been optimized to look good in the saved PNG. To get a proper aligned caption in the Rmd you may need to optimize the hjust value in <code>theme((plot.caption = element_text()))</code></p>
<p>To save the plot as PNG:</p>
<div class="sourceCode" id="cb354"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb354-1"><a href="complete-protocols.html#cb354-1" tabindex="-1"></a> <span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_06.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb354-2"><a href="complete-protocols.html#cb354-2" tabindex="-1"></a>   p</span>
<span id="cb354-3"><a href="complete-protocols.html#cb354-3" tabindex="-1"></a> <span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-7" class="section level2 hasAnchor" number="4.7">
<h2><span class="header-section-number">4.7</span> Protocol 7 - Ridgeline plot<a href="complete-protocols.html#protocol-7" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In protocol 6 we have seen how time traces can be visualized with a heatmap style data visualization.
The advantage is that huge amounts of data can be simultaneously visualized. However, it is difficult to relate the color to numbers.
As such the color coded values in a heatmap give a qualitative view of heterogeneity.
In this protocol, we use a method that allows more data than an ordinary lineplot, but keeps the advantage of this plot, i.e. allowing a quantitative comparison.
To do this, an offset between curves is introduced. A package to this directly in ggplot2 is available (<code>{ggridges}</code>) but it may not work well when the values show a negative and positive deviation from baseline values (usually the baseline is zero).
In addition, we have more flexibility (and insight) when we do this ourselves.</p>
<p>Load the <code>{tidyverse}</code> package and the data from a CSV file:</p>
<div class="sourceCode" id="cb356"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb356-1"><a href="complete-protocols.html#cb356-1" tabindex="-1"></a><span class="fu">require</span>(tidyverse)</span>
<span id="cb356-2"><a href="complete-protocols.html#cb356-2" tabindex="-1"></a>df1 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;data/Rac_S1P.csv&quot;</span>)</span>
<span id="cb356-3"><a href="complete-protocols.html#cb356-3" tabindex="-1"></a><span class="fu">head</span>(df1)</span></code></pre></div>
<pre><code>       Time    Cell.1    Cell.2    Cell.3    Cell.4    Cell.5    Cell.6
1 0.0000000 1.0012160 1.0026460 1.0022090 0.9917870 0.9935569 0.9961453
2 0.1666667 0.9994997 0.9928106 0.9997658 0.9975348 1.0018910 1.0039790
3 0.3333333 0.9908362 0.9964057 0.9905094 0.9946743 0.9961497 0.9953369
4 0.5000000 0.9991967 0.9972504 0.9972806 1.0074250 1.0060510 1.0062390
5 0.6666667 1.0093450 1.0109910 1.0103590 1.0084080 1.0022130 0.9982496
6 0.8333333 0.9941078 0.9940830 0.9990720 1.0181230 1.0110220 1.0139400
     Cell.7    Cell.8    Cell.9   Cell.10   Cell.11   Cell.12   Cell.13
1 0.9964277 1.0006770 0.9999106 1.0043270 1.0086590 1.0074670 0.9875549
2 1.0098780 1.0015050 1.0014380 0.9938114 0.9791933 0.9925425 1.0106440
3 0.9917276 1.0000020 0.9997350 1.0049370 1.0250660 1.0079290 1.0062720
4 1.0039710 1.0046840 0.9976190 0.9971925 1.0008340 0.9975954 1.0050230
5 0.9979655 0.9930890 1.0012990 0.9997435 0.9865567 0.9944269 0.9907017
6 1.0087770 0.9916067 1.0025310 0.9919835 0.9930975 1.0012420 1.0025800
    Cell.14   Cell.15   Cell.16   Cell.17   Cell.18   Cell.19   Cell.20
1 0.9941667 0.9931722 0.9918481 0.9973316 0.9976466 0.9982940 0.9954924
2 1.0024420 0.9980425 0.9974822 0.9976164 0.9881339 1.0122250 1.0132220
3 1.0096800 1.0074450 1.0073690 0.9980090 0.9980346 0.9923310 0.9891400
4 1.0007800 0.9995486 1.0018860 1.0064760 1.0073410 0.9959588 0.9984370
5 0.9930816 1.0018040 1.0014300 1.0006490 1.0088410 1.0013600 1.0040700
6 1.0117190 1.0082190 1.0092750 1.0018410 0.9976196 1.0008680 1.0013100
    Cell.21   Cell.22   Cell.23   Cell.24   Cell.25   Cell.26   Cell.27
1 1.0024980 0.9986917 0.9987728 0.9988987 1.0019450 0.9954534 1.0025060
2 1.0022340 1.0023420 0.9963444 0.9969288 0.9990684 0.9973828 0.9944580
3 1.0003150 1.0014660 0.9985501 0.9983435 0.9944611 0.9987581 0.9935282
4 0.9969226 0.9956030 1.0020300 1.0063250 1.0036700 1.0056470 1.0065540
5 0.9980916 1.0018990 1.0042990 0.9995825 1.0009020 1.0028340 1.0031070
6 0.9993777 1.0041500 0.9984087 1.0016190 1.0007090 1.0022360 1.0007950
    Cell.28   Cell.29   Cell.30   Cell.31   Cell.32
1 0.9957569 0.9852318 1.0007450 0.9927866 0.9871355
2 0.9952635 0.9931840 1.0030060 0.9977890 1.0028110
3 0.9976221 1.0010230 0.9998598 1.0029210 0.9985340
4 1.0068090 1.0139990 0.9968880 1.0026060 1.0039670
5 1.0046560 1.0070850 0.9995161 1.0039100 1.0076550
6 1.0073290 1.0128910 1.0000250 1.0025260 0.9998251</code></pre>
<p>The data is not tidy, so it needs to be re-arranged:</p>
<div class="sourceCode" id="cb358"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb358-1"><a href="complete-protocols.html#cb358-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(df1, <span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(Time), <span class="at">names_to =</span> <span class="st">&quot;Cell&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;Activity&quot;</span>)</span>
<span id="cb358-2"><a href="complete-protocols.html#cb358-2" tabindex="-1"></a><span class="fu">head</span>(df_tidy)</span></code></pre></div>
<pre><code># A tibble: 6 × 3
   Time Cell   Activity
  &lt;dbl&gt; &lt;chr&gt;     &lt;dbl&gt;
1     0 Cell.1    1.00 
2     0 Cell.2    1.00 
3     0 Cell.3    1.00 
4     0 Cell.4    0.992
5     0 Cell.5    0.994
6     0 Cell.6    0.996</code></pre>
<p>In the next step, we create a new dataframe ‘df_rank’ to order the traces. We group the data by ‘Cell’ and extract the data from a specified time window with <code>filter()</code>. The filtered data is used to integrate the activity by using the function <code>sum()</code>. This summed value is used to generate a rank, ranging from 0 to 1:</p>
<div class="sourceCode" id="cb360"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb360-1"><a href="complete-protocols.html#cb360-1" tabindex="-1"></a>df_rank <span class="ot">&lt;-</span> df_tidy <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Cell) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Time<span class="sc">&gt;=</span><span class="dv">2</span> <span class="sc">&amp;</span> Time <span class="sc">&lt;=</span><span class="dv">10</span>) <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(<span class="at">amplitude=</span><span class="fu">sum</span>(Activity)) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">rank=</span><span class="fu">percent_rank</span>(amplitude))</span>
<span id="cb360-2"><a href="complete-protocols.html#cb360-2" tabindex="-1"></a><span class="fu">head</span>(df_rank)</span></code></pre></div>
<pre><code># A tibble: 6 × 3
  Cell    amplitude  rank
  &lt;chr&gt;       &lt;dbl&gt; &lt;dbl&gt;
1 Cell.1       50.3 0.742
2 Cell.10      50.9 0.806
3 Cell.11      49.8 0.645
4 Cell.12      50.3 0.774
5 Cell.13      49.6 0.548
6 Cell.14      49.4 0.484</code></pre>
<p>We can add the rank information from ‘df_rank’ to the ‘df_tidy’ dataframe:</p>
<div class="sourceCode" id="cb362"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb362-1"><a href="complete-protocols.html#cb362-1" tabindex="-1"></a>df_7 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">full_join</span>(df_tidy, df_rank,<span class="at">by=</span><span class="st">&quot;Cell&quot;</span>))</span></code></pre></div>
<p>This data is saved:</p>
<div class="sourceCode" id="cb363"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb363-1"><a href="complete-protocols.html#cb363-1" tabindex="-1"></a>df_7 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_7.csv&quot;</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>Let’s make a lineplot of this data and use the rank to shift the data plotted on the y-axis:</p>
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb364-1"><a href="complete-protocols.html#cb364-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_7, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>Activity<span class="sc">+</span>(rank<span class="sc">*</span><span class="dv">1</span>), <span class="at">group=</span>Cell, <span class="at">height =</span> rank)) <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">size=</span><span class="fl">0.5</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-594-1.png" width="672" /></p>
<p>We can use the rank to shift the plot also in the horizontal direction:</p>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb365-1"><a href="complete-protocols.html#cb365-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_7, <span class="fu">aes</span>(<span class="at">x=</span>Time<span class="sc">+</span>(rank<span class="sc">*</span><span class="dv">10</span>), <span class="at">y=</span>(Activity<span class="sc">+</span>(rank<span class="sc">*</span><span class="dv">1</span>)), <span class="at">group=</span>Cell)) <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">size=</span><span class="fl">0.5</span>)</span>
<span id="cb365-2"><a href="complete-protocols.html#cb365-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-595-1.png" width="672" /></p>
<p>Add labels:</p>
<div class="sourceCode" id="cb366"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb366-1"><a href="complete-protocols.html#cb366-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb366-2"><a href="complete-protocols.html#cb366-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb366-3"><a href="complete-protocols.html#cb366-3" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Activities measured in single cells over time&quot;</span>,</span>
<span id="cb366-4"><a href="complete-protocols.html#cb366-4" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;Sorting of the cells is based on integrated activity&quot;</span>,</span>
<span id="cb366-5"><a href="complete-protocols.html#cb366-5" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Time [min]&quot;</span>,</span>
<span id="cb366-6"><a href="complete-protocols.html#cb366-6" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Activity [arbitrary units]&quot;</span>,</span>
<span id="cb366-7"><a href="complete-protocols.html#cb366-7" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart | data from Reinhard et al; doi: 10.1091/mbc.E17-03-0136&quot;</span>,</span>
<span id="cb366-8"><a href="complete-protocols.html#cb366-8" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 7&quot;</span></span>
<span id="cb366-9"><a href="complete-protocols.html#cb366-9" tabindex="-1"></a>  )</span></code></pre></div>
<p>If we would like to use color, this would be a way to do that:</p>
<div class="sourceCode" id="cb367"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb367-1"><a href="complete-protocols.html#cb367-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">color=</span><span class="fu">as.factor</span>(rank))) <span class="sc">+</span></span>
<span id="cb367-2"><a href="complete-protocols.html#cb367-2" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-597-1.png" width="672" /></p>
<p>Although it looks flashy, we do not really need color here. So we stick to black and white and make some adjustments to the layout by tweaking the theme settings. To remove the grid and show the axis:</p>
<div class="sourceCode" id="cb368"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb368-1"><a href="complete-protocols.html#cb368-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">16</span>)</span>
<span id="cb368-2"><a href="complete-protocols.html#cb368-2" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb368-3"><a href="complete-protocols.html#cb368-3" tabindex="-1"></a>               <span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>),</span>
<span id="cb368-4"><a href="complete-protocols.html#cb368-4" tabindex="-1"></a>           <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb368-5"><a href="complete-protocols.html#cb368-5" tabindex="-1"></a>           <span class="cn">NULL</span>)</span>
<span id="cb368-6"><a href="complete-protocols.html#cb368-6" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-598-1.png" width="672" /></p>
<p>To give it more of a 3D feel we can add a third axis by defining a line:</p>
<div class="sourceCode" id="cb369"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb369-1"><a href="complete-protocols.html#cb369-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;segment&quot;</span>, <span class="at">x=</span><span class="dv">0</span>,<span class="at">y=</span><span class="dv">1</span>,<span class="at">xend=</span><span class="dv">10</span>,<span class="at">yend=</span><span class="dv">2</span>, <span class="at">size=</span><span class="dv">1</span>)</span></code></pre></div>
<p>The next step is to remove the original x-axis, which is a bit too long and also replace that with a line that runs until 20 (minutes):</p>
<div class="sourceCode" id="cb370"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb370-1"><a href="complete-protocols.html#cb370-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.line.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb370-2"><a href="complete-protocols.html#cb370-2" tabindex="-1"></a>               <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.3</span>)</span>
<span id="cb370-3"><a href="complete-protocols.html#cb370-3" tabindex="-1"></a>               ) <span class="sc">+</span></span>
<span id="cb370-4"><a href="complete-protocols.html#cb370-4" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;segment&quot;</span>, <span class="at">x=</span><span class="dv">0</span>,<span class="at">y=</span><span class="dv">1</span>,<span class="at">xend=</span><span class="dv">20</span>,<span class="at">yend=</span><span class="dv">1</span>, <span class="at">size=</span><span class="dv">1</span>)</span>
<span id="cb370-5"><a href="complete-protocols.html#cb370-5" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-600-1.png" width="672" /></p>
<p>Finally, we can use <code>scale_x_continuous()</code> to improve two aspects of the x-axis. First, the labels run up to 30, but we can set the scale to 0-20 with <code>breaks = c(0,20)</code>. Second, the default in ggplot is add a bit of margin to the plot, that’s why the x-axis does not touch the y-axis. This can be solved by using <code>expand = c(0, 0)</code>:
We can manually add two</p>
<div class="sourceCode" id="cb371"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb371-1"><a href="complete-protocols.html#cb371-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">10</span>,<span class="dv">20</span>), <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>))</span>
<span id="cb371-2"><a href="complete-protocols.html#cb371-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-601-1.png" width="672" /></p>
<p>To save the plot as PNG:</p>
<div class="sourceCode" id="cb372"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb372-1"><a href="complete-protocols.html#cb372-1" tabindex="-1"></a> <span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_07.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb372-2"><a href="complete-protocols.html#cb372-2" tabindex="-1"></a>   p</span>
<span id="cb372-3"><a href="complete-protocols.html#cb372-3" tabindex="-1"></a> <span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-8" class="section level2 hasAnchor" number="4.8">
<h2><span class="header-section-number">4.8</span> Protocol 8 - Plotting data in a 96-wells layout<a href="complete-protocols.html#protocol-8" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This protocol displays data from a 96-wells plate in the same format as the plate. This data visualization of 96-well plate data is also used in the <a href="https://doi.org/10.12688/f1000research.73641.1">plotXpress app</a>.</p>
<p>We start by loading a package that we need:</p>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb374-1"><a href="complete-protocols.html#cb374-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>A good example of data wrangling related to 96-wells data is given in <a href="complete-protocols.html#protocol-4">protocol 4</a>. Here we start from the tidy data that was the result of data tiyding in the section <a href="#Data-in-96-wells-format">Data in 96-wells format</a>:</p>
<div class="sourceCode" id="cb375"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb375-1"><a href="complete-protocols.html#cb375-1" tabindex="-1"></a>df_8 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;data/df_tidy_wells.csv&quot;</span>)</span>
<span id="cb375-2"><a href="complete-protocols.html#cb375-2" tabindex="-1"></a><span class="fu">head</span>(df_8)</span></code></pre></div>
<pre><code>  X column row Well Intensity
1 1      1   A   A1      2010
2 2      1   B   B1      3210
3 3      1   C   C1      1965
4 4      1   D   D1      2381
5 5      1   E   E1      1292
6 6      1   F   F1       991</code></pre>
<p>We can construct a basic plot where each well is represented by a large dot and the color used to fill the dot represents the intensity. This is much like a heatmap, but it uses a dot instead of a tile:</p>
<div class="sourceCode" id="cb377"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb377-1"><a href="complete-protocols.html#cb377-1" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>df_8, <span class="fu">aes</span>(<span class="at">x=</span>column, <span class="at">y=</span>row)) <span class="sc">+</span></span>
<span id="cb377-2"><a href="complete-protocols.html#cb377-2" tabindex="-1"></a><span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color=</span>Intensity), <span class="at">size=</span><span class="dv">10</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-620-1.png" width="672" />
There is a couple of things that need to be fixed. First, we invert the row names, to put the row ‘A’ at the top. In addition, the column names are depicted as factors (this can also be achieved by mutating the dataframe):</p>
<div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb378-1"><a href="complete-protocols.html#cb378-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>df_8, <span class="fu">aes</span>(<span class="at">x=</span><span class="fu">as.factor</span>(column), <span class="at">y=</span><span class="fu">fct_rev</span>(row))) <span class="sc">+</span></span>
<span id="cb378-2"><a href="complete-protocols.html#cb378-2" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color=</span>Intensity), <span class="at">size=</span><span class="dv">10</span>)</span>
<span id="cb378-3"><a href="complete-protocols.html#cb378-3" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-621-1.png" width="672" /></p>
<p>Let’s change the color palette that displays intensity to viridis. The numbers in the legend are high and that is why they are difficult to read quickly. So we use the <code>scales</code> package to use the scientific notation:</p>
<div class="sourceCode" id="cb379"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb379-1"><a href="complete-protocols.html#cb379-1" tabindex="-1"></a>  p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_x_discrete</span>(<span class="at">position =</span> <span class="st">&quot;top&quot;</span>) <span class="sc">+</span></span>
<span id="cb379-2"><a href="complete-protocols.html#cb379-2" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">label =</span> scales<span class="sc">::</span>scientific)</span>
<span id="cb379-3"><a href="complete-protocols.html#cb379-3" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-622-1.png" width="672" /></p>
<p>In my opinion, the scientific notation does not work well, so let’s try to use commas as a thousands separator:</p>
<div class="sourceCode" id="cb380"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb380-1"><a href="complete-protocols.html#cb380-1" tabindex="-1"></a>  p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_x_discrete</span>(<span class="at">position =</span> <span class="st">&quot;top&quot;</span>) <span class="sc">+</span></span>
<span id="cb380-2"><a href="complete-protocols.html#cb380-2" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">label =</span> scales<span class="sc">::</span>comma)</span></code></pre></div>
<pre><code>Scale for x is already present.
Adding another scale for x, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb382-1"><a href="complete-protocols.html#cb382-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-623-1.png" width="672" /></p>
<p>This is a bit better, but the numbers need to be right aligned:</p>
<div class="sourceCode" id="cb383"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb383-1"><a href="complete-protocols.html#cb383-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.text.align =</span> <span class="dv">1</span>)</span>
<span id="cb383-2"><a href="complete-protocols.html#cb383-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-624-1.png" width="672" />
This looks better, now add the labels (and we skip the labels for the x- and y-axis):</p>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb384-1"><a href="complete-protocols.html#cb384-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb384-2"><a href="complete-protocols.html#cb384-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb384-3"><a href="complete-protocols.html#cb384-3" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Readings from a 96-wells plate&quot;</span>,</span>
<span id="cb384-4"><a href="complete-protocols.html#cb384-4" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="cn">NULL</span>,</span>
<span id="cb384-5"><a href="complete-protocols.html#cb384-5" tabindex="-1"></a>    <span class="at">x =</span> <span class="cn">NULL</span>,</span>
<span id="cb384-6"><a href="complete-protocols.html#cb384-6" tabindex="-1"></a>    <span class="at">y =</span> <span class="cn">NULL</span>,</span>
<span id="cb384-7"><a href="complete-protocols.html#cb384-7" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;</span><span class="sc">\n</span><span class="st">@joachimgoedhart</span><span class="sc">\n</span><span class="st">based on data from Brandorff et al., DOI: 10.1101/2021.07.08.451595&quot;</span>,</span>
<span id="cb384-8"><a href="complete-protocols.html#cb384-8" tabindex="-1"></a>    <span class="at">color=</span> <span class="st">&#39;Intensity [a.u.]&#39;</span>,</span>
<span id="cb384-9"><a href="complete-protocols.html#cb384-9" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 8&quot;</span></span>
<span id="cb384-10"><a href="complete-protocols.html#cb384-10" tabindex="-1"></a>  ) </span>
<span id="cb384-11"><a href="complete-protocols.html#cb384-11" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-625-1.png" width="672" /></p>
<p>Set the theme and font size:</p>
<div class="sourceCode" id="cb385"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb385-1"><a href="complete-protocols.html#cb385-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">14</span>)</span></code></pre></div>
<p>Note that this overrides any previously defined modifications of the <code>theme()</code> function, such as the alignment of the legend labels. So we need to define this again.</p>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb386-1"><a href="complete-protocols.html#cb386-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.text.align =</span> <span class="dv">1</span>)</span></code></pre></div>
<p>We also adjust the theme settings for the other elements:</p>
<div class="sourceCode" id="cb387"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb387-1"><a href="complete-protocols.html#cb387-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>),</span>
<span id="cb387-2"><a href="complete-protocols.html#cb387-2" tabindex="-1"></a>               <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey50&quot;</span>, <span class="at">face =</span> <span class="st">&quot;italic&quot;</span>),</span>
<span id="cb387-3"><a href="complete-protocols.html#cb387-3" tabindex="-1"></a>               <span class="co">#Remove the grid</span></span>
<span id="cb387-4"><a href="complete-protocols.html#cb387-4" tabindex="-1"></a>               <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb387-5"><a href="complete-protocols.html#cb387-5" tabindex="-1"></a>               <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>()</span>
<span id="cb387-6"><a href="complete-protocols.html#cb387-6" tabindex="-1"></a>)</span></code></pre></div>
<p>Let’s look at the result:</p>
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb388-1"><a href="complete-protocols.html#cb388-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-629-1.png" width="672" /></p>
<p>To save the plot as a png file:</p>
<div class="sourceCode" id="cb389"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb389-1"><a href="complete-protocols.html#cb389-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_08.png&quot;</span>), <span class="at">width =</span> <span class="dv">3000</span>, <span class="at">height =</span> <span class="dv">2000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb389-2"><a href="complete-protocols.html#cb389-2" tabindex="-1"></a> p <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span>
<span id="cb389-3"><a href="complete-protocols.html#cb389-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-9" class="section level2 hasAnchor" number="4.9">
<h2><span class="header-section-number">4.9</span> Protocol 9 - A dose response curve with fit<a href="complete-protocols.html#protocol-9" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this protocol we visualize the data that was used to make a ‘dose reponse curve’. The data consists of some response that is measured at different concentrations of a compound that induces the effect (agonist). This is a typical experiment in pharmacology to understand the relation between the dose and the response. One of the relevant parameters is the ‘Half maximal effective concentration’ abbreviated as EC50. This is the concentration at which 50% of the maximal response is measured and we obtain this value by curve fitting. In this protocol we use data acquired from single cells from different biological replicates. We generate a ‘superplot’ <a href="https://doi.org/10.1083/jcb.202001064">(Lord et al., 2021)</a> to distuinguish technical and sample replicates. More details about the experimental approach are published by <a href="https://doi.org/10.1242/jcs.259685">Chavez-Abiega et al. (2022)</a>.</p>
<p>We start by loading the <code>{tidyverse}</code> package:</p>
<div class="sourceCode" id="cb391"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb391-1"><a href="complete-protocols.html#cb391-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>Read the data:</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb392-1"><a href="complete-protocols.html#cb392-1" tabindex="-1"></a>df_DRC <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;data/DRC_Histamine.csv&quot;</span>, <span class="at">stringsAsFactors =</span> <span class="cn">TRUE</span>)</span>
<span id="cb392-2"><a href="complete-protocols.html#cb392-2" tabindex="-1"></a><span class="fu">head</span>(df_DRC)</span></code></pre></div>
<pre><code>  X Unique_Unique_Object  Experiment  Aktnac  ERKnac Condition Concen
1 1                   12 20200122 P1  0.0558 -0.2106       PTx      0
2 2                   15 20200122 P1  0.2458 -0.1424       PTx      0
3 3                   16 20200122 P1  0.0412 -0.2493       PTx      0
4 4                   22 20200122 P1 -0.1194 -0.1980       PTx      0
5 5                   23 20200122 P1  0.1838 -0.1540       PTx      0
6 6                   25 20200122 P1  0.3291 -0.7078       PTx      0</code></pre>
<p>The concentration is listed in the column ‘Concen’, the measured response is in the column ‘ERKnac’ and the column ‘Experiment’ identifies the replicates. Let’s rename the columns for easier identification of what they represent:</p>
<div class="sourceCode" id="cb394"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb394-1"><a href="complete-protocols.html#cb394-1" tabindex="-1"></a>df_DRC <span class="ot">&lt;-</span> df_DRC <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">Concentration=</span>Concen, <span class="at">Response=</span>ERKnac)</span>
<span id="cb394-2"><a href="complete-protocols.html#cb394-2" tabindex="-1"></a><span class="fu">head</span>(df_DRC)</span></code></pre></div>
<pre><code>  X Unique_Unique_Object  Experiment  Aktnac Response Condition Concentration
1 1                   12 20200122 P1  0.0558  -0.2106       PTx             0
2 2                   15 20200122 P1  0.2458  -0.1424       PTx             0
3 3                   16 20200122 P1  0.0412  -0.2493       PTx             0
4 4                   22 20200122 P1 -0.1194  -0.1980       PTx             0
5 5                   23 20200122 P1  0.1838  -0.1540       PTx             0
6 6                   25 20200122 P1  0.3291  -0.7078       PTx             0</code></pre>
<p>To change the column ‘Experiment’ from text into a number that represents the replicate, we can convert it using <code>as.numeric()</code>. Since we need these numbers as qualitative data (a label), we convert the numbers to a factor with <code>as.factor()</code>:</p>
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb396-1"><a href="complete-protocols.html#cb396-1" tabindex="-1"></a>df_DRC <span class="ot">&lt;-</span> df_DRC <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Replicate=</span><span class="fu">as.factor</span>(<span class="fu">as.integer</span>(Experiment)))</span>
<span id="cb396-2"><a href="complete-protocols.html#cb396-2" tabindex="-1"></a><span class="fu">head</span>(df_DRC)</span></code></pre></div>
<pre><code>  X Unique_Unique_Object  Experiment  Aktnac Response Condition Concentration
1 1                   12 20200122 P1  0.0558  -0.2106       PTx             0
2 2                   15 20200122 P1  0.2458  -0.1424       PTx             0
3 3                   16 20200122 P1  0.0412  -0.2493       PTx             0
4 4                   22 20200122 P1 -0.1194  -0.1980       PTx             0
5 5                   23 20200122 P1  0.1838  -0.1540       PTx             0
6 6                   25 20200122 P1  0.3291  -0.7078       PTx             0
  Replicate
1         1
2         1
3         1
4         1
5         1
6         1</code></pre>
<p>The range of concentrations at which the compound is examined spans a few orders of magnitude and therefore a log scale is used to display the concentrations. The minimal response is usually measured at a concentration of 0, but the logarithm of 0 is undefined. Therefore, plotting 0 on a logscale will give an error. The logarithm of 0 can be approached by minus infinity. Therefore, we convert the concentration of 0 to a low value, in this case 0.01:</p>
<div class="sourceCode" id="cb398"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb398-1"><a href="complete-protocols.html#cb398-1" tabindex="-1"></a>df_DRC <span class="ot">&lt;-</span> df_DRC  <span class="sc">%&gt;%</span></span>
<span id="cb398-2"><a href="complete-protocols.html#cb398-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Concentration =</span> <span class="fu">ifelse</span>((Concentration <span class="sc">==</span> <span class="dv">0</span>),</span>
<span id="cb398-3"><a href="complete-protocols.html#cb398-3" tabindex="-1"></a>                         <span class="at">yes =</span> <span class="fl">0.01</span>,</span>
<span id="cb398-4"><a href="complete-protocols.html#cb398-4" tabindex="-1"></a>                         <span class="at">no =</span> Concentration)</span>
<span id="cb398-5"><a href="complete-protocols.html#cb398-5" tabindex="-1"></a>  )</span></code></pre></div>
<p>Next we take the ‘Response’ and calculate the average per concentration and for each biological replicate and store that information in a new dataframe:</p>
<div class="sourceCode" id="cb399"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb399-1"><a href="complete-protocols.html#cb399-1" tabindex="-1"></a>df_summary <span class="ot">&lt;-</span> df_DRC <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Concentration, Replicate) <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(<span class="at">mean_Response=</span><span class="fu">mean</span>(Response))</span></code></pre></div>
<pre><code>`summarise()` has grouped output by &#39;Concentration&#39;. You can override using the
`.groups` argument.</code></pre>
<p>We can define a plot that shows the data with <code>geom_jitter()</code> and the average with a large dot with <code>geom_point()</code>. Each replicate has its own color and the data is plotted on a log scale with <code>scale_x_log10()</code>:</p>
<div class="sourceCode" id="cb401"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb401-1"><a href="complete-protocols.html#cb401-1" tabindex="-1"></a>p <span class="ot">&lt;-</span>  <span class="fu">ggplot</span>(<span class="at">data =</span> df_DRC, <span class="fu">aes</span>(<span class="at">x =</span> Concentration, <span class="at">y =</span> Response)) <span class="sc">+</span> </span>
<span id="cb401-2"><a href="complete-protocols.html#cb401-2" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(<span class="at">x =</span> Concentration, <span class="at">y =</span> Response, <span class="at">color=</span>Replicate), <span class="at">width=</span><span class="fl">0.2</span>, <span class="at">size=</span><span class="dv">2</span>, <span class="at">shape=</span><span class="dv">16</span>, <span class="at">alpha=</span><span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb401-3"><a href="complete-protocols.html#cb401-3" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>df_summary, <span class="fu">aes</span>(<span class="at">x =</span> Concentration, <span class="at">y =</span> mean_Response, <span class="at">fill=</span>Replicate), <span class="at">size=</span><span class="dv">8</span>, <span class="at">shape=</span><span class="dv">21</span>, <span class="at">alpha=</span><span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb401-4"><a href="complete-protocols.html#cb401-4" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>()</span>
<span id="cb401-5"><a href="complete-protocols.html#cb401-5" tabindex="-1"></a></span>
<span id="cb401-6"><a href="complete-protocols.html#cb401-6" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-652-1.png" width="672" /></p>
<p>Adjusting the theme and the y-axis scale improves the plot:</p>
<div class="sourceCode" id="cb402"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb402-1"><a href="complete-protocols.html#cb402-1" tabindex="-1"></a>p <span class="ot">&lt;-</span>  p <span class="sc">+</span></span>
<span id="cb402-2"><a href="complete-protocols.html#cb402-2" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb402-3"><a href="complete-protocols.html#cb402-3" tabindex="-1"></a>  <span class="fu">theme_light</span>(<span class="dv">16</span>)</span>
<span id="cb402-4"><a href="complete-protocols.html#cb402-4" tabindex="-1"></a></span>
<span id="cb402-5"><a href="complete-protocols.html#cb402-5" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-653-1.png" width="672" /></p>
<p>Add labels:</p>
<div class="sourceCode" id="cb403"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb403-1"><a href="complete-protocols.html#cb403-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb403-2"><a href="complete-protocols.html#cb403-2" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;Responses of individual cells to Histamine&quot;</span>,</span>
<span id="cb403-3"><a href="complete-protocols.html#cb403-3" tabindex="-1"></a>  <span class="at">x =</span> <span class="st">&quot;Histamine concentration [µM]&quot;</span>,</span>
<span id="cb403-4"><a href="complete-protocols.html#cb403-4" tabindex="-1"></a>  <span class="at">y =</span> <span class="st">&quot;Response [arbitrary units]&quot;</span>,</span>
<span id="cb403-5"><a href="complete-protocols.html#cb403-5" tabindex="-1"></a>  <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">based on data from Chavez-Abiega et al., DOI: 10.1242/jcs.259685&quot;</span>,</span>
<span id="cb403-6"><a href="complete-protocols.html#cb403-6" tabindex="-1"></a>  <span class="at">tag =</span> <span class="st">&quot;Protocol 9&quot;</span></span>
<span id="cb403-7"><a href="complete-protocols.html#cb403-7" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb403-8"><a href="complete-protocols.html#cb403-8" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>))</span></code></pre></div>
<p>And to label the different replicates with a colorblind friendly palette, we define a set of colors that were proposed by Paul Tol: <a href="https://personal.sron.nl/~pault/" class="uri">https://personal.sron.nl/~pault/</a></p>
<div class="sourceCode" id="cb404"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb404-1"><a href="complete-protocols.html#cb404-1" tabindex="-1"></a>Tol_bright <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;66CCEE&#39;</span>, <span class="st">&#39;#BBBBBB&#39;</span>, <span class="st">&#39;#CCBB44&#39;</span>,<span class="st">&#39;#AA3377&#39;</span>,<span class="st">&#39;#228833&#39;</span>, <span class="st">&#39;#4477AA&#39;</span>)</span></code></pre></div>
<p>To use these colors we define manual color scales for both ‘fill’ (used for <code>geom_point()</code>) and ‘color’ (used for <code>geom_jitter()</code>):</p>
<div class="sourceCode" id="cb405"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb405-1"><a href="complete-protocols.html#cb405-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> Tol_bright) <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values =</span> Tol_bright)</span>
<span id="cb405-2"><a href="complete-protocols.html#cb405-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-656-1.png" width="672" /></p>
<p>The result is a dose response curve in which the replicates are clearly identified with colorblind friendly colors. The EC50 can be determined from fitting a curve. To this end, we use the function <code>nls()</code>, which needs an equation, the data and estimates of the values. It will perform a ‘nonlinear least squares’ optimization to find parameters for the equation that best fit with the data of ‘df_summary’. The parameters are ‘bot’, ‘top’, ‘EC50’ and ‘slope’:</p>
<div class="sourceCode" id="cb406"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb406-1"><a href="complete-protocols.html#cb406-1" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">nls</span>(mean_Response <span class="sc">~</span> top<span class="sc">+</span>(bot<span class="sc">-</span>top)<span class="sc">/</span>(<span class="dv">1</span><span class="sc">+</span>(Concentration<span class="sc">/</span>EC50)<span class="sc">^</span>slope),</span>
<span id="cb406-2"><a href="complete-protocols.html#cb406-2" tabindex="-1"></a>           <span class="at">data =</span> df_summary,</span>
<span id="cb406-3"><a href="complete-protocols.html#cb406-3" tabindex="-1"></a>           <span class="at">start=</span><span class="fu">list</span>(<span class="at">bot=</span><span class="sc">-</span><span class="dv">2</span>, <span class="at">top=</span><span class="dv">8</span>, <span class="at">EC50=</span><span class="dv">1</span>, <span class="at">slope=</span><span class="dv">1</span>))</span></code></pre></div>
<p>The result is stored in the object ‘fit’ and the parameters can be listed:</p>
<div class="sourceCode" id="cb407"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb407-1"><a href="complete-protocols.html#cb407-1" tabindex="-1"></a><span class="fu">coef</span>(fit)</span></code></pre></div>
<pre><code>       bot        top       EC50      slope 
-0.1990749  3.2915666  0.4968077  0.7141987 </code></pre>
<p>To plot the fitted data, we use <code>augment()</code> function that is part of the <code>{broom}</code> package, which we need to load:</p>
<div class="sourceCode" id="cb409"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb409-1"><a href="complete-protocols.html#cb409-1" tabindex="-1"></a><span class="fu">library</span>(broom)</span>
<span id="cb409-2"><a href="complete-protocols.html#cb409-2" tabindex="-1"></a></span>
<span id="cb409-3"><a href="complete-protocols.html#cb409-3" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="at">data =</span> <span class="fu">augment</span>(fit), <span class="fu">aes</span>(<span class="at">x=</span>Concentration, <span class="at">y=</span>.fitted), <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size=</span><span class="dv">1</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-659-1.png" width="672" /></p>
<p>Instead of <code>geom_lin()</code> for plotting the fit, we can also use <code>geom_smooth()</code>:</p>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb410-1"><a href="complete-protocols.html#cb410-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">data =</span> <span class="fu">augment</span>(fit), <span class="fu">aes</span>(<span class="at">x=</span>Concentration, <span class="at">y=</span>.fitted), <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size=</span><span class="dv">1</span>)</span>
<span id="cb410-2"><a href="complete-protocols.html#cb410-2" tabindex="-1"></a>p</span></code></pre></div>
<pre><code>`geom_smooth()` using method = &#39;loess&#39; and formula = &#39;y ~ x&#39;</code></pre>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-660-1.png" width="672" /></p>
<p>An alternative way to do the curve fit is provided through the <code>{drc}</code> package. To demonstrate this, we load the package and add the curve fit to the data:</p>
<div class="sourceCode" id="cb412"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb412-1"><a href="complete-protocols.html#cb412-1" tabindex="-1"></a><span class="fu">library</span>(drc)</span></code></pre></div>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: &#39;MASS&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:dplyr&#39;:

    select</code></pre>
<pre><code>
&#39;drc&#39; has been loaded.</code></pre>
<pre><code>Please cite R and &#39;drc&#39; if used for a publication,</code></pre>
<pre><code>for references type &#39;citation()&#39; and &#39;citation(&#39;drc&#39;)&#39;.</code></pre>
<pre><code>
Attaching package: &#39;drc&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    gaussian, getInitial</code></pre>
<div class="sourceCode" id="cb421"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb421-1"><a href="complete-protocols.html#cb421-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">data=</span>df_summary, <span class="fu">aes</span>(<span class="at">x =</span> Concentration, <span class="at">y =</span> mean_Response), <span class="at">color=</span><span class="st">&#39;blue&#39;</span>, <span class="at">method =</span> drm, <span class="at">method.args =</span> <span class="fu">list</span>(<span class="at">fct =</span> <span class="fu">L.4</span>()), <span class="at">se =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>`geom_smooth()` using method = &#39;loess&#39; and formula = &#39;y ~ x&#39;</code></pre>
<pre><code>`geom_smooth()` using formula = &#39;y ~ x&#39;</code></pre>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-661-1.png" width="672" /></p>
<p>The <code>{drc}</code> package was used to fit the data in the paper by <a href="https://doi.org/10.1242/jcs.259685">Chavez-Abiega et al. (2022)</a>. But the approach followed above with <code>nls()</code> is easier to adapt for data that requires other models for the fit.</p>
<p>Finally, we can save the plot:</p>
<div class="sourceCode" id="cb424"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb424-1"><a href="complete-protocols.html#cb424-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_09.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb424-2"><a href="complete-protocols.html#cb424-2" tabindex="-1"></a> p</span></code></pre></div>
<pre><code>`geom_smooth()` using method = &#39;loess&#39; and formula = &#39;y ~ x&#39;</code></pre>
<div class="sourceCode" id="cb426"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb426-1"><a href="complete-protocols.html#cb426-1" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-10" class="section level2 hasAnchor" number="4.10">
<h2><span class="header-section-number">4.10</span> Protocol 10 - Plotting data that was harvested with a Google form<a href="complete-protocols.html#protocol-10" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The goal of this script is to read data that was generated by a Google form, transform it into a tidy format and plot the data. The form contains data that is submitted by students after doing a course on cell biology. The data that is submitted are length measurements on cells and in the end, we will plot these date.</p>
<p>First, we need the <code>{tidyverse}</code> package:</p>
<div class="sourceCode" id="cb428"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb428-1"><a href="complete-protocols.html#cb428-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>The data from the Google form are stored in a Google sheet that was published on the web as a CSV. This was achieved by ‘File &gt; Share &gt; Publish to web’ and selecting to ‘Publish’ as ‘Comma-separated values’.
After clicking on ‘Publish’ a link is generated that can be read as a csv, and any empty cells are converted to <code>NA</code>. Below is the <em>outcommented</em> code that loads the online google sheet. To make sure that this script works even when the google sheet is no longer active and to ensure a reproducible data visualization, I downloaded a snapshot (made on 21st of April, 2022) of the google sheet as a csv file and I’ll load that insetad:</p>
<div class="sourceCode" id="cb429"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb429-1"><a href="complete-protocols.html#cb429-1" tabindex="-1"></a><span class="co"># df_sheet &lt;- read.csv(&quot;https://docs.google.com/spreadsheets/d/e/2PACX-1vSc-nI1-s_u-XkNXEn_u2l6wkBafxJMHQ_Cd3kStrnToh7kawqjQU3y2l_1riLigKRkIqlNOqPrgkdW/pub?output=csv&quot;, na.strings = &quot;&quot;)</span></span>
<span id="cb429-2"><a href="complete-protocols.html#cb429-2" tabindex="-1"></a></span>
<span id="cb429-3"><a href="complete-protocols.html#cb429-3" tabindex="-1"></a>df_sheet <span class="ot">&lt;-</span>  <span class="fu">read.csv</span>(<span class="st">&quot;data/20220421_Resultaat-metingen.csv&quot;</span>, <span class="at">na.strings =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb429-4"><a href="complete-protocols.html#cb429-4" tabindex="-1"></a></span>
<span id="cb429-5"><a href="complete-protocols.html#cb429-5" tabindex="-1"></a><span class="fu">head</span>(df_sheet)</span></code></pre></div>
<pre><code>          Tijdstempel Groep
1                &lt;NA&gt;  &lt;NA&gt;
2                &lt;NA&gt;  &lt;NA&gt;
3                &lt;NA&gt;  &lt;NA&gt;
4                &lt;NA&gt;  &lt;NA&gt;
5                &lt;NA&gt;  &lt;NA&gt;
6 01/03/2021 15:42:23     B
  Resultaten.van.de.10.metingen..in.µm...van.de.celkleuring..methyleenblauw.
1                                                                       &lt;NA&gt;
2                                                                       &lt;NA&gt;
3                                                                       &lt;NA&gt;
4                                                                       &lt;NA&gt;
5                                                                       &lt;NA&gt;
6                 64.6, 52.5, 79.4, 48.4, 66.0, 64.4, 69.8, 71.2, 50.7, 51.0
  Resultaten.van.de.10.metingen..in.µm...van.de.kernkleuring..DAPI.
1                                                              &lt;NA&gt;
2                                                              &lt;NA&gt;
3                                                              &lt;NA&gt;
4                                                              &lt;NA&gt;
5                                                              &lt;NA&gt;
6          11.8, 7.8, 7.0, 12.4, 12.7, 14.0, 11.9, 12.9, 11.7, 13.2</code></pre>
<p>There are a number of issues with this data. The first 5 rows are empty. I deleted the values in these rows, since they were dummy values from testing the form. Let’s remove rows without data:</p>
<div class="sourceCode" id="cb431"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb431-1"><a href="complete-protocols.html#cb431-1" tabindex="-1"></a>df_sheet <span class="ot">&lt;-</span> df_sheet <span class="sc">%&gt;%</span> <span class="fu">na.omit</span>()</span></code></pre></div>
<p>The header names are still in the original language (Dutch) and the third and fourth header names with the relevant data are very long. Let’s change the names:</p>
<div class="sourceCode" id="cb432"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb432-1"><a href="complete-protocols.html#cb432-1" tabindex="-1"></a><span class="fu">colnames</span>(df_sheet) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Timestamp&quot;</span>, <span class="st">&quot;Group&quot;</span>, <span class="st">&quot;Cell&quot;</span>, <span class="st">&quot;Nucleus&quot;</span>)</span></code></pre></div>
<p>All the length measurements are distributed over two columns, let’s make it tidy and move it to one column:</p>
<div class="sourceCode" id="cb433"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb433-1"><a href="complete-protocols.html#cb433-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span></span>
<span id="cb433-2"><a href="complete-protocols.html#cb433-2" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb433-3"><a href="complete-protocols.html#cb433-3" tabindex="-1"></a>    df_sheet,</span>
<span id="cb433-4"><a href="complete-protocols.html#cb433-4" tabindex="-1"></a>    <span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(<span class="st">&quot;Timestamp&quot;</span>, <span class="st">&quot;Group&quot;</span>),</span>
<span id="cb433-5"><a href="complete-protocols.html#cb433-5" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">&quot;Sample&quot;</span>,</span>
<span id="cb433-6"><a href="complete-protocols.html#cb433-6" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">&quot;Size&quot;</span></span>
<span id="cb433-7"><a href="complete-protocols.html#cb433-7" tabindex="-1"></a>  )</span>
<span id="cb433-8"><a href="complete-protocols.html#cb433-8" tabindex="-1"></a><span class="fu">head</span>(df_tidy)</span></code></pre></div>
<pre><code># A tibble: 6 × 4
  Timestamp           Group Sample  Size                                        
  &lt;chr&gt;               &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;                                       
1 01/03/2021 15:42:23 B     Cell    64.6, 52.5, 79.4, 48.4, 66.0, 64.4, 69.8, 7…
2 01/03/2021 15:42:23 B     Nucleus 11.8, 7.8, 7.0, 12.4, 12.7, 14.0, 11.9, 12.…
3 01/03/2021 23:50:53 A     Cell    57.7, 64.9, 44.8, 60.8, 46.8, 70.1, 67.8, 7…
4 01/03/2021 23:50:53 A     Nucleus 9.6, 7.1, 8.7, 9.5, 8.1, 6, 8.3, 5.5, 7.8, …
5 02/03/2021 17:09:52 B     Cell    14.839,17.578,9.054,16.218,14.724,12.410,14…
6 02/03/2021 17:09:52 B     Nucleus 2.792,1.819,2.275,2.426,2.263,2.159,3.056,2…</code></pre>
<p>The column with measurements contains up to 10 length measurements. The values are separated by a comma and often a space. First, the space is removed and then each row is split into multiple rows, where each row has a single measurement. The comma is used to separate the values:</p>
<div class="sourceCode" id="cb435"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb435-1"><a href="complete-protocols.html#cb435-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> df_tidy <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Size =</span> <span class="fu">gsub</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;&quot;</span>, Size)) <span class="sc">%&gt;%</span> <span class="fu">separate_rows</span>(Size, <span class="at">sep=</span><span class="st">&quot;,&quot;</span>) </span></code></pre></div>
<p>The ‘Size’ column values are converted to the class numeric. The advantage is that anything that is not a number will be converted to <code>NA</code>. These non-numeric data can stem from incorrectly enetered data. Examples of these incorrect data are numbers with units or with wrong separators.</p>
<div class="sourceCode" id="cb436"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb436-1"><a href="complete-protocols.html#cb436-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> df_tidy <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Size =</span> <span class="fu">as.numeric</span>(Size))</span></code></pre></div>
<p>Finally, we can filter any values that fall outside the range of expected values (and this will also remove the cells with <code>NA</code>):</p>
<div class="sourceCode" id="cb437"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb437-1"><a href="complete-protocols.html#cb437-1" tabindex="-1"></a>df_size <span class="ot">&lt;-</span> df_tidy <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Size<span class="sc">&gt;</span><span class="dv">0</span> <span class="sc">&amp;</span> Size<span class="sc">&lt;</span><span class="dv">1000</span>)</span>
<span id="cb437-2"><a href="complete-protocols.html#cb437-2" tabindex="-1"></a><span class="fu">head</span>(df_size)</span></code></pre></div>
<pre><code># A tibble: 6 × 4
  Timestamp           Group Sample  Size
  &lt;chr&gt;               &lt;chr&gt; &lt;chr&gt;  &lt;dbl&gt;
1 01/03/2021 15:42:23 B     Cell    64.6
2 01/03/2021 15:42:23 B     Cell    52.5
3 01/03/2021 15:42:23 B     Cell    79.4
4 01/03/2021 15:42:23 B     Cell    48.4
5 01/03/2021 15:42:23 B     Cell    66  
6 01/03/2021 15:42:23 B     Cell    64.4</code></pre>
<p>The dataframe is cleaned and tidy and so it is ready for visualization. The primary information is the distribution of size measurements for the measurements of cells and cell nuclei. This can be achieved by using <code>geom_density()</code>. Given the range of sizes, it makes sense to use a log10 scale for the x-axis:</p>
<div class="sourceCode" id="cb439"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb439-1"><a href="complete-protocols.html#cb439-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_size, <span class="fu">aes</span>(<span class="at">x=</span>Size, <span class="at">fill=</span>Sample))<span class="sc">+</span><span class="fu">geom_density</span>(<span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb439-2"><a href="complete-protocols.html#cb439-2" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>()</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-689-1.png" width="672" /></p>
<p>The <code>geom_density()</code> produces a smooth distribution of the data. I actually prefer to see the real data and that’s why we turn to <code>geom_histogram</code>. I also will split the data according to the sample and groups and this can be achieved with <code>facet_grid()</code>. Here, we use the <code>labeller()</code> function to combine the column name ‘Group’ with the the factor it presents, as this makes it easier to understand what is plotted:</p>
<div class="sourceCode" id="cb440"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb440-1"><a href="complete-protocols.html#cb440-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_size, <span class="fu">aes</span>(<span class="at">x=</span>Size, <span class="at">fill=</span>Sample))<span class="sc">+</span><span class="fu">geom_histogram</span>(<span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb440-2"><a href="complete-protocols.html#cb440-2" tabindex="-1"></a>  <span class="fu">facet_grid</span>(Sample<span class="sc">~</span>Group, <span class="at">labeller =</span> <span class="fu">labeller</span>(<span class="at">Group=</span>label_both)) <span class="sc">+</span></span>
<span id="cb440-3"><a href="complete-protocols.html#cb440-3" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>()</span></code></pre></div>
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-690-1.png" width="672" /></p>
<p>The use of color is redundant, since the data is split in Cell and Nucleus anyway, so we remove that. The difference between groups is not large and for the final data visualization I chose to combine the data for all the groups:</p>
<div class="sourceCode" id="cb442"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb442-1"><a href="complete-protocols.html#cb442-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_size, <span class="fu">aes</span>(<span class="at">x=</span>Size)) <span class="sc">+</span></span>
<span id="cb442-2"><a href="complete-protocols.html#cb442-2" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">alpha=</span>.<span class="dv">5</span>, <span class="at">bins=</span><span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb442-3"><a href="complete-protocols.html#cb442-3" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="sc">~</span>Sample) <span class="sc">+</span></span>
<span id="cb442-4"><a href="complete-protocols.html#cb442-4" tabindex="-1"></a>  <span class="fu">annotation_logticks</span>(<span class="at">sides=</span><span class="st">&quot;b&quot;</span>, <span class="at">outside =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb442-5"><a href="complete-protocols.html#cb442-5" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>()</span>
<span id="cb442-6"><a href="complete-protocols.html#cb442-6" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-691-1.png" width="672" /></p>
<p>The data is presented in a nice way now, and we can improve the axis labels and add a title and caption:</p>
<div class="sourceCode" id="cb443"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb443-1"><a href="complete-protocols.html#cb443-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb443-2"><a href="complete-protocols.html#cb443-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb443-3"><a href="complete-protocols.html#cb443-3" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Distribution of size measurements&quot;</span>,</span>
<span id="cb443-4"><a href="complete-protocols.html#cb443-4" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;Sizes of human cells and nuclei of cells&quot;</span>,</span>
<span id="cb443-5"><a href="complete-protocols.html#cb443-5" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Size [µm] - log scale&quot;</span>,</span>
<span id="cb443-6"><a href="complete-protocols.html#cb443-6" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Count&quot;</span>,</span>
<span id="cb443-7"><a href="complete-protocols.html#cb443-7" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart | data submitted by 4 groups of students&quot;</span>,</span>
<span id="cb443-8"><a href="complete-protocols.html#cb443-8" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 10&quot;</span></span>
<span id="cb443-9"><a href="complete-protocols.html#cb443-9" tabindex="-1"></a>  )</span></code></pre></div>
<p>Optimizing the layout:</p>
<div class="sourceCode" id="cb444"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb444-1"><a href="complete-protocols.html#cb444-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb444-2"><a href="complete-protocols.html#cb444-2" tabindex="-1"></a>  <span class="co">#Set text size</span></span>
<span id="cb444-3"><a href="complete-protocols.html#cb444-3" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">16</span>) <span class="sc">+</span></span>
<span id="cb444-4"><a href="complete-protocols.html#cb444-4" tabindex="-1"></a>  <span class="co"># Change the color and position of the caption</span></span>
<span id="cb444-5"><a href="complete-protocols.html#cb444-5" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb444-6"><a href="complete-protocols.html#cb444-6" tabindex="-1"></a>    <span class="at">plot.caption =</span> <span class="fu">element_text</span>(</span>
<span id="cb444-7"><a href="complete-protocols.html#cb444-7" tabindex="-1"></a>      <span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>,</span>
<span id="cb444-8"><a href="complete-protocols.html#cb444-8" tabindex="-1"></a>      <span class="at">hjust =</span> <span class="dv">1</span></span>
<span id="cb444-9"><a href="complete-protocols.html#cb444-9" tabindex="-1"></a>    )</span>
<span id="cb444-10"><a href="complete-protocols.html#cb444-10" tabindex="-1"></a>  )</span>
<span id="cb444-11"><a href="complete-protocols.html#cb444-11" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-693-1.png" width="672" /></p>
<p>This looks very decent. But we can edit change the x-axis to display logarithmic space ticks. In addition, I’d like the bars of the histogram to start from the axis and not ‘float’ as in the graph above:</p>
<div class="sourceCode" id="cb445"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb445-1"><a href="complete-protocols.html#cb445-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb445-2"><a href="complete-protocols.html#cb445-2" tabindex="-1"></a>  <span class="co">#Force the y-axis to start at zero</span></span>
<span id="cb445-3"><a href="complete-protocols.html#cb445-3" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="cn">NA</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">200</span>)) <span class="sc">+</span></span>
<span id="cb445-4"><a href="complete-protocols.html#cb445-4" tabindex="-1"></a>  <span class="co">#Apply a logarithmic scale to the x-axis and set the numbers for the scale</span></span>
<span id="cb445-5"><a href="complete-protocols.html#cb445-5" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">10</span>,<span class="dv">100</span>), <span class="at">limits =</span> <span class="fu">c</span>(.<span class="dv">5</span>,<span class="dv">200</span>)) <span class="sc">+</span></span>
<span id="cb445-6"><a href="complete-protocols.html#cb445-6" tabindex="-1"></a>  <span class="co">#Remove minor gridlines</span></span>
<span id="cb445-7"><a href="complete-protocols.html#cb445-7" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb445-8"><a href="complete-protocols.html#cb445-8" tabindex="-1"></a>  <span class="co">#Add ticks to the bottom, outside</span></span>
<span id="cb445-9"><a href="complete-protocols.html#cb445-9" tabindex="-1"></a>  <span class="fu">annotation_logticks</span>(<span class="at">sides=</span><span class="st">&quot;b&quot;</span>, <span class="at">outside =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb445-10"><a href="complete-protocols.html#cb445-10" tabindex="-1"></a>  <span class="co">#Give a little more space to the log-ticks by adding margin to the top of the x-axis text</span></span>
<span id="cb445-11"><a href="complete-protocols.html#cb445-11" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">margin =</span> <span class="fu">margin</span>(<span class="at">t=</span><span class="dv">8</span>))) <span class="sc">+</span></span>
<span id="cb445-12"><a href="complete-protocols.html#cb445-12" tabindex="-1"></a>  <span class="co">#Needed to see the tcks outside the plot panel</span></span>
<span id="cb445-13"><a href="complete-protocols.html#cb445-13" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">clip =</span> <span class="st">&quot;off&quot;</span>)</span></code></pre></div>
<pre><code>Scale for x is already present.
Adding another scale for x, which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb447"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb447-1"><a href="complete-protocols.html#cb447-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-694-1.png" width="672" /></p>
<p>To save the plot as a PNG file:</p>
<div class="sourceCode" id="cb448"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb448-1"><a href="complete-protocols.html#cb448-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_10.png&quot;</span>), <span class="at">width =</span> <span class="dv">3000</span>, <span class="at">height =</span> <span class="dv">2000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb448-2"><a href="complete-protocols.html#cb448-2" tabindex="-1"></a>  p</span>
<span id="cb448-3"><a href="complete-protocols.html#cb448-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-11" class="section level2 hasAnchor" number="4.11">
<h2><span class="header-section-number">4.11</span> Protocol 11 - Plotting time data from a Google form<a href="complete-protocols.html#protocol-11" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The goal of this script is to read data that was generated by a Google form and visualize the distribution of the hours at which the form was submitted. We need the tidyverse package:</p>
<div class="sourceCode" id="cb450"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb450-1"><a href="complete-protocols.html#cb450-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>The data from the Google form are stored in a sheet that was published on the web as a CSV. This was achieved by ‘File &gt; Share &gt; Publish to web’ and selecting to ‘Publish’ as ‘Comma-separated values’.
After clicking on ‘Publish’ a link is generated that can be read as a csv. Any empty cells are converted to <code>NA</code>. As explained in protocol 10, we will load a snapshot of the sheet (the code to read the actual sheet is outcommented here):</p>
<div class="sourceCode" id="cb451"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb451-1"><a href="complete-protocols.html#cb451-1" tabindex="-1"></a><span class="co"># df_sheet &lt;- read.csv(&quot;https://docs.google.com/spreadsheets/d/e/2PACX-1vSc-nI1-s_u-XkNXEn_u2l6wkBafxJMHQ_Cd3kStrnToh7kawqjQU3y2l_1riLigKRkIqlNOqPrgkdW/pub?output=csv&quot;, na.strings = &quot;&quot;)</span></span>
<span id="cb451-2"><a href="complete-protocols.html#cb451-2" tabindex="-1"></a></span>
<span id="cb451-3"><a href="complete-protocols.html#cb451-3" tabindex="-1"></a>df_sheet <span class="ot">&lt;-</span>  <span class="fu">read.csv</span>(<span class="st">&quot;data/20220421_Resultaat-metingen.csv&quot;</span>, <span class="at">na.strings =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb451-4"><a href="complete-protocols.html#cb451-4" tabindex="-1"></a><span class="fu">head</span>(df_sheet)</span></code></pre></div>
<pre><code>          Tijdstempel Groep
1                &lt;NA&gt;  &lt;NA&gt;
2                &lt;NA&gt;  &lt;NA&gt;
3                &lt;NA&gt;  &lt;NA&gt;
4                &lt;NA&gt;  &lt;NA&gt;
5                &lt;NA&gt;  &lt;NA&gt;
6 01/03/2021 15:42:23     B
  Resultaten.van.de.10.metingen..in.µm...van.de.celkleuring..methyleenblauw.
1                                                                       &lt;NA&gt;
2                                                                       &lt;NA&gt;
3                                                                       &lt;NA&gt;
4                                                                       &lt;NA&gt;
5                                                                       &lt;NA&gt;
6                 64.6, 52.5, 79.4, 48.4, 66.0, 64.4, 69.8, 71.2, 50.7, 51.0
  Resultaten.van.de.10.metingen..in.µm...van.de.kernkleuring..DAPI.
1                                                              &lt;NA&gt;
2                                                              &lt;NA&gt;
3                                                              &lt;NA&gt;
4                                                              &lt;NA&gt;
5                                                              &lt;NA&gt;
6          11.8, 7.8, 7.0, 12.4, 12.7, 14.0, 11.9, 12.9, 11.7, 13.2</code></pre>
<p>Here, we are only interested in the first column with the time data. So we select that column and replace the dutch column name by ‘Timestamp’:</p>
<div class="sourceCode" id="cb453"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb453-1"><a href="complete-protocols.html#cb453-1" tabindex="-1"></a>df_sheet <span class="ot">&lt;-</span> df_sheet <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="st">&quot;Timestamp&quot;</span><span class="ot">=</span><span class="dv">1</span>) <span class="sc">%&gt;%</span> <span class="fu">na.omit</span>()</span></code></pre></div>
<p>We can seperate the Timestamp column in date and time:</p>
<div class="sourceCode" id="cb454"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb454-1"><a href="complete-protocols.html#cb454-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> df_sheet <span class="sc">%&gt;%</span> <span class="fu">separate</span>(<span class="st">&#39;Timestamp&#39;</span>, <span class="fu">c</span>(<span class="st">&quot;Date&quot;</span>, <span class="st">&quot;Time&quot;</span>), <span class="at">sep=</span><span class="st">&quot; &quot;</span>)</span></code></pre></div>
<p>Correct handling of date and time data is quite challenging and a topic on its own. The code to format the ‘Date’ column as date (note that we need a capital <code>Y</code> as the year is presented in 4 digits) would be:</p>
<div class="sourceCode" id="cb455"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb455-1"><a href="complete-protocols.html#cb455-1" tabindex="-1"></a>df_tidy <span class="sc">%&gt;%</span></span>
<span id="cb455-2"><a href="complete-protocols.html#cb455-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Date=</span><span class="fu">as.Date</span>(Date, <span class="st">&quot;%d/%m/%Y&quot;</span>)) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code></pre></div>
<pre><code>         Date     Time
6  2021-03-01 15:42:23
7  2021-03-01 23:50:53
8  2021-03-02 17:09:52
9  2021-03-03 09:30:43
10 2021-03-03 16:21:32
11 2021-03-03 21:00:23</code></pre>
<p>Here, we avoid the <code>Date</code> format and use the individual numbers. The ‘Date’ column is seperated into three columns for year, month and day. The ‘Time’ column can be split into hour, minute and second:</p>
<div class="sourceCode" id="cb457"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb457-1"><a href="complete-protocols.html#cb457-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span>  df_tidy <span class="sc">%&gt;%</span></span>
<span id="cb457-2"><a href="complete-protocols.html#cb457-2" tabindex="-1"></a>  <span class="fu">separate</span>(<span class="st">&#39;Date&#39;</span>, <span class="fu">c</span>(<span class="st">&quot;day&quot;</span>, <span class="st">&quot;month&quot;</span>, <span class="st">&quot;year&quot;</span>), <span class="at">sep=</span><span class="st">&quot;/&quot;</span>, <span class="at">convert =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb457-3"><a href="complete-protocols.html#cb457-3" tabindex="-1"></a>  <span class="fu">separate</span>(<span class="st">&#39;Time&#39;</span>, <span class="fu">c</span>(<span class="st">&quot;hour&quot;</span>, <span class="st">&quot;min&quot;</span>, <span class="st">&quot;sec&quot;</span>), <span class="at">sep=</span><span class="st">&quot;:&quot;</span>, <span class="at">convert =</span> <span class="cn">TRUE</span>)</span>
<span id="cb457-4"><a href="complete-protocols.html#cb457-4" tabindex="-1"></a><span class="fu">head</span>(df_tidy)</span></code></pre></div>
<pre><code>   day month year hour min sec
6    1     3 2021   15  42  23
7    1     3 2021   23  50  53
8    2     3 2021   17   9  52
9    3     3 2021    9  30  43
10   3     3 2021   16  21  32
11   3     3 2021   21   0  23</code></pre>
<p>Note that we use <code>convert = TRUE</code> to ensure that the data is treated as numbers (not as characters). We can use the data to check at what time the data was submitted:</p>
<div class="sourceCode" id="cb459"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb459-1"><a href="complete-protocols.html#cb459-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_tidy, <span class="fu">aes</span>(<span class="at">x=</span>hour))<span class="sc">+</span><span class="fu">geom_histogram</span>(<span class="at">alpha=</span>.<span class="dv">8</span>)</span></code></pre></div>
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-719-1.png" width="672" /></p>
<p>The default number of ‘bins’ is 30, but it makes more sense to use 24 here as there are 24 hours in a day:</p>
<div class="sourceCode" id="cb461"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb461-1"><a href="complete-protocols.html#cb461-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_tidy, <span class="fu">aes</span>(<span class="at">x=</span>hour))<span class="sc">+</span><span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">24</span>, <span class="at">alpha=</span>.<span class="dv">8</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-720-1.png" width="672" />
Since ‘hour’ is a number that cycles, with 0 followed by 23, it is nice to connect the data from 23 with 0 and we can do that with <code>coord_polar()</code>:</p>
<div class="sourceCode" id="cb462"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb462-1"><a href="complete-protocols.html#cb462-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_tidy, <span class="fu">aes</span>(<span class="at">x=</span>hour)) <span class="sc">+</span> </span>
<span id="cb462-2"><a href="complete-protocols.html#cb462-2" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">24</span>, <span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb462-3"><a href="complete-protocols.html#cb462-3" tabindex="-1"></a>  <span class="fu">coord_polar</span>() <span class="sc">+</span> </span>
<span id="cb462-4"><a href="complete-protocols.html#cb462-4" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">24</span>, <span class="at">by=</span><span class="dv">2</span>))</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-721-1.png" width="672" /></p>
<p>To center the plot at 0, we need to rotate it by -7.5 degrees (1 hour equal 15 degrees). We can use the <code>start</code> option within <code>coord_polar()</code>, but it only accepts ‘radians’. To convert degrees to radians the value is multiplied by pi/180:</p>
<div class="sourceCode" id="cb463"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb463-1"><a href="complete-protocols.html#cb463-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_tidy, <span class="fu">aes</span>(<span class="at">x=</span>hour)) <span class="sc">+</span> </span>
<span id="cb463-2"><a href="complete-protocols.html#cb463-2" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">24</span>, <span class="at">alpha=</span>.<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb463-3"><a href="complete-protocols.html#cb463-3" tabindex="-1"></a>  <span class="fu">coord_polar</span>(<span class="at">start =</span> (<span class="sc">-</span><span class="fl">7.5</span><span class="sc">*</span>pi<span class="sc">/</span><span class="dv">180</span>)) <span class="sc">+</span> </span>
<span id="cb463-4"><a href="complete-protocols.html#cb463-4" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">24</span>, <span class="at">by=</span><span class="dv">3</span>))</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-722-1.png" width="672" />
The plot can be rotated to center it around 15h as it reflects the middle of the afternoon, when most activity takes place. To achieve this, we need to shift the start another 15*15 degrees:</p>
<div class="sourceCode" id="cb464"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb464-1"><a href="complete-protocols.html#cb464-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_tidy, <span class="fu">aes</span>(<span class="at">x=</span>hour)) <span class="sc">+</span> </span>
<span id="cb464-2"><a href="complete-protocols.html#cb464-2" tabindex="-1"></a>        <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">24</span>, <span class="at">alpha=</span>.<span class="dv">8</span>, <span class="at">color=</span><span class="st">&#39;black&#39;</span>) <span class="sc">+</span></span>
<span id="cb464-3"><a href="complete-protocols.html#cb464-3" tabindex="-1"></a>        <span class="fu">coord_polar</span>(<span class="at">start =</span> (<span class="sc">-</span><span class="fl">7.5</span> <span class="sc">-</span> <span class="dv">15</span><span class="sc">*</span><span class="dv">15</span>)<span class="sc">*</span>pi<span class="sc">/</span><span class="dv">180</span>) <span class="sc">+</span> </span>
<span id="cb464-4"><a href="complete-protocols.html#cb464-4" tabindex="-1"></a>        <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">24</span>, <span class="at">by=</span><span class="dv">3</span>))</span>
<span id="cb464-5"><a href="complete-protocols.html#cb464-5" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-723-1.png" width="672" />
This looks good, let’s first improve on the layout by changing the theme. I do not like the grey background, so any theme with a white background is an improvement, for instance <code>theme_minimal()</code>:</p>
<div class="sourceCode" id="cb465"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb465-1"><a href="complete-protocols.html#cb465-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">theme_minimal</span>()</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-724-1.png" width="672" /></p>
<p>I do not like the labels that show the count. Since it is clear that the length of the bar indicates an amount and since the absolute amount is not so interesting, these labels can be removed. The labels for the hours are helpful and their sice needs to be increased. Finally, I do not like the grid running through the numbers. I did not find a way in the theme setting to get rid of this, so we define the grid ourselves by using repetitive horizontal and vertical lines:</p>
<div class="sourceCode" id="cb466"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb466-1"><a href="complete-protocols.html#cb466-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb466-2"><a href="complete-protocols.html#cb466-2" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>), <span class="at">colour =</span> <span class="st">&quot;grey90&quot;</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb466-3"><a href="complete-protocols.html#cb466-3" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">21</span>, <span class="at">by=</span><span class="dv">3</span>), <span class="at">colour =</span> <span class="st">&quot;grey90&quot;</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb466-4"><a href="complete-protocols.html#cb466-4" tabindex="-1"></a>  </span>
<span id="cb466-5"><a href="complete-protocols.html#cb466-5" tabindex="-1"></a>  <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">16</span>) <span class="sc">+</span></span>
<span id="cb466-6"><a href="complete-protocols.html#cb466-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb466-7"><a href="complete-protocols.html#cb466-7" tabindex="-1"></a>        <span class="at">panel.grid.major.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb466-8"><a href="complete-protocols.html#cb466-8" tabindex="-1"></a>        <span class="at">panel.grid.minor.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb466-9"><a href="complete-protocols.html#cb466-9" tabindex="-1"></a>        <span class="at">panel.grid.minor.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb466-10"><a href="complete-protocols.html#cb466-10" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_blank</span>()</span>
<span id="cb466-11"><a href="complete-protocols.html#cb466-11" tabindex="-1"></a>  )</span>
<span id="cb466-12"><a href="complete-protocols.html#cb466-12" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-725-1.png" width="672" /></p>
<p>The grid looks good, but the bars have to be on top of this layer, so we redefine the plot in the correct order:</p>
<div class="sourceCode" id="cb467"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb467-1"><a href="complete-protocols.html#cb467-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_tidy, <span class="fu">aes</span>(<span class="at">x=</span>hour)) <span class="sc">+</span></span>
<span id="cb467-2"><a href="complete-protocols.html#cb467-2" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>), <span class="at">colour =</span> <span class="st">&quot;grey90&quot;</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb467-3"><a href="complete-protocols.html#cb467-3" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">21</span>, <span class="at">by=</span><span class="dv">3</span>), <span class="at">colour =</span> <span class="st">&quot;grey90&quot;</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb467-4"><a href="complete-protocols.html#cb467-4" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">24</span>, <span class="at">alpha=</span>.<span class="dv">8</span>, <span class="at">color=</span><span class="st">&#39;black&#39;</span>) <span class="sc">+</span></span>
<span id="cb467-5"><a href="complete-protocols.html#cb467-5" tabindex="-1"></a>  <span class="fu">coord_polar</span>(<span class="at">start =</span> (<span class="sc">-</span><span class="fl">7.5</span> <span class="sc">-</span> <span class="dv">15</span><span class="sc">*</span><span class="dv">15</span>)<span class="sc">*</span>pi<span class="sc">/</span><span class="dv">180</span>) <span class="sc">+</span> </span>
<span id="cb467-6"><a href="complete-protocols.html#cb467-6" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">24</span>, <span class="at">by=</span><span class="dv">3</span>)) <span class="sc">+</span></span>
<span id="cb467-7"><a href="complete-protocols.html#cb467-7" tabindex="-1"></a>  <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">16</span>) <span class="sc">+</span></span>
<span id="cb467-8"><a href="complete-protocols.html#cb467-8" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb467-9"><a href="complete-protocols.html#cb467-9" tabindex="-1"></a>        <span class="at">panel.grid.major.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb467-10"><a href="complete-protocols.html#cb467-10" tabindex="-1"></a>        <span class="at">panel.grid.minor.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb467-11"><a href="complete-protocols.html#cb467-11" tabindex="-1"></a>        <span class="at">panel.grid.minor.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb467-12"><a href="complete-protocols.html#cb467-12" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_blank</span>()</span>
<span id="cb467-13"><a href="complete-protocols.html#cb467-13" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb467-14"><a href="complete-protocols.html#cb467-14" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb467-15"><a href="complete-protocols.html#cb467-15" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Counting the hours...&quot;</span>,</span>
<span id="cb467-16"><a href="complete-protocols.html#cb467-16" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;that a Google form was submitted&quot;</span>,</span>
<span id="cb467-17"><a href="complete-protocols.html#cb467-17" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb467-18"><a href="complete-protocols.html#cb467-18" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb467-19"><a href="complete-protocols.html#cb467-19" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart | data submitted by students&quot;</span>,</span>
<span id="cb467-20"><a href="complete-protocols.html#cb467-20" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 11&quot;</span></span>
<span id="cb467-21"><a href="complete-protocols.html#cb467-21" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb467-22"><a href="complete-protocols.html#cb467-22" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>, <span class="at">hjust =</span> <span class="dv">1</span>))</span>
<span id="cb467-23"><a href="complete-protocols.html#cb467-23" tabindex="-1"></a>  </span>
<span id="cb467-24"><a href="complete-protocols.html#cb467-24" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-726-1.png" width="672" /></p>
<p>A centered title and subtitle looks nice on this plot:</p>
<div class="sourceCode" id="cb468"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb468-1"><a href="complete-protocols.html#cb468-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(</span>
<span id="cb468-2"><a href="complete-protocols.html#cb468-2" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb468-3"><a href="complete-protocols.html#cb468-3" tabindex="-1"></a>        <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>)</span>
<span id="cb468-4"><a href="complete-protocols.html#cb468-4" tabindex="-1"></a>)</span>
<span id="cb468-5"><a href="complete-protocols.html#cb468-5" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-727-1.png" width="672" /></p>
<p>To save the plot as a PNG file:</p>
<div class="sourceCode" id="cb469"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb469-1"><a href="complete-protocols.html#cb469-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_11.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb469-2"><a href="complete-protocols.html#cb469-2" tabindex="-1"></a>  p</span>
<span id="cb469-3"><a href="complete-protocols.html#cb469-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-12" class="section level2 hasAnchor" number="4.12">
<h2><span class="header-section-number">4.12</span> Protocol 12 - Plotting grouped data<a href="complete-protocols.html#protocol-12" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This protocol explains how data can be plotted side-by-side. It is a remake of a figure that was published by <a href="https://doi.org/10.7554/eLife.66440">de Man et al. (2021)</a>. The source data is published with the paper and can be downloaded. Next to the <code>tidyverse</code> package we need <code>readxl</code> to import the excel file:</p>
<div class="sourceCode" id="cb471"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb471-1"><a href="complete-protocols.html#cb471-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb471-2"><a href="complete-protocols.html#cb471-2" tabindex="-1"></a><span class="fu">library</span>(readr)</span></code></pre></div>
<p>Next, we download the xls file from the url, store it as a (temporary) file, and use <code>read_excel()</code> to import the data (note that unlike the <code>read.csv()</code> we cannot directly download the data from a url):</p>
<div class="sourceCode" id="cb472"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb472-1"><a href="complete-protocols.html#cb472-1" tabindex="-1"></a>url_xls <span class="ot">&lt;-</span> <span class="st">&quot;https://cdn.elifesciences.org/articles/66440/elife-66440-fig5-data1-v2.xlsx&quot;</span></span>
<span id="cb472-2"><a href="complete-protocols.html#cb472-2" tabindex="-1"></a><span class="co"># file_xls &lt;- basename(url_xls)</span></span>
<span id="cb472-3"><a href="complete-protocols.html#cb472-3" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> url_xls, <span class="at">destfile =</span> <span class="st">&#39;temp_file&#39;</span>)</span>
<span id="cb472-4"><a href="complete-protocols.html#cb472-4" tabindex="-1"></a>df_download <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">&#39;temp_file&#39;</span>)</span></code></pre></div>
<pre><code>New names:
• `` -&gt; `...2`
• `` -&gt; `...3`
• `` -&gt; `...4`
• `` -&gt; `...5`
• `` -&gt; `...6`
• `` -&gt; `...7`
• `` -&gt; `...8`
• `` -&gt; `...9`
• `` -&gt; `...10`
• `` -&gt; `...11`</code></pre>
<p>Inspection of the dataframe shows that we can remove the first 11 rows and that the name of the columns is in the 12th row. First we fix the column names:</p>
<div class="sourceCode" id="cb474"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb474-1"><a href="complete-protocols.html#cb474-1" tabindex="-1"></a><span class="fu">colnames</span>(df_download) <span class="ot">&lt;-</span> df_download <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="dv">12</span>)</span></code></pre></div>
<p>Then, we can remove the first 12 frames and select only the first 3 columns (I use <code>dplyr::</code> here to make explicit that the <code>select()</code> function is taken from the <code>{dplyr}</code> package. This may not be necessary when you run your own code):</p>
<div class="sourceCode" id="cb475"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb475-1"><a href="complete-protocols.html#cb475-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> df_download <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>)) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>)</span>
<span id="cb475-2"><a href="complete-protocols.html#cb475-2" tabindex="-1"></a><span class="fu">head</span>(df_tidy)</span></code></pre></div>
<pre><code># A tibble: 6 × 3
  Sample               `Measurement date` `Total SGFP2-CTNNB1 concentration`
  &lt;chr&gt;                &lt;chr&gt;              &lt;chr&gt;                             
1 B10 WNT3A c1 cyto    180926             250.188126820392                  
2 B10 WNT3A c2 cyto    180926             243.40459000116601                
3 B10 WNT3A c3 cyto    180926             233.140028430596                  
4 B10 WNT3A c3 nucleus 180926             377.33830720940603                
5 B10 WNT3A c4 cyto    180926             256.42702997341098                
6 B10 WNT3A c4 nucleus 180926             379.44318769572499                </code></pre>
<p>The third column, containing the measurement data has the class <code>&lt;chr&gt;</code>, but the correct class is numeric. Here, we mutate the column and rename it. To keep the data that is untouched, we add <code>.keep = "unused"</code>:</p>
<div class="sourceCode" id="cb477"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb477-1"><a href="complete-protocols.html#cb477-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> df_tidy <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="st">`</span><span class="at">SGFP2-CTNNB1</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">as.numeric</span>(<span class="st">`</span><span class="at">Total SGFP2-CTNNB1 concentration</span><span class="st">`</span>), <span class="at">.keep =</span> <span class="st">&quot;unused&quot;</span>)</span></code></pre></div>
<p>Finally, we need to separate the first column, since it contains the relevant data on the conditions and the grouping of the data:</p>
<div class="sourceCode" id="cb478"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb478-1"><a href="complete-protocols.html#cb478-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> df_tidy <span class="sc">%&gt;%</span> <span class="fu">separate</span>(<span class="st">`</span><span class="at">Sample</span><span class="st">`</span>, <span class="fu">c</span>(<span class="st">&quot;Celltype&quot;</span>, <span class="st">&quot;ligand&quot;</span>, <span class="st">&quot;number&quot;</span>, <span class="st">&quot;compartment&quot;</span>), <span class="at">sep=</span><span class="st">&quot; &quot;</span>)</span>
<span id="cb478-2"><a href="complete-protocols.html#cb478-2" tabindex="-1"></a><span class="fu">head</span>(df_tidy)</span></code></pre></div>
<pre><code># A tibble: 6 × 6
  Celltype ligand number compartment `Measurement date` `SGFP2-CTNNB1`
  &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;       &lt;chr&gt;                       &lt;dbl&gt;
1 B10      WNT3A  c1     cyto        180926                       250.
2 B10      WNT3A  c2     cyto        180926                       243.
3 B10      WNT3A  c3     cyto        180926                       233.
4 B10      WNT3A  c3     nucleus     180926                       377.
5 B10      WNT3A  c4     cyto        180926                       256.
6 B10      WNT3A  c4     nucleus     180926                       379.</code></pre>
<p>There are two compartments (nucleus and cytoplasm) and two ligands (BSA as a control and WNT3A). To avoid abbreviated labels in the plot, we will rename the ‘cyto’ condition:</p>
<div class="sourceCode" id="cb480"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb480-1"><a href="complete-protocols.html#cb480-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> df_tidy <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">compartment =</span></span>
<span id="cb480-2"><a href="complete-protocols.html#cb480-2" tabindex="-1"></a>                                <span class="fu">case_when</span>(compartment <span class="sc">==</span> <span class="st">&quot;cyto&quot;</span> <span class="sc">~</span> <span class="st">&quot;cytoplasm&quot;</span>,</span>
<span id="cb480-3"><a href="complete-protocols.html#cb480-3" tabindex="-1"></a>                                          <span class="cn">TRUE</span> <span class="sc">~</span> compartment)</span>
<span id="cb480-4"><a href="complete-protocols.html#cb480-4" tabindex="-1"></a>                              )</span>
<span id="cb480-5"><a href="complete-protocols.html#cb480-5" tabindex="-1"></a><span class="fu">head</span>(df_tidy)</span></code></pre></div>
<pre><code># A tibble: 6 × 6
  Celltype ligand number compartment `Measurement date` `SGFP2-CTNNB1`
  &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;       &lt;chr&gt;                       &lt;dbl&gt;
1 B10      WNT3A  c1     cytoplasm   180926                       250.
2 B10      WNT3A  c2     cytoplasm   180926                       243.
3 B10      WNT3A  c3     cytoplasm   180926                       233.
4 B10      WNT3A  c3     nucleus     180926                       377.
5 B10      WNT3A  c4     cytoplasm   180926                       256.
6 B10      WNT3A  c4     nucleus     180926                       379.</code></pre>
<p>This data can be plotted in a number of different ways. The first way is by comparing the different compartments for each ligand:</p>
<div class="sourceCode" id="cb482"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb482-1"><a href="complete-protocols.html#cb482-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_tidy, <span class="fu">aes</span>(<span class="at">x=</span>ligand, <span class="at">y=</span><span class="st">`</span><span class="at">SGFP2-CTNNB1</span><span class="st">`</span>, <span class="at">fill =</span> compartment)) <span class="sc">+</span></span>
<span id="cb482-2"><a href="complete-protocols.html#cb482-2" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.color =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb482-3"><a href="complete-protocols.html#cb482-3" tabindex="-1"></a>  <span class="fu">geom_jitter</span>() </span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-754-1.png" width="672" /></p>
<p>However, in the original figure, the data is split between the two compartments and the two ligands are plotted next to eachother:</p>
<div class="sourceCode" id="cb483"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb483-1"><a href="complete-protocols.html#cb483-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_tidy, <span class="fu">aes</span>(<span class="at">x=</span>ligand, <span class="at">y=</span><span class="st">`</span><span class="at">SGFP2-CTNNB1</span><span class="st">`</span>, <span class="at">fill =</span> compartment)) <span class="sc">+</span></span>
<span id="cb483-2"><a href="complete-protocols.html#cb483-2" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.color =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb483-3"><a href="complete-protocols.html#cb483-3" tabindex="-1"></a>  <span class="fu">geom_jitter</span>() <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>compartment)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-755-1.png" width="672" /></p>
<p>This is close to the original figure, but the colors in that figure reflect the ligand, not the compartment:</p>
<div class="sourceCode" id="cb484"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb484-1"><a href="complete-protocols.html#cb484-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_tidy, <span class="fu">aes</span>(<span class="at">x=</span>ligand, <span class="at">y=</span><span class="st">`</span><span class="at">SGFP2-CTNNB1</span><span class="st">`</span>, <span class="at">fill =</span> ligand)) <span class="sc">+</span></span>
<span id="cb484-2"><a href="complete-protocols.html#cb484-2" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.color =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb484-3"><a href="complete-protocols.html#cb484-3" tabindex="-1"></a>  <span class="fu">geom_jitter</span>() <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>compartment)</span>
<span id="cb484-4"><a href="complete-protocols.html#cb484-4" tabindex="-1"></a></span>
<span id="cb484-5"><a href="complete-protocols.html#cb484-5" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-756-1.png" width="672" /></p>
<p>This plot looks good. Now let’s improve the theme.</p>
<div class="sourceCode" id="cb485"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb485-1"><a href="complete-protocols.html#cb485-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_linedraw</span>(<span class="at">base_size =</span> <span class="dv">16</span>) <span class="sc">+</span> </span>
<span id="cb485-2"><a href="complete-protocols.html#cb485-2" tabindex="-1"></a>          <span class="fu">theme</span>(<span class="at">panel.spacing =</span> <span class="fu">unit</span>(<span class="dv">0</span>, <span class="st">&quot;points&quot;</span>))</span>
<span id="cb485-3"><a href="complete-protocols.html#cb485-3" tabindex="-1"></a></span>
<span id="cb485-4"><a href="complete-protocols.html#cb485-4" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-757-1.png" width="672" /></p>
<p>The theme is pretty nice, but the black strips are very distracting. Let’s change the fill and color (border) of the strips and change the text to black. At the same time, we remove the grid and the legend:</p>
<div class="sourceCode" id="cb486"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb486-1"><a href="complete-protocols.html#cb486-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="cn">NA</span>, <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size =</span> .<span class="dv">5</span>),</span>
<span id="cb486-2"><a href="complete-protocols.html#cb486-2" tabindex="-1"></a>               <span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&#39;black&#39;</span>),</span>
<span id="cb486-3"><a href="complete-protocols.html#cb486-3" tabindex="-1"></a>               <span class="at">panel.grid =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb486-4"><a href="complete-protocols.html#cb486-4" tabindex="-1"></a>               <span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb486-5"><a href="complete-protocols.html#cb486-5" tabindex="-1"></a>               <span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color=</span><span class="st">&#39;grey80&#39;</span>)</span>
<span id="cb486-6"><a href="complete-protocols.html#cb486-6" tabindex="-1"></a>               )</span>
<span id="cb486-7"><a href="complete-protocols.html#cb486-7" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-758-1.png" width="672" /></p>
<p>Finally, we fix the labels:</p>
<div class="sourceCode" id="cb487"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb487-1"><a href="complete-protocols.html#cb487-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&quot;&quot;</span>,</span>
<span id="cb487-2"><a href="complete-protocols.html#cb487-2" tabindex="-1"></a>              <span class="at">y=</span><span class="st">&quot;Concentration [nM]&quot;</span>,</span>
<span id="cb487-3"><a href="complete-protocols.html#cb487-3" tabindex="-1"></a>              <span class="at">title =</span> <span class="st">&quot;Total SGFP2-CTNNB1&quot;</span>,</span>
<span id="cb487-4"><a href="complete-protocols.html#cb487-4" tabindex="-1"></a>              <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart | based on data from de Man et al., DOI: 10.7554/eLife.66440&quot;</span>,</span>
<span id="cb487-5"><a href="complete-protocols.html#cb487-5" tabindex="-1"></a>              <span class="at">tag =</span> <span class="st">&quot;Protocol 12&quot;</span></span>
<span id="cb487-6"><a href="complete-protocols.html#cb487-6" tabindex="-1"></a>              )</span>
<span id="cb487-7"><a href="complete-protocols.html#cb487-7" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-759-1.png" width="672" /></p>
<p>To save the plot as a PNG file:</p>
<div class="sourceCode" id="cb488"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb488-1"><a href="complete-protocols.html#cb488-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_12.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb488-2"><a href="complete-protocols.html#cb488-2" tabindex="-1"></a>  p</span>
<span id="cb488-3"><a href="complete-protocols.html#cb488-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-13" class="section level2 hasAnchor" number="4.13">
<h2><span class="header-section-number">4.13</span> Protocol 13 - Plotting multiple plots side-by-side<a href="complete-protocols.html#protocol-13" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This protocol is a little different, in that it plots the same data in three different ways. I use this figure to illustrate that barplots are a poor choice for plotting multiple measurements per condition as the actual data is concealed. The alternatives that show all, i.e. jittered dotplots and beeswarmplots (or sinaplots) are shown next to the barplot.</p>
<p>In addition to the <code>{tidyverse}</code> package, we need two additional packages. The <code>{ggbeeswarm}</code> package to plot the beeswarm plot and <code>{patchwork}</code> to stitch the plots together:</p>
<div class="sourceCode" id="cb490"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb490-1"><a href="complete-protocols.html#cb490-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb490-2"><a href="complete-protocols.html#cb490-2" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span></code></pre></div>
<pre><code>
Attaching package: &#39;patchwork&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:MASS&#39;:

    area</code></pre>
<div class="sourceCode" id="cb493"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb493-1"><a href="complete-protocols.html#cb493-1" tabindex="-1"></a><span class="fu">library</span>(ggbeeswarm)</span></code></pre></div>
<p>Next, the wide data is loaded and converted into a tidy format:</p>
<div class="sourceCode" id="cb494"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb494-1"><a href="complete-protocols.html#cb494-1" tabindex="-1"></a>df_wide <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;data/Area_in_um-GEFs.csv&quot;</span>)</span>
<span id="cb494-2"><a href="complete-protocols.html#cb494-2" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(df_wide, <span class="at">cols =</span> <span class="fu">everything</span>(), <span class="at">names_to =</span> <span class="st">&quot;Condition&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;Area&quot;</span>)</span>
<span id="cb494-3"><a href="complete-protocols.html#cb494-3" tabindex="-1"></a><span class="fu">head</span>(df_tidy)</span></code></pre></div>
<pre><code># A tibble: 6 × 2
  Condition  Area
  &lt;chr&gt;     &lt;dbl&gt;
1 LARG      1064.
2 wt        1571.
3 TIAM      3058.
4 LARG      1374.
5 wt        1946.
6 TIAM      1767.</code></pre>
<p>Next, we define the order of factors in the column with Conditions:</p>
<div class="sourceCode" id="cb496"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb496-1"><a href="complete-protocols.html#cb496-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> df_tidy <span class="sc">%&gt;%</span></span>
<span id="cb496-2"><a href="complete-protocols.html#cb496-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Condition =</span> <span class="fu">fct_relevel</span>(Condition, <span class="fu">c</span>(<span class="st">&quot;LARG&quot;</span>, <span class="st">&quot;wt&quot;</span>, <span class="st">&quot;TIAM&quot;</span>)))</span></code></pre></div>
<p>To plot the barplot, we calculate the summary statistics and store them in a separate dataframe:</p>
<div class="sourceCode" id="cb497"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb497-1"><a href="complete-protocols.html#cb497-1" tabindex="-1"></a>df_summary <span class="ot">&lt;-</span> df_tidy <span class="sc">%&gt;%</span> <span class="fu">na.omit</span>() <span class="sc">%&gt;%</span></span>
<span id="cb497-2"><a href="complete-protocols.html#cb497-2" tabindex="-1"></a>  <span class="fu">group_by</span>(Condition) <span class="sc">%&gt;%</span></span>
<span id="cb497-3"><a href="complete-protocols.html#cb497-3" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n=</span><span class="fu">n</span>(),</span>
<span id="cb497-4"><a href="complete-protocols.html#cb497-4" tabindex="-1"></a>            <span class="at">mean=</span><span class="fu">mean</span>(Area),</span>
<span id="cb497-5"><a href="complete-protocols.html#cb497-5" tabindex="-1"></a>            <span class="at">sd=</span><span class="fu">sd</span>(Area)</span>
<span id="cb497-6"><a href="complete-protocols.html#cb497-6" tabindex="-1"></a>            ) <span class="sc">%&gt;%</span></span>
<span id="cb497-7"><a href="complete-protocols.html#cb497-7" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sem=</span>sd<span class="sc">/</span><span class="fu">sqrt</span>(n<span class="dv">-1</span>))</span></code></pre></div>
<p>Now, we do things a bit differently than we’ve done in other protocols. First, we define a plotting canvas and its theme settings. We use this object <code>p</code> as the basis for the three plots that differ in the data visualization and not in the layout.</p>
<div class="sourceCode" id="cb498"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb498-1"><a href="complete-protocols.html#cb498-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_tidy, (<span class="fu">aes</span>(<span class="at">x=</span>Condition))) <span class="sc">+</span></span>
<span id="cb498-2"><a href="complete-protocols.html#cb498-2" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">12</span>) <span class="sc">+</span></span>
<span id="cb498-3"><a href="complete-protocols.html#cb498-3" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb498-4"><a href="complete-protocols.html#cb498-4" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">4200</span>)) <span class="sc">+</span></span>
<span id="cb498-5"><a href="complete-protocols.html#cb498-5" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="cn">NULL</span>)</span>
<span id="cb498-6"><a href="complete-protocols.html#cb498-6" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-781-1.png" width="672" />
The first plot is a barplot (the length of the bar reflects the mean value) with the standard error of the mean as errorbar:</p>
<div class="sourceCode" id="cb499"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb499-1"><a href="complete-protocols.html#cb499-1" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_errorbar</span>(<span class="at">data=</span>df_summary,</span>
<span id="cb499-2"><a href="complete-protocols.html#cb499-2" tabindex="-1"></a>                        <span class="fu">aes</span>(<span class="at">x=</span>Condition,<span class="at">ymin=</span>mean<span class="sc">-</span>sem, <span class="at">ymax=</span>(mean<span class="sc">+</span>sem)),</span>
<span id="cb499-3"><a href="complete-protocols.html#cb499-3" tabindex="-1"></a>                        <span class="at">width=</span><span class="fl">0.3</span>, <span class="at">size=</span><span class="dv">1</span>, <span class="at">alpha=</span><span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb499-4"><a href="complete-protocols.html#cb499-4" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">data=</span>df_summary, <span class="fu">aes</span>(<span class="at">y=</span>mean), <span class="at">stat =</span> <span class="st">&#39;identity&#39;</span>, <span class="at">width =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb499-5"><a href="complete-protocols.html#cb499-5" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="fu">expression</span>(<span class="st">&quot;Area [µm&quot;</span><span class="sc">^</span><span class="dv">2</span><span class="sc">*</span><span class="st">&quot;]&quot;</span>), <span class="at">title=</span><span class="st">&quot;Clean | Data hidden&quot;</span>)</span></code></pre></div>
<p>The second plot shows the data as jittered dots:</p>
<div class="sourceCode" id="cb500"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb500-1"><a href="complete-protocols.html#cb500-1" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_jitter</span>(<span class="at">data =</span> df_tidy,</span>
<span id="cb500-2"><a href="complete-protocols.html#cb500-2" tabindex="-1"></a>                      <span class="fu">aes</span>(<span class="at">x=</span>Condition, <span class="at">y=</span>Area),</span>
<span id="cb500-3"><a href="complete-protocols.html#cb500-3" tabindex="-1"></a>                      <span class="at">position=</span><span class="fu">position_jitter</span>(<span class="fl">0.3</span>), <span class="at">cex=</span><span class="dv">2</span>, <span class="at">alpha=</span><span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb500-4"><a href="complete-protocols.html#cb500-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Messy | Data shown&quot;</span>)</span></code></pre></div>
<p>The third plot is a beeswarm-type plot that organizes the dots according to the data distribution:</p>
<div class="sourceCode" id="cb501"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb501-1"><a href="complete-protocols.html#cb501-1" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_quasirandom</span>(<span class="at">data =</span> df_tidy,</span>
<span id="cb501-2"><a href="complete-protocols.html#cb501-2" tabindex="-1"></a>                           <span class="fu">aes</span>(<span class="at">x=</span>Condition, <span class="at">y=</span>Area),</span>
<span id="cb501-3"><a href="complete-protocols.html#cb501-3" tabindex="-1"></a>                           <span class="at">varwidth =</span> <span class="cn">TRUE</span>, <span class="at">cex=</span><span class="dv">2</span>, <span class="at">alpha=</span><span class="fl">0.4</span>)<span class="sc">+</span></span>
<span id="cb501-4"><a href="complete-protocols.html#cb501-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">&quot;Clean | Data shown&quot;</span>)</span></code></pre></div>
<p>The patchwork package has a simple syntax for combining the plots, for examples check the <a href="https://patchwork.data-imaginist.com/articles/patchwork.html">package website</a>. A vertical composition is made by using <code>p1/p2/p3</code>. But this won’t look good, so combine the plots horizontally using <code>p1+p2+p3</code>:</p>
<div class="sourceCode" id="cb502"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb502-1"><a href="complete-protocols.html#cb502-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p1<span class="sc">+</span>p2<span class="sc">+</span>p3</span>
<span id="cb502-2"><a href="complete-protocols.html#cb502-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-785-1.png" width="672" /></p>
<p>The result is a set of plots that are nicely aligned, side-by-side. We have titles for individual plots and it would be nice to have a title and caption for the entire panel. We can use the <code>plot_annotation()</code> function from the <code>{patchwork}</code> package. The ‘theme’ specified in this function is only applied to the main title and caption:</p>
<div class="sourceCode" id="cb503"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb503-1"><a href="complete-protocols.html#cb503-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">plot_annotation</span>(<span class="at">title=</span><span class="st">&quot;Protocol 13&quot;</span>,</span>
<span id="cb503-2"><a href="complete-protocols.html#cb503-2" tabindex="-1"></a>                           <span class="at">caption=</span><span class="st">&quot;@joachimgoedhart&quot;</span>,</span>
<span id="cb503-3"><a href="complete-protocols.html#cb503-3" tabindex="-1"></a>                           <span class="at">theme =</span> <span class="fu">theme</span>(<span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>, <span class="at">color=</span><span class="st">&quot;grey80&quot;</span>),</span>
<span id="cb503-4"><a href="complete-protocols.html#cb503-4" tabindex="-1"></a>                                         <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>))</span>
<span id="cb503-5"><a href="complete-protocols.html#cb503-5" tabindex="-1"></a>                          )</span>
<span id="cb503-6"><a href="complete-protocols.html#cb503-6" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-786-1.png" width="672" /></p>
<p>To save the composite plot:</p>
<div class="sourceCode" id="cb504"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb504-1"><a href="complete-protocols.html#cb504-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;protocol_13.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb504-2"><a href="complete-protocols.html#cb504-2" tabindex="-1"></a>p</span>
<span id="cb504-3"><a href="complete-protocols.html#cb504-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-14" class="section level2 hasAnchor" number="4.14">
<h2><span class="header-section-number">4.14</span> Protocol 14 - Volcano plot<a href="complete-protocols.html#protocol-14" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The results of comparative omics experiments are (very) large datasets. The datasets contain information on the fold change and level of significance for every gene or protein in a genome. The volcano plot is a scatterplot that visualizes both the (<span class="math inline">\(log_2\)</span>) fold change and level of significance (usually <span class="math inline">\(-log_{10}[p-value]\)</span>). This data visualization enables the display of a large amount of data in a concise way. Typically, only a handful of datapoints in the volcano plot are of interest, also known as hits. The hits are datapoints that surpass a threshold that is defined by the user for both the significance and fold change. Annotation of hits (with names) is used to draw attention to these most relevant or interesting datapoints.</p>
<p>This protocol showcases the construction of a volcano plot, including the filtering of significant datapoints and the annotation of top candidates. The code that is used here is similar to what is used in the web app <a href="https://doi.org/10.1038/s41598-020-76603-3">VolcaNoseR</a> which is dedicated to making volcano plots. In fact, the data visualization shown here is very close to the standard output of VolcaNoseR and uses the same example data.</p>
<p>We start by loading the <code>{tidyverse}</code> package for plotting:</p>
<div class="sourceCode" id="cb506"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb506-1"><a href="complete-protocols.html#cb506-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>The measured data is read from a CSV file. These data are originally from <a href="https://doi.org/10.1016/j.celrep.2018.12.094">Becares et al, DOI: 10.1016/j.celrep.2018.12.094</a>:</p>
<div class="sourceCode" id="cb507"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb507-1"><a href="complete-protocols.html#cb507-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;data/Becares-diffgenes_HFHC.csv&quot;</span>)</span></code></pre></div>
<pre><code>Rows: 3001 Columns: 9
── Column specification ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
chr (1): Gene
dbl (8): baseMean, log2_FoldChange, FC, lfcSE, stat, pvalue, padj, minus_log...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<div class="sourceCode" id="cb509"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb509-1"><a href="complete-protocols.html#cb509-1" tabindex="-1"></a><span class="fu">head</span>(df_tidy)</span></code></pre></div>
<pre><code># A tibble: 6 × 9
  Gene  baseMean log2_FoldChange    FC lfcSE  stat    pvalue     padj
  &lt;chr&gt;    &lt;dbl&gt;           &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 A1bg      39.2           1.25   2.38 0.299  4.19 0.0000278 0.00170 
2 A1cf    1183.            0.745  1.68 0.192  3.88 0.000102  0.00456 
3 Aadac   3196.            0.688  1.61 0.157  4.39 0.0000111 0.000817
4 Aadat   1481.            0.603  1.52 0.181  3.33 0.000878  0.0232  
5 Aaed1    432.            0.422  1.34 0.193  2.19 0.0288    0.235   
6 Aamdc    165.            0.332  1.26 0.191  1.74 0.0827    0.419   
# ℹ 1 more variable: minus_log10_pvalue &lt;dbl&gt;</code></pre>
<p>The basic volcano plot is a scatterplot that displays <code>minus_log10_pvalue</code> versus <code>log2_FoldChange</code>:</p>
<div class="sourceCode" id="cb511"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb511-1"><a href="complete-protocols.html#cb511-1" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> df_tidy, <span class="fu">aes</span>(<span class="at">x=</span>log2_FoldChange, <span class="at">y=</span>minus_log10_pvalue)) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.5</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-803-1.png" width="672" /></p>
<p>To improve the visualization, we will 1) set thresholds, 2) display the thresholds, 3) filter interesting candidates based on the thresholds and 4) annotate the top-10 candidates.</p>
<div id="step-1---define-thresholds" class="section level3 hasAnchor" number="4.14.1">
<h3><span class="header-section-number">4.14.1</span> Step 1 - Define thresholds<a href="complete-protocols.html#step-1---define-thresholds" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To identify data that is of interest, thresholds are defined for both the effect size and the significance. We define the thresholds, &lt;-1.5 and &gt;1.5 for the effect size and &gt;5 for the significance. When both thresholds are exceeded the data is considered as significant (worthy of a closer look).</p>
</div>
<div id="step-2---display-thresholds" class="section level3 hasAnchor" number="4.14.2">
<h3><span class="header-section-number">4.14.2</span> Step 2 - Display thresholds<a href="complete-protocols.html#step-2---display-thresholds" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To plot the thresholds, we use <code>geom_hline()</code> and <code>geom_vline()</code>. Note that we can specify two intercepts at once, which is used below for <code>geom_vline()</code>:</p>
<div class="sourceCode" id="cb512"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb512-1"><a href="complete-protocols.html#cb512-1" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> df_tidy, <span class="fu">aes</span>(<span class="at">x=</span>log2_FoldChange, <span class="at">y=</span>minus_log10_pvalue)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb512-2"><a href="complete-protocols.html#cb512-2" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">1.5</span>,<span class="fl">1.5</span>), <span class="at">linetype=</span><span class="st">&quot;dashed&quot;</span>) <span class="sc">+</span></span>
<span id="cb512-3"><a href="complete-protocols.html#cb512-3" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">5</span>, <span class="at">linetype=</span><span class="st">&quot;dashed&quot;</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-804-1.png" width="672" /></p>
</div>
<div id="step-3---filtering-with-thresholds" class="section level3 hasAnchor" number="4.14.3">
<h3><span class="header-section-number">4.14.3</span> Step 3 - Filtering with thresholds<a href="complete-protocols.html#step-3---filtering-with-thresholds" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this step, we add another column to the data, which defines, for each row, whether a gene is significantly lower, higher or unchanged. This is based on the thresholds:</p>
<div class="sourceCode" id="cb513"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb513-1"><a href="complete-protocols.html#cb513-1" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> df_tidy <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(</span>
<span id="cb513-2"><a href="complete-protocols.html#cb513-2" tabindex="-1"></a>  <span class="at">Change =</span> <span class="fu">case_when</span>(</span>
<span id="cb513-3"><a href="complete-protocols.html#cb513-3" tabindex="-1"></a>    <span class="st">`</span><span class="at">log2_FoldChange</span><span class="st">`</span> <span class="sc">&gt;</span> <span class="fl">1.5</span> <span class="sc">&amp;</span> <span class="st">`</span><span class="at">minus_log10_pvalue</span><span class="st">`</span> <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">&quot;Increased&quot;</span>,</span>
<span id="cb513-4"><a href="complete-protocols.html#cb513-4" tabindex="-1"></a>    <span class="st">`</span><span class="at">log2_FoldChange</span><span class="st">`</span> <span class="sc">&lt;</span> <span class="sc">-</span><span class="fl">1.5</span> <span class="sc">&amp;</span> <span class="st">`</span><span class="at">minus_log10_pvalue</span><span class="st">`</span> <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">~</span> <span class="st">&quot;Decreased&quot;</span>,</span>
<span id="cb513-5"><a href="complete-protocols.html#cb513-5" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">&quot;Unchanged&quot;</span></span>
<span id="cb513-6"><a href="complete-protocols.html#cb513-6" tabindex="-1"></a>    )</span>
<span id="cb513-7"><a href="complete-protocols.html#cb513-7" tabindex="-1"></a>  )</span>
<span id="cb513-8"><a href="complete-protocols.html#cb513-8" tabindex="-1"></a></span>
<span id="cb513-9"><a href="complete-protocols.html#cb513-9" tabindex="-1"></a><span class="fu">head</span>(df_tidy)</span></code></pre></div>
<pre><code># A tibble: 6 × 10
  Gene  baseMean log2_FoldChange    FC lfcSE  stat    pvalue     padj
  &lt;chr&gt;    &lt;dbl&gt;           &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 A1bg      39.2           1.25   2.38 0.299  4.19 0.0000278 0.00170 
2 A1cf    1183.            0.745  1.68 0.192  3.88 0.000102  0.00456 
3 Aadac   3196.            0.688  1.61 0.157  4.39 0.0000111 0.000817
4 Aadat   1481.            0.603  1.52 0.181  3.33 0.000878  0.0232  
5 Aaed1    432.            0.422  1.34 0.193  2.19 0.0288    0.235   
6 Aamdc    165.            0.332  1.26 0.191  1.74 0.0827    0.419   
# ℹ 2 more variables: minus_log10_pvalue &lt;dbl&gt;, Change &lt;chr&gt;</code></pre>
<p>We can now use the column ‘Change’ to color code the data:</p>
<div class="sourceCode" id="cb515"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb515-1"><a href="complete-protocols.html#cb515-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> df_tidy, <span class="fu">aes</span>(<span class="at">x=</span>log2_FoldChange, <span class="at">y=</span>minus_log10_pvalue)) <span class="sc">+</span> </span>
<span id="cb515-2"><a href="complete-protocols.html#cb515-2" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">1.5</span>,<span class="fl">1.5</span>), <span class="at">linetype=</span><span class="st">&quot;dashed&quot;</span>) <span class="sc">+</span></span>
<span id="cb515-3"><a href="complete-protocols.html#cb515-3" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">5</span>, <span class="at">linetype=</span><span class="st">&quot;dashed&quot;</span>)</span>
<span id="cb515-4"><a href="complete-protocols.html#cb515-4" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> Change))</span>
<span id="cb515-5"><a href="complete-protocols.html#cb515-5" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-806-1.png" width="672" /></p>
</div>
<div id="step-4---annotations" class="section level3 hasAnchor" number="4.14.4">
<h3><span class="header-section-number">4.14.4</span> Step 4 - Annotations<a href="complete-protocols.html#step-4---annotations" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The ‘top hits’ are the datapoints that are the furthest from the origin. Here, we only consider genes that are either ‘Decreased’ or ‘Increased’, so we remove the ‘Unchanged’. To determine the distance from the origin for each datapoint, we sum the x- and y- position. Note that this simple summation is known as the ‘Manhattan distance’. After calculation, the top 10 is selected and stored in another dataframe, df_top. :</p>
<div class="sourceCode" id="cb516"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb516-1"><a href="complete-protocols.html#cb516-1" tabindex="-1"></a>df_top <span class="ot">&lt;-</span> df_tidy <span class="sc">%&gt;%</span></span>
<span id="cb516-2"><a href="complete-protocols.html#cb516-2" tabindex="-1"></a>  <span class="fu">filter</span>(Change <span class="sc">!=</span> <span class="st">&#39;Unchanged&#39;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb516-3"><a href="complete-protocols.html#cb516-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">distance =</span> minus_log10_pvalue <span class="sc">+</span> <span class="fu">abs</span>(log2_FoldChange)) <span class="sc">%&gt;%</span></span>
<span id="cb516-4"><a href="complete-protocols.html#cb516-4" tabindex="-1"></a>  <span class="fu">top_n</span>(<span class="dv">10</span>,distance)</span></code></pre></div>
<p>The new dataframe can be used for labeling the top-10 hits:</p>
<div class="sourceCode" id="cb517"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb517-1"><a href="complete-protocols.html#cb517-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_text</span>(</span>
<span id="cb517-2"><a href="complete-protocols.html#cb517-2" tabindex="-1"></a>    <span class="at">data =</span> df_top,</span>
<span id="cb517-3"><a href="complete-protocols.html#cb517-3" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">label =</span> Gene)</span>
<span id="cb517-4"><a href="complete-protocols.html#cb517-4" tabindex="-1"></a>    )</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-808-1.png" width="672" /></p>
<p>Wwe see that the labels overlap and this is often the case when a lot of data is displayed. There is a very handy packgae <code>{ggrepel}</code> that will reduce the overlap between labels:</p>
<div class="sourceCode" id="cb518"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb518-1"><a href="complete-protocols.html#cb518-1" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb518-2"><a href="complete-protocols.html#cb518-2" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_text_repel</span>(</span>
<span id="cb518-3"><a href="complete-protocols.html#cb518-3" tabindex="-1"></a>    <span class="at">data =</span> df_top,</span>
<span id="cb518-4"><a href="complete-protocols.html#cb518-4" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">label =</span> Gene),</span>
<span id="cb518-5"><a href="complete-protocols.html#cb518-5" tabindex="-1"></a>    <span class="at">size=</span><span class="dv">5</span>, <span class="co"># increase font size</span></span>
<span id="cb518-6"><a href="complete-protocols.html#cb518-6" tabindex="-1"></a>    <span class="at">min.segment.length =</span> <span class="dv">0</span>, <span class="co"># draw all line segments</span></span>
<span id="cb518-7"><a href="complete-protocols.html#cb518-7" tabindex="-1"></a>    <span class="at">box.padding =</span> <span class="fl">0.5</span><span class="co"># give some space for drawing the lines</span></span>
<span id="cb518-8"><a href="complete-protocols.html#cb518-8" tabindex="-1"></a>    </span>
<span id="cb518-9"><a href="complete-protocols.html#cb518-9" tabindex="-1"></a>    )</span>
<span id="cb518-10"><a href="complete-protocols.html#cb518-10" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-809-1.png" width="672" /></p>
<p>Now that we have implemeted the features of a volcano plot, we can focus on improving the data visualization. We can emphasize the hits by adding a black outline to these points. The default shape for ‘geom_point()’ is a filled circle. Below we use shape 21, which accepts both a color for the outline and filling. The black outline is defined by ‘color’:</p>
<div class="sourceCode" id="cb519"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb519-1"><a href="complete-protocols.html#cb519-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">data =</span> df_top, <span class="fu">aes</span>(<span class="at">x=</span>log2_FoldChange, <span class="at">y=</span>minus_log10_pvalue), <span class="at">shape=</span><span class="dv">21</span>, <span class="at">color=</span><span class="st">&#39;black&#39;</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-810-1.png" width="672" /></p>
<p>The line that connects the dot with the label is a bit too long and is visible within the dot, which is a bit ugly. To solve this we can fill these dots which will hide the line. We can also make the dots a bit bigger to emphasize the hits:</p>
<div class="sourceCode" id="cb520"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb520-1"><a href="complete-protocols.html#cb520-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">data =</span> df_top, <span class="fu">aes</span>(<span class="at">x=</span>log2_FoldChange, <span class="at">y=</span>minus_log10_pvalue, <span class="at">fill=</span>Change), <span class="at">shape=</span><span class="dv">21</span>, <span class="at">color=</span><span class="st">&#39;black&#39;</span>, <span class="at">size=</span><span class="dv">2</span>)</span>
<span id="cb520-2"><a href="complete-protocols.html#cb520-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-811-1.png" width="672" /></p>
<p>Let’s change the colors to make the uninteresting datapoints less pronounced:</p>
<div class="sourceCode" id="cb521"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb521-1"><a href="complete-protocols.html#cb521-1" tabindex="-1"></a>newColors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;dodgerblue&quot;</span>, <span class="st">&quot;orange&quot;</span>, <span class="st">&quot;grey&quot;</span>)</span>
<span id="cb521-2"><a href="complete-protocols.html#cb521-2" tabindex="-1"></a></span>
<span id="cb521-3"><a href="complete-protocols.html#cb521-3" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values =</span> newColors) <span class="sc">+</span> <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> newColors)</span>
<span id="cb521-4"><a href="complete-protocols.html#cb521-4" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-812-1.png" width="672" />
Finally, we can change the styling by adjusting the labels and the theme. The labels for x- and y-axis use the function <code>expression()</code> to add subscript text to the labels:</p>
<div class="sourceCode" id="cb522"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb522-1"><a href="complete-protocols.html#cb522-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb522-2"><a href="complete-protocols.html#cb522-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb522-3"><a href="complete-protocols.html#cb522-3" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Differentially expressed genes&quot;</span>,</span>
<span id="cb522-4"><a href="complete-protocols.html#cb522-4" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">expression</span>(<span class="st">&#39;Fold Change (&#39;</span><span class="sc">*</span>Log[<span class="dv">2</span>]<span class="sc">*</span><span class="st">&#39;)&#39;</span>),</span>
<span id="cb522-5"><a href="complete-protocols.html#cb522-5" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">expression</span>(<span class="st">&#39;Significance (&#39;</span><span class="sc">*-</span>Log[<span class="dv">10</span>]<span class="sc">*</span><span class="st">&#39;)&#39;</span>),</span>
<span id="cb522-6"><a href="complete-protocols.html#cb522-6" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">(based on data from Becares et al, DOI: 10.1016/j.celrep.2018.12.094)&quot;</span>,</span>
<span id="cb522-7"><a href="complete-protocols.html#cb522-7" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 14&quot;</span></span>
<span id="cb522-8"><a href="complete-protocols.html#cb522-8" tabindex="-1"></a>  )</span></code></pre></div>
<p>Modify the layout by adjusting the theme:</p>
<div class="sourceCode" id="cb523"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb523-1"><a href="complete-protocols.html#cb523-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb523-2"><a href="complete-protocols.html#cb523-2" tabindex="-1"></a>  <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">16</span>) <span class="sc">+</span></span>
<span id="cb523-3"><a href="complete-protocols.html#cb523-3" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb523-4"><a href="complete-protocols.html#cb523-4" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb523-5"><a href="complete-protocols.html#cb523-5" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p>Change the x-axis scale the give the labels a bit more room and display the result:</p>
<div class="sourceCode" id="cb524"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb524-1"><a href="complete-protocols.html#cb524-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">xlim</span>(<span class="sc">-</span><span class="dv">3</span>,<span class="dv">3</span>)</span>
<span id="cb524-2"><a href="complete-protocols.html#cb524-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-815-1.png" width="672" /></p>
<p>To save the plot as a png file:</p>
<div class="sourceCode" id="cb525"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb525-1"><a href="complete-protocols.html#cb525-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_14.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb525-2"><a href="complete-protocols.html#cb525-2" tabindex="-1"></a> p</span>
<span id="cb525-3"><a href="complete-protocols.html#cb525-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
</div>
<div id="protocol-15" class="section level2 hasAnchor" number="4.15">
<h2><span class="header-section-number">4.15</span> Protocol 15 - Timeline<a href="complete-protocols.html#protocol-15" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this protocol we will make a timeline. The timeline organizes events along a line that consists of dates, in this case years. The years are treated as continuous data and the events are text and hence qualitative data.</p>
<p>We start by loading the <code>{tidyverse}</code> package, mainly for the use of ggplot2:</p>
<div class="sourceCode" id="cb527"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb527-1"><a href="complete-protocols.html#cb527-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>The data for the timeline is stored in a text file that has tabs as delimiters:</p>
<div class="sourceCode" id="cb528"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb528-1"><a href="complete-protocols.html#cb528-1" tabindex="-1"></a>df_timeline <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;data/GFP-timeline.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&#39;</span><span class="sc">\t</span><span class="st">&#39;</span> )</span>
<span id="cb528-2"><a href="complete-protocols.html#cb528-2" tabindex="-1"></a><span class="fu">head</span>(df_timeline)</span></code></pre></div>
<pre><code>  Year                          DOI
1 1962       10.1002/jcp.1030590302
2 1971       10.1002/jcp.1040770305
3 1974          10.1021/bi00709a028
4 1979 10.1016/0014-5793(79)80818-2
5 1992 10.1016/0378-1119(92)90691-h
6 1994      10.1126/science.8303295
                                                    Event
1 Osamu Shimomura isolates a green protein from jellyfish
2  First appearance of the name Green Fluorescent Protein
3                      Crystallization of GFP, QY is 0.72
4              Structure of the GFP chromophore is solved
5                   Douglas Prasher clones the AvGFP gene
6    Martin Chalfie expresses AvGFP in bacteria and worms</code></pre>
<p>Let’s first make a basic plot to visualize the data in the column ‘Year’:</p>
<div class="sourceCode" id="cb530"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb530-1"><a href="complete-protocols.html#cb530-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_timeline, <span class="fu">aes</span>(<span class="at">x=</span><span class="dv">1</span>,<span class="at">y=</span>Year)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-837-1.png" width="672" /></p>
<p>Each point reflects a year in which some event happened. The events can be plotted as text:</p>
<div class="sourceCode" id="cb531"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb531-1"><a href="complete-protocols.html#cb531-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_timeline, <span class="fu">aes</span>(<span class="at">x=</span><span class="dv">1</span>,<span class="at">y=</span>Year)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> Event))</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-838-1.png" width="672" /></p>
<p>We need to do a couple of things. First, make a nice timeline that shows each year as a dot and the years with events as a more pronounced dot. Second, we need to align the text and reduce the overlap.</p>
<div id="improve-the-timeline" class="section level3 hasAnchor" number="4.15.1">
<h3><span class="header-section-number">4.15.1</span> Improve the timeline<a href="complete-protocols.html#improve-the-timeline" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To make a nice timeline, I’d like to show every year as dot and therefore I define a new dataframe:</p>
<div class="sourceCode" id="cb532"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb532-1"><a href="complete-protocols.html#cb532-1" tabindex="-1"></a>yrs <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Year=</span><span class="fu">min</span>(df_timeline<span class="sc">$</span>Year)<span class="sc">:</span><span class="fu">max</span>(df_timeline<span class="sc">$</span>Year))</span></code></pre></div>
<div class="sourceCode" id="cb533"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb533-1"><a href="complete-protocols.html#cb533-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_timeline, <span class="fu">aes</span>(<span class="at">x=</span><span class="dv">1</span>,<span class="at">y=</span>Year)) <span class="sc">+</span></span>
<span id="cb533-2"><a href="complete-protocols.html#cb533-2" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">shape=</span><span class="dv">21</span>, <span class="at">size=</span><span class="dv">3</span>, <span class="at">fill=</span><span class="st">&quot;white&quot;</span>) <span class="sc">+</span></span>
<span id="cb533-3"><a href="complete-protocols.html#cb533-3" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">data =</span>yrs, <span class="fu">aes</span>(<span class="at">x=</span><span class="dv">1</span>), <span class="at">size=</span><span class="fl">0.5</span>, <span class="at">color=</span><span class="st">&quot;grey20&quot;</span>) </span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-840-1.png" width="672" /></p>
<p>Let’s add a line to make it a time<em>line</em> and do some formatting of the theme to get rid of the legend:</p>
<div class="sourceCode" id="cb534"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb534-1"><a href="complete-protocols.html#cb534-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_timeline, <span class="fu">aes</span>(<span class="at">x=</span><span class="dv">1</span>, <span class="at">y=</span>Year)) <span class="sc">+</span></span>
<span id="cb534-2"><a href="complete-protocols.html#cb534-2" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">1</span>, <span class="at">size=</span><span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb534-3"><a href="complete-protocols.html#cb534-3" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">shape=</span><span class="dv">21</span>, <span class="at">size=</span><span class="dv">3</span>, <span class="at">fill=</span><span class="st">&quot;white&quot;</span>) <span class="sc">+</span></span>
<span id="cb534-4"><a href="complete-protocols.html#cb534-4" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">data=</span>yrs, <span class="fu">aes</span>(<span class="at">x=</span><span class="dv">1</span>), <span class="at">size=</span><span class="fl">0.5</span>, <span class="at">color=</span><span class="st">&quot;grey20&quot;</span>) <span class="sc">+</span></span>
<span id="cb534-5"><a href="complete-protocols.html#cb534-5" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-841-1.png" width="672" /></p>
<p>We will remove the grid and text for the x-axis. We add the ticks for the y-axis and we specify the labels for the years using <code>breaks</code> in <code>scale_y_continuous()</code>:</p>
<div class="sourceCode" id="cb535"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb535-1"><a href="complete-protocols.html#cb535-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_timeline, <span class="fu">aes</span>(<span class="at">x=</span><span class="dv">1</span>, <span class="at">y=</span>Year)) <span class="sc">+</span></span>
<span id="cb535-2"><a href="complete-protocols.html#cb535-2" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">1</span>, <span class="at">size=</span><span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb535-3"><a href="complete-protocols.html#cb535-3" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">shape=</span><span class="dv">21</span>, <span class="at">size=</span><span class="dv">3</span>, <span class="at">fill=</span><span class="st">&quot;white&quot;</span>) <span class="sc">+</span></span>
<span id="cb535-4"><a href="complete-protocols.html#cb535-4" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>yrs, <span class="fu">aes</span>(<span class="at">x=</span><span class="dv">1</span>), <span class="at">size=</span><span class="fl">0.5</span>, <span class="at">color=</span><span class="st">&quot;grey20&quot;</span>) <span class="sc">+</span></span>
<span id="cb535-5"><a href="complete-protocols.html#cb535-5" tabindex="-1"></a>  <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">16</span>) <span class="sc">+</span></span>
<span id="cb535-6"><a href="complete-protocols.html#cb535-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>) <span class="sc">+</span></span>
<span id="cb535-7"><a href="complete-protocols.html#cb535-7" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.ticks.y =</span> <span class="fu">element_line</span>(<span class="at">color=</span><span class="st">&quot;grey&quot;</span>, <span class="at">linetype =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="dv">1</span>),</span>
<span id="cb535-8"><a href="complete-protocols.html#cb535-8" tabindex="-1"></a>        <span class="at">panel.grid =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb535-9"><a href="complete-protocols.html#cb535-9" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb535-10"><a href="complete-protocols.html#cb535-10" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb535-11"><a href="complete-protocols.html#cb535-11" tabindex="-1"></a>        <span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>, <span class="at">hjust =</span> <span class="dv">0</span>)</span>
<span id="cb535-12"><a href="complete-protocols.html#cb535-12" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb535-13"><a href="complete-protocols.html#cb535-13" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">1962</span>,<span class="dv">2025</span>), <span class="at">breaks =</span> <span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">1965</span>, <span class="dv">2024</span>, <span class="at">by=</span><span class="dv">5</span>)))</span>
<span id="cb535-14"><a href="complete-protocols.html#cb535-14" tabindex="-1"></a></span>
<span id="cb535-15"><a href="complete-protocols.html#cb535-15" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-842-1.png" width="672" /></p>
<p>Allright, things start to look better now. The next step is to add text. We will use labels and the <code>geom_label_repel()</code> function from the <code>{ggrepel}</code> package to avoid overlap. First we load the package:</p>
<div class="sourceCode" id="cb536"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb536-1"><a href="complete-protocols.html#cb536-1" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span></code></pre></div>
</div>
<div id="add-and-format-text" class="section level3 hasAnchor" number="4.15.2">
<h3><span class="header-section-number">4.15.2</span> Add and format text<a href="complete-protocols.html#add-and-format-text" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now, we can add labels to the plotting object:</p>
<div class="sourceCode" id="cb537"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb537-1"><a href="complete-protocols.html#cb537-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_label_repel</span>(<span class="fu">aes</span>(<span class="at">label=</span>Event))</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-844-1.png" width="672" /></p>
<p>Let’s change the x-axis scale, to make room for the text on the right side of the line. Also, we can left-align the text vertically and add some space between the points and the text. This is achieved by optimizing the ‘nudge_x’, ‘hjust’ and ‘direction’:</p>
<div class="sourceCode" id="cb538"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb538-1"><a href="complete-protocols.html#cb538-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_label_repel</span>(<span class="fu">aes</span>(<span class="at">label=</span>Event),</span>
<span id="cb538-2"><a href="complete-protocols.html#cb538-2" tabindex="-1"></a>                     <span class="at">nudge_x   =</span> <span class="fl">0.05</span>,</span>
<span id="cb538-3"><a href="complete-protocols.html#cb538-3" tabindex="-1"></a>                     <span class="at">hjust     =</span> <span class="dv">0</span>,</span>
<span id="cb538-4"><a href="complete-protocols.html#cb538-4" tabindex="-1"></a>                     <span class="at">direction =</span> <span class="st">&quot;y&quot;</span></span>
<span id="cb538-5"><a href="complete-protocols.html#cb538-5" tabindex="-1"></a>                     ) <span class="sc">+</span></span>
<span id="cb538-6"><a href="complete-protocols.html#cb538-6" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="fl">0.01</span>, <span class="dv">0</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="fl">1.0</span>,<span class="fl">1.5</span>)) </span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-845-1.png" width="672" /></p>
<p>Finally, we can adjust the font size for the label. We also add colors to the different years. I like to use the ‘viridis’ color scale here. because it runs from dark to bright. However, if we would use the full range the final color is bright yellow which has a poor contrast on white. To solve this, we define the ‘end’ of the viridis scale at 80%:</p>
<div class="sourceCode" id="cb539"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb539-1"><a href="complete-protocols.html#cb539-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">color=</span>Year) <span class="sc">+</span></span>
<span id="cb539-2"><a href="complete-protocols.html#cb539-2" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(<span class="fu">aes</span>(<span class="at">label=</span>Event),</span>
<span id="cb539-3"><a href="complete-protocols.html#cb539-3" tabindex="-1"></a>                     <span class="at">nudge_x   =</span> <span class="fl">0.05</span>,</span>
<span id="cb539-4"><a href="complete-protocols.html#cb539-4" tabindex="-1"></a>                     <span class="at">hjust     =</span> <span class="dv">0</span>,</span>
<span id="cb539-5"><a href="complete-protocols.html#cb539-5" tabindex="-1"></a>                     <span class="at">direction =</span> <span class="st">&quot;y&quot;</span>,</span>
<span id="cb539-6"><a href="complete-protocols.html#cb539-6" tabindex="-1"></a>                     <span class="at">size      =</span> <span class="dv">3</span></span>
<span id="cb539-7"><a href="complete-protocols.html#cb539-7" tabindex="-1"></a>                     ) <span class="sc">+</span></span>
<span id="cb539-8"><a href="complete-protocols.html#cb539-8" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="fl">0.01</span>, <span class="dv">0</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="fl">1.0</span>,<span class="fl">1.5</span>)) <span class="sc">+</span></span>
<span id="cb539-9"><a href="complete-protocols.html#cb539-9" tabindex="-1"></a></span>
<span id="cb539-10"><a href="complete-protocols.html#cb539-10" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">end =</span> <span class="fl">0.8</span>)<span class="sc">+</span></span>
<span id="cb539-11"><a href="complete-protocols.html#cb539-11" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;60 years of GFP discovery and engineering&quot;</span>,</span>
<span id="cb539-12"><a href="complete-protocols.html#cb539-12" tabindex="-1"></a>       <span class="at">tag =</span> <span class="st">&quot;Protocol 15&quot;</span>,</span>
<span id="cb539-13"><a href="complete-protocols.html#cb539-13" tabindex="-1"></a>       <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart&quot;</span>) <span class="sc">+</span></span>
<span id="cb539-14"><a href="complete-protocols.html#cb539-14" tabindex="-1"></a>  <span class="cn">NULL</span></span>
<span id="cb539-15"><a href="complete-protocols.html#cb539-15" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-846-1.png" width="768" /></p>
<p>To save the plot as a png file:</p>
<div class="sourceCode" id="cb540"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb540-1"><a href="complete-protocols.html#cb540-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_15.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb540-2"><a href="complete-protocols.html#cb540-2" tabindex="-1"></a> p</span>
<span id="cb540-3"><a href="complete-protocols.html#cb540-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
</div>
<div id="protocol-16" class="section level2 hasAnchor" number="4.16">
<h2><span class="header-section-number">4.16</span> Protocol 16 - Scatterplots and correlation<a href="complete-protocols.html#protocol-16" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this protocol we will look at the data from a co-expression experiment. Images of cells co-expressing two fluorescent proteins were acquired with a fluorescence microscope. The relative intensity of each fluorescent protein was measured. The data can be plotted in a scatter plot to visualize the correlation between the two expressed proteins. To quantify the variation in co-expression, a correlation coefficient can be determined. More experimental details can be found in the original publication <a href="https://doi.org/10.1371/journal.pone.0027321">Goedhart et al. (2011)</a>.</p>
<p>We start by loading the <code>{tidyverse}</code> package:</p>
<div class="sourceCode" id="cb542"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb542-1"><a href="complete-protocols.html#cb542-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>Read the data:</p>
<div class="sourceCode" id="cb543"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb543-1"><a href="complete-protocols.html#cb543-1" tabindex="-1"></a>df_co <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;data/coexpression.csv&quot;</span>, <span class="at">stringsAsFactors =</span> <span class="cn">TRUE</span>)</span>
<span id="cb543-2"><a href="complete-protocols.html#cb543-2" tabindex="-1"></a><span class="fu">head</span>(df_co)</span></code></pre></div>
<pre><code>  BCY_YFP BCY_CFP BYC_YFP BYC_CFP TCY_YFP TCY_CFP TYC_YFP TYC_CFP IYC_YFP
1  88.841  113.92  88.841  113.92  107.36  124.57  148.75  165.71  290.32
2  95.053  118.53  95.053  118.53  129.05  158.97  166.32  192.10  362.47
3  99.563  208.09  99.563  208.09  132.55  160.52  170.43  215.41  395.74
4 105.630  275.01 105.630  275.01  140.84  166.70  191.14  225.23  420.21
5 124.370  162.51 124.370  162.51  144.45  166.72  198.11  255.85  476.79
6 142.130  225.24 142.130  225.24  152.47  146.71  213.19  230.92  495.70
  IYC_CFP ICY_YFP ICY_CFP
1  122.01  60.505  196.97
2  151.95 124.820  625.77
3  159.96 145.510  672.24
4  194.17 143.980  703.17
5  161.08 143.450  717.70
6  225.22 150.590  775.40</code></pre>
<p>The data was stored in a spreadsheet, and is in a wide format. Pairs of columns belong to the same measurement. The two fluorescent proteins were CFP and YFP and this indicated in the column name after the underscore. Several co-expression systems were analyzed and this is indicated in the string before the underscore. We can plot one condition to see what the data looks like in a scatter plot:</p>
<div class="sourceCode" id="cb545"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb545-1"><a href="complete-protocols.html#cb545-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_co, <span class="fu">aes</span>(<span class="at">x=</span>BCY_YFP, <span class="at">y=</span>BCY_CFP)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-865-1.png" width="672" /></p>
<div id="linear-fit" class="section level3 hasAnchor" number="4.16.1">
<h3><span class="header-section-number">4.16.1</span> Linear fit<a href="complete-protocols.html#linear-fit" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>A linear fit can be added with the <code>geom_smooth</code> function. The method is speficied as <code>lm</code> which means linear model:</p>
<div class="sourceCode" id="cb546"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb546-1"><a href="complete-protocols.html#cb546-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_co, <span class="fu">aes</span>(<span class="at">x=</span>BCY_YFP, <span class="at">y=</span>BCY_CFP)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&#39;lm&#39;</span>)</span></code></pre></div>
<pre><code>`geom_smooth()` using formula = &#39;y ~ x&#39;</code></pre>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-866-1.png" width="672" /></p>
<p>The linear model has an intercept, but suppose that I want to fit the data without offset. To this end, a custom formula can added to overrule the default:</p>
<div class="sourceCode" id="cb548"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb548-1"><a href="complete-protocols.html#cb548-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_co, <span class="fu">aes</span>(<span class="at">x=</span>BCY_YFP, <span class="at">y=</span>BCY_CFP)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&#39;lm&#39;</span>, <span class="at">formula =</span> y<span class="sc">~</span><span class="dv">0</span><span class="sc">+</span>x)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-867-1.png" width="672" /></p>
<p>To get to the coefficients of the fit we can use the <code>lm()</code> function fits data to a linear model:</p>
<div class="sourceCode" id="cb549"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb549-1"><a href="complete-protocols.html#cb549-1" tabindex="-1"></a><span class="fu">lm</span>(BCY_CFP <span class="sc">~</span> BCY_YFP <span class="sc">+</span> <span class="dv">0</span>, <span class="at">data =</span> df_co)</span></code></pre></div>
<pre><code>
Call:
lm(formula = BCY_CFP ~ BCY_YFP + 0, data = df_co)

Coefficients:
BCY_YFP  
   1.14  </code></pre>
<p>To get a summary of the fit we use the <code>summary()</code> function and assign the results to ‘fit_summary’:</p>
<div class="sourceCode" id="cb551"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb551-1"><a href="complete-protocols.html#cb551-1" tabindex="-1"></a>fit_summary <span class="ot">&lt;-</span> <span class="fu">summary</span>(<span class="fu">lm</span>(BCY_CFP <span class="sc">~</span> BCY_YFP <span class="sc">+</span> <span class="dv">0</span>, <span class="at">data =</span> df_co))</span>
<span id="cb551-2"><a href="complete-protocols.html#cb551-2" tabindex="-1"></a>fit_summary</span></code></pre></div>
<pre><code>
Call:
lm(formula = BCY_CFP ~ BCY_YFP + 0, data = df_co)

Residuals:
    Min      1Q  Median      3Q     Max 
-3622.9  -140.6    45.6   284.5  4270.0 

Coefficients:
        Estimate Std. Error t value Pr(&gt;|t|)    
BCY_YFP  1.13983    0.02218   51.39   &lt;2e-16 ***
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

Residual standard error: 572 on 733 degrees of freedom
  (106 observations deleted due to missingness)
Multiple R-squared:  0.7828,    Adjusted R-squared:  0.7825 
F-statistic:  2641 on 1 and 733 DF,  p-value: &lt; 2.2e-16</code></pre>
<p>The object ‘fit_summary’ is a list and we can extract the R-squared value from this list:</p>
<div class="sourceCode" id="cb553"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb553-1"><a href="complete-protocols.html#cb553-1" tabindex="-1"></a>fit_summary<span class="sc">$</span>r.squared</span></code></pre></div>
<pre><code>[1] 0.7827556</code></pre>
</div>
<div id="multiple-scatterplots" class="section level3 hasAnchor" number="4.16.2">
<h3><span class="header-section-number">4.16.2</span> Multiple scatterplots<a href="complete-protocols.html#multiple-scatterplots" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>So far so good; we have plotted the data from a single condition and the relevant parameters can be displayed. Now, let’s see if we can do the same for the entire dataset. First, we will need to restructure the data into a long format:</p>
<div class="sourceCode" id="cb555"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb555-1"><a href="complete-protocols.html#cb555-1" tabindex="-1"></a>df_longer_co <span class="ot">&lt;-</span> df_co <span class="sc">%&gt;%</span> <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">everything</span>(), <span class="at">names_to =</span> <span class="st">&quot;Condition&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;Intensity&quot;</span>)</span>
<span id="cb555-2"><a href="complete-protocols.html#cb555-2" tabindex="-1"></a><span class="fu">head</span>(df_longer_co)</span></code></pre></div>
<pre><code># A tibble: 6 × 2
  Condition Intensity
  &lt;chr&gt;         &lt;dbl&gt;
1 BCY_YFP        88.8
2 BCY_CFP       114. 
3 BYC_YFP        88.8
4 BYC_CFP       114. 
5 TCY_YFP       107. 
6 TCY_CFP       125. </code></pre>
<p>This is a ‘true’ long format, but we will need to have the CFP and YFP data side-by-side to generate the scatterplot. Let’s first split the column with conditions in the Condition and the fluorescent protein (CFP or YFP):</p>
<div class="sourceCode" id="cb557"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb557-1"><a href="complete-protocols.html#cb557-1" tabindex="-1"></a>df_longer_co <span class="ot">&lt;-</span> df_longer_co <span class="sc">%&gt;%</span> <span class="fu">separate</span>(Condition, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">&quot;Plasmid&quot;</span>, <span class="st">&quot;Protein&quot;</span>), <span class="at">sep =</span> <span class="st">&quot;_&quot;</span>)</span>
<span id="cb557-2"><a href="complete-protocols.html#cb557-2" tabindex="-1"></a><span class="fu">head</span>(df_longer_co)</span></code></pre></div>
<pre><code># A tibble: 6 × 3
  Plasmid Protein Intensity
  &lt;chr&gt;   &lt;chr&gt;       &lt;dbl&gt;
1 BCY     YFP          88.8
2 BCY     CFP         114. 
3 BYC     YFP          88.8
4 BYC     CFP         114. 
5 TCY     YFP         107. 
6 TCY     CFP         125. </code></pre>
<p>Now we can make the data ‘wider’ with the <code>pivot_wider()</code> function:</p>
<div class="sourceCode" id="cb559"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb559-1"><a href="complete-protocols.html#cb559-1" tabindex="-1"></a>df_wider_co <span class="ot">&lt;-</span> df_longer_co <span class="sc">%&gt;%</span> <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> Protein, <span class="at">values_from =</span> Intensity)</span>
<span id="cb559-2"><a href="complete-protocols.html#cb559-2" tabindex="-1"></a><span class="fu">head</span>(df_wider_co)</span></code></pre></div>
<pre><code># A tibble: 6 × 3
  Plasmid YFP         CFP        
  &lt;chr&gt;   &lt;list&gt;      &lt;list&gt;     
1 BCY     &lt;dbl [840]&gt; &lt;dbl [840]&gt;
2 BYC     &lt;dbl [840]&gt; &lt;dbl [840]&gt;
3 TCY     &lt;dbl [840]&gt; &lt;dbl [840]&gt;
4 TYC     &lt;dbl [840]&gt; &lt;dbl [840]&gt;
5 IYC     &lt;dbl [840]&gt; &lt;dbl [840]&gt;
6 ICY     &lt;dbl [840]&gt; &lt;dbl [840]&gt;</code></pre>
<p>This will throw a warning message, since some rows have duplicated values. This is not an issue here, but the result is a ‘nested’ dataframe with the data in lists instead of a single value per cell. To fix this, we use <code>unnest()</code>:</p>
<div class="sourceCode" id="cb561"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb561-1"><a href="complete-protocols.html#cb561-1" tabindex="-1"></a>df_co_all <span class="ot">&lt;-</span> df_wider_co <span class="sc">%&gt;%</span> <span class="fu">unnest</span>(<span class="at">cols =</span> <span class="fu">c</span>(YFP, CFP))</span>
<span id="cb561-2"><a href="complete-protocols.html#cb561-2" tabindex="-1"></a><span class="fu">head</span>(df_co_all)</span></code></pre></div>
<pre><code># A tibble: 6 × 3
  Plasmid   YFP   CFP
  &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;
1 BCY      88.8  114.
2 BCY      95.1  119.
3 BCY      99.6  208.
4 BCY     106.   275.
5 BCY     124.   163.
6 BCY     142.   225.</code></pre>
<p>Now, we have the data in the right shape and we can make scatterplots for each ‘Plasmid’ with <code>facet_wrap()</code>:</p>
<div class="sourceCode" id="cb563"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb563-1"><a href="complete-protocols.html#cb563-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_co_all, <span class="fu">aes</span>(<span class="at">x=</span>YFP, <span class="at">y=</span>CFP)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>Plasmid, <span class="at">nrow =</span> <span class="dv">2</span> )</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-875-1.png" width="672" /></p>
<p>This looks great, but the names for the column ‘Plasmids’ are not informative so let’s fix that in the dataframe. The last two letter, either YC or CY indicate the order of the fluorescent proteins. The first letter indicates the system that was used for the co-expressing, with B = dual promotor, I = Ires and T = 2A.
First, we split the column ‘Plasmid’:</p>
<div class="sourceCode" id="cb564"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb564-1"><a href="complete-protocols.html#cb564-1" tabindex="-1"></a>df_co_all <span class="sc">%&gt;%</span> <span class="fu">separate</span>(Plasmid, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">&#39;System&#39;</span>, <span class="st">&#39;Order&#39;</span>), <span class="at">sep =</span> <span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code></pre></div>
<pre><code># A tibble: 6 × 4
  System Order   YFP   CFP
  &lt;chr&gt;  &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;
1 &quot;&quot;     B      88.8  114.
2 &quot;&quot;     B      95.1  119.
3 &quot;&quot;     B      99.6  208.
4 &quot;&quot;     B     106.   275.
5 &quot;&quot;     B     124.   163.
6 &quot;&quot;     B     142.   225.</code></pre>
<p>This doesn’t look right, since the first column is empty, so let’s try again with:</p>
<div class="sourceCode" id="cb566"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb566-1"><a href="complete-protocols.html#cb566-1" tabindex="-1"></a>df_co_all_split <span class="ot">&lt;-</span> df_co_all <span class="sc">%&gt;%</span> <span class="fu">separate</span>(Plasmid, <span class="at">into =</span> <span class="fu">c</span>(<span class="cn">NA</span>, <span class="st">&#39;System&#39;</span>, <span class="st">&#39;Order&#39;</span>), <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb566-2"><a href="complete-protocols.html#cb566-2" tabindex="-1"></a><span class="fu">head</span>(df_co_all_split)</span></code></pre></div>
<pre><code># A tibble: 6 × 4
  System Order   YFP   CFP
  &lt;chr&gt;  &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;
1 B      C      88.8  114.
2 B      C      95.1  119.
3 B      C      99.6  208.
4 B      C     106.   275.
5 B      C     124.   163.
6 B      C     142.   225.</code></pre>
<p>This looks good, and so we can replace the characters with names:</p>
<div class="sourceCode" id="cb568"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb568-1"><a href="complete-protocols.html#cb568-1" tabindex="-1"></a>df <span class="ot">&lt;-</span> df_co_all_split <span class="sc">%&gt;%</span></span>
<span id="cb568-2"><a href="complete-protocols.html#cb568-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">System =</span> <span class="fu">case_when</span>(System <span class="sc">==</span> <span class="st">&#39;B&#39;</span> <span class="sc">~</span> <span class="st">&quot; Dual-promotor&quot;</span>,</span>
<span id="cb568-3"><a href="complete-protocols.html#cb568-3" tabindex="-1"></a>                            System <span class="sc">==</span> <span class="st">&#39;I&#39;</span> <span class="sc">~</span> <span class="st">&quot; IRES&quot;</span>,</span>
<span id="cb568-4"><a href="complete-protocols.html#cb568-4" tabindex="-1"></a>                            System <span class="sc">==</span> <span class="st">&#39;T&#39;</span> <span class="sc">~</span> <span class="st">&#39;2A-peptide&#39;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb568-5"><a href="complete-protocols.html#cb568-5" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Order =</span> <span class="fu">case_when</span>(Order <span class="sc">==</span> <span class="st">&#39;C&#39;</span> <span class="sc">~</span> <span class="st">&quot;CFP first&quot;</span>,</span>
<span id="cb568-6"><a href="complete-protocols.html#cb568-6" tabindex="-1"></a>                           Order <span class="sc">==</span> <span class="st">&#39;Y&#39;</span> <span class="sc">~</span> <span class="st">&quot;YFP first&quot;</span>))</span>
<span id="cb568-7"><a href="complete-protocols.html#cb568-7" tabindex="-1"></a></span>
<span id="cb568-8"><a href="complete-protocols.html#cb568-8" tabindex="-1"></a><span class="fu">head</span>(df)</span></code></pre></div>
<pre><code># A tibble: 6 × 4
  System           Order       YFP   CFP
  &lt;chr&gt;            &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt;
1 &quot; Dual-promotor&quot; CFP first  88.8  114.
2 &quot; Dual-promotor&quot; CFP first  95.1  119.
3 &quot; Dual-promotor&quot; CFP first  99.6  208.
4 &quot; Dual-promotor&quot; CFP first 106.   275.
5 &quot; Dual-promotor&quot; CFP first 124.   163.
6 &quot; Dual-promotor&quot; CFP first 142.   225.</code></pre>
<div class="sourceCode" id="cb570"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb570-1"><a href="complete-protocols.html#cb570-1" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>YFP, <span class="at">y=</span>CFP)) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha=</span><span class="fl">0.5</span>) <span class="sc">+</span> <span class="fu">facet_grid</span>(Order<span class="sc">~</span>System)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-879-1.png" width="672" /></p>
<p>This looks a lot like the original figure in the paper (figure 1: <a href="https://doi.org/10.1371/journal.pone.0027321" class="uri">https://doi.org/10.1371/journal.pone.0027321</a>), but we still need to add the fits and optimize the layout. So let’s first assign the plot to an object and optimize the theme:</p>
<div class="sourceCode" id="cb571"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb571-1"><a href="complete-protocols.html#cb571-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>YFP, <span class="at">y=</span>CFP)) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha=</span><span class="fl">0.5</span>) <span class="sc">+</span> <span class="fu">facet_grid</span>(Order<span class="sc">~</span>System)</span>
<span id="cb571-2"><a href="complete-protocols.html#cb571-2" tabindex="-1"></a></span>
<span id="cb571-3"><a href="complete-protocols.html#cb571-3" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">formula =</span> y<span class="sc">~</span>x<span class="sc">+</span><span class="dv">0</span>)</span>
<span id="cb571-4"><a href="complete-protocols.html#cb571-4" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_minimal</span>(<span class="dv">14</span>)</span>
<span id="cb571-5"><a href="complete-protocols.html#cb571-5" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-880-1.png" width="672" /></p>
<p>Again, I’m not the biggest fan of grids, so I renmove the grids here. To indicate the area of the individual plots, I add a light blueish background with a thin grey border and the ‘strips’ with labels for the facets are a bit darker shade of blue:</p>
<div class="sourceCode" id="cb572"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb572-1"><a href="complete-protocols.html#cb572-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb572-2"><a href="complete-protocols.html#cb572-2" tabindex="-1"></a>               <span class="at">panel.grid =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb572-3"><a href="complete-protocols.html#cb572-3" tabindex="-1"></a>               <span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="st">&#39;#E0E0FF&#39;</span>, <span class="at">color=</span><span class="st">&quot;grey90&quot;</span>, <span class="at">size =</span> .<span class="dv">5</span>),</span>
<span id="cb572-4"><a href="complete-protocols.html#cb572-4" tabindex="-1"></a>               <span class="at">panel.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="st">&#39;#F4F4FF&#39;</span>, <span class="at">color=</span><span class="st">&quot;grey90&quot;</span>),</span>
<span id="cb572-5"><a href="complete-protocols.html#cb572-5" tabindex="-1"></a>               <span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color=</span><span class="st">&#39;grey80&#39;</span>, <span class="at">hjust=</span><span class="dv">1</span>),</span>
<span id="cb572-6"><a href="complete-protocols.html#cb572-6" tabindex="-1"></a>               <span class="cn">NULL</span>)</span>
<span id="cb572-7"><a href="complete-protocols.html#cb572-7" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-881-1.png" width="672" />
I’m happy with the layout, so it’s time to adjust the labels. The units for the intensity are arbitrary, so we indicate this with a.u. for arbitrary units:</p>
<div class="sourceCode" id="cb573"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb573-1"><a href="complete-protocols.html#cb573-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&quot;yellow fluorescence [a.u.]&quot;</span>,</span>
<span id="cb573-2"><a href="complete-protocols.html#cb573-2" tabindex="-1"></a>              <span class="at">y=</span><span class="st">&quot;cyan fluorescence [a.u.]&quot;</span>,</span>
<span id="cb573-3"><a href="complete-protocols.html#cb573-3" tabindex="-1"></a>              <span class="at">title =</span> <span class="st">&quot;Correlations for co-expression systems&quot;</span>,</span>
<span id="cb573-4"><a href="complete-protocols.html#cb573-4" tabindex="-1"></a>              <span class="at">caption =</span> <span class="st">&quot;</span><span class="sc">\n</span><span class="st">@joachimgoedhart</span><span class="sc">\n</span><span class="st">based on data from Goedhart et al., DOI: 10.1371/journal.pone.0027321&quot;</span>,</span>
<span id="cb573-5"><a href="complete-protocols.html#cb573-5" tabindex="-1"></a>              <span class="at">tag =</span> <span class="st">&quot;Protocol 16&quot;</span></span>
<span id="cb573-6"><a href="complete-protocols.html#cb573-6" tabindex="-1"></a>              )</span>
<span id="cb573-7"><a href="complete-protocols.html#cb573-7" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-882-1.png" width="672" /></p>
</div>
<div id="fit-parameters" class="section level3 hasAnchor" number="4.16.3">
<h3><span class="header-section-number">4.16.3</span> Fit parameters<a href="complete-protocols.html#fit-parameters" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This looks great, but we do not have the numbers for the correlations yet. So let’s calculate those and add them to a new dataframe ‘df_cor’:</p>
<div class="sourceCode" id="cb574"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb574-1"><a href="complete-protocols.html#cb574-1" tabindex="-1"></a>df_cor <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> </span>
<span id="cb574-2"><a href="complete-protocols.html#cb574-2" tabindex="-1"></a>  <span class="fu">group_by</span>(System, Order) <span class="sc">%&gt;%</span> <span class="fu">drop_na</span>() <span class="sc">%&gt;%</span></span>
<span id="cb574-3"><a href="complete-protocols.html#cb574-3" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">n=</span><span class="fu">n</span>(), <span class="at">R_squared =</span> <span class="fu">cor</span>(CFP, YFP, <span class="at">method =</span> <span class="st">&quot;pearson&quot;</span>)<span class="sc">^</span><span class="dv">2</span>)</span></code></pre></div>
<pre><code>`summarise()` has grouped output by &#39;System&#39;. You can override using the
`.groups` argument.</code></pre>
<div class="sourceCode" id="cb576"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb576-1"><a href="complete-protocols.html#cb576-1" tabindex="-1"></a>df_cor</span></code></pre></div>
<pre><code># A tibble: 6 × 4
# Groups:   System [3]
  System           Order         n R_squared
  &lt;chr&gt;            &lt;chr&gt;     &lt;int&gt;     &lt;dbl&gt;
1 &quot; Dual-promotor&quot; CFP first   734     0.483
2 &quot; Dual-promotor&quot; YFP first   338     0.613
3 &quot; IRES&quot;          CFP first   294     0.873
4 &quot; IRES&quot;          YFP first   497     0.785
5 &quot;2A-peptide&quot;     CFP first   629     0.944
6 &quot;2A-peptide&quot;     YFP first   840     0.968</code></pre>
<p>We can use the dataframe with correlation values to add these to the plot. We use <code>geom_text()</code> with the ‘df_cor’ dataframe. The main challenges are (i) to position the text at the correct place and (ii) to add a nice label to indicate that the value is R squared.
Positioning of the label is defined by the x- and y-coordinate in combination with hjust and vjust. My approach to get the label at the correct place is usually trial and error. We can use ‘x=Inf’, to assign an infinite value to x, implying that it will be located at the extreme right of the plot. To position the label on the left side of the x-coordinate, we can use ‘hjust=1’. But to give it a bit more space, I prefer here ‘hjust=1.1’.
Similarly, we use 0 for y-coordinate and ‘vjust=0’ to define that the middle position of the label is at this coordinate (‘vjust=1’ moves the label down and ‘vjust=-1’ moves the label up.).</p>
<p>To add a styled R-squared to the value we use the <code>paste()</code> function to add text preceding the actual value. We will round down the value to two digits using <code>round()</code>. Finally we can set ‘parse = TRUE’ to make sure that the label is styled according to the <code>plotmath()</code> convention. Here that means that the caret produces superscript and that the double equal sign is reduced to a single equal sign:</p>
<div class="sourceCode" id="cb578"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb578-1"><a href="complete-protocols.html#cb578-1" tabindex="-1"></a><span class="co"># Code to add r-squared to the plot</span></span>
<span id="cb578-2"><a href="complete-protocols.html#cb578-2" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="at">data=</span>df_cor, </span>
<span id="cb578-3"><a href="complete-protocols.html#cb578-3" tabindex="-1"></a>                   <span class="at">x=</span><span class="cn">Inf</span>,</span>
<span id="cb578-4"><a href="complete-protocols.html#cb578-4" tabindex="-1"></a>                   <span class="at">y=</span><span class="dv">0</span>,</span>
<span id="cb578-5"><a href="complete-protocols.html#cb578-5" tabindex="-1"></a>                   <span class="at">hjust=</span><span class="fl">1.1</span>,</span>
<span id="cb578-6"><a href="complete-protocols.html#cb578-6" tabindex="-1"></a>                   <span class="at">vjust=</span><span class="dv">0</span>,</span>
<span id="cb578-7"><a href="complete-protocols.html#cb578-7" tabindex="-1"></a>                   <span class="fu">aes</span>(<span class="at">label=</span><span class="fu">paste0</span>(<span class="st">&#39;R^2 == &#39;</span>,<span class="fu">round</span>(R_squared,<span class="dv">2</span>))),</span>
<span id="cb578-8"><a href="complete-protocols.html#cb578-8" tabindex="-1"></a>                   <span class="at">parse =</span> <span class="cn">TRUE</span>,</span>
<span id="cb578-9"><a href="complete-protocols.html#cb578-9" tabindex="-1"></a>                   <span class="at">color=</span><span class="st">&#39;darkslateblue&#39;</span></span>
<span id="cb578-10"><a href="complete-protocols.html#cb578-10" tabindex="-1"></a>                   )</span>
<span id="cb578-11"><a href="complete-protocols.html#cb578-11" tabindex="-1"></a></span>
<span id="cb578-12"><a href="complete-protocols.html#cb578-12" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-884-1.png" width="672" /></p>
<p>Finally, we can save the plot:</p>
<div class="sourceCode" id="cb579"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb579-1"><a href="complete-protocols.html#cb579-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_16.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb579-2"><a href="complete-protocols.html#cb579-2" tabindex="-1"></a> p</span>
<span id="cb579-3"><a href="complete-protocols.html#cb579-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
</div>
<div id="protocol-17" class="section level2 hasAnchor" number="4.17">
<h2><span class="header-section-number">4.17</span> Protocol 17 - Animated plots of data from optogenetics experiments<a href="complete-protocols.html#protocol-17" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To study dynamic processes, data is measured at multiple points in time. This type of data is often presented in lineplots (see for instance protocols 3 &amp; 7). To better visualize the dynamics, animated versions of lineplots can be made. These are nice for presentations or websites. This protocol explains how to generate an animated lineplot from timeseries data. The data is originally published by <a href="https://www.biorxiv.org/content/10.1101/2022.10.17.512253v1">Mahlandt et al (2022)</a>.</p>
<p>Load the required packages:</p>
<div class="sourceCode" id="cb581"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb581-1"><a href="complete-protocols.html#cb581-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb581-2"><a href="complete-protocols.html#cb581-2" tabindex="-1"></a><span class="fu">library</span>(gganimate)</span></code></pre></div>
<p>First, we load the data from an excel file:</p>
<div class="sourceCode" id="cb582"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb582-1"><a href="complete-protocols.html#cb582-1" tabindex="-1"></a>df <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">&#39;data/210222_OptoG08_02_Sheldon_TIAM_tidy_02.xlsx&#39;</span>)</span>
<span id="cb582-2"><a href="complete-protocols.html#cb582-2" tabindex="-1"></a><span class="fu">head</span>(df)</span></code></pre></div>
<pre><code># A tibble: 6 × 6
  unique_id    id          Time Sample Value type   
  &lt;chr&gt;        &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;  
1 Sheldon_A1_R Sheldon 0.000355 A1_R    836. Control
2 Sheldon_A1_R Sheldon 0.00761  A1_R    836. Control
3 Sheldon_A1_R Sheldon 0.0104   A1_R    835. Control
4 Sheldon_A1_R Sheldon 0.0132   A1_R    834. Control
5 Sheldon_A1_R Sheldon 0.0159   A1_R    834. Control
6 Sheldon_A1_R Sheldon 0.0187   A1_R    833. Control</code></pre>
<p>The entire dataset is quite large and to simplify the presentation, we select a timerange between 17.75 and 18.8378 (this is the time in hours). In this timerange the dynamics can be nicely observed:</p>
<div class="sourceCode" id="cb584"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb584-1"><a href="complete-protocols.html#cb584-1" tabindex="-1"></a>df_sub <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Time <span class="sc">&gt;=</span> <span class="fl">17.75</span> <span class="sc">&amp;</span> Time <span class="sc">&lt;=</span> <span class="fl">18.8378</span>)</span></code></pre></div>
<p>There are two factors in the column ‘type’, namely Control and OptoTiam. We calculate the summary of the measurements and labele the new column ‘Value’:</p>
<div class="sourceCode" id="cb585"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb585-1"><a href="complete-protocols.html#cb585-1" tabindex="-1"></a>df_sub_summary <span class="ot">&lt;-</span> df_sub <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(type, Time) <span class="sc">%&gt;%</span> <span class="fu">summarize</span>(<span class="at">mean=</span><span class="fu">mean</span>(Value))</span></code></pre></div>
<pre><code>`summarise()` has grouped output by &#39;type&#39;. You can override using the
`.groups` argument.</code></pre>
<p>Let’s make a first plot to visualize the data:</p>
<div class="sourceCode" id="cb587"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb587-1"><a href="complete-protocols.html#cb587-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_sub_summary, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>mean, <span class="at">color =</span> type)) <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="at">linewidth=</span><span class="dv">2</span>) </span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-914-1.png" width="672" /></p>
<p>This looks good, so let’s define an object that contains the plot:</p>
<div class="sourceCode" id="cb588"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb588-1"><a href="complete-protocols.html#cb588-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_sub_summary, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>mean, <span class="at">color =</span> type)) <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="at">linewidth=</span><span class="dv">2</span>) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">fill=</span>type), <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size=</span><span class="dv">4</span>, <span class="at">shape=</span><span class="dv">21</span>)</span></code></pre></div>
<p>Note that we add a dot as well, by ‘geom_point()’. This will looks messy in the static plot, but it is a nice addition for the dynamic plot.</p>
<div class="sourceCode" id="cb589"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb589-1"><a href="complete-protocols.html#cb589-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_light</span>(<span class="at">base_size=</span><span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb589-2"><a href="complete-protocols.html#cb589-2" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb589-3"><a href="complete-protocols.html#cb589-3" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">&quot;none&quot;</span>)</span>
<span id="cb589-4"><a href="complete-protocols.html#cb589-4" tabindex="-1"></a></span>
<span id="cb589-5"><a href="complete-protocols.html#cb589-5" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-916-1.png" width="672" /></p>
<p>In this experiment, cells are exposed to blue light (optogenetics) and this triggers a change in the measured value. So below, we define two lightblue rectangles that indicate the illumination with blue light:</p>
<div class="sourceCode" id="cb590"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb590-1"><a href="complete-protocols.html#cb590-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="fl">17.8333</span>, <span class="at">xmax =</span> <span class="fl">18.0833</span> , <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymax =</span> <span class="cn">Inf</span>, <span class="at">fill=</span><span class="st">&quot;blue&quot;</span>, <span class="at">alpha=</span><span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb590-2"><a href="complete-protocols.html#cb590-2" tabindex="-1"></a>   <span class="fu">annotate</span>(<span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="fl">18.3333</span>, <span class="at">xmax =</span> <span class="fl">18.5833</span> , <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymax =</span> <span class="cn">Inf</span>, <span class="at">fill=</span><span class="st">&quot;blue&quot;</span>, <span class="at">alpha=</span><span class="fl">0.1</span>)</span>
<span id="cb590-3"><a href="complete-protocols.html#cb590-3" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-917-1.png" width="672" /></p>
<p>Now, we change the colors of the two lines and dots:</p>
<div class="sourceCode" id="cb591"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb591-1"><a href="complete-protocols.html#cb591-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;#BEBEBE&quot;</span>, <span class="st">&quot;darkorchid&quot;</span>)) <span class="sc">+</span></span>
<span id="cb591-2"><a href="complete-protocols.html#cb591-2" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;#BEBEBE&quot;</span>, <span class="st">&quot;darkorchid&quot;</span>))</span>
<span id="cb591-3"><a href="complete-protocols.html#cb591-3" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-918-1.png" width="672" /></p>
<p>Next, we can add labels and titles:</p>
<div class="sourceCode" id="cb592"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb592-1"><a href="complete-protocols.html#cb592-1" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb592-2"><a href="complete-protocols.html#cb592-2" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb592-3"><a href="complete-protocols.html#cb592-3" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Reversible increase of resistance in endothelial cells by optogenetics&quot;</span>,</span>
<span id="cb592-4"><a href="complete-protocols.html#cb592-4" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Time [h]&quot;</span>,</span>
<span id="cb592-5"><a href="complete-protocols.html#cb592-5" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Resistance [Ω]&quot;</span>,</span>
<span id="cb592-6"><a href="complete-protocols.html#cb592-6" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">(based on data from Mahlandt et al, DOI: https://doi.org/10.1101/2022.10.17.512253)&quot;</span>,</span>
<span id="cb592-7"><a href="complete-protocols.html#cb592-7" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 17&quot;</span></span>
<span id="cb592-8"><a href="complete-protocols.html#cb592-8" tabindex="-1"></a>  ) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color=</span><span class="st">&#39;grey80&#39;</span>, <span class="at">hjust=</span><span class="dv">1</span>))</span></code></pre></div>
<p>This is the version that I save as a preview of the animation:</p>
<div class="sourceCode" id="cb593"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb593-1"><a href="complete-protocols.html#cb593-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_17.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb593-2"><a href="complete-protocols.html#cb593-2" tabindex="-1"></a> p</span>
<span id="cb593-3"><a href="complete-protocols.html#cb593-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
<p>For the animated version, we can add a label for each of the two conditions. Note that (similar to the dots) it is plotted for each datapoint. But, in the animated version, only the leading (newly drawn) datapoint and label is shown.</p>
<div class="sourceCode" id="cb595"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb595-1"><a href="complete-protocols.html#cb595-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_label</span>(<span class="fu">aes</span>(<span class="at">x =</span> Time, <span class="at">y=</span>mean, <span class="at">label=</span>type), <span class="at">nudge_x =</span> <span class="fl">0.02</span>, <span class="at">size=</span><span class="dv">4</span>, <span class="at">hjust=</span><span class="dv">0</span>)</span></code></pre></div>
<p>Finally, we can turn this into an animated object with the function <code>transition_reveal()</code> from the <code>{gganimate}</code> package:</p>
<div class="sourceCode" id="cb596"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb596-1"><a href="complete-protocols.html#cb596-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">transition_reveal</span>(Time)</span></code></pre></div>
<p>To generate an animated plot, we used <code>animate()</code> and define the number of frames:</p>
<div class="sourceCode" id="cb597"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb597-1"><a href="complete-protocols.html#cb597-1" tabindex="-1"></a>animated_plot <span class="ot">&lt;-</span> <span class="fu">animate</span>(a, <span class="at">nframes=</span><span class="dv">100</span>, <span class="at">width =</span> <span class="dv">1000</span>, <span class="at">height =</span> <span class="dv">750</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">100</span>)</span></code></pre></div>
<p>The result can be displayed by typing <code>animated_plot</code> in the command line:</p>
<p><img src="https://raw.githubusercontent.com/JoachimGoedhart/DataViz-protocols/main/Protocols/Protocol_17.gif" /></p>
<p>To save the animated object <code>a</code> to disk, use:</p>
<div class="sourceCode" id="cb598"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb598-1"><a href="complete-protocols.html#cb598-1" tabindex="-1"></a><span class="fu">animate</span>(a, <span class="dv">70</span>, <span class="at">renderer =</span> <span class="fu">gifski_renderer</span>(<span class="st">&quot;Protocol_17.gif&quot;</span>), <span class="at">width =</span> <span class="dv">1000</span>, <span class="at">height =</span> <span class="dv">750</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">100</span>)</span></code></pre></div>
</div>
<div id="protocol-18" class="section level2 hasAnchor" number="4.18">
<h2><span class="header-section-number">4.18</span> Protocol 18 - Combining animated plots with a movie<a href="complete-protocols.html#protocol-18" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In <a href="complete-protocols.html#protocol-17">protocol 17</a> an animated plot is generated from time series data. Here, we will see how an animated plot can be combined with a movie from imaging data. The data is originally published by <a href="https://www.biorxiv.org/content/10.1101/2022.10.17.512253v1">Mahlandt et al (2022)</a>. We will plot the cell area over time and the movie that shows the area of individual cells.</p>
<p>Load the required packages:</p>
<div class="sourceCode" id="cb599"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb599-1"><a href="complete-protocols.html#cb599-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb599-2"><a href="complete-protocols.html#cb599-2" tabindex="-1"></a><span class="fu">library</span>(gganimate)</span></code></pre></div>
<p>Read the excel file with the data:</p>
<div class="sourceCode" id="cb600"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb600-1"><a href="complete-protocols.html#cb600-1" tabindex="-1"></a>df <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">&#39;data/Summary_GlobalActivation_TimeTrace_TIAM_01.xlsx&#39;</span>)</span>
<span id="cb600-2"><a href="complete-protocols.html#cb600-2" tabindex="-1"></a><span class="fu">head</span>(df)</span></code></pre></div>
<pre><code># A tibble: 6 × 45
  Experiment          Construct Cell_Experiment Cell   Time Frame AreaShape_Area
  &lt;chr&gt;               &lt;chr&gt;     &lt;chr&gt;           &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;          &lt;dbl&gt;
1 210429_OptoG011_03… OptoTIAM  Cell_01         Cell…    15     1          32774
2 210429_OptoG011_03… OptoTIAM  Cell_01         Cell…    30     2          32514
3 210429_OptoG011_03… OptoTIAM  Cell_01         Cell…    45     3          32371
4 210429_OptoG011_03… OptoTIAM  Cell_01         Cell…    60     4          32082
5 210429_OptoG011_03… OptoTIAM  Cell_01         Cell…    75     5          31799
6 210429_OptoG011_03… OptoTIAM  Cell_01         Cell…    90     6          31296
# ℹ 38 more variables: AreaShape_BoundingBoxArea &lt;dbl&gt;,
#   AreaShape_BoundingBoxMaximum_X &lt;dbl&gt;, AreaShape_BoundingBoxMaximum_Y &lt;dbl&gt;,
#   AreaShape_BoundingBoxMinimum_X &lt;dbl&gt;, AreaShape_BoundingBoxMinimum_Y &lt;dbl&gt;,
#   AreaShape_Center_X &lt;dbl&gt;, AreaShape_Center_Y &lt;dbl&gt;,
#   AreaShape_Compactness &lt;dbl&gt;, AreaShape_Eccentricity &lt;dbl&gt;,
#   AreaShape_EquivalentDiameter &lt;dbl&gt;, AreaShape_EulerNumber &lt;dbl&gt;,
#   AreaShape_Extent &lt;dbl&gt;, AreaShape_FormFactor &lt;dbl&gt;, …</code></pre>
<p>The first timepoint is 15 and to set that to zero:</p>
<div class="sourceCode" id="cb602"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb602-1"><a href="complete-protocols.html#cb602-1" tabindex="-1"></a>df <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Time=</span>Time<span class="dv">-15</span>)</span></code></pre></div>
<p>The dataframe ‘df’ has data for a number of cells, time and the area with the column names ‘Cell’, ‘Time’ and ‘AreaShape_Area’. To compare the area of the cells at the start of the experiment, we normalize the ‘AreaShape_Area’ to the initial value dividing all values of a single cell trace by the value of the first timepoint:</p>
<div class="sourceCode" id="cb603"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb603-1"><a href="complete-protocols.html#cb603-1" tabindex="-1"></a>df <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Cell) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(Time) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Area_norm=</span>AreaShape_Area<span class="sc">/</span>AreaShape_Area[<span class="dv">1</span>]) <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>()</span></code></pre></div>
<p>Next, we calculate the average response and store this in a new dataframe:</p>
<div class="sourceCode" id="cb604"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb604-1"><a href="complete-protocols.html#cb604-1" tabindex="-1"></a>df_summary <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Construct, Time) <span class="sc">%&gt;%</span> <span class="fu">summarize</span>(<span class="at">mean=</span><span class="fu">mean</span>(Area_norm))</span></code></pre></div>
<pre><code>`summarise()` has grouped output by &#39;Construct&#39;. You can override using the
`.groups` argument.</code></pre>
<p>Let’s make a plot of the average:</p>
<div class="sourceCode" id="cb606"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb606-1"><a href="complete-protocols.html#cb606-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_summary, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>mean)) <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="at">size=</span><span class="dv">2</span>, <span class="at">color=</span><span class="st">&#39;darkorchid4&#39;</span>) </span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-947-1.png" width="672" /></p>
<p>This is an optogenetic experiment, where the light is turned on between timepoints 135 and 720, and we can define a new column ‘light’ that stores this information:</p>
<div class="sourceCode" id="cb607"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb607-1"><a href="complete-protocols.html#cb607-1" tabindex="-1"></a>df_summary <span class="ot">&lt;-</span> df_summary <span class="sc">%&gt;%</span></span>
<span id="cb607-2"><a href="complete-protocols.html#cb607-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">light =</span> <span class="fu">case_when</span>(Time <span class="sc">&gt;=</span> <span class="dv">135</span> <span class="sc">&amp;</span> Time <span class="sc">&lt;=</span> <span class="dv">720</span>  <span class="sc">~</span> <span class="st">&quot;ON&quot;</span>,</span>
<span id="cb607-3"><a href="complete-protocols.html#cb607-3" tabindex="-1"></a>                           <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">&quot;OFF&quot;</span></span>
<span id="cb607-4"><a href="complete-protocols.html#cb607-4" tabindex="-1"></a>                           )</span>
<span id="cb607-5"><a href="complete-protocols.html#cb607-5" tabindex="-1"></a>  )</span></code></pre></div>
<p>This information on the light condition needs to be plotted ‘behind’ the data, and so we define it as the first layer:</p>
<div class="sourceCode" id="cb608"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb608-1"><a href="complete-protocols.html#cb608-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_summary, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>mean)) <span class="sc">+</span></span>
<span id="cb608-2"><a href="complete-protocols.html#cb608-2" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="at">data=</span>df_summary,</span>
<span id="cb608-3"><a href="complete-protocols.html#cb608-3" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span><span class="dv">1</span>,</span>
<span id="cb608-4"><a href="complete-protocols.html#cb608-4" tabindex="-1"></a>                <span class="at">height =</span> <span class="cn">Inf</span>,</span>
<span id="cb608-5"><a href="complete-protocols.html#cb608-5" tabindex="-1"></a>                <span class="at">width =</span> <span class="dv">15</span>,</span>
<span id="cb608-6"><a href="complete-protocols.html#cb608-6" tabindex="-1"></a>                <span class="at">fill =</span> light,</span>
<span id="cb608-7"><a href="complete-protocols.html#cb608-7" tabindex="-1"></a>                <span class="at">group =</span> <span class="fu">seq_along</span>(Time)),</span>
<span id="cb608-8"><a href="complete-protocols.html#cb608-8" tabindex="-1"></a>                <span class="at">alpha=</span><span class="dv">1</span>)</span>
<span id="cb608-9"><a href="complete-protocols.html#cb608-9" tabindex="-1"></a></span>
<span id="cb608-10"><a href="complete-protocols.html#cb608-10" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-949-1.png" width="672" /></p>
<p>Note that we use here <code>group = seq_along(Time))</code>. This will be explained below, when the animated object is generated.</p>
<p>The blocks in the plot are indicating the different light conditions. The colors can be changed, so that a grey color is displayed when the light is off, and a cyan rectangle shows when the light was switched on:</p>
<div class="sourceCode" id="cb609"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb609-1"><a href="complete-protocols.html#cb609-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;cyan&quot;</span>))</span>
<span id="cb609-2"><a href="complete-protocols.html#cb609-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-950-1.png" width="672" /></p>
<p>Now, let’s add the data for both the mean and the individual cells:</p>
<div class="sourceCode" id="cb610"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb610-1"><a href="complete-protocols.html#cb610-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="at">size=</span><span class="dv">2</span>, <span class="at">color=</span><span class="st">&#39;darkorchid4&#39;</span>)  <span class="sc">+</span></span>
<span id="cb610-2"><a href="complete-protocols.html#cb610-2" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">data=</span>df, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>Area_norm, <span class="at">group=</span>Cell), <span class="at">alpha=</span><span class="fl">0.3</span>, <span class="at">color=</span><span class="st">&#39;darkorchid4&#39;</span>)</span>
<span id="cb610-3"><a href="complete-protocols.html#cb610-3" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-951-1.png" width="672" /></p>
<p>Some styling:</p>
<div class="sourceCode" id="cb611"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb611-1"><a href="complete-protocols.html#cb611-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_light</span>(<span class="at">base_size=</span><span class="dv">18</span>) <span class="sc">+</span></span>
<span id="cb611-2"><a href="complete-protocols.html#cb611-2" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb611-3"><a href="complete-protocols.html#cb611-3" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">&quot;top&quot;</span>,</span>
<span id="cb611-4"><a href="complete-protocols.html#cb611-4" tabindex="-1"></a>        <span class="at">legend.justification =</span> <span class="st">&quot;left&quot;</span>,</span>
<span id="cb611-5"><a href="complete-protocols.html#cb611-5" tabindex="-1"></a>        <span class="at">legend.box.margin =</span> <span class="fu">margin</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="sc">-</span><span class="dv">40</span>),</span>
<span id="cb611-6"><a href="complete-protocols.html#cb611-6" tabindex="-1"></a>        <span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color=</span><span class="st">&#39;grey80&#39;</span>, <span class="at">hjust=</span><span class="dv">1</span>))</span>
<span id="cb611-7"><a href="complete-protocols.html#cb611-7" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-952-1.png" width="672" /></p>
<p>To add a dot that is like the tip of a pen, drawing the line, we can use geom_point(). Here it is only added for the average curve:</p>
<div class="sourceCode" id="cb612"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb612-1"><a href="complete-protocols.html#cb612-1" tabindex="-1"></a>  p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">color=</span><span class="st">&#39;darkorchid4&#39;</span>, <span class="at">size=</span><span class="dv">4</span>)</span>
<span id="cb612-2"><a href="complete-protocols.html#cb612-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-953-1.png" width="672" /></p>
<p>In the plot, the dot is shown for each timepoint, but in the animation it will only be drawn for the active frame. This is unlike the line, which remains visible. More detail about this behavior can be found below.</p>
<p>Finally, let’s change the labels for the axes and set the scale. We use <code>expand = FALSE</code> to use the exact limits that we have specified:</p>
<div class="sourceCode" id="cb613"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb613-1"><a href="complete-protocols.html#cb613-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&#39;Time [s]&#39;</span>,</span>
<span id="cb613-2"><a href="complete-protocols.html#cb613-2" tabindex="-1"></a>              <span class="at">y=</span><span class="st">&#39;Area [µm]&#39;</span>,</span>
<span id="cb613-3"><a href="complete-protocols.html#cb613-3" tabindex="-1"></a>              <span class="at">caption=</span><span class="st">&#39;</span><span class="sc">\n</span><span class="st">@joachimgoedhart</span><span class="sc">\n</span><span class="st">based on data from Mahlandt et al., DOI: 10.1101/2022.10.17.512253&#39;</span>) <span class="sc">+</span></span>
<span id="cb613-4"><a href="complete-protocols.html#cb613-4" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1305</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="fl">0.8</span>,<span class="fl">1.45</span>), <span class="at">expand =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>For the animation, we need to know the number of timepoints, as we will use this number to generate the same amount of ‘frames’ in the animation:</p>
<div class="sourceCode" id="cb614"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb614-1"><a href="complete-protocols.html#cb614-1" tabindex="-1"></a><span class="fu">length</span>(df_summary<span class="sc">$</span>Time)</span></code></pre></div>
<pre><code>[1] 88</code></pre>
<p>So the number of frames should be 88. We use to plot object <code>p</code> as the input for the animation and generate the animation with the function <code>transition_reveal()</code>, which will reveal the data over time. The result is stored in the object <code>a</code>:</p>
<div class="sourceCode" id="cb616"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb616-1"><a href="complete-protocols.html#cb616-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">transition_reveal</span>(Time)</span></code></pre></div>
<p>Now, we can render the animation with a defined number of frames and size. Since we will display the movie and plot side-by-side, the height of both panels should be identical. The movie has a height of 900 pixels, so we use this to set the dimension of the animated plot:</p>
<div class="sourceCode" id="cb617"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb617-1"><a href="complete-protocols.html#cb617-1" tabindex="-1"></a>animated_plot <span class="ot">&lt;-</span> <span class="fu">animate</span>(<span class="at">plot =</span> a,</span>
<span id="cb617-2"><a href="complete-protocols.html#cb617-2" tabindex="-1"></a>                         <span class="at">nframes =</span> <span class="dv">88</span>,</span>
<span id="cb617-3"><a href="complete-protocols.html#cb617-3" tabindex="-1"></a>                         <span class="at">width =</span> <span class="dv">900</span>,</span>
<span id="cb617-4"><a href="complete-protocols.html#cb617-4" tabindex="-1"></a>                         <span class="at">height =</span> <span class="dv">900</span>,</span>
<span id="cb617-5"><a href="complete-protocols.html#cb617-5" tabindex="-1"></a>                         <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">100</span>,</span>
<span id="cb617-6"><a href="complete-protocols.html#cb617-6" tabindex="-1"></a>                         <span class="at">renderer =</span> <span class="fu">magick_renderer</span>())</span></code></pre></div>
<pre><code>`geom_line()`: Each group consists of only one observation.
ℹ Do you need to adjust the group aesthetic?
`geom_line()`: Each group consists of only one observation.
ℹ Do you need to adjust the group aesthetic?
`geom_line()`: Each group consists of only one observation.
ℹ Do you need to adjust the group aesthetic?
`geom_line()`: Each group consists of only one observation.
ℹ Do you need to adjust the group aesthetic?</code></pre>
<p>The result is <code>animated_plot</code> which is a “magick-image”:</p>
<div class="sourceCode" id="cb619"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb619-1"><a href="complete-protocols.html#cb619-1" tabindex="-1"></a><span class="fu">class</span>(animated_plot)</span></code></pre></div>
<pre><code>[1] &quot;magick-image&quot;</code></pre>
<p>This will allow us to combine it with a GIF using the <code>{magick}</code> package. So it is important to define the ‘renderer’ in the <code>animate()</code>function.</p>
<div id="intermezzo" class="section level3 hasAnchor" number="4.18.1">
<h3><span class="header-section-number">4.18.1</span> Intermezzo<a href="complete-protocols.html#intermezzo" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>A few words on the animated objects. Using the function <code>transition_reveal()</code>, the lines are revealed over time. Objects like the dot and the mean are only displayed at the location that is defined by a specific time and do not stay. A similar behaviour would be the default for the <code>geom_tile()</code>. However, here we want the area to be drawn, which means that the area that was drawn before should remain visible. To achieve that, we added <code>group = seq_along(Time)</code> to <code>geom_tile()</code>. The same could be done for the <code>geom_point()</code> when the points should remain visible for the entire trace.</p>
</div>
<div id="time-for-some-magick" class="section level3 hasAnchor" number="4.18.2">
<h3><span class="header-section-number">4.18.2</span> Time for some magick<a href="complete-protocols.html#time-for-some-magick" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now it’s time to load the movie. We will use the <code>{magick}</code> package to load the GIF and to combine them:</p>
<div class="sourceCode" id="cb621"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb621-1"><a href="complete-protocols.html#cb621-1" tabindex="-1"></a><span class="fu">library</span>(magick)</span></code></pre></div>
<pre><code>Linking to ImageMagick 6.9.12.93
Enabled features: cairo, fontconfig, freetype, heic, lcms, pango, raw, rsvg, webp
Disabled features: fftw, ghostscript, x11</code></pre>
<div class="sourceCode" id="cb623"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb623-1"><a href="complete-protocols.html#cb623-1" tabindex="-1"></a>movie <span class="ot">&lt;-</span> <span class="fu">image_read</span>(<span class="st">&quot;data/OptoTIAM_movie.gif&quot;</span>)</span></code></pre></div>
<p>To test whether we can combine the movie and plot, we start by combining the first frame and verify that the resulting image looks right:</p>
<div class="sourceCode" id="cb624"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb624-1"><a href="complete-protocols.html#cb624-1" tabindex="-1"></a>combined_gif <span class="ot">&lt;-</span> <span class="fu">image_append</span>(<span class="fu">c</span>(movie[<span class="dv">1</span>], animated_plot[<span class="dv">1</span>]))</span>
<span id="cb624-2"><a href="complete-protocols.html#cb624-2" tabindex="-1"></a></span>
<span id="cb624-3"><a href="complete-protocols.html#cb624-3" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">as.raster</span>(combined_gif))</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-960-1.png" width="672" /></p>
<p>To combine all other frames and add them to the object ‘combined_gif’ we use a for loop:</p>
<div class="sourceCode" id="cb625"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb625-1"><a href="complete-protocols.html#cb625-1" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">88</span>) {</span>
<span id="cb625-2"><a href="complete-protocols.html#cb625-2" tabindex="-1"></a>   combined_panel <span class="ot">&lt;-</span> <span class="fu">image_append</span>(<span class="fu">c</span>(movie[i], animated_plot[i]))</span>
<span id="cb625-3"><a href="complete-protocols.html#cb625-3" tabindex="-1"></a>   combined_gif <span class="ot">&lt;-</span> <span class="fu">c</span>(combined_gif, combined_panel)</span>
<span id="cb625-4"><a href="complete-protocols.html#cb625-4" tabindex="-1"></a>}</span></code></pre></div>
<p>To show the combined GIF:</p>
<div class="sourceCode" id="cb626"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb626-1"><a href="complete-protocols.html#cb626-1" tabindex="-1"></a>combined_gif</span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/JoachimGoedhart/DataViz-protocols/main/Protocols/Protocol_18.gif" /></p>
<p>To save a preview, we can selected one frame of the movie and save it:</p>
<div class="sourceCode" id="cb627"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb627-1"><a href="complete-protocols.html#cb627-1" tabindex="-1"></a>montage <span class="ot">&lt;-</span> <span class="fu">image_montage</span>(<span class="fu">c</span>(combined_gif[<span class="dv">22</span>],combined_gif[<span class="dv">44</span>],combined_gif[<span class="dv">66</span>],combined_gif[<span class="dv">88</span>]), <span class="at">geometry =</span> <span class="st">&quot;x400+0+100&quot;</span>, <span class="at">tile =</span> <span class="st">&#39;2x2&#39;</span>)</span>
<span id="cb627-2"><a href="complete-protocols.html#cb627-2" tabindex="-1"></a></span>
<span id="cb627-3"><a href="complete-protocols.html#cb627-3" tabindex="-1"></a><span class="fu">image_write_gif</span>(montage, <span class="st">&#39;Protocol_18.png&#39;</span>)</span></code></pre></div>
<p>To write the combined GIF to a file:</p>
<div class="sourceCode" id="cb628"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb628-1"><a href="complete-protocols.html#cb628-1" tabindex="-1"></a><span class="fu">image_write_gif</span>(combined_gif, <span class="st">&#39;Protocol_18.gif&#39;</span>)</span></code></pre></div>
</div>
</div>
<div id="protocol-19" class="section level2 hasAnchor" number="4.19">
<h2><span class="header-section-number">4.19</span> Protocol 19 - Plotting ratiometric FRET data<a href="complete-protocols.html#protocol-19" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this protocol, we plot the intensity data that are quantified by ImageJ/FIJI. The image analysis is originally reported in a preprint by <a href="https://doi.org/10.1101/2021.02.24.432684">Mahlandt and Goedhart (2021)</a> in section 3.2.2. The original R-script is available on <a href="https://github.com/JoachimGoedhart/TimeLapse-Imaging-DataViz/tree/main/3_2_2_RatioImaging">Github</a>. Here, all steps are explained in protocol style.</p>
<div class="sourceCode" id="cb629"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb629-1"><a href="complete-protocols.html#cb629-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>We can load the output from FIJI data from the Github repository:</p>
<div class="sourceCode" id="cb630"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb630-1"><a href="complete-protocols.html#cb630-1" tabindex="-1"></a>df_CFP <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;https://raw.githubusercontent.com/JoachimGoedhart/TimeLapse-Imaging-DataViz/main/3_2_2_RatioImaging/Results-CFP.csv&quot;</span>)</span>
<span id="cb630-2"><a href="complete-protocols.html#cb630-2" tabindex="-1"></a>df_YFP <span class="ot">&lt;-</span>  <span class="fu">read.csv</span>(<span class="st">&quot;https://raw.githubusercontent.com/JoachimGoedhart/TimeLapse-Imaging-DataViz/main/3_2_2_RatioImaging/Results-YFP.csv&quot;</span>)</span></code></pre></div>
<p>To avoid the repetition of actions on both dataframes, we merge them:</p>
<div class="sourceCode" id="cb631"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb631-1"><a href="complete-protocols.html#cb631-1" tabindex="-1"></a>df_ratio <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(df_CFP, df_YFP)</span>
<span id="cb631-2"><a href="complete-protocols.html#cb631-2" tabindex="-1"></a><span class="fu">head</span>(df_ratio)</span></code></pre></div>
<pre><code>  X                                               Label  Mean Slice
1 1 CFP:Cell-01:180228_YCaM3.60_His_Py-02_w1YCam-CFP_t1 12.31     1
2 2 CFP:Cell-02:180228_YCaM3.60_His_Py-02_w1YCam-CFP_t1 14.04     1
3 3 CFP:Cell-03:180228_YCaM3.60_His_Py-02_w1YCam-CFP_t1 16.08     1
4 4 CFP:Cell-04:180228_YCaM3.60_His_Py-02_w1YCam-CFP_t1 18.39     1
5 5 CFP:Cell-01:180228_YCaM3.60_His_Py-02_w1YCam-CFP_t2 12.23     2
6 6 CFP:Cell-02:180228_YCaM3.60_His_Py-02_w1YCam-CFP_t2 13.90     2</code></pre>
<p>And let’s look at the last rows of this dataframe</p>
<div class="sourceCode" id="cb633"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb633-1"><a href="complete-protocols.html#cb633-1" tabindex="-1"></a><span class="fu">tail</span>(df_ratio)</span></code></pre></div>
<pre><code>      X                                                 Label   Mean Slice
963 479 YFP:Cell-03:180228_YCaM3.60_His_Py-02_w2YCam-YFP_t120 161.79   120
964 480 YFP:Cell-04:180228_YCaM3.60_His_Py-02_w2YCam-YFP_t120 191.84   120
965 481 YFP:Cell-01:180228_YCaM3.60_His_Py-02_w2YCam-YFP_t121 127.41   121
966 482 YFP:Cell-02:180228_YCaM3.60_His_Py-02_w2YCam-YFP_t121 148.50   121
967 483 YFP:Cell-03:180228_YCaM3.60_His_Py-02_w2YCam-YFP_t121 161.95   121
968 484 YFP:Cell-04:180228_YCaM3.60_His_Py-02_w2YCam-YFP_t121 191.58   121</code></pre>
<p>This looks right and the format of the dataframe is in a long format. However, the first column contains multiple conditions, or IDs, separated by a colon, so we split those with <code>separate()</code>:</p>
<div class="sourceCode" id="cb635"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb635-1"><a href="complete-protocols.html#cb635-1" tabindex="-1"></a>df_split_ratio <span class="ot">&lt;-</span> df_ratio <span class="sc">%&gt;%</span> <span class="fu">separate</span>(Label,<span class="fu">c</span>(<span class="st">&quot;filename&quot;</span>, <span class="st">&quot;Sample&quot;</span>,<span class="st">&quot;Number&quot;</span>),<span class="at">sep=</span><span class="st">&#39;:&#39;</span>)</span>
<span id="cb635-2"><a href="complete-protocols.html#cb635-2" tabindex="-1"></a><span class="fu">str</span>(df_split_ratio)</span></code></pre></div>
<pre><code>&#39;data.frame&#39;:   968 obs. of  6 variables:
 $ X       : int  1 2 3 4 5 6 7 8 9 10 ...
 $ filename: chr  &quot;CFP&quot; &quot;CFP&quot; &quot;CFP&quot; &quot;CFP&quot; ...
 $ Sample  : chr  &quot;Cell-01&quot; &quot;Cell-02&quot; &quot;Cell-03&quot; &quot;Cell-04&quot; ...
 $ Number  : chr  &quot;180228_YCaM3.60_His_Py-02_w1YCam-CFP_t1&quot; &quot;180228_YCaM3.60_His_Py-02_w1YCam-CFP_t1&quot; &quot;180228_YCaM3.60_His_Py-02_w1YCam-CFP_t1&quot; &quot;180228_YCaM3.60_His_Py-02_w1YCam-CFP_t1&quot; ...
 $ Mean    : num  12.3 14 16.1 18.4 12.2 ...
 $ Slice   : int  1 1 1 1 2 2 2 2 3 3 ...</code></pre>
<p>We can have a first glance at the data:</p>
<div class="sourceCode" id="cb637"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb637-1"><a href="complete-protocols.html#cb637-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_split_ratio, <span class="fu">aes</span>(<span class="at">x=</span>Slice, <span class="at">y=</span>Mean, <span class="at">by=</span>Sample)) <span class="sc">+</span></span>
<span id="cb637-2"><a href="complete-protocols.html#cb637-2" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>filename)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-995-1.png" width="672" /></p>
<p>To get rid of cell-to-cell variation in expression levels, we normalize the data by dividing all data by the average value of the first 5 timepoints (a.k.a. the baseline):</p>
<div class="sourceCode" id="cb638"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb638-1"><a href="complete-protocols.html#cb638-1" tabindex="-1"></a>df_split_ratio <span class="ot">&lt;-</span> df_split_ratio <span class="sc">%&gt;%</span></span>
<span id="cb638-2"><a href="complete-protocols.html#cb638-2" tabindex="-1"></a>  <span class="fu">group_by</span>(Sample, filename) <span class="sc">%&gt;%</span></span>
<span id="cb638-3"><a href="complete-protocols.html#cb638-3" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Mean_norm =</span> Mean<span class="sc">/</span><span class="fu">mean</span>(Mean[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]))</span>
<span id="cb638-4"><a href="complete-protocols.html#cb638-4" tabindex="-1"></a></span>
<span id="cb638-5"><a href="complete-protocols.html#cb638-5" tabindex="-1"></a><span class="fu">ggplot</span>(df_split_ratio, <span class="fu">aes</span>(<span class="at">x=</span>Slice, <span class="at">y=</span>Mean_norm, <span class="at">by=</span>Sample)) <span class="sc">+</span></span>
<span id="cb638-6"><a href="complete-protocols.html#cb638-6" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>filename)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-996-1.png" width="672" /></p>
<p>This looks good. To display the FRET ratio the data from the YFP channel are divided by that of the CFP channel. To achieve that, I will first make the data wider with <code>pivot_wider()</code> by defining two columns (this is close to what’s done in protocol 16). One with the YFP intensities and one with the CFP intensities:</p>
<div class="sourceCode" id="cb639"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb639-1"><a href="complete-protocols.html#cb639-1" tabindex="-1"></a>df_wider_ratio <span class="ot">&lt;-</span> df_split_ratio <span class="sc">%&gt;%</span> <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> filename, <span class="at">values_from =</span> Mean_norm)</span>
<span id="cb639-2"><a href="complete-protocols.html#cb639-2" tabindex="-1"></a><span class="fu">head</span>(df_wider_ratio)</span></code></pre></div>
<pre><code># A tibble: 6 × 7
# Groups:   Sample [4]
      X Sample  Number                                   Mean Slice   CFP   YFP
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                                   &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt;
1     1 Cell-01 180228_YCaM3.60_His_Py-02_w1YCam-CFP_t1  12.3     1  1.03    NA
2     2 Cell-02 180228_YCaM3.60_His_Py-02_w1YCam-CFP_t1  14.0     1  1.03    NA
3     3 Cell-03 180228_YCaM3.60_His_Py-02_w1YCam-CFP_t1  16.1     1  1.02    NA
4     4 Cell-04 180228_YCaM3.60_His_Py-02_w1YCam-CFP_t1  18.4     1  1.03    NA
5     5 Cell-01 180228_YCaM3.60_His_Py-02_w1YCam-CFP_t2  12.2     2  1.03    NA
6     6 Cell-02 180228_YCaM3.60_His_Py-02_w1YCam-CFP_t2  13.9     2  1.02    NA</code></pre>
<p>This is not what I wanted, as the YFP column shows <code>NA</code> . The reason is that the columns “Number” and “Mean” are not identical for the CFP and YFP data. So I will do this again, but first get rid of these columns:</p>
<div class="sourceCode" id="cb641"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb641-1"><a href="complete-protocols.html#cb641-1" tabindex="-1"></a>df_wider_ratio <span class="ot">&lt;-</span> df_split_ratio <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(Number, Mean)) <span class="sc">%&gt;%</span></span>
<span id="cb641-2"><a href="complete-protocols.html#cb641-2" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> filename, <span class="at">values_from =</span> Mean_norm)</span>
<span id="cb641-3"><a href="complete-protocols.html#cb641-3" tabindex="-1"></a></span>
<span id="cb641-4"><a href="complete-protocols.html#cb641-4" tabindex="-1"></a><span class="fu">head</span>(df_wider_ratio)</span></code></pre></div>
<pre><code># A tibble: 6 × 5
# Groups:   Sample [4]
      X Sample  Slice   CFP   YFP
  &lt;int&gt; &lt;chr&gt;   &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt;
1     1 Cell-01     1  1.03 0.991
2     2 Cell-02     1  1.03 1.01 
3     3 Cell-03     1  1.02 1.00 
4     4 Cell-04     1  1.03 1.00 
5     5 Cell-01     2  1.03 0.999
6     6 Cell-02     2  1.02 1.00 </code></pre>
<p>The data is ow in the right format and the calculation the ratio is straightforward:</p>
<div class="sourceCode" id="cb643"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb643-1"><a href="complete-protocols.html#cb643-1" tabindex="-1"></a>df_wider_ratio <span class="ot">&lt;-</span> df_wider_ratio <span class="sc">%&gt;%</span> <span class="fu">mutate</span> (<span class="at">Ratio =</span> YFP<span class="sc">/</span>CFP)</span>
<span id="cb643-2"><a href="complete-protocols.html#cb643-2" tabindex="-1"></a><span class="fu">head</span>(df_wider_ratio)</span></code></pre></div>
<pre><code># A tibble: 6 × 6
# Groups:   Sample [4]
      X Sample  Slice   CFP   YFP Ratio
  &lt;int&gt; &lt;chr&gt;   &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     1 Cell-01     1  1.03 0.991 0.959
2     2 Cell-02     1  1.03 1.01  0.980
3     3 Cell-03     1  1.02 1.00  0.980
4     4 Cell-04     1  1.03 1.00  0.976
5     5 Cell-01     2  1.03 0.999 0.973
6     6 Cell-02     2  1.02 1.00  0.987</code></pre>
<p>From the experimental settings, we know that each slice takes two seconds, so we can add a column with ‘Time’ data:</p>
<div class="sourceCode" id="cb645"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb645-1"><a href="complete-protocols.html#cb645-1" tabindex="-1"></a>df_wider_ratio <span class="ot">&lt;-</span> df_wider_ratio <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Time=</span>Slice<span class="sc">*</span><span class="dv">2</span>)</span></code></pre></div>
<p>To plot the ratio data:</p>
<div class="sourceCode" id="cb646"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb646-1"><a href="complete-protocols.html#cb646-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_wider_ratio, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>Ratio, <span class="at">group=</span>Sample, <span class="at">color=</span>Sample)) <span class="sc">+</span></span>
<span id="cb646-2"><a href="complete-protocols.html#cb646-2" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb646-3"><a href="complete-protocols.html#cb646-3" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">2</span>)</span>
<span id="cb646-4"><a href="complete-protocols.html#cb646-4" tabindex="-1"></a></span>
<span id="cb646-5"><a href="complete-protocols.html#cb646-5" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1001-1.png" width="672" /></p>
<p>The data is in correct shape and we can start to work on the layout of the plot. The response that is observed is triggered by receptor activation at time point t=25 until t=175. We can display this time window by adding a rectangle with the function <code>annotate()</code>. We also change the theme:</p>
<div class="sourceCode" id="cb647"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb647-1"><a href="complete-protocols.html#cb647-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb647-2"><a href="complete-protocols.html#cb647-2" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;rect&quot;</span>,<span class="at">xmin=</span><span class="dv">25</span>,<span class="at">xmax=</span><span class="dv">175</span>,<span class="at">ymin=</span><span class="sc">-</span><span class="cn">Inf</span>,<span class="at">ymax=</span><span class="cn">Inf</span>,<span class="at">alpha=</span><span class="fl">0.1</span>,<span class="at">fill=</span><span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb647-3"><a href="complete-protocols.html#cb647-3" tabindex="-1"></a>  <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">16</span>)</span>
<span id="cb647-4"><a href="complete-protocols.html#cb647-4" tabindex="-1"></a></span>
<span id="cb647-5"><a href="complete-protocols.html#cb647-5" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1002-1.png" width="672" /></p>
<p>Let’s remove the grid and change the labels of the axes:</p>
<div class="sourceCode" id="cb648"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb648-1"><a href="complete-protocols.html#cb648-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb648-2"><a href="complete-protocols.html#cb648-2" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Time [s]&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Normalized Ratio&quot;</span>) <span class="sc">+</span></span>
<span id="cb648-3"><a href="complete-protocols.html#cb648-3" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Calcium oscillations induced by histamine&quot;</span>,</span>
<span id="cb648-4"><a href="complete-protocols.html#cb648-4" tabindex="-1"></a>       <span class="at">tag =</span> <span class="st">&quot;Protocol 19&quot;</span>) <span class="sc">+</span></span>
<span id="cb648-5"><a href="complete-protocols.html#cb648-5" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span>
<span id="cb648-6"><a href="complete-protocols.html#cb648-6" tabindex="-1"></a></span>
<span id="cb648-7"><a href="complete-protocols.html#cb648-7" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1003-1.png" width="672" /></p>
<p>The heterogeneity is best display by zooming in (to a timerange of 0-200s) and by displaying the plots individually:</p>
<div class="sourceCode" id="cb649"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb649-1"><a href="complete-protocols.html#cb649-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">190</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="fl">0.8</span>,<span class="fl">5.9</span>), <span class="at">expand =</span> <span class="cn">FALSE</span>) <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>Sample)</span>
<span id="cb649-2"><a href="complete-protocols.html#cb649-2" tabindex="-1"></a></span>
<span id="cb649-3"><a href="complete-protocols.html#cb649-3" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1004-1.png" width="672" /></p>
<p>The ‘Sample’ is indicatd twice in this plot and we can get rid of at least one label. However, since I’m interested in displaying the heterogeneity and not in connecting the plots to a specific sample, I’ll get rid of both labels:</p>
<p>To remove the labels and strips for the facets:</p>
<div class="sourceCode" id="cb650"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb650-1"><a href="complete-protocols.html#cb650-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">strip.text.x =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p>To remove the legend and add a :</p>
<div class="sourceCode" id="cb651"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb651-1"><a href="complete-protocols.html#cb651-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span>
<span id="cb651-2"><a href="complete-protocols.html#cb651-2" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1006-1.png" width="672" /></p>
<p>To save the plot:</p>
<div class="sourceCode" id="cb652"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb652-1"><a href="complete-protocols.html#cb652-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_19.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb652-2"><a href="complete-protocols.html#cb652-2" tabindex="-1"></a> p</span>
<span id="cb652-3"><a href="complete-protocols.html#cb652-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-20" class="section level2 hasAnchor" number="4.20">
<h2><span class="header-section-number">4.20</span> Protocol 20 - On- and off-kinetics<a href="complete-protocols.html#protocol-20" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This document explains how to perform an exponential fit in R. It relies heavy on this blog by Douglas Watson:
<a href="https://douglas-watson.github.io/post/2018-09_exponential_curve_fitting/" class="uri">https://douglas-watson.github.io/post/2018-09_exponential_curve_fitting/</a>
We use data that was originally published in figure 1 of <a href="https://www.biorxiv.org/content/10.1101/2022.10.17.512253v1">Mahlandt et al (2022)</a>.</p>
<p>Load packages needed to wrangle the data and for plotting:</p>
<div class="sourceCode" id="cb654"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb654-1"><a href="complete-protocols.html#cb654-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>We can also define custom colors and set the theme for the plots:</p>
<div class="sourceCode" id="cb655"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb655-1"><a href="complete-protocols.html#cb655-1" tabindex="-1"></a><span class="co">#Nature themed color palette by Eike =D</span></span>
<span id="cb655-2"><a href="complete-protocols.html#cb655-2" tabindex="-1"></a><span class="co">#rapamycin would be springgreen4, magnets are cornflowerblue and iLID is darkorchid4</span></span>
<span id="cb655-3"><a href="complete-protocols.html#cb655-3" tabindex="-1"></a>newColors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;cornflowerblue&quot;</span>, <span class="st">&quot;darkorchid4&quot;</span>, <span class="st">&quot;springgreen4&quot;</span>)</span></code></pre></div>
<div id="reading-and-processing-the-data" class="section level3 hasAnchor" number="4.20.1">
<h3><span class="header-section-number">4.20.1</span> Reading and processing the data<a href="complete-protocols.html#reading-and-processing-the-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Read the data. Since this dataset has “N.A.” that need to be converted to ‘NA’ for proper handling, we define this in ‘na.strings’ and after that the rows with ‘NA’ are removed:</p>
<div class="sourceCode" id="cb656"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb656-1"><a href="complete-protocols.html#cb656-1" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;data/OptoG003_Kinetics_Summary_timeZero.csv&quot;</span>, <span class="at">na.strings =</span> <span class="st">&quot;N.A.&quot;</span>) <span class="sc">%&gt;%</span> <span class="fu">na.omit</span>()</span>
<span id="cb656-2"><a href="complete-protocols.html#cb656-2" tabindex="-1"></a><span class="fu">head</span>(df)</span></code></pre></div>
<pre><code>  Replicate                                   Experiment Construct    Cell
1         1 211215_OptoG003_01_4500_4516_iLID_P2_Cell_01      iLID Cell_01
2         1 211215_OptoG003_01_4500_4516_iLID_P2_Cell_01      iLID Cell_01
3         1 211215_OptoG003_01_4500_4516_iLID_P2_Cell_01      iLID Cell_01
4         1 211215_OptoG003_01_4500_4516_iLID_P2_Cell_01      iLID Cell_01
5         1 211215_OptoG003_01_4500_4516_iLID_P2_Cell_01      iLID Cell_01
6         1 211215_OptoG003_01_4500_4516_iLID_P2_Cell_01      iLID Cell_01
  Time_adjusted Frame Membrane Cytosol     Ratio
1        -12.95     1   200.23  356.14 0.5622227
2        -10.36     2   202.58  362.64 0.5586256
3         -7.77     3   212.70  357.81 0.5944496
4         -5.18     4   218.74  353.31 0.6191164
5         -2.59     5   205.28  347.05 0.5914998
6          0.00     6   215.72  348.98 0.6181443</code></pre>
<p>Two of the columns are renamed. And the order of the factors in the new column ‘System’ is defined (this will define the order in which the plots are shown):</p>
<div class="sourceCode" id="cb658"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb658-1"><a href="complete-protocols.html#cb658-1" tabindex="-1"></a>df <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">Time=</span>Time_adjusted, <span class="at">System=</span>Construct) <span class="sc">%&gt;%</span></span>
<span id="cb658-2"><a href="complete-protocols.html#cb658-2" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">`</span><span class="at">System</span><span class="st">`</span> <span class="ot">=</span> forcats<span class="sc">::</span><span class="fu">fct_relevel</span>(<span class="st">`</span><span class="at">System</span><span class="st">`</span>, <span class="fu">c</span>(<span class="st">&quot;magnets&quot;</span>, <span class="st">&quot;iLID&quot;</span>, <span class="st">&quot;rapamycin&quot;</span>)))</span></code></pre></div>
<p>Let’s have a look at the data:</p>
<div class="sourceCode" id="cb659"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb659-1"><a href="complete-protocols.html#cb659-1" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>Ratio))<span class="sc">+</span><span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">group=</span>Experiment), <span class="at">alpha=</span><span class="fl">0.5</span>)<span class="sc">+</span><span class="fu">facet_wrap</span>(<span class="sc">~</span>System)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1031-1.png" width="672" /></p>
<p>To summarise the data and get the mean for each condition:</p>
<div class="sourceCode" id="cb660"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb660-1"><a href="complete-protocols.html#cb660-1" tabindex="-1"></a>df_summary <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Time, System) <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(<span class="at">n=</span><span class="fu">n</span>(),</span>
<span id="cb660-2"><a href="complete-protocols.html#cb660-2" tabindex="-1"></a>                                                          <span class="at">mean=</span><span class="fu">mean</span>(Ratio))</span></code></pre></div>
<pre><code>`summarise()` has grouped output by &#39;Time&#39;. You can override using the
`.groups` argument.</code></pre>
<div class="sourceCode" id="cb662"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb662-1"><a href="complete-protocols.html#cb662-1" tabindex="-1"></a><span class="fu">head</span>(df_summary)</span></code></pre></div>
<pre><code># A tibble: 6 × 4
# Groups:   Time [2]
   Time System        n  mean
  &lt;dbl&gt; &lt;fct&gt;     &lt;int&gt; &lt;dbl&gt;
1 -13.0 magnets      12 0.604
2 -13.0 iLID         17 0.553
3 -13.0 rapamycin    11 0.781
4 -10.4 magnets      12 0.581
5 -10.4 iLID         17 0.550
6 -10.4 rapamycin    11 0.788</code></pre>
<p>Let’s plot the average values:</p>
<div class="sourceCode" id="cb664"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb664-1"><a href="complete-protocols.html#cb664-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_summary, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>mean))<span class="sc">+</span><span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">2</span>, <span class="at">alpha=</span><span class="fl">0.8</span>)<span class="sc">+</span><span class="fu">facet_wrap</span>(<span class="sc">~</span>System)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1033-1.png" width="672" /></p>
<p>The data shows an increase and decrease of signal. The increase in the signal is due to blue-light stimulation and starts at t=0. Note that the systems ‘iLID’ and ‘magnets’ are reversible and have an on-rate and an off-rate. The rapamycin system is irreversible and only has an on-rate. To determine the rates, we will filter the relevant time-window.</p>
</div>
<div id="trial-fitting-on-a-single-trace" class="section level3 hasAnchor" number="4.20.2">
<h3><span class="header-section-number">4.20.2</span> Trial fitting on a single trace<a href="complete-protocols.html#trial-fitting-on-a-single-trace" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>First we define a dataframe that only has the iLID data from t=0 until t=100:</p>
<div class="sourceCode" id="cb665"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb665-1"><a href="complete-protocols.html#cb665-1" tabindex="-1"></a>df_ilid <span class="ot">&lt;-</span> df_summary <span class="sc">%&gt;%</span> <span class="fu">filter</span>(System <span class="sc">==</span> <span class="st">&quot;iLID&quot;</span>) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Time<span class="sc">&gt;=</span><span class="dv">0</span> <span class="sc">&amp;</span> Time<span class="sc">&lt;</span><span class="dv">100</span>)</span></code></pre></div>
<p>To fit the data we use the function <code>nls()</code>. First, let’s look at an example for a linear fit with the unknown variables ‘a’ and ‘b’:</p>
<div class="sourceCode" id="cb666"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb666-1"><a href="complete-protocols.html#cb666-1" tabindex="-1"></a><span class="co">#linear fit</span></span>
<span id="cb666-2"><a href="complete-protocols.html#cb666-2" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">nls</span>(mean <span class="sc">~</span> a <span class="sc">*</span> Time <span class="sc">+</span> b, <span class="at">data =</span> df_ilid, <span class="at">start =</span> <span class="fu">list</span>(<span class="at">a =</span> <span class="dv">1</span>, <span class="at">b =</span> <span class="dv">0</span>))</span></code></pre></div>
<p>The coefficients of the fit can be retrieved by:</p>
<div class="sourceCode" id="cb667"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb667-1"><a href="complete-protocols.html#cb667-1" tabindex="-1"></a><span class="fu">coef</span>(fit)</span></code></pre></div>
<pre><code>         a          b 
0.00201496 0.87887972 </code></pre>
<p>This will not be very accurate, but it illustrates how it works. Now let’s use a better model and fit an exponential:</p>
<div class="sourceCode" id="cb669"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb669-1"><a href="complete-protocols.html#cb669-1" tabindex="-1"></a><span class="co">#exponential fit: a0 is starting value, a is the amplitude, k is the rate constant</span></span>
<span id="cb669-2"><a href="complete-protocols.html#cb669-2" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">nls</span>(mean <span class="sc">~</span> a0 <span class="sc">+</span> a<span class="sc">*</span>(<span class="dv">1</span><span class="sc">-</span> <span class="fu">exp</span>(<span class="sc">-</span>k <span class="sc">*</span> Time)),</span>
<span id="cb669-3"><a href="complete-protocols.html#cb669-3" tabindex="-1"></a>           <span class="at">data =</span> df_ilid, </span>
<span id="cb669-4"><a href="complete-protocols.html#cb669-4" tabindex="-1"></a>           <span class="at">start =</span> <span class="fu">list</span>(<span class="at">a0 =</span> <span class="dv">0</span>, <span class="at">a =</span> <span class="dv">2</span>, <span class="at">k =</span> .<span class="dv">1</span>)</span>
<span id="cb669-5"><a href="complete-protocols.html#cb669-5" tabindex="-1"></a>           )</span></code></pre></div>
<p>The variables are ‘a0’, ‘a’ and ‘k’, which reflect the value at t=0, the amplitude and the rate. The values are:</p>
<div class="sourceCode" id="cb670"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb670-1"><a href="complete-protocols.html#cb670-1" tabindex="-1"></a><span class="fu">coef</span>(fit)</span></code></pre></div>
<pre><code>       a0         a         k 
0.5505354 0.4681990 0.1349511 </code></pre>
<p>To plot the fit together with the data, we use the function <code>augment()</code> from the broom package:</p>
<div class="sourceCode" id="cb672"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb672-1"><a href="complete-protocols.html#cb672-1" tabindex="-1"></a><span class="fu">library</span>(broom)</span>
<span id="cb672-2"><a href="complete-protocols.html#cb672-2" tabindex="-1"></a></span>
<span id="cb672-3"><a href="complete-protocols.html#cb672-3" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">augment</span>(fit), <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>mean)) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">4</span>, <span class="at">alpha=</span>.<span class="dv">5</span>) <span class="sc">+</span> </span>
<span id="cb672-4"><a href="complete-protocols.html#cb672-4" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y=</span>.fitted), <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size=</span><span class="dv">1</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1039-1.png" width="672" /></p>
<p>This looks like a good fit. Now let’s determine the half time, which is defines as ln2/k, where ln2 is the natural logarithm of 2. By default the <code>log()</code> function calculates the natural logarithm, so we use <code>log(2)</code>. The value of ‘k’ is the third variable of the list that is returned by <code>coef(fit)</code>:</p>
<div class="sourceCode" id="cb673"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb673-1"><a href="complete-protocols.html#cb673-1" tabindex="-1"></a>var <span class="ot">&lt;-</span> <span class="fu">coef</span>(fit)</span>
<span id="cb673-2"><a href="complete-protocols.html#cb673-2" tabindex="-1"></a><span class="fu">paste</span>(<span class="st">&quot;The halftime is&quot;</span>,<span class="fu">log</span>(<span class="dv">2</span>)<span class="sc">/</span>var[<span class="dv">3</span>])</span></code></pre></div>
<pre><code>[1] &quot;The halftime is 5.13628563844601&quot;</code></pre>
</div>
<div id="on-kinetics---alltogether" class="section level3 hasAnchor" number="4.20.3">
<h3><span class="header-section-number">4.20.3</span> On kinetics - alltogether<a href="complete-protocols.html#on-kinetics---alltogether" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Let’s try to do the fitting on all three curves at once. First we need to filter the data to get the relevant part of the trace for determining the on-kinetics and we store this is a new dataframe <code>df_fit_on</code>:</p>
<div class="sourceCode" id="cb675"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb675-1"><a href="complete-protocols.html#cb675-1" tabindex="-1"></a>df_fit_on <span class="ot">&lt;-</span> df_summary <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Time<span class="sc">&gt;=</span><span class="dv">0</span>)</span>
<span id="cb675-2"><a href="complete-protocols.html#cb675-2" tabindex="-1"></a>df_fit_on <span class="ot">&lt;-</span> df_fit_on <span class="sc">%&gt;%</span> <span class="fu">filter</span>(System<span class="sc">==</span><span class="st">&quot;iLID&quot;</span> <span class="sc">&amp;</span> Time<span class="sc">&lt;</span><span class="dv">100</span> <span class="sc">|</span> System<span class="sc">==</span><span class="st">&quot;magnets&quot;</span> <span class="sc">&amp;</span> Time<span class="sc">&lt;</span><span class="dv">100</span> <span class="sc">|</span> System<span class="sc">==</span><span class="st">&quot;rapamycin&quot;</span>)</span>
<span id="cb675-3"><a href="complete-protocols.html#cb675-3" tabindex="-1"></a></span>
<span id="cb675-4"><a href="complete-protocols.html#cb675-4" tabindex="-1"></a><span class="fu">ggplot</span>(df_fit_on, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>mean))<span class="sc">+</span><span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">2</span>, <span class="at">alpha=</span><span class="fl">0.8</span>)<span class="sc">+</span><span class="fu">facet_wrap</span>(<span class="sc">~</span>System)<span class="sc">+</span><span class="fu">xlim</span>(<span class="dv">0</span>,<span class="dv">200</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1041-1.png" width="672" /></p>
<p>This looks good. To fit the data, we use an approach that uses <code>nest()</code> to generate a nested dataframe. The <code>map()</code> function is used to apply the fit to each System.</p>
<div class="sourceCode" id="cb676"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb676-1"><a href="complete-protocols.html#cb676-1" tabindex="-1"></a>fitted <span class="ot">&lt;-</span> df_fit_on <span class="sc">%&gt;%</span> </span>
<span id="cb676-2"><a href="complete-protocols.html#cb676-2" tabindex="-1"></a>  <span class="fu">nest</span>(<span class="sc">-</span>System) <span class="sc">%&gt;%</span></span>
<span id="cb676-3"><a href="complete-protocols.html#cb676-3" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb676-4"><a href="complete-protocols.html#cb676-4" tabindex="-1"></a>    <span class="at">fit =</span> <span class="fu">map</span>(data, <span class="sc">~</span><span class="fu">nls</span>(mean <span class="sc">~</span> a0 <span class="sc">+</span> a<span class="sc">*</span>(<span class="dv">1</span><span class="sc">-</span> <span class="fu">exp</span>(<span class="sc">-</span>k <span class="sc">*</span> Time)),</span>
<span id="cb676-5"><a href="complete-protocols.html#cb676-5" tabindex="-1"></a>           <span class="at">data =</span> ., </span>
<span id="cb676-6"><a href="complete-protocols.html#cb676-6" tabindex="-1"></a>           <span class="at">start =</span> <span class="fu">list</span>(<span class="at">a0 =</span> <span class="dv">0</span>, <span class="at">a =</span> <span class="dv">2</span>, <span class="at">k =</span> .<span class="dv">1</span>)</span>
<span id="cb676-7"><a href="complete-protocols.html#cb676-7" tabindex="-1"></a>           )),</span>
<span id="cb676-8"><a href="complete-protocols.html#cb676-8" tabindex="-1"></a>    <span class="at">tidied =</span> <span class="fu">map</span>(fit, tidy),</span>
<span id="cb676-9"><a href="complete-protocols.html#cb676-9" tabindex="-1"></a>    <span class="at">augmented =</span> <span class="fu">map</span>(fit, augment),</span>
<span id="cb676-10"><a href="complete-protocols.html#cb676-10" tabindex="-1"></a>  )</span>
<span id="cb676-11"><a href="complete-protocols.html#cb676-11" tabindex="-1"></a>fitted</span></code></pre></div>
<pre><code># A tibble: 3 × 5
  System    data                fit    tidied           augmented        
  &lt;fct&gt;     &lt;list&gt;              &lt;list&gt; &lt;list&gt;           &lt;list&gt;           
1 magnets   &lt;gropd_df [39 × 3]&gt; &lt;nls&gt;  &lt;tibble [3 × 5]&gt; &lt;tibble [39 × 4]&gt;
2 iLID      &lt;gropd_df [39 × 3]&gt; &lt;nls&gt;  &lt;tibble [3 × 5]&gt; &lt;tibble [39 × 4]&gt;
3 rapamycin &lt;gropd_df [89 × 3]&gt; &lt;nls&gt;  &lt;tibble [3 × 5]&gt; &lt;tibble [89 × 4]&gt;</code></pre>
<p>The resulting dataframe <code>fitted</code> holds the estimated variables and these can be shown by using the function <code>unnest()</code> as follows:</p>
<div class="sourceCode" id="cb678"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb678-1"><a href="complete-protocols.html#cb678-1" tabindex="-1"></a>fitted <span class="sc">%&gt;%</span> </span>
<span id="cb678-2"><a href="complete-protocols.html#cb678-2" tabindex="-1"></a>  <span class="fu">unnest</span>(tidied) <span class="sc">%&gt;%</span> </span>
<span id="cb678-3"><a href="complete-protocols.html#cb678-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(System, term, estimate) <span class="sc">%&gt;%</span> </span>
<span id="cb678-4"><a href="complete-protocols.html#cb678-4" tabindex="-1"></a>  <span class="fu">spread</span>(term, estimate) <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">Amplitude=</span>a) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="st">`</span><span class="at">t1/2</span><span class="st">`</span><span class="ot">=</span><span class="fu">log</span>(<span class="dv">2</span>)<span class="sc">/</span>k)</span></code></pre></div>
<pre><code># A tibble: 3 × 5
  System    Amplitude    a0      k `t1/2`
  &lt;fct&gt;         &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 magnets       0.298 0.599 0.175    3.97
2 iLID          0.468 0.551 0.135    5.14
3 rapamycin     1.98  0.573 0.0161  43.1 </code></pre>
<p>Finally, we can extract the fit and plot it on top of the experimental data:</p>
<div class="sourceCode" id="cb680"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb680-1"><a href="complete-protocols.html#cb680-1" tabindex="-1"></a>augmented <span class="ot">&lt;-</span> fitted <span class="sc">%&gt;%</span> </span>
<span id="cb680-2"><a href="complete-protocols.html#cb680-2" tabindex="-1"></a>  <span class="fu">unnest</span>(augmented)</span>
<span id="cb680-3"><a href="complete-protocols.html#cb680-3" tabindex="-1"></a></span>
<span id="cb680-4"><a href="complete-protocols.html#cb680-4" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> augmented, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>mean, <span class="at">colour =</span> System)) <span class="sc">+</span> </span>
<span id="cb680-5"><a href="complete-protocols.html#cb680-5" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">4</span>, <span class="at">alpha=</span>.<span class="dv">5</span>) <span class="sc">+</span> </span>
<span id="cb680-6"><a href="complete-protocols.html#cb680-6" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y=</span>.fitted), <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>System) <span class="sc">+</span> <span class="fu">xlim</span>(<span class="dv">0</span>,<span class="dv">100</span>) <span class="sc">+</span></span>
<span id="cb680-7"><a href="complete-protocols.html#cb680-7" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> newColors)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1044-1.png" width="672" /></p>
</div>
<div id="off-kinetics" class="section level3 hasAnchor" number="4.20.4">
<h3><span class="header-section-number">4.20.4</span> Off kinetics<a href="complete-protocols.html#off-kinetics" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In a similar way, we can fit the off-kinetics. First we need to select the right time window:</p>
<div class="sourceCode" id="cb681"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb681-1"><a href="complete-protocols.html#cb681-1" tabindex="-1"></a>df_fit_off <span class="ot">&lt;-</span> df_summary <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Time=</span>Time<span class="dv">-121</span>)</span>
<span id="cb681-2"><a href="complete-protocols.html#cb681-2" tabindex="-1"></a>df_fit_off <span class="ot">&lt;-</span> df_fit_off <span class="sc">%&gt;%</span> <span class="fu">filter</span>(System<span class="sc">==</span><span class="st">&quot;iLID&quot;</span> <span class="sc">&amp;</span> Time<span class="sc">&gt;=</span><span class="dv">0</span> <span class="sc">|</span> System<span class="sc">==</span><span class="st">&quot;magnets&quot;</span> <span class="sc">&amp;</span> Time<span class="sc">&gt;=</span><span class="dv">0</span>)</span>
<span id="cb681-3"><a href="complete-protocols.html#cb681-3" tabindex="-1"></a></span>
<span id="cb681-4"><a href="complete-protocols.html#cb681-4" tabindex="-1"></a><span class="fu">ggplot</span>(df_fit_off, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>mean))<span class="sc">+</span><span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">2</span>, <span class="at">alpha=</span><span class="fl">0.8</span>)<span class="sc">+</span><span class="fu">facet_wrap</span>(<span class="sc">~</span>System)<span class="sc">+</span><span class="fu">xlim</span>(<span class="dv">0</span>,<span class="dv">200</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1045-1.png" width="672" /></p>
<p>Next, we can do the fitting:</p>
<div class="sourceCode" id="cb682"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb682-1"><a href="complete-protocols.html#cb682-1" tabindex="-1"></a>fitted_off <span class="ot">&lt;-</span> df_fit_off <span class="sc">%&gt;%</span> </span>
<span id="cb682-2"><a href="complete-protocols.html#cb682-2" tabindex="-1"></a>  <span class="fu">nest</span>(<span class="sc">-</span>System) <span class="sc">%&gt;%</span></span>
<span id="cb682-3"><a href="complete-protocols.html#cb682-3" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb682-4"><a href="complete-protocols.html#cb682-4" tabindex="-1"></a>    <span class="at">fit =</span> <span class="fu">map</span>(data, <span class="sc">~</span><span class="fu">nls</span>(mean <span class="sc">~</span> a0 <span class="sc">+</span> a<span class="sc">*</span>(<span class="dv">1</span><span class="sc">-</span> <span class="fu">exp</span>(<span class="sc">-</span>k <span class="sc">*</span> Time)),</span>
<span id="cb682-5"><a href="complete-protocols.html#cb682-5" tabindex="-1"></a>           <span class="at">data =</span> ., </span>
<span id="cb682-6"><a href="complete-protocols.html#cb682-6" tabindex="-1"></a>           <span class="at">start =</span> <span class="fu">list</span>(<span class="at">a0 =</span> <span class="dv">2</span>, <span class="at">a =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">k =</span> .<span class="dv">1</span>)</span>
<span id="cb682-7"><a href="complete-protocols.html#cb682-7" tabindex="-1"></a>           )),</span>
<span id="cb682-8"><a href="complete-protocols.html#cb682-8" tabindex="-1"></a>    <span class="at">tidied =</span> <span class="fu">map</span>(fit, tidy),</span>
<span id="cb682-9"><a href="complete-protocols.html#cb682-9" tabindex="-1"></a>    <span class="at">augmented =</span> <span class="fu">map</span>(fit, augment),</span>
<span id="cb682-10"><a href="complete-protocols.html#cb682-10" tabindex="-1"></a>  )</span></code></pre></div>
<p>To get the fit parameters:</p>
<div class="sourceCode" id="cb683"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb683-1"><a href="complete-protocols.html#cb683-1" tabindex="-1"></a>fit_results <span class="ot">&lt;-</span> fitted_off <span class="sc">%&gt;%</span> </span>
<span id="cb683-2"><a href="complete-protocols.html#cb683-2" tabindex="-1"></a>  <span class="fu">unnest</span>(tidied) <span class="sc">%&gt;%</span> </span>
<span id="cb683-3"><a href="complete-protocols.html#cb683-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(System, term, estimate) <span class="sc">%&gt;%</span> </span>
<span id="cb683-4"><a href="complete-protocols.html#cb683-4" tabindex="-1"></a>  <span class="fu">spread</span>(term, estimate) <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">Amplitude=</span>a) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="st">`</span><span class="at">t1/2</span><span class="st">`</span><span class="ot">=</span><span class="fu">log</span>(<span class="dv">2</span>)<span class="sc">/</span>k)</span>
<span id="cb683-5"><a href="complete-protocols.html#cb683-5" tabindex="-1"></a>fit_results</span></code></pre></div>
<pre><code># A tibble: 2 × 5
  System  Amplitude    a0      k `t1/2`
  &lt;fct&gt;       &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 magnets    -0.264 0.810 0.0373   18.6
2 iLID       -0.342 0.948 0.0342   20.3</code></pre>
<p>And finally, we can inspect the fit:</p>
<div class="sourceCode" id="cb685"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb685-1"><a href="complete-protocols.html#cb685-1" tabindex="-1"></a>augmented <span class="ot">&lt;-</span> fitted_off <span class="sc">%&gt;%</span> </span>
<span id="cb685-2"><a href="complete-protocols.html#cb685-2" tabindex="-1"></a>  <span class="fu">unnest</span>(augmented)</span>
<span id="cb685-3"><a href="complete-protocols.html#cb685-3" tabindex="-1"></a></span>
<span id="cb685-4"><a href="complete-protocols.html#cb685-4" tabindex="-1"></a></span>
<span id="cb685-5"><a href="complete-protocols.html#cb685-5" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> augmented, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>mean, <span class="at">colour =</span> System)) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">4</span>, <span class="at">alpha=</span>.<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb685-6"><a href="complete-protocols.html#cb685-6" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y=</span>.fitted), <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>System)  <span class="sc">+</span></span>
<span id="cb685-7"><a href="complete-protocols.html#cb685-7" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> newColors)</span>
<span id="cb685-8"><a href="complete-protocols.html#cb685-8" tabindex="-1"></a></span>
<span id="cb685-9"><a href="complete-protocols.html#cb685-9" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1048-1.png" width="672" /></p>
<p>Finally, we can adjust the layout and titles:</p>
<div class="sourceCode" id="cb686"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb686-1"><a href="complete-protocols.html#cb686-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb686-2"><a href="complete-protocols.html#cb686-2" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;Off-kinetics of two optogenetic systems&quot;</span>,</span>
<span id="cb686-3"><a href="complete-protocols.html#cb686-3" tabindex="-1"></a>  <span class="at">x =</span> <span class="st">&quot;Time [s]&quot;</span>,</span>
<span id="cb686-4"><a href="complete-protocols.html#cb686-4" tabindex="-1"></a>  <span class="at">y =</span> <span class="st">&quot;Response [arbitrary units]&quot;</span>,</span>
<span id="cb686-5"><a href="complete-protocols.html#cb686-5" tabindex="-1"></a>  <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">based on data from Mahlandt et al., DOI: 10.1101/2022.10.17.512253&quot;</span>,</span>
<span id="cb686-6"><a href="complete-protocols.html#cb686-6" tabindex="-1"></a>  <span class="at">tag =</span> <span class="st">&quot;Protocol 20&quot;</span></span>
<span id="cb686-7"><a href="complete-protocols.html#cb686-7" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb686-8"><a href="complete-protocols.html#cb686-8" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>),</span>
<span id="cb686-9"><a href="complete-protocols.html#cb686-9" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span>
<span id="cb686-10"><a href="complete-protocols.html#cb686-10" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1049-1.png" width="672" /></p>
<p>The data and fit look good, let’s add the t1/2 values as well. Protocol 16 explains how to add labels to plots that uses ‘facets’ and we use that approach here too:</p>
<div class="sourceCode" id="cb687"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb687-1"><a href="complete-protocols.html#cb687-1" tabindex="-1"></a><span class="co"># Code to add r-squared to the plot</span></span>
<span id="cb687-2"><a href="complete-protocols.html#cb687-2" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="at">data=</span>fit_results, </span>
<span id="cb687-3"><a href="complete-protocols.html#cb687-3" tabindex="-1"></a>                   <span class="at">x=</span><span class="cn">Inf</span>,</span>
<span id="cb687-4"><a href="complete-protocols.html#cb687-4" tabindex="-1"></a>                   <span class="at">y=</span><span class="fl">0.9</span>,</span>
<span id="cb687-5"><a href="complete-protocols.html#cb687-5" tabindex="-1"></a>                   <span class="at">hjust=</span><span class="fl">1.1</span>,</span>
<span id="cb687-6"><a href="complete-protocols.html#cb687-6" tabindex="-1"></a>                   <span class="at">vjust=</span><span class="dv">0</span>,</span>
<span id="cb687-7"><a href="complete-protocols.html#cb687-7" tabindex="-1"></a>                   <span class="fu">aes</span>(<span class="at">label=</span><span class="fu">paste0</span>(<span class="st">&#39;t½ = &#39;</span>,<span class="fu">round</span>(<span class="st">`</span><span class="at">t1/2</span><span class="st">`</span>,<span class="dv">1</span>), <span class="st">&#39;s&#39;</span>)),</span>
<span id="cb687-8"><a href="complete-protocols.html#cb687-8" tabindex="-1"></a>                  <span class="at">size=</span><span class="dv">5</span>,</span>
<span id="cb687-9"><a href="complete-protocols.html#cb687-9" tabindex="-1"></a>                   <span class="at">color=</span><span class="st">&#39;grey40&#39;</span></span>
<span id="cb687-10"><a href="complete-protocols.html#cb687-10" tabindex="-1"></a>                   )</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1050-1.png" width="672" /></p>
<p>Let’s save the result:</p>
<div class="sourceCode" id="cb688"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb688-1"><a href="complete-protocols.html#cb688-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_20.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb688-2"><a href="complete-protocols.html#cb688-2" tabindex="-1"></a> p</span>
<span id="cb688-3"><a href="complete-protocols.html#cb688-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
</div>
<div id="protocol-21" class="section level2 hasAnchor" number="4.21">
<h2><span class="header-section-number">4.21</span> Protocol 21 - A spiral plot<a href="complete-protocols.html#protocol-21" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this protocol, we will plot time-series data in a spiral plot. This protocol is inspired by another application of a <a href="https://flowingdata.com/2022/01/10/a-quick-and-easy-way-to-make-spiral-charts-in-r/">spiral plot</a> to visualize data from several years.</p>
<p>We will use the data that is supplied by <a href="http://www.addgene.org">addgene</a> on requests for plasmids that we have deposited.
We first load the <code>{tidyverse}</code> package:</p>
<div class="sourceCode" id="cb690"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb690-1"><a href="complete-protocols.html#cb690-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>The data comes in a csv format, let’s load that and check its first 6 rows:</p>
<div class="sourceCode" id="cb691"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb691-1"><a href="complete-protocols.html#cb691-1" tabindex="-1"></a>df_addgene <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;data/Addgene-Requests-for-Materials_MolCyto_March_2024.csv&quot;</span>, <span class="at">stringsAsFactors =</span> <span class="cn">TRUE</span>)</span>
<span id="cb691-2"><a href="complete-protocols.html#cb691-2" tabindex="-1"></a><span class="fu">head</span>(df_addgene)</span></code></pre></div>
<pre><code>      ID                   Material Requesting.Country Date.Ordered
1 189773      pmScarlet3-Giantin_C1      UNITED STATES   03/18/2024
2 189768 pmScarlet3-alphaTubulin_C1            DENMARK   03/18/2024
3 189754              pDx_mScarlet3      UNITED STATES   03/15/2024
4 189767      pLifeAct-mScarlet3_N1      UNITED STATES   03/15/2024
5 176098       dimericTomato-2xrGBD             FRANCE   03/15/2024
6  36201       pLifeAct-mTurquoise2             ISRAEL   03/14/2024</code></pre>
<p>The column with dates should not be a factor as it is now, but rather be in a date format. Alternatively, and that is what we will do here, we can split this column into day/month/year. The argument <code>convert = TRUE</code> ensures conversion of the values into a integer:</p>
<div class="sourceCode" id="cb693"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb693-1"><a href="complete-protocols.html#cb693-1" tabindex="-1"></a>df_addgene <span class="ot">&lt;-</span> df_addgene <span class="sc">%&gt;%</span></span>
<span id="cb693-2"><a href="complete-protocols.html#cb693-2" tabindex="-1"></a>  <span class="fu">separate</span>(Date.Ordered, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">&quot;Month&quot;</span>,<span class="st">&quot;Day&quot;</span>,<span class="st">&quot;Year&quot;</span>), <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>, <span class="at">convert =</span> <span class="cn">TRUE</span>)</span>
<span id="cb693-3"><a href="complete-protocols.html#cb693-3" tabindex="-1"></a><span class="fu">head</span>(df_addgene)</span></code></pre></div>
<pre><code>      ID                   Material Requesting.Country Month Day Year
1 189773      pmScarlet3-Giantin_C1      UNITED STATES     3  18 2024
2 189768 pmScarlet3-alphaTubulin_C1            DENMARK     3  18 2024
3 189754              pDx_mScarlet3      UNITED STATES     3  15 2024
4 189767      pLifeAct-mScarlet3_N1      UNITED STATES     3  15 2024
5 176098       dimericTomato-2xrGBD             FRANCE     3  15 2024
6  36201       pLifeAct-mTurquoise2             ISRAEL     3  14 2024</code></pre>
<p>Let’s calculate the accumulated number of requests. In this dataframe, each row is a single request:</p>
<div class="sourceCode" id="cb695"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb695-1"><a href="complete-protocols.html#cb695-1" tabindex="-1"></a>total_requests <span class="ot">&lt;-</span> <span class="fu">nrow</span>(df_addgene)</span>
<span id="cb695-2"><a href="complete-protocols.html#cb695-2" tabindex="-1"></a>df_addgene<span class="sc">$</span>total <span class="ot">&lt;-</span> <span class="fu">seq</span>(total_requests, <span class="dv">1</span>)</span></code></pre></div>
<p>Let’s plot how the number of requests evolved over months:</p>
<div class="sourceCode" id="cb696"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb696-1"><a href="complete-protocols.html#cb696-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_addgene, <span class="fu">aes</span>(Month, total, <span class="at">color=</span><span class="fu">as.factor</span>(Year))) <span class="sc">+</span> <span class="fu">geom_line</span>()</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1083-1.png" width="672" /></p>
<p>The plot shows the accumulation of requests over the years. The zig-zag appears because there are multiple values for each month. We can simplify this dataframe to show only the maximal value for each month (per year):</p>
<div class="sourceCode" id="cb697"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb697-1"><a href="complete-protocols.html#cb697-1" tabindex="-1"></a>df_addgene_month <span class="ot">&lt;-</span> df_addgene <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Year, Month) <span class="sc">%&gt;%</span></span>
<span id="cb697-2"><a href="complete-protocols.html#cb697-2" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">total_max =</span> <span class="fu">max</span>(total)) <span class="sc">%&gt;%</span></span>
<span id="cb697-3"><a href="complete-protocols.html#cb697-3" tabindex="-1"></a>  <span class="fu">ungroup</span>()</span></code></pre></div>
<pre><code>`summarise()` has grouped output by &#39;Year&#39;. You can override using the
`.groups` argument.</code></pre>
<p>Let’s plot these data again:</p>
<div class="sourceCode" id="cb699"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb699-1"><a href="complete-protocols.html#cb699-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_addgene_month, <span class="fu">aes</span>(Month, total_max, <span class="at">color=</span><span class="fu">as.factor</span>(Year))) <span class="sc">+</span> <span class="fu">geom_line</span>()</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1085-1.png" width="672" /></p>
<p>This looks better, and we can now plot this in a circular fahsion with <code>coord_polar()</code>:</p>
<div class="sourceCode" id="cb700"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb700-1"><a href="complete-protocols.html#cb700-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_addgene_month, <span class="fu">aes</span>(Month, total_max, <span class="at">color=</span><span class="fu">as.factor</span>(Year))) <span class="sc">+</span></span>
<span id="cb700-2"><a href="complete-protocols.html#cb700-2" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">linewidth=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb700-3"><a href="complete-protocols.html#cb700-3" tabindex="-1"></a>  <span class="fu">coord_polar</span>(<span class="at">clip =</span> <span class="st">&#39;off&#39;</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1086-1.png" width="672" /></p>
<p>This looks good, but the lines show a step at the 0 degrees angle. This is because the data for the different years are connected to join all these data. There is a nice solution described here that we will use:
<a href="https://stackoverflow.com/questions/41842249/join-gap-in-polar-line-ggplot-plot" class="uri">https://stackoverflow.com/questions/41842249/join-gap-in-polar-line-ggplot-plot</a></p>
<div class="sourceCode" id="cb701"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb701-1"><a href="complete-protocols.html#cb701-1" tabindex="-1"></a>bridge <span class="ot">&lt;-</span> df_addgene_month[df_addgene_month<span class="sc">$</span>Month<span class="sc">==</span><span class="dv">12</span>,]</span>
<span id="cb701-2"><a href="complete-protocols.html#cb701-2" tabindex="-1"></a>bridge<span class="sc">$</span>Year <span class="ot">&lt;-</span> bridge<span class="sc">$</span>Year <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb701-3"><a href="complete-protocols.html#cb701-3" tabindex="-1"></a>bridge<span class="sc">$</span>Month <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb701-4"><a href="complete-protocols.html#cb701-4" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(df_addgene_month, bridge)</span></code></pre></div>
<p>Now, plot again and change the colors:</p>
<div class="sourceCode" id="cb702"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb702-1"><a href="complete-protocols.html#cb702-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df, <span class="fu">aes</span>(Month, total_max, <span class="at">color=</span><span class="fu">as.factor</span>(Year), <span class="at">fill=</span><span class="fu">as.factor</span>(Year))) <span class="sc">+</span></span>
<span id="cb702-2"><a href="complete-protocols.html#cb702-2" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="at">linewidth=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb702-3"><a href="complete-protocols.html#cb702-3" tabindex="-1"></a>  <span class="fu">coord_polar</span>()</span>
<span id="cb702-4"><a href="complete-protocols.html#cb702-4" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1088-1.png" width="672" /></p>
<p>I do not like the girds and axis labels for these types of plots, so let’s get rid of those:</p>
<div class="sourceCode" id="cb703"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb703-1"><a href="complete-protocols.html#cb703-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb703-2"><a href="complete-protocols.html#cb703-2" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&quot;&quot;</span>, <span class="at">y=</span><span class="st">&quot;&quot;</span>, <span class="at">color=</span><span class="st">&quot;Year&quot;</span>) <span class="sc">+</span></span>
<span id="cb703-3"><a href="complete-protocols.html#cb703-3" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb703-4"><a href="complete-protocols.html#cb703-4" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(<span class="at">reverse =</span> <span class="cn">TRUE</span>)) <span class="sc">+</span></span>
<span id="cb703-5"><a href="complete-protocols.html#cb703-5" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb703-6"><a href="complete-protocols.html#cb703-6" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb703-7"><a href="complete-protocols.html#cb703-7" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb703-8"><a href="complete-protocols.html#cb703-8" tabindex="-1"></a>        <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span>.<span class="dv">8</span>),</span>
<span id="cb703-9"><a href="complete-protocols.html#cb703-9" tabindex="-1"></a>        <span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color=</span><span class="st">&#39;grey80&#39;</span>, <span class="at">hjust=</span><span class="dv">1</span>, <span class="at">vjust=</span><span class="dv">5</span>))</span>
<span id="cb703-10"><a href="complete-protocols.html#cb703-10" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1089-1.png" width="672" /></p>
<p>Adding titles:</p>
<div class="sourceCode" id="cb704"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb704-1"><a href="complete-protocols.html#cb704-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Evolution of the total number of plasmids&quot;</span>,</span>
<span id="cb704-2"><a href="complete-protocols.html#cb704-2" tabindex="-1"></a>              <span class="at">subtitle =</span> <span class="st">&quot;shared through addgene.org&quot;</span>,</span>
<span id="cb704-3"><a href="complete-protocols.html#cb704-3" tabindex="-1"></a>              <span class="at">tag =</span> <span class="st">&quot;Protocol 21&quot;</span>,</span>
<span id="cb704-4"><a href="complete-protocols.html#cb704-4" tabindex="-1"></a>              <span class="at">caption=</span><span class="st">&quot;data provided by addgene.org - March 2024&quot;</span>)</span>
<span id="cb704-5"><a href="complete-protocols.html#cb704-5" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1090-1.png" width="672" /></p>
<p>The legend is quite long as it lists all the years. We could list a subset by using <code>scale_color_discrete()</code>:</p>
<div class="sourceCode" id="cb705"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb705-1"><a href="complete-protocols.html#cb705-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">scale_color_discrete</span>(<span class="at">breaks=</span><span class="fu">c</span>(<span class="st">&quot;2011&quot;</span>, <span class="st">&quot;2014&quot;</span>, <span class="st">&quot;2017&quot;</span>, <span class="st">&quot;2020&quot;</span>, <span class="st">&quot;2023&quot;</span>)) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1091-1.png" width="672" /></p>
<p>The legend is less cluttered, but it is less clear how many years are shown and which line belongs to which year (especially if you are colorblind, like I am). Another way the make the legend less prominent is by reducing the size:</p>
<div class="sourceCode" id="cb706"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb706-1"><a href="complete-protocols.html#cb706-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>),</span>
<span id="cb706-2"><a href="complete-protocols.html#cb706-2" tabindex="-1"></a>          <span class="at">legend.key.height=</span> <span class="fu">unit</span>(<span class="dv">12</span>, <span class="st">&#39;points&#39;</span>))</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1092-1.png" width="672" /></p>
<p>Yep, that is better. Let’s see if we can also do some direct labeling. We need a separate dataframe to get there:</p>
<div class="sourceCode" id="cb707"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb707-1"><a href="complete-protocols.html#cb707-1" tabindex="-1"></a>df_label <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Month<span class="sc">==</span><span class="dv">12</span>)</span></code></pre></div>
<p>Now we take a subset, because we only want to label every nth line (every 5th in this example), to reduce the number of labels. We use <code>slice()</code> and since we want to include the most recent year, we first sort the order of the years with <code>arrange()</code>:</p>
<div class="sourceCode" id="cb708"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb708-1"><a href="complete-protocols.html#cb708-1" tabindex="-1"></a>df_label <span class="ot">&lt;-</span> df_label <span class="sc">%&gt;%</span></span>
<span id="cb708-2"><a href="complete-protocols.html#cb708-2" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(Year)) <span class="sc">%&gt;%</span></span>
<span id="cb708-3"><a href="complete-protocols.html#cb708-3" tabindex="-1"></a>  <span class="fu">slice</span>(<span class="fu">seq</span>(<span class="dv">1</span>, <span class="fu">nrow</span>(df_label), <span class="dv">5</span>))</span>
<span id="cb708-4"><a href="complete-protocols.html#cb708-4" tabindex="-1"></a></span>
<span id="cb708-5"><a href="complete-protocols.html#cb708-5" tabindex="-1"></a>df_label</span></code></pre></div>
<pre><code># A tibble: 3 × 3
   Year Month total_max
  &lt;dbl&gt; &lt;dbl&gt;     &lt;int&gt;
1  2023    12      9881
2  2018    12      4394
3  2013    12       691</code></pre>
<p>The <code>df_label</code> is now exactly as I wanted it. Let’s add the labels from this dataframe:</p>
<div class="sourceCode" id="cb710"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb710-1"><a href="complete-protocols.html#cb710-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_label</span>(<span class="at">data=</span>df_label, <span class="fu">aes</span>(<span class="at">x=</span>Month, <span class="at">y=</span>total_max, <span class="at">label=</span><span class="fu">paste0</span>(Year,<span class="st">&quot;: &quot;</span>,total_max)), <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size=</span><span class="dv">4</span>, <span class="at">label.size =</span> <span class="cn">NA</span>, <span class="at">alpha=</span><span class="fl">0.8</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1095-1.png" width="672" /></p>
<p>Nice. We do not need the legend anymore. And the labels are a bit large, so let’s make them smaller and move them a bit upwards. By adding a triangle (this is defined by <code>shape=25</code> within <code>geom_point()</code>), that is also slightly moved upwards in the plot, the labels nicely point to the correct line:</p>
<div class="sourceCode" id="cb711"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb711-1"><a href="complete-protocols.html#cb711-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>, <span class="at">axis.text =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb711-2"><a href="complete-protocols.html#cb711-2" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>df_label,</span>
<span id="cb711-3"><a href="complete-protocols.html#cb711-3" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">x=</span>Month, <span class="at">y=</span>total_max<span class="sc">+</span><span class="dv">600</span>), <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">shape=</span><span class="dv">25</span>, <span class="at">size=</span><span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb711-4"><a href="complete-protocols.html#cb711-4" tabindex="-1"></a>  <span class="fu">geom_label</span>(<span class="at">data=</span>df_label,</span>
<span id="cb711-5"><a href="complete-protocols.html#cb711-5" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">x=</span>Month, <span class="at">y=</span>total_max<span class="sc">+</span><span class="dv">1200</span>, <span class="at">label=</span><span class="fu">paste0</span>(Year,<span class="st">&quot;: &quot;</span>,total_max)),</span>
<span id="cb711-6"><a href="complete-protocols.html#cb711-6" tabindex="-1"></a>             <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size=</span><span class="dv">4</span>,</span>
<span id="cb711-7"><a href="complete-protocols.html#cb711-7" tabindex="-1"></a>             <span class="co"># label.size = NA,</span></span>
<span id="cb711-8"><a href="complete-protocols.html#cb711-8" tabindex="-1"></a>             <span class="at">alpha=</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb711-9"><a href="complete-protocols.html#cb711-9" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="cn">NA</span>), <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>))</span></code></pre></div>
<pre><code>Scale for y is already present.
Adding another scale for y, which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb713"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb713-1"><a href="complete-protocols.html#cb713-1" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1096-1.png" width="672" /></p>
<p>Let’s also add a label that shows the current, maximal value. First, we define the dataframe:</p>
<div class="sourceCode" id="cb714"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb714-1"><a href="complete-protocols.html#cb714-1" tabindex="-1"></a>df_label_last <span class="ot">&lt;-</span> df_addgene_month <span class="sc">%&gt;%</span> <span class="fu">tail</span>(<span class="dv">1</span>)</span></code></pre></div>
<p>We use this dataframe to add a label to the endpoint:</p>
<div class="sourceCode" id="cb715"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb715-1"><a href="complete-protocols.html#cb715-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb715-2"><a href="complete-protocols.html#cb715-2" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>df_label_last,</span>
<span id="cb715-3"><a href="complete-protocols.html#cb715-3" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">x=</span>Month, <span class="at">y=</span>total_max), <span class="at">size=</span><span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb715-4"><a href="complete-protocols.html#cb715-4" tabindex="-1"></a>  <span class="fu">geom_label</span>(<span class="at">data=</span>df_label_last,</span>
<span id="cb715-5"><a href="complete-protocols.html#cb715-5" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">x=</span>Month, <span class="at">y=</span>total_max<span class="sc">+</span><span class="dv">2000</span>, <span class="at">label=</span><span class="fu">paste0</span>(total_max)),</span>
<span id="cb715-6"><a href="complete-protocols.html#cb715-6" tabindex="-1"></a>             <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size=</span><span class="dv">4</span>,</span>
<span id="cb715-7"><a href="complete-protocols.html#cb715-7" tabindex="-1"></a>             <span class="co"># label.size = NA,</span></span>
<span id="cb715-8"><a href="complete-protocols.html#cb715-8" tabindex="-1"></a>             <span class="at">alpha=</span><span class="dv">1</span>)</span>
<span id="cb715-9"><a href="complete-protocols.html#cb715-9" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1098-1.png" width="672" /></p>
<p>Finally, we can save the plot:</p>
<div class="sourceCode" id="cb716"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb716-1"><a href="complete-protocols.html#cb716-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_21.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb716-2"><a href="complete-protocols.html#cb716-2" tabindex="-1"></a> p</span>
<span id="cb716-3"><a href="complete-protocols.html#cb716-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-22" class="section level2 hasAnchor" number="4.22">
<h2><span class="header-section-number">4.22</span> Protocol 22 - Frequencies of discrete data<a href="complete-protocols.html#protocol-22" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this protocol we will look at the data supplied by <a href="http://www.addgene.org">addgene</a> on requests for plasmids that we have deposited. We first load the <code>{tidyverse}</code> package:</p>
<div class="sourceCode" id="cb718"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb718-1"><a href="complete-protocols.html#cb718-1" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>Let’s load the data:</p>
<div class="sourceCode" id="cb719"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb719-1"><a href="complete-protocols.html#cb719-1" tabindex="-1"></a>df_addgene <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;data/Addgene-Requests-for-Materials-Joachim-Goedhart-2005-2023-2023-Jan-31-05-05.csv&quot;</span>, <span class="at">stringsAsFactors =</span> <span class="cn">TRUE</span>)</span>
<span id="cb719-2"><a href="complete-protocols.html#cb719-2" tabindex="-1"></a><span class="fu">head</span>(df_addgene)</span></code></pre></div>
<pre><code>      ID                        Material Requesting.Country Date.Ordered
1 176106             mNeonGreen-Paxillin      UNITED STATES   01/25/2023
2 191450 CMVdel-dimericTomato-WASp(CRIB)             FRANCE   01/25/2023
3  36205              pmTurquoise2-Golgi             ISRAEL   01/25/2023
4 176102                iSH-iRFP670-sspB            GERMANY   01/20/2023
5 176113         SspB-HaloTag-DHPH-ITSN1            GERMANY   01/20/2023
6 176114         SspB-HaloTag-DHPH-TIAM1            GERMANY   01/20/2023</code></pre>
<p>There are several columns of data (note that the original data contains a column with names of the ‘Requesting PI’ which I have removed for privacy reasons). The <code>Material</code>, <code>Requesting.Country</code> and <code>Date.ordered</code> are of type <code>factor</code>. We can learn more about these factors with the function <code>str()</code></p>
<div class="sourceCode" id="cb721"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb721-1"><a href="complete-protocols.html#cb721-1" tabindex="-1"></a><span class="fu">str</span>(df_addgene)</span></code></pre></div>
<pre><code>&#39;data.frame&#39;:   3740 obs. of  4 variables:
 $ ID                : int  176106 191450 36205 176102 176113 176114 176116 176125 176130 191453 ...
 $ Material          : Factor w/ 107 levels &quot;3xmNeonGreen-3xrGBD&quot;,..: 36 4 73 26 98 99 100 29 62 43 ...
 $ Requesting.Country: Factor w/ 54 levels &quot;ARGENTINA&quot;,&quot;AUSTRALIA&quot;,..: 54 16 24 17 17 17 17 17 17 17 ...
 $ Date.Ordered      : Factor w/ 1655 levels &quot;01/01/2013&quot;,&quot;01/02/2013&quot;,..: 114 114 114 90 90 90 90 90 90 90 ...</code></pre>
<p>There are 3740 rows, which reflect the number of requests. There are 107 different values for <code>Material</code> and 54 unique countries. Now, the ID is a unique identifier of the material as described in the column <code>Material</code>.</p>
<div class="sourceCode" id="cb723"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb723-1"><a href="complete-protocols.html#cb723-1" tabindex="-1"></a>df_addgene <span class="ot">&lt;-</span> df_addgene <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">ID=</span><span class="fu">as.factor</span>(ID))</span>
<span id="cb723-2"><a href="complete-protocols.html#cb723-2" tabindex="-1"></a><span class="fu">head</span>(df_addgene)</span></code></pre></div>
<pre><code>      ID                        Material Requesting.Country Date.Ordered
1 176106             mNeonGreen-Paxillin      UNITED STATES   01/25/2023
2 191450 CMVdel-dimericTomato-WASp(CRIB)             FRANCE   01/25/2023
3  36205              pmTurquoise2-Golgi             ISRAEL   01/25/2023
4 176102                iSH-iRFP670-sspB            GERMANY   01/20/2023
5 176113         SspB-HaloTag-DHPH-ITSN1            GERMANY   01/20/2023
6 176114         SspB-HaloTag-DHPH-TIAM1            GERMANY   01/20/2023</code></pre>
<p>Let’s look at the number of requests for each plasmid:</p>
<div class="sourceCode" id="cb725"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb725-1"><a href="complete-protocols.html#cb725-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_addgene, <span class="fu">aes</span>(<span class="at">x=</span>Material)) <span class="sc">+</span> <span class="fu">geom_bar</span>()</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1126-1.png" width="672" /></p>
<p>Clearly, there are several popular plasmids and a large majority that has been requested only a few times. We generate a summary that shows has the count for each plasmid:</p>
<div class="sourceCode" id="cb726"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb726-1"><a href="complete-protocols.html#cb726-1" tabindex="-1"></a>df_count <span class="ot">&lt;-</span> df_addgene <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Material) <span class="sc">%&gt;%</span> <span class="fu">tally</span>(<span class="at">sort =</span> <span class="cn">TRUE</span>)</span>
<span id="cb726-2"><a href="complete-protocols.html#cb726-2" tabindex="-1"></a><span class="fu">head</span>(df_count)</span></code></pre></div>
<pre><code># A tibble: 6 × 2
  Material                   n
  &lt;fct&gt;                  &lt;int&gt;
1 pmTurquoise2-Mito        352
2 pmTurquoise2-Golgi       340
3 pPalmitoyl-mTurquoise2   340
4 pLifeAct-mTurquoise2     330
5 pmTurquoise2-ER          268
6 pmTurquoise2-Tubulin     207</code></pre>
<p>A column with the count data is generated. This can be used to sort the data:</p>
<div class="sourceCode" id="cb728"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb728-1"><a href="complete-protocols.html#cb728-1" tabindex="-1"></a>df_count <span class="ot">&lt;-</span> df_count <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Material =</span> <span class="fu">fct_reorder</span>(Material, n))</span>
<span id="cb728-2"><a href="complete-protocols.html#cb728-2" tabindex="-1"></a><span class="fu">head</span>(df_count)</span></code></pre></div>
<pre><code># A tibble: 6 × 2
  Material                   n
  &lt;fct&gt;                  &lt;int&gt;
1 pmTurquoise2-Mito        352
2 pmTurquoise2-Golgi       340
3 pPalmitoyl-mTurquoise2   340
4 pLifeAct-mTurquoise2     330
5 pmTurquoise2-ER          268
6 pmTurquoise2-Tubulin     207</code></pre>
<p>The resulting dataframe shows the sorted count <code>n</code> for each <code>Material</code>. We can use these values in combination with <code>geom_col()</code> to plot these data.</p>
<div class="sourceCode" id="cb730"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb730-1"><a href="complete-protocols.html#cb730-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_count, <span class="fu">aes</span>(<span class="at">x=</span>Material, <span class="at">y=</span>n)) <span class="sc">+</span> <span class="fu">geom_col</span>()</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1129-1.png" width="672" /></p>
<p>Ok, this looks better, but it’s probably more informative to look at a subset. Let’s say anything that has more than 10 requests. We also rotate the plot, which makes the plasmid names easier to read:</p>
<div class="sourceCode" id="cb731"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb731-1"><a href="complete-protocols.html#cb731-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_count <span class="sc">%&gt;%</span> <span class="fu">filter</span>(n<span class="sc">&gt;</span><span class="dv">10</span>), <span class="fu">aes</span>(<span class="at">x=</span>Material, <span class="at">y=</span>n)) <span class="sc">+</span> <span class="fu">geom_col</span>() <span class="sc">+</span> <span class="fu">coord_flip</span>()</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1130-1.png" width="672" /></p>
<p>Addgene uses different colors of flames to indicate the popularity of the plasmids. For more than 20 request a red flame, more than 50 requests is a green flame and more than 100 requests is a blue flame. We can use these categories and colors also in the bar chart, but we first need to add this to the dataframe:</p>
<div class="sourceCode" id="cb732"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb732-1"><a href="complete-protocols.html#cb732-1" tabindex="-1"></a>df_count <span class="ot">&lt;-</span> df_count <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">flame =</span> <span class="fu">case_when</span>(n<span class="sc">&lt;</span><span class="dv">20</span> <span class="sc">~</span> <span class="st">&quot;grey80&quot;</span>,</span>
<span id="cb732-2"><a href="complete-protocols.html#cb732-2" tabindex="-1"></a>                                                  (n<span class="sc">&gt;=</span><span class="dv">20</span> <span class="sc">&amp;</span> n<span class="sc">&lt;</span><span class="dv">50</span>) <span class="sc">~</span> <span class="st">&quot;#D96C6D&quot;</span>,  <span class="co">#using a &#39;softer red&#39;</span></span>
<span id="cb732-3"><a href="complete-protocols.html#cb732-3" tabindex="-1"></a>                                                  (n<span class="sc">&gt;=</span><span class="dv">50</span> <span class="sc">&amp;</span> n <span class="sc">&lt;</span><span class="dv">100</span>) <span class="sc">~</span> <span class="st">&quot;#E19E37&quot;</span>,</span>
<span id="cb732-4"><a href="complete-protocols.html#cb732-4" tabindex="-1"></a>                                                  n<span class="sc">&gt;=</span><span class="dv">100</span> <span class="sc">~</span> <span class="st">&quot;#50B0E8&quot;</span>)</span>
<span id="cb732-5"><a href="complete-protocols.html#cb732-5" tabindex="-1"></a>                                )</span></code></pre></div>
<p>Let’s replot these data:</p>
<div class="sourceCode" id="cb733"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb733-1"><a href="complete-protocols.html#cb733-1" tabindex="-1"></a><span class="fu">ggplot</span>(df_count <span class="sc">%&gt;%</span> <span class="fu">filter</span>(n<span class="sc">&gt;</span><span class="dv">15</span>), <span class="fu">aes</span>(<span class="at">x=</span>Material, <span class="at">y=</span>n, <span class="at">fill=</span>flame)) <span class="sc">+</span> <span class="fu">geom_col</span>() <span class="sc">+</span> <span class="fu">coord_flip</span>()</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1132-1.png" width="672" /></p>
<p>The colors do not match with the data in the <code>flame</code> column. To use the actual colors from this column, we need the function <code>scale_fill_identity()</code>:</p>
<div class="sourceCode" id="cb734"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb734-1"><a href="complete-protocols.html#cb734-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_count <span class="sc">%&gt;%</span> <span class="fu">filter</span>(n<span class="sc">&gt;</span><span class="dv">15</span>), <span class="fu">aes</span>(<span class="at">x=</span>Material, <span class="at">y=</span>n, <span class="at">fill=</span>flame)) <span class="sc">+</span></span>
<span id="cb734-2"><a href="complete-protocols.html#cb734-2" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span></span>
<span id="cb734-3"><a href="complete-protocols.html#cb734-3" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb734-4"><a href="complete-protocols.html#cb734-4" tabindex="-1"></a>  <span class="fu">scale_fill_identity</span>()</span>
<span id="cb734-5"><a href="complete-protocols.html#cb734-5" tabindex="-1"></a></span>
<span id="cb734-6"><a href="complete-protocols.html#cb734-6" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1133-1.png" width="672" /></p>
<p>Let’s tweak the layout:</p>
<div class="sourceCode" id="cb735"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb735-1"><a href="complete-protocols.html#cb735-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">12</span>) <span class="sc">+</span></span>
<span id="cb735-2"><a href="complete-protocols.html#cb735-2" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">&quot;&quot;</span>,</span>
<span id="cb735-3"><a href="complete-protocols.html#cb735-3" tabindex="-1"></a>       <span class="at">y=</span><span class="st">&quot;Number of requests&quot;</span>,</span>
<span id="cb735-4"><a href="complete-protocols.html#cb735-4" tabindex="-1"></a>       <span class="at">title=</span><span class="st">&quot;Requests for our plasmids at Addgene.org&quot;</span>,</span>
<span id="cb735-5"><a href="complete-protocols.html#cb735-5" tabindex="-1"></a>       <span class="at">caption=</span><span class="st">&quot;@joachimgoedhart | Data from addgene&quot;</span>) <span class="sc">+</span></span>
<span id="cb735-6"><a href="complete-protocols.html#cb735-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>) <span class="sc">+</span></span>
<span id="cb735-7"><a href="complete-protocols.html#cb735-7" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb735-8"><a href="complete-protocols.html#cb735-8" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color=</span><span class="st">&#39;grey80&#39;</span>, <span class="at">hjust=</span><span class="dv">1</span>),</span>
<span id="cb735-9"><a href="complete-protocols.html#cb735-9" tabindex="-1"></a>        <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb735-10"><a href="complete-protocols.html#cb735-10" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb735-11"><a href="complete-protocols.html#cb735-11" tabindex="-1"></a>        <span class="at">panel.grid.major.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb735-12"><a href="complete-protocols.html#cb735-12" tabindex="-1"></a>        <span class="at">axis.ticks.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb735-13"><a href="complete-protocols.html#cb735-13" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">8</span>))</span>
<span id="cb735-14"><a href="complete-protocols.html#cb735-14" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-1134-1.png" width="672" /></p>
<p>Note that the font size for labels on the vertical axis is rather small, to avoid overlap between the labels. The text size can only be invreased when the number of labels is reduced (or when the plot area is increased).</p>
<p>To save the plot:</p>
<div class="sourceCode" id="cb736"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb736-1"><a href="complete-protocols.html#cb736-1" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_22.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb736-2"><a href="complete-protocols.html#cb736-2" tabindex="-1"></a> p</span>
<span id="cb736-3"><a href="complete-protocols.html#cb736-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
<p>Note that this plot is also used on the addgene dashboard that is available here:</p>
<p><a href="https://amsterdamstudygroup.shinyapps.io/Addgene_dashboard/" class="uri">https://amsterdamstudygroup.shinyapps.io/Addgene_dashboard/</a></p>
<p>You can upload your own addgene data, which will a plot similar to the one that is described here.</p>
<hr />
<div id="in-progress" class="section level4 unnumbered hasAnchor">
<h4>In progress<a href="complete-protocols.html#in-progress" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>This chapter is ‘in progress’ and I plan to add more protocols in the future.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="plotting-the-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="questions-and-answers.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": false
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
