<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Complete protocols | DataViz protocols</title>
  <meta name="description" content="This is an introduction to data visualization protocols for wet lab scientists" />
  <meta name="generator" content="bookdown 0.27 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Complete protocols | DataViz protocols" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is an introduction to data visualization protocols for wet lab scientists" />
  <meta name="github-repo" content="JoachimGoedhart/DataViz-protocols" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Complete protocols | DataViz protocols" />
  
  <meta name="twitter:description" content="This is an introduction to data visualization protocols for wet lab scientists" />
  

<meta name="author" content="Joachim Goedhart" />


<meta name="date" content="2022-07-05" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="plotting-the-data.html"/>
<link rel="next" href="questions-and-answers.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>

<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-ZDBZZ5FR0E"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-ZDBZZ5FR0E');
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DataViz Protocols</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#a-toast"><i class="fa fa-check"></i>A toast</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>1</b> Getting started with R</a>
<ul>
<li class="chapter" data-level="1.1" data-path="getting-started.html"><a href="getting-started.html#running-r"><i class="fa fa-check"></i><b>1.1</b> Running R</a></li>
<li class="chapter" data-level="1.2" data-path="getting-started.html"><a href="getting-started.html#using-the-command-line"><i class="fa fa-check"></i><b>1.2</b> Using the command line</a></li>
<li class="chapter" data-level="1.3" data-path="getting-started.html"><a href="getting-started.html#help"><i class="fa fa-check"></i><b>1.3</b> ?Help</a></li>
<li class="chapter" data-level="1.4" data-path="getting-started.html"><a href="getting-started.html#installing-packages"><i class="fa fa-check"></i><b>1.4</b> Installing packages</a></li>
<li class="chapter" data-level="1.5" data-path="getting-started.html"><a href="getting-started.html#multiline-code"><i class="fa fa-check"></i><b>1.5</b> Multiline code</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="read-and-reshape.html"><a href="read-and-reshape.html"><i class="fa fa-check"></i><b>2</b> Reading and reshaping data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="read-and-reshape.html"><a href="read-and-reshape.html#introduction"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="read-and-reshape.html"><a href="read-and-reshape.html#types-of-data"><i class="fa fa-check"></i><b>2.2</b> Types of data</a></li>
<li class="chapter" data-level="2.3" data-path="read-and-reshape.html"><a href="read-and-reshape.html#reading-data"><i class="fa fa-check"></i><b>2.3</b> Reading data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="read-and-reshape.html"><a href="read-and-reshape.html#loading-data-from-a-text-or-csv-file"><i class="fa fa-check"></i><b>2.3.1</b> Loading data from a text or csv file</a></li>
<li class="chapter" data-level="2.3.2" data-path="read-and-reshape.html"><a href="read-and-reshape.html#loading-data-from-a-url"><i class="fa fa-check"></i><b>2.3.2</b> Loading data from a URL</a></li>
<li class="chapter" data-level="2.3.3" data-path="read-and-reshape.html"><a href="read-and-reshape.html#retrieving-data-from-excel"><i class="fa fa-check"></i><b>2.3.3</b> Retrieving data from Excel</a></li>
<li class="chapter" data-level="2.3.4" data-path="read-and-reshape.html"><a href="read-and-reshape.html#retrieving-data-from-multiple-files"><i class="fa fa-check"></i><b>2.3.4</b> Retrieving data from multiple files</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="read-and-reshape.html"><a href="read-and-reshape.html#tidy-data"><i class="fa fa-check"></i><b>2.4</b> Reshaping data</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="read-and-reshape.html"><a href="read-and-reshape.html#quantitative-data-discrete-conditions"><i class="fa fa-check"></i><b>2.4.1</b> Quantitative data, discrete conditions</a></li>
<li class="chapter" data-level="2.4.2" data-path="read-and-reshape.html"><a href="read-and-reshape.html#multiple-discrete-conditions"><i class="fa fa-check"></i><b>2.4.2</b> Multiple discrete conditions</a></li>
<li class="chapter" data-level="2.4.3" data-path="read-and-reshape.html"><a href="read-and-reshape.html#double-quantitative-data"><i class="fa fa-check"></i><b>2.4.3</b> Double quantitative data</a></li>
<li class="chapter" data-level="2.4.4" data-path="read-and-reshape.html"><a href="read-and-reshape.html#data-from-multiple-files"><i class="fa fa-check"></i><b>2.4.4</b> Data from multiple files</a></li>
<li class="chapter" data-level="2.4.5" data-path="read-and-reshape.html"><a href="read-and-reshape.html#data-in-96-wells-format-data-in-96-wells-format"><i class="fa fa-check"></i><b>2.4.5</b> Data in 96-wells format {data-in-96-wells-format}</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="plotting-the-data.html"><a href="plotting-the-data.html"><i class="fa fa-check"></i><b>3</b> Plotting the Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="plotting-the-data.html"><a href="plotting-the-data.html#data-over-time-continuous-vs.-continuous"><i class="fa fa-check"></i><b>3.1</b> Data over time (continuous vs. continuous)</a></li>
<li class="chapter" data-level="3.2" data-path="plotting-the-data.html"><a href="plotting-the-data.html#discrete-conditions"><i class="fa fa-check"></i><b>3.2</b> Discrete conditions</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="plotting-the-data.html"><a href="plotting-the-data.html#x-axis-data-qualitative-versus-quantitative-data"><i class="fa fa-check"></i><b>3.2.1</b> X-axis data: qualitative versus quantitative data</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="plotting-the-data.html"><a href="plotting-the-data.html#statistics"><i class="fa fa-check"></i><b>3.3</b> Statistics</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="plotting-the-data.html"><a href="plotting-the-data.html#introduction-1"><i class="fa fa-check"></i><b>3.3.1</b> Introduction</a></li>
<li class="chapter" data-level="3.3.2" data-path="plotting-the-data.html"><a href="plotting-the-data.html#data-summaries-directly-added-as-a-plot-layer"><i class="fa fa-check"></i><b>3.3.2</b> Data summaries directly added as a plot layer</a></li>
<li class="chapter" data-level="3.3.3" data-path="plotting-the-data.html"><a href="plotting-the-data.html#data-summaries-from-a-dataframe"><i class="fa fa-check"></i><b>3.3.3</b> Data summaries from a dataframe</a></li>
<li class="chapter" data-level="3.3.4" data-path="plotting-the-data.html"><a href="plotting-the-data.html#data-summaries-for-continuous-x-axis-data"><i class="fa fa-check"></i><b>3.3.4</b> Data summaries for continuous x-axis data</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="plotting-the-data.html"><a href="plotting-the-data.html#plot-a-lot---discrete-data"><i class="fa fa-check"></i><b>3.4</b> Plot-a-lot - discrete data</a></li>
<li class="chapter" data-level="3.5" data-path="plotting-the-data.html"><a href="plotting-the-data.html#optimizing-the-data-visualization"><i class="fa fa-check"></i><b>3.5</b> Optimizing the data visualization</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="plotting-the-data.html"><a href="plotting-the-data.html#rotation"><i class="fa fa-check"></i><b>3.5.1</b> Rotation</a></li>
<li class="chapter" data-level="3.5.2" data-path="plotting-the-data.html"><a href="plotting-the-data.html#ordering-conditions"><i class="fa fa-check"></i><b>3.5.2</b> Ordering conditions</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="plotting-the-data.html"><a href="plotting-the-data.html#adjusting-the-layout"><i class="fa fa-check"></i><b>3.6</b> Adjusting the layout</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="plotting-the-data.html"><a href="plotting-the-data.html#themes"><i class="fa fa-check"></i><b>3.6.1</b> Themes</a></li>
<li class="chapter" data-level="3.6.2" data-path="plotting-the-data.html"><a href="plotting-the-data.html#legend"><i class="fa fa-check"></i><b>3.6.2</b> Legend</a></li>
<li class="chapter" data-level="3.6.3" data-path="plotting-the-data.html"><a href="plotting-the-data.html#grids"><i class="fa fa-check"></i><b>3.6.3</b> Grids</a></li>
<li class="chapter" data-level="3.6.4" data-path="plotting-the-data.html"><a href="plotting-the-data.html#labelstitles"><i class="fa fa-check"></i><b>3.6.4</b> Labels/Titles</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="plotting-the-data.html"><a href="plotting-the-data.html#plot-a-lot---continuous-data"><i class="fa fa-check"></i><b>3.7</b> Plot-a-lot - continuous data</a>
<ul>
<li class="chapter" data-level="3.7.1" data-path="plotting-the-data.html"><a href="plotting-the-data.html#small-multiples"><i class="fa fa-check"></i><b>3.7.1</b> Small multiples</a></li>
<li class="chapter" data-level="3.7.2" data-path="plotting-the-data.html"><a href="plotting-the-data.html#heatmaps"><i class="fa fa-check"></i><b>3.7.2</b> Heatmaps</a></li>
<li class="chapter" data-level="3.7.3" data-path="plotting-the-data.html"><a href="plotting-the-data.html#custom-objects"><i class="fa fa-check"></i><b>3.7.3</b> Custom objects</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="complete-protocols.html"><a href="complete-protocols.html"><i class="fa fa-check"></i><b>4</b> Complete protocols</a>
<ul>
<li class="chapter" data-level="4.1" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-1"><i class="fa fa-check"></i><b>4.1</b> Protocol 1 - Spectra of fluorescent proteins</a></li>
<li class="chapter" data-level="4.2" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-2"><i class="fa fa-check"></i><b>4.2</b> Protocol 2 - A superplot of calcium concentrations</a></li>
<li class="chapter" data-level="4.3" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-3"><i class="fa fa-check"></i><b>4.3</b> Protocol 3 - small multiples of time courses</a></li>
<li class="chapter" data-level="4.4" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-4"><i class="fa fa-check"></i><b>4.4</b> Protocol 4 - Plotting data from a 96-wells experiment</a></li>
<li class="chapter" data-level="4.5" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-5"><i class="fa fa-check"></i><b>4.5</b> Protocol 5 - A map of amino acids</a></li>
<li class="chapter" data-level="4.6" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-6"><i class="fa fa-check"></i><b>4.6</b> Protocol 6 - Heatmap style visualization of timelapse data</a></li>
<li class="chapter" data-level="4.7" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-7"><i class="fa fa-check"></i><b>4.7</b> Protocol 7 - Ridgeline plot</a></li>
<li class="chapter" data-level="4.8" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-8"><i class="fa fa-check"></i><b>4.8</b> Protocol 8 - Plotting data in a 96-wells layout</a></li>
<li class="chapter" data-level="4.9" data-path="complete-protocols.html"><a href="complete-protocols.html#protocol-9"><i class="fa fa-check"></i><b>4.9</b> Protocol 9 - A dose response curve with fit</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="questions-and-answers.html"><a href="questions-and-answers.html"><i class="fa fa-check"></i><b>5</b> Questions and Answers</a>
<ul>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#what-is-the-difference-between-a-data.frame-and-a-tibble"><i class="fa fa-check"></i>What is the difference between a data.frame and a tibble?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#whats-up-with-spaces-in-variable-names-and-in-files-names"><i class="fa fa-check"></i>What’s up with spaces in variable names and in files names?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#what-are-the-rules-for-naming-dataframes-or-variables"><i class="fa fa-check"></i>What are the rules for naming dataframes or variables?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#what-is-the-difference-between---and-for-assigning-variables"><i class="fa fa-check"></i>What is the difference between <code>&lt;-</code> and <code>=</code> for assigning variables?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#what-does-na-mean-in-a-dataframe"><i class="fa fa-check"></i>What does ‘NA’ mean in a dataframe?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#what-is-the-beste-way-to-type-a-operator"><i class="fa fa-check"></i>What is the beste way to type a ‘%&gt;%’ operator?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#where-do-i-find-the-example-data"><i class="fa fa-check"></i>Where do I find the example data?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#is-there-a-way-to-re-use-or-adjust-the-protocols"><i class="fa fa-check"></i>Is there a way to re-use or adjust the protocols?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#which-packages-are-included-in-the-tidyverse-package"><i class="fa fa-check"></i>Which packages are included in the <code>{tidyverse}</code> package?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#what-is-the-difference-between-require-and-library-for-loading-a-package"><i class="fa fa-check"></i>What is the difference between <code>require()</code> and <code>library()</code> for loading a package?</a></li>
<li class="chapter" data-level="" data-path="questions-and-answers.html"><a href="questions-and-answers.html#is-it-possible-to-have-interactive-file-selection"><i class="fa fa-check"></i>Is it possible to have interactive file selection?</a></li>
<li><a href="questions-and-answers.html#section"></a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DataViz protocols</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="complete-protocols" class="section level1 hasAnchor" number="4">
<h1><span class="header-section-number">Chapter 4</span> Complete protocols<a href="complete-protocols.html#complete-protocols" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>This chapter showcases several <em>complete</em> protocols for different kinds of data visualizations. Each protocol starts with the raw data and ends with a publication quality plot. The data is available from Github, just follow <a href="https://github.com/JoachimGoedhart/DataViz-protocols/tree/main/Protocols">this link</a>. In the same folder you can find the R markdown files for each protocol. To reproduce the data visualizations you can either take the data and follow the instructions in the chapter. Alternatively you can download the R markdown file (and the data) and run it step by step. The R markdown files can also be used as a starting point to apply the same visualization to your own data.</p>
<p>An overview of the data visualizations that are generated by the protocols:</p>
<p><img src="Protocols/overview_of_protocols.png" /></p>
<div id="protocol-1" class="section level2 hasAnchor" number="4.1">
<h2><span class="header-section-number">4.1</span> Protocol 1 - Spectra of fluorescent proteins<a href="complete-protocols.html#protocol-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This protocol describes how you can turn a csv with spectral data that is obtained from FPbase.org into a plot of those spectra. First, we load the required package:</p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="complete-protocols.html#cb206-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>For this data visualization, I selected several spectra from fluorescent proteins at FPbase.org: <a href="https://www.fpbase.org/spectra/?s=1746,6551,101,102,123,124,1604,1606&amp;showY=0&amp;showX=1&amp;showGrid=0&amp;areaFill=1&amp;logScale=0&amp;scaleEC=0&amp;scaleQY=0&amp;shareTooltip=1&amp;palette=wavelength">https://www.fpbase.org/spectra/?s=1746,6551,101,102,123,124,1604,1606&amp;showY=0&amp;showX=1&amp;showGrid=0&amp;areaFill=1&amp;logScale=0&amp;scaleEC=0&amp;scaleQY=0&amp;shareTooltip=1&amp;palette=wavelength</a></p>
<p>The data was downloaded in CSV format (by clicking on the button in the lower right corner of the webpage) and renamed to ‘FPbase_Spectra_4FPs.csv’.
We read the data from the CSV by using the <code>read_csv()</code> function. This function is part of the tidy verse and loads the data as a tibble. It also guesses type of data for each column. To hide that information, we use <code>show_col_types = FALSE</code> here.</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="complete-protocols.html#cb207-1" aria-hidden="true" tabindex="-1"></a>df_raw <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;FPbase_Spectra_4FPs.csv&quot;</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Let’s briefly look at what we have loaded:</p>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="complete-protocols.html#cb208-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(df_raw)</span></code></pre></div>
<pre><code>Rows: 512
Columns: 9
$ Wavelength       &lt;dbl&gt; 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310…
$ `mTurquoise2 EM` &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `mTurquoise2 EX` &lt;dbl&gt; 0.2484, 0.2266, 0.2048, 0.1852, 0.1634, 0.1482, 0.132…
$ `mNeonGreen EM`  &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `mNeonGreen EX`  &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `mScarlet-I EM`  &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `mScarlet-I EX`  &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `miRFP670 EM`    &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `miRFP670 EX`    &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…</code></pre>
<p>The data needs conversion to a tidy format before plotting. Since we have a single continuous data column with Wavelength information that is used for the x-axis, it is excluded from the operation:</p>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb210-1"><a href="complete-protocols.html#cb210-1" aria-hidden="true" tabindex="-1"></a>df_1 <span class="ot">&lt;-</span></span>
<span id="cb210-2"><a href="complete-protocols.html#cb210-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb210-3"><a href="complete-protocols.html#cb210-3" aria-hidden="true" tabindex="-1"></a>    df_raw,</span>
<span id="cb210-4"><a href="complete-protocols.html#cb210-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="sc">-</span>Wavelength,</span>
<span id="cb210-5"><a href="complete-protocols.html#cb210-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">&quot;Sample&quot;</span>,</span>
<span id="cb210-6"><a href="complete-protocols.html#cb210-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">&quot;Intensity&quot;</span></span>
<span id="cb210-7"><a href="complete-protocols.html#cb210-7" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>There are several rows that have NA values for Intensity and this is how to get rid of that:</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="complete-protocols.html#cb211-1" aria-hidden="true" tabindex="-1"></a>df_1 <span class="ot">&lt;-</span> df_1 <span class="sc">%&gt;%</span> <span class="fu">drop_na</span>(Intensity)</span></code></pre></div>
<p>The column ‘Sample’ has labels for the fluorescent protein and the type of spectrum. We can separate that column into two different columns that we name ‘Fluorescent Protein’ and ‘Spectrum’:</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="complete-protocols.html#cb212-1" aria-hidden="true" tabindex="-1"></a>df_1 <span class="ot">&lt;-</span> df_1 <span class="sc">%&gt;%</span></span>
<span id="cb212-2"><a href="complete-protocols.html#cb212-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(Sample, <span class="fu">c</span>(<span class="st">&quot;Fluorescent Protein&quot;</span>, <span class="st">&quot;Spectrum&quot;</span>), <span class="at">sep =</span> <span class="st">&quot; &quot;</span>)</span></code></pre></div>
<p>Let’s do a first attempt and plot the data:</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="complete-protocols.html#cb213-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb213-2"><a href="complete-protocols.html#cb213-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> df_1,</span>
<span id="cb213-3"><a href="complete-protocols.html#cb213-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">x =</span> Wavelength, <span class="at">y =</span> Intensity, <span class="at">color =</span> <span class="st">`</span><span class="at">Fluorescent Protein</span><span class="st">`</span>)</span>
<span id="cb213-4"><a href="complete-protocols.html#cb213-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb213-5"><a href="complete-protocols.html#cb213-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">linetype =</span> Spectrum), <span class="at">size =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-177-1.png" width="672" /></p>
<p>This looks pretty good already. Now let’s change the order of the fluorescent proteins to their order in the plot:</p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="complete-protocols.html#cb214-1" aria-hidden="true" tabindex="-1"></a>df_1 <span class="ot">&lt;-</span> df_1 <span class="sc">%&gt;%</span></span>
<span id="cb214-2"><a href="complete-protocols.html#cb214-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">`</span><span class="at">Fluorescent Protein</span><span class="st">`</span> <span class="ot">=</span> forcats<span class="sc">::</span><span class="fu">fct_relevel</span>(</span>
<span id="cb214-3"><a href="complete-protocols.html#cb214-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">Fluorescent Protein</span><span class="st">`</span>,</span>
<span id="cb214-4"><a href="complete-protocols.html#cb214-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(<span class="st">&quot;mTurquoise2&quot;</span>, <span class="st">&quot;mNeonGreen&quot;</span>, <span class="st">&quot;mScarlet-I&quot;</span>, <span class="st">&quot;miRFP670&quot;</span>)</span>
<span id="cb214-5"><a href="complete-protocols.html#cb214-5" aria-hidden="true" tabindex="-1"></a>  ))</span></code></pre></div>
<p>The data is in the right shape now, so let’s save it:</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="complete-protocols.html#cb215-1" aria-hidden="true" tabindex="-1"></a>df_1 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_1.csv&quot;</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>We define the plot object and add an extra geometry, <code>geom_area()</code> to fill the area under the curves:</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="complete-protocols.html#cb216-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb216-2"><a href="complete-protocols.html#cb216-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb216-3"><a href="complete-protocols.html#cb216-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> df_1,</span>
<span id="cb216-4"><a href="complete-protocols.html#cb216-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb216-5"><a href="complete-protocols.html#cb216-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> Wavelength, <span class="at">y =</span> Intensity,</span>
<span id="cb216-6"><a href="complete-protocols.html#cb216-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">fill =</span> <span class="st">`</span><span class="at">Fluorescent Protein</span><span class="st">`</span></span>
<span id="cb216-7"><a href="complete-protocols.html#cb216-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb216-8"><a href="complete-protocols.html#cb216-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb216-9"><a href="complete-protocols.html#cb216-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">linetype =</span> Spectrum),</span>
<span id="cb216-10"><a href="complete-protocols.html#cb216-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.5</span></span>
<span id="cb216-11"><a href="complete-protocols.html#cb216-11" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb216-12"><a href="complete-protocols.html#cb216-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_area</span>(</span>
<span id="cb216-13"><a href="complete-protocols.html#cb216-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">linetype =</span> Spectrum),</span>
<span id="cb216-14"><a href="complete-protocols.html#cb216-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="cn">NA</span>,</span>
<span id="cb216-15"><a href="complete-protocols.html#cb216-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="st">&quot;identity&quot;</span>,</span>
<span id="cb216-16"><a href="complete-protocols.html#cb216-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="dv">1</span>,</span>
<span id="cb216-17"><a href="complete-protocols.html#cb216-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha =</span> <span class="fl">0.5</span></span>
<span id="cb216-18"><a href="complete-protocols.html#cb216-18" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>Let’s check the result:</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="complete-protocols.html#cb217-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-181-1.png" width="672" /></p>
<p>Next, we set the limits of the axis and force the y-axis to start at 0</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="complete-protocols.html#cb218-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb218-2"><a href="complete-protocols.html#cb218-2" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">1.1</span>)) <span class="sc">+</span> </span>
<span id="cb218-3"><a href="complete-protocols.html#cb218-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">350</span>, <span class="dv">810</span>))</span></code></pre></div>
<p>Add labels:</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="complete-protocols.html#cb219-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb219-2"><a href="complete-protocols.html#cb219-2" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb219-3"><a href="complete-protocols.html#cb219-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Spectra of Fluorescent Proteins&quot;</span>,</span>
<span id="cb219-4"><a href="complete-protocols.html#cb219-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Wavelength [nm]&quot;</span>,</span>
<span id="cb219-5"><a href="complete-protocols.html#cb219-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Normalized Intensity [a.u.]&quot;</span>,</span>
<span id="cb219-6"><a href="complete-protocols.html#cb219-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">(based on data from FPbase.org)&quot;</span>,</span>
<span id="cb219-7"><a href="complete-protocols.html#cb219-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 1&quot;</span></span>
<span id="cb219-8"><a href="complete-protocols.html#cb219-8" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>Modify the layout by adjusting the theme. Comments are used to explain effect of the individual lines of code:</p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="complete-protocols.html#cb220-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb220-2"><a href="complete-protocols.html#cb220-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Set text size</span></span>
<span id="cb220-3"><a href="complete-protocols.html#cb220-3" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span> <span class="fu">theme</span>(</span>
<span id="cb220-4"><a href="complete-protocols.html#cb220-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">plot.caption =</span> <span class="fu">element_text</span>(</span>
<span id="cb220-5"><a href="complete-protocols.html#cb220-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>,</span>
<span id="cb220-6"><a href="complete-protocols.html#cb220-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">hjust =</span> <span class="dv">1</span></span>
<span id="cb220-7"><a href="complete-protocols.html#cb220-7" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb220-8"><a href="complete-protocols.html#cb220-8" aria-hidden="true" tabindex="-1"></a>    <span class="co">#Remove grid</span></span>
<span id="cb220-9"><a href="complete-protocols.html#cb220-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb220-10"><a href="complete-protocols.html#cb220-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb220-11"><a href="complete-protocols.html#cb220-11" aria-hidden="true" tabindex="-1"></a>    <span class="co">#Set position of legend</span></span>
<span id="cb220-12"><a href="complete-protocols.html#cb220-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.position =</span> <span class="st">&quot;top&quot;</span>,</span>
<span id="cb220-13"><a href="complete-protocols.html#cb220-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.justification =</span> <span class="st">&quot;left&quot;</span></span>
<span id="cb220-14"><a href="complete-protocols.html#cb220-14" aria-hidden="true" tabindex="-1"></a>    <span class="co">#Define the legend layout</span></span>
<span id="cb220-15"><a href="complete-protocols.html#cb220-15" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> <span class="fu">guides</span>(</span>
<span id="cb220-16"><a href="complete-protocols.html#cb220-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">linetype =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb220-17"><a href="complete-protocols.html#cb220-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="cn">NULL</span>, <span class="at">label.position =</span> <span class="st">&quot;right&quot;</span>)</span>
<span id="cb220-18"><a href="complete-protocols.html#cb220-18" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="complete-protocols.html#cb221-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-185-1.png" width="672" /></p>
<p>We are almost there, except that the colors of the plot do not match with the natural colors of the fluorescent proteins. Let’s fix that by defining a custom color palette. The order of the colors matches with the order of the fluorescent proteins that was defined earlier:</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="complete-protocols.html#cb222-1" aria-hidden="true" tabindex="-1"></a>custom_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;green&quot;</span>, <span class="st">&quot;orange&quot;</span>, <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p>To apply the custom colors to the filled area:</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="complete-protocols.html#cb223-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> custom_colors)</span></code></pre></div>
<p>This is the result:</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="complete-protocols.html#cb224-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-188-1.png" width="672" /></p>
<p>To save this plot as a PNG file:</p>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="complete-protocols.html#cb225-1" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_1.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb225-2"><a href="complete-protocols.html#cb225-2" aria-hidden="true" tabindex="-1"></a>  p</span>
<span id="cb225-3"><a href="complete-protocols.html#cb225-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-2" class="section level2 hasAnchor" number="4.2">
<h2><span class="header-section-number">4.2</span> Protocol 2 - A superplot of calcium concentrations<a href="complete-protocols.html#protocol-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This protocol is used to create a superplot which differentiates between technical and biological replicates. The concept of superplots has been reported by Lord and colleagues <a href="https://doi.org/10.1083/jcb.202001064">(Lord et al., 2021)</a>. We will use the data that was used to create figure 5e in a publication by <a href="https://doi.org/10.1038/s41467-021-27249-w">van der Linden et al. (2021)</a>.</p>
<p>The figure in the publication summarizes the data from all experiments and does not identify the biological replicates. Below, we will differentiate the biological replicates, by treating each batch of neutrophils as a biological replicate.</p>
<p>We start by loading the required tidyverse package:</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="complete-protocols.html#cb227-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>We define the confidence level as 95%:</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="complete-protocols.html#cb228-1" aria-hidden="true" tabindex="-1"></a>Confidence_level <span class="ot">&lt;-</span>  <span class="fl">0.95</span></span></code></pre></div>
<p>The data is stored in an excel sheet and we read it, skipping the first 6 lines which contain comments:</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="complete-protocols.html#cb229-1" aria-hidden="true" tabindex="-1"></a>df_raw <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">&#39;figure5.xlsx&#39;</span>, <span class="at">skip=</span><span class="dv">6</span>)</span></code></pre></div>
<p>Let’s look at the data:</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="complete-protocols.html#cb230-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_raw)</span></code></pre></div>
<pre><code># A tibble: 6 × 7
  `Experimental day` `Replicate no.` `Neutrophil no.` `Batch no. neutrop…` Stage
               &lt;dbl&gt;           &lt;dbl&gt;            &lt;dbl&gt;                &lt;dbl&gt; &lt;chr&gt;
1                  1               1                1                    1 befo…
2                  1               1                1                    1 befo…
3                  1               1                1                    1 befo…
4                  1               1                1                    1 befo…
5                  1               1                1                    1 befo…
6                  1               1                1                    1 befo…
# … with 2 more variables: `dF/F0` &lt;chr&gt;, `Calcium (uM)` &lt;dbl&gt;</code></pre>
<p>The data is already in a tidy format. The column with ‘Stage’ has the four different conditions for which we will compare the data in the column ‘Calcium (uM)’. We change the name of the column ‘Batch no. neutrophils’ to ‘Replicate’ and make sure the different replicates are treated as factors (qualitative data):</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="complete-protocols.html#cb232-1" aria-hidden="true" tabindex="-1"></a>df_raw <span class="ot">&lt;-</span> df_raw <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Replicate =</span> <span class="fu">as.factor</span>(<span class="st">`</span><span class="at">Batch no. neutrophils</span><span class="st">`</span>))</span></code></pre></div>
<p>Let’s look at the data, and identify the biological replicates, as suggested in the original publication on Superplot by <a href="https://doi.org/10.1083/jcb.202001064">(Lord et al., 2021)</a>. In this example a color code is used to label the replicates:</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="complete-protocols.html#cb233-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>df_raw, <span class="fu">aes</span>(<span class="at">x=</span>Stage)) <span class="sc">+</span></span>
<span id="cb233-2"><a href="complete-protocols.html#cb233-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">data=</span>df_raw, <span class="fu">aes</span>(<span class="at">x=</span>Stage, <span class="at">y=</span><span class="st">`</span><span class="at">Calcium (uM)</span><span class="st">`</span>, <span class="at">color=</span>Replicate))</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-195-1.png" width="672" /></p>
<p>To display the statistics for the individual biological replicates, we define a new dataframe. To this end, we group the data for the different stages and biological replicates:</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="complete-protocols.html#cb234-1" aria-hidden="true" tabindex="-1"></a>df_summary <span class="ot">&lt;-</span> df_raw <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Stage, Replicate) <span class="sc">%&gt;%</span></span>
<span id="cb234-2"><a href="complete-protocols.html#cb234-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n=</span><span class="fu">n</span>(), <span class="at">mean=</span><span class="fu">mean</span>(<span class="st">`</span><span class="at">Calcium (uM)</span><span class="st">`</span>))</span></code></pre></div>
<pre><code>`summarise()` has grouped output by &#39;Stage&#39;. You can override using the
`.groups` argument.</code></pre>
<p>Next, we use ‘df_summary’ which holds the averages of each biological replicate, and we calculate the statistics for the different conditions:</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="complete-protocols.html#cb236-1" aria-hidden="true" tabindex="-1"></a>df_summary_replicas <span class="ot">&lt;-</span> df_summary <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Stage) <span class="sc">%&gt;%</span></span>
<span id="cb236-2"><a href="complete-protocols.html#cb236-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">n_rep=</span><span class="fu">n</span>(), <span class="at">mean_rep=</span><span class="fu">mean</span>(mean), <span class="at">sd_rep =</span> <span class="fu">sd</span>(mean))  <span class="sc">%&gt;%</span></span>
<span id="cb236-3"><a href="complete-protocols.html#cb236-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sem =</span> sd_rep <span class="sc">/</span> <span class="fu">sqrt</span>(n_rep <span class="sc">-</span> <span class="dv">1</span>),</span>
<span id="cb236-4"><a href="complete-protocols.html#cb236-4" aria-hidden="true" tabindex="-1"></a>         <span class="st">`</span><span class="at">95%CI_lo</span><span class="st">`</span> <span class="ot">=</span> mean_rep <span class="sc">+</span> <span class="fu">qt</span>((<span class="dv">1</span><span class="sc">-</span>Confidence_level)<span class="sc">/</span><span class="dv">2</span>, n_rep <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">*</span> sem,</span>
<span id="cb236-5"><a href="complete-protocols.html#cb236-5" aria-hidden="true" tabindex="-1"></a>         <span class="st">`</span><span class="at">95%CI_hi</span><span class="st">`</span> <span class="ot">=</span> mean_rep <span class="sc">-</span> <span class="fu">qt</span>((<span class="dv">1</span><span class="sc">-</span>Confidence_level)<span class="sc">/</span><span class="dv">2</span>, n_rep <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">*</span> sem,</span>
<span id="cb236-6"><a href="complete-protocols.html#cb236-6" aria-hidden="true" tabindex="-1"></a>         <span class="cn">NULL</span>)</span></code></pre></div>
<p>The dataframe has the summary of the conditions and note that each condition has a summary of 4 biological replicates:</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="complete-protocols.html#cb237-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_summary_replicas)</span></code></pre></div>
<pre><code># A tibble: 6 × 10
# Groups:   Stage [2]
  Stage    Replicate     n   mean n_rep mean_rep  sd_rep     sem `95%CI_lo`
  &lt;chr&gt;    &lt;fct&gt;     &lt;int&gt;  &lt;dbl&gt; &lt;int&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;
1 before   1            38 0.0305     4   0.0301 0.00414 0.00239     0.0225
2 before   2            67 0.0270     4   0.0301 0.00414 0.00239     0.0225
3 before   3            56 0.0358     4   0.0301 0.00414 0.00239     0.0225
4 before   4            55 0.0270     4   0.0301 0.00414 0.00239     0.0225
5 crawling 1             7 0.0296     4   0.0317 0.00339 0.00196     0.0255
6 crawling 2            29 0.0289     4   0.0317 0.00339 0.00196     0.0255
# … with 1 more variable: `95%CI_hi` &lt;dbl&gt;</code></pre>
<p>We can now add or ‘bind’ the data of ‘df_summary_replicas’ to the original dataframe ‘df’ and store this as a dataframe ‘df_2’:</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="complete-protocols.html#cb239-1" aria-hidden="true" tabindex="-1"></a>df_2 <span class="ot">&lt;-</span> df_raw <span class="sc">%&gt;%</span> <span class="fu">left_join</span>(df_summary_replicas, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;Stage&quot;</span>,<span class="st">&quot;Replicate&quot;</span>))</span></code></pre></div>
<p>Let’s save this data:</p>
<div class="sourceCode" id="cb240"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb240-1"><a href="complete-protocols.html#cb240-1" aria-hidden="true" tabindex="-1"></a>df_2 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_2.csv&quot;</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>Let’s first define a basic plot with all of the data for each stage shown as a violinplot:</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="complete-protocols.html#cb241-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>df_2, <span class="fu">aes</span>(<span class="at">x=</span>Stage)) <span class="sc">+</span></span>
<span id="cb241-2"><a href="complete-protocols.html#cb241-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">data=</span>df_2, <span class="fu">aes</span>(<span class="at">x=</span>Stage, <span class="at">y=</span><span class="st">`</span><span class="at">Calcium (uM)</span><span class="st">`</span>), <span class="at">color=</span><span class="cn">NA</span>, <span class="at">fill=</span><span class="st">&quot;grey80&quot;</span>)</span></code></pre></div>
<p>This is what it looks like:</p>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="complete-protocols.html#cb242-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-202-1.png" width="672" /></p>
<p>We add the 95% confidence interval from the summary of the biological replicates as a line:</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="complete-protocols.html#cb243-1" aria-hidden="true" tabindex="-1"></a> p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_linerange</span>(<span class="at">data =</span> df_2, <span class="fu">aes</span>(<span class="at">ymin=</span><span class="st">`</span><span class="at">95%CI_lo</span><span class="st">`</span>, <span class="at">ymax=</span><span class="st">`</span><span class="at">95%CI_hi</span><span class="st">`</span>), <span class="at">size=</span><span class="dv">1</span>, <span class="at">alpha=</span><span class="fl">0.8</span>) </span></code></pre></div>
<p>And we add the mean value of each replicate as a dot. Here, the size of the dot is reflecting n:</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="complete-protocols.html#cb244-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">data=</span>df_2, <span class="fu">aes</span>(<span class="at">x=</span>Stage, <span class="at">y=</span>mean, <span class="at">size=</span>n, <span class="at">fill=</span>Replicate),  <span class="at">shape=</span><span class="dv">21</span>, <span class="at">stroke =</span> <span class="dv">1</span>)</span></code></pre></div>
<p>The function <code>scale_size_area()</code> ensures that 0 is represented as an area of 0 and allows to to define that an n of 10,50 and 90 is shown in the legend:</p>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="complete-protocols.html#cb245-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_size_area</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">50</span>, <span class="dv">90</span>), <span class="at">max_size =</span> <span class="dv">6</span>)</span></code></pre></div>
<p>This is what that looks like:</p>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="complete-protocols.html#cb246-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-206-1.png" width="672" /></p>
<p>Next, one of my favorite tweaks for discrete conditions is to rotate the plot 90 degrees. At the same time, the limits are defined.</p>
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="complete-protocols.html#cb247-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">coord_flip</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="fl">0.02</span>,<span class="fl">0.09</span>)) <span class="sc">+</span></span>
<span id="cb247-2"><a href="complete-protocols.html#cb247-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># This ensures correct order of conditions when plot is rotated 90 degrees</span></span>
<span id="cb247-3"><a href="complete-protocols.html#cb247-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">limits =</span> rev)</span></code></pre></div>
<p>Rotation improves readability of the labels for the conditions, even when they are long. It also easier to read the different calcium levels:</p>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="complete-protocols.html#cb248-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-208-1.png" width="672" /></p>
<p>To guide the interpretation, a line is added as a threshold of 0.06 µM (=60 nM):</p>
<div class="sourceCode" id="cb249"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb249-1"><a href="complete-protocols.html#cb249-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">0.060</span>, <span class="at">linetype=</span><span class="st">&#39;dotted&#39;</span>)</span></code></pre></div>
<p>Adjusting the axis labels and adding a title and caption:</p>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="complete-protocols.html#cb250-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb250-2"><a href="complete-protocols.html#cb250-2" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb250-3"><a href="complete-protocols.html#cb250-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Calcium concentrations are less than 60 nM&quot;</span>,</span>
<span id="cb250-4"><a href="complete-protocols.html#cb250-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;at different stages of transendothelial migration&quot;</span>,</span>
<span id="cb250-5"><a href="complete-protocols.html#cb250-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Stage&quot;</span>,</span>
<span id="cb250-6"><a href="complete-protocols.html#cb250-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Calcium [µM]&quot;</span>,</span>
<span id="cb250-7"><a href="complete-protocols.html#cb250-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">(based on data from van der Linden, DOI: 10.1101/2021.06.21.449214)&quot;</span>,</span>
<span id="cb250-8"><a href="complete-protocols.html#cb250-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 2&quot;</span></span>
<span id="cb250-9"><a href="complete-protocols.html#cb250-9" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>The layout it further optimized. The most tricky part is positioning of the label for the different conditions. It is placed right above the conditions, which I really like. However, getting this right involves a bit of trial and error and I recommend playing with the parameters to see how it affects the positioning. Something similar applies to the legend which is moved into the lower right corner of the plot, although this is eassier to accomplish. The comments explain the effect of the different lines:</p>
<div class="sourceCode" id="cb251"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb251-1"><a href="complete-protocols.html#cb251-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb251-2"><a href="complete-protocols.html#cb251-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Set text size</span></span>
<span id="cb251-3"><a href="complete-protocols.html#cb251-3" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">16</span>) <span class="sc">+</span> <span class="fu">theme</span>(</span>
<span id="cb251-4"><a href="complete-protocols.html#cb251-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">plot.caption =</span> <span class="fu">element_text</span>(</span>
<span id="cb251-5"><a href="complete-protocols.html#cb251-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>,</span>
<span id="cb251-6"><a href="complete-protocols.html#cb251-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">hjust =</span> <span class="dv">1</span></span>
<span id="cb251-7"><a href="complete-protocols.html#cb251-7" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb251-8"><a href="complete-protocols.html#cb251-8" aria-hidden="true" tabindex="-1"></a>    <span class="co">#Set position of legend to lower right corner</span></span>
<span id="cb251-9"><a href="complete-protocols.html#cb251-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.88</span>,<span class="fl">0.15</span>),</span>
<span id="cb251-10"><a href="complete-protocols.html#cb251-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb251-11"><a href="complete-protocols.html#cb251-11" aria-hidden="true" tabindex="-1"></a>    <span class="co">#This line positions the label (&#39;title&#39;) of the conditions</span></span>
<span id="cb251-12"><a href="complete-protocols.html#cb251-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="fl">0.98</span>, <span class="at">angle =</span> <span class="dv">0</span>, <span class="at">margin=</span><span class="fu">margin</span>(<span class="at">l=</span><span class="dv">70</span>)),</span>
<span id="cb251-13"><a href="complete-protocols.html#cb251-13" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb251-14"><a href="complete-protocols.html#cb251-14" aria-hidden="true" tabindex="-1"></a>    <span class="co">#This line positions the names of the conditions</span></span>
<span id="cb251-15"><a href="complete-protocols.html#cb251-15" aria-hidden="true" tabindex="-1"></a>    <span class="co">#A negative margin is needed for aligning the y-axis &#39;title&#39; with the &#39;text&#39;</span></span>
<span id="cb251-16"><a href="complete-protocols.html#cb251-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>, <span class="at">angle =</span> <span class="dv">0</span>, <span class="at">margin=</span><span class="fu">margin</span>(<span class="at">l=</span><span class="sc">-</span><span class="dv">90</span>, <span class="at">r=</span><span class="dv">5</span>)),</span>
<span id="cb251-17"><a href="complete-protocols.html#cb251-17" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb251-18"><a href="complete-protocols.html#cb251-18" aria-hidden="true" tabindex="-1"></a>    <span class="co">#Move &#39;tag&#39;, so its position partially overlaps with the conditions</span></span>
<span id="cb251-19"><a href="complete-protocols.html#cb251-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">plot.tag.position =</span> <span class="fu">c</span>(<span class="fl">0.06</span>,<span class="fl">0.99</span>)</span>
<span id="cb251-20"><a href="complete-protocols.html#cb251-20" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">fill =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb251-21"><a href="complete-protocols.html#cb251-21" aria-hidden="true" tabindex="-1"></a>               <span class="at">size =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">&#39;n per replicate&#39;</span>, <span class="at">label.position =</span> <span class="st">&quot;left&quot;</span>)</span>
<span id="cb251-22"><a href="complete-protocols.html#cb251-22" aria-hidden="true" tabindex="-1"></a>               )</span></code></pre></div>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="complete-protocols.html#cb252-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-212-1.png" width="672" /></p>
<p>To save the plot as a PNG file:</p>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb253-1"><a href="complete-protocols.html#cb253-1" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_2.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb253-2"><a href="complete-protocols.html#cb253-2" aria-hidden="true" tabindex="-1"></a>  p</span>
<span id="cb253-3"><a href="complete-protocols.html#cb253-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-3" class="section level2 hasAnchor" number="4.3">
<h2><span class="header-section-number">4.3</span> Protocol 3 - small multiples of time courses<a href="complete-protocols.html#protocol-3" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This protocol displays a number of different timecourses as ‘small multiples’. Small multiples, as the name suggests, displays many small plot separately as a stamp collection. By stressing the data, rather than the labels and grids, this can be a powerful visualization strategy.</p>
<p>The data is taken from a publication by <a href="https://doi.org/10.3389/fimmu.2021.667213">Arts et al. (2021)</a> and we recreate figure panel 1F. The original figure is in small multiple format, but we tweak it a bit more to increase the focus on the data.</p>
<p>Let’s first load the necessary package:</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb255-1"><a href="complete-protocols.html#cb255-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>The data comes from an excel file:</p>
<div class="sourceCode" id="cb256"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb256-1"><a href="complete-protocols.html#cb256-1" aria-hidden="true" tabindex="-1"></a>df_raw <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">&quot;Data_Arts_Circularity.xlsx&quot;</span>)</span>
<span id="cb256-2"><a href="complete-protocols.html#cb256-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_raw)</span></code></pre></div>
<pre><code># A tibble: 6 × 13
   time `neutro 1` `neutro 2` `neutro 3` `neutro 4` `neutro 5` `neutro 6`
  &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
1     0       0.53       0.54       0.55       0.59       0.58       0.34
2    10       0.44       0.6        0.5        0.4        0.54       0.4 
3    20       0.33       0.55       0.64       0.3        0.48       0.41
4    30       0.35       0.54       0.69       0.28       0.52       0.32
5    40       0.41       0.53       0.57       0.23       0.44       0.29
6    50       0.32       0.4        0.4        0.26       0.43       0.25
# … with 6 more variables: `neutro 7` &lt;dbl&gt;, `neutro 8` &lt;dbl&gt;,
#   `neutro 9` &lt;dbl&gt;, `neutro 10` &lt;dbl&gt;, `neutro 11` &lt;dbl&gt;, `neutro 12` &lt;dbl&gt;</code></pre>
<p>It is in a wide format, so we need to make it tidy. The parameter that was measured over time is the ‘roundness’ of cells:</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="complete-protocols.html#cb258-1" aria-hidden="true" tabindex="-1"></a>df_3 <span class="ot">&lt;-</span></span>
<span id="cb258-2"><a href="complete-protocols.html#cb258-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb258-3"><a href="complete-protocols.html#cb258-3" aria-hidden="true" tabindex="-1"></a>    df_raw,</span>
<span id="cb258-4"><a href="complete-protocols.html#cb258-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="sc">-</span>time,</span>
<span id="cb258-5"><a href="complete-protocols.html#cb258-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">&quot;Cell&quot;</span>,</span>
<span id="cb258-6"><a href="complete-protocols.html#cb258-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">&quot;Roundness&quot;</span></span>
<span id="cb258-7"><a href="complete-protocols.html#cb258-7" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>The data is in the right shape now, so let’s save it:</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="complete-protocols.html#cb259-1" aria-hidden="true" tabindex="-1"></a>df_3 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_3.csv&quot;</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>First we create a line plot of all the data:</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="complete-protocols.html#cb260-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_3, <span class="fu">aes</span>(<span class="at">x=</span>time, <span class="at">y=</span>Roundness, <span class="at">group=</span>Cell)) <span class="sc">+</span> <span class="fu">geom_line</span>()</span>
<span id="cb260-2"><a href="complete-protocols.html#cb260-2" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-218-1.png" width="672" /></p>
<p>With the facet_wrap() function, we turn this into a small multiple:</p>
<div class="sourceCode" id="cb261"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb261-1"><a href="complete-protocols.html#cb261-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>Cell)</span>
<span id="cb261-2"><a href="complete-protocols.html#cb261-2" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-219-1.png" width="672" /></p>
<p>Set the limits of the axis and force the y-axis to start at 0</p>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="complete-protocols.html#cb262-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb262-2"><a href="complete-protocols.html#cb262-2" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">1.0</span>)) <span class="sc">+</span> </span>
<span id="cb262-3"><a href="complete-protocols.html#cb262-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">300</span>))</span></code></pre></div>
<p>Use a minimal theme and remove the strips and grid to increase focus on the data:</p>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="complete-protocols.html#cb263-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">14</span>)</span>
<span id="cb263-2"><a href="complete-protocols.html#cb263-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb263-3"><a href="complete-protocols.html#cb263-3" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">strip.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb263-4"><a href="complete-protocols.html#cb263-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">strip.text =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb263-5"><a href="complete-protocols.html#cb263-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>),</span>
<span id="cb263-6"><a href="complete-protocols.html#cb263-6" aria-hidden="true" tabindex="-1"></a>               <span class="co">#Remove grid</span></span>
<span id="cb263-7"><a href="complete-protocols.html#cb263-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb263-8"><a href="complete-protocols.html#cb263-8" aria-hidden="true" tabindex="-1"></a>               <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>()</span>
<span id="cb263-9"><a href="complete-protocols.html#cb263-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb263-10"><a href="complete-protocols.html#cb263-10" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-221-1.png" width="672" /></p>
<p>I do not like the repeated axis for the different plots. We can remove those:</p>
<div class="sourceCode" id="cb264"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb264-1"><a href="complete-protocols.html#cb264-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(        </span>
<span id="cb264-2"><a href="complete-protocols.html#cb264-2" aria-hidden="true" tabindex="-1"></a>        <span class="co">#Remove axis labels</span></span>
<span id="cb264-3"><a href="complete-protocols.html#cb264-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text =</span> <span class="fu">element_blank</span>()</span>
<span id="cb264-4"><a href="complete-protocols.html#cb264-4" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb264-5"><a href="complete-protocols.html#cb264-5" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-222-1.png" width="672" /></p>
<p>This is a very minimal plot, focusing entirely on the data. It may work well, but it is informative to add some information about the scaling of the x- and y-axis. To achieve this, I add lines to the lower left plot, which correspond to the data of ‘neutro 6’ (you can see this in the small multiple plot where each plot was labeled). I define a new dataframe with the x- and y-scale for ‘neutro 6’ to do just that:</p>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="complete-protocols.html#cb265-1" aria-hidden="true" tabindex="-1"></a>ann_line<span class="ot">&lt;-</span><span class="fu">data.frame</span>(<span class="at">xmin=</span><span class="dv">0</span>,<span class="at">xmax=</span><span class="dv">300</span>,<span class="at">ymin=</span><span class="dv">0</span>,<span class="at">ymax=</span><span class="dv">1</span>,</span>
<span id="cb265-2"><a href="complete-protocols.html#cb265-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">Cell=</span><span class="fu">factor</span>(<span class="st">&quot;neutro 6&quot;</span>,<span class="at">levels=</span><span class="fu">c</span>(<span class="st">&quot;neutro 6&quot;</span>)))</span>
<span id="cb265-3"><a href="complete-protocols.html#cb265-3" aria-hidden="true" tabindex="-1"></a>ann_line</span></code></pre></div>
<pre><code>  xmin xmax ymin ymax     Cell
1    0  300    0    1 neutro 6</code></pre>
<p>This dataframe can now be used to draw two lines with <code>geom_segment()</code>:</p>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="complete-protocols.html#cb267-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb267-2"><a href="complete-protocols.html#cb267-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Vertical line</span></span>
<span id="cb267-3"><a href="complete-protocols.html#cb267-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">data=</span>ann_line, <span class="fu">aes</span>(<span class="at">x=</span>xmin,<span class="at">xend=</span>xmin,<span class="at">y=</span>ymin,<span class="at">yend=</span>ymax), <span class="at">size=</span><span class="dv">2</span>, <span class="at">color=</span><span class="st">&#39;grey40&#39;</span>) <span class="sc">+</span></span>
<span id="cb267-4"><a href="complete-protocols.html#cb267-4" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Horizontal line</span></span>
<span id="cb267-5"><a href="complete-protocols.html#cb267-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">data=</span>ann_line, <span class="fu">aes</span>(<span class="at">x=</span>xmin,<span class="at">xend=</span>xmax,<span class="at">y=</span>ymin,<span class="at">yend=</span>ymin), <span class="at">size=</span><span class="dv">2</span>, <span class="at">color=</span><span class="st">&#39;grey40&#39;</span>) <span class="sc">+</span></span>
<span id="cb267-6"><a href="complete-protocols.html#cb267-6" aria-hidden="true" tabindex="-1"></a>  <span class="cn">NULL</span></span>
<span id="cb267-7"><a href="complete-protocols.html#cb267-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb267-8"><a href="complete-protocols.html#cb267-8" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-224-1.png" width="672" /></p>
<p>The plot is now in black and white which gives it a strong contrast. We can make it a bit more soft and pleasant to look at by changing to shades of grey. Also, the labels of the axes are moved next to the lines:</p>
<div class="sourceCode" id="cb268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb268-1"><a href="complete-protocols.html#cb268-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb268-2"><a href="complete-protocols.html#cb268-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="st">&#39;grey98&#39;</span>, <span class="at">color=</span><span class="cn">NA</span>),</span>
<span id="cb268-3"><a href="complete-protocols.html#cb268-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">color=</span><span class="st">&#39;grey90&#39;</span>, <span class="at">fill=</span><span class="cn">NA</span>),</span>
<span id="cb268-4"><a href="complete-protocols.html#cb268-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>, <span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">color=</span><span class="st">&#39;grey40&#39;</span>),</span>
<span id="cb268-5"><a href="complete-protocols.html#cb268-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>, <span class="at">vjust =</span> <span class="dv">0</span>, <span class="at">hjust=</span><span class="dv">0</span>, <span class="at">angle =</span> <span class="dv">90</span>, <span class="at">color=</span><span class="st">&#39;grey40&#39;</span>),</span>
<span id="cb268-6"><a href="complete-protocols.html#cb268-6" aria-hidden="true" tabindex="-1"></a>  ) </span></code></pre></div>
<p>Finally, we add a title, caption, and labels (and a scale in brackets):</p>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="complete-protocols.html#cb269-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb269-2"><a href="complete-protocols.html#cb269-2" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb269-3"><a href="complete-protocols.html#cb269-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Changes in the shape of migrating cells over time&quot;</span>,</span>
<span id="cb269-4"><a href="complete-protocols.html#cb269-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Time [300s]&quot;</span>,</span>
<span id="cb269-5"><a href="complete-protocols.html#cb269-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Circularity [0-1]&quot;</span>,</span>
<span id="cb269-6"><a href="complete-protocols.html#cb269-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">(based on data from Arts et al., DOI: 10.3389/fimmu.2021.667213)&quot;</span>,</span>
<span id="cb269-7"><a href="complete-protocols.html#cb269-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 3&quot;</span></span>
<span id="cb269-8"><a href="complete-protocols.html#cb269-8" aria-hidden="true" tabindex="-1"></a>  ) </span></code></pre></div>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="complete-protocols.html#cb270-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-227-1.png" width="672" /></p>
<div class="sourceCode" id="cb271"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb271-1"><a href="complete-protocols.html#cb271-1" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_3.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb271-2"><a href="complete-protocols.html#cb271-2" aria-hidden="true" tabindex="-1"></a> p</span>
<span id="cb271-3"><a href="complete-protocols.html#cb271-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-4" class="section level2 hasAnchor" number="4.4">
<h2><span class="header-section-number">4.4</span> Protocol 4 - Plotting data from a 96-wells experiment<a href="complete-protocols.html#protocol-4" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This protocol showcases some serious data wrangling and tidying. One reason is that the data is acquired with a 96-wells plate reader and the data is stored according to the layout of the plate. This makes total sense from a human perspective, but it is not well suited for data visualization.</p>
<p>In addition, the plate is measured twice. One measurement is the luminescence from a luciferase and the other measurement is the luminescence from Renilla. The latter reading serves as a reference and therefore, the luciferase data is divided by the Renilla intensities. A final step before the data is visualized is a normalization to a control condition.</p>
<p>The code that is shown here is also the basis for the <a href="https://doi.org/10.12688/f1000research.73641.1">plotXpress app</a> that can be used to process and visualize the data. In fact, the data visualization is very close to the standard output of plotXpress and uses the same example data.</p>
<p>We start by loading a package that we need:</p>
<div class="sourceCode" id="cb273"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb273-1"><a href="complete-protocols.html#cb273-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>The measured data is read from an excel sheet. Note that this is the raw data that is stored by the software that operates the plate reader:</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="complete-protocols.html#cb274-1" aria-hidden="true" tabindex="-1"></a>df_raw <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">&quot;DualLuc_example_data.xlsx&quot;</span>, <span class="at">sheet =</span> <span class="st">&quot;Results&quot;</span>)</span></code></pre></div>
<pre><code>New names:
• `` -&gt; `...1`
• `` -&gt; `...2`
• `` -&gt; `...3`
• `` -&gt; `...5`
• `` -&gt; `...6`
• `` -&gt; `...7`
• `` -&gt; `...8`
• `` -&gt; `...9`
• `` -&gt; `...10`
• `` -&gt; `...11`
• `` -&gt; `...12`
• `` -&gt; `...13`
• `` -&gt; `...14`
• `` -&gt; `...15`
• `` -&gt; `...16`
• `` -&gt; `...17`</code></pre>
<p>The experimental conditions for each well are stored in a separate CSV file, generated by the experimentalist that did the experiment:</p>
<div class="sourceCode" id="cb276"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb276-1"><a href="complete-protocols.html#cb276-1" aria-hidden="true" tabindex="-1"></a>df_design <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;Tidy_design.csv&quot;</span>)</span>
<span id="cb276-2"><a href="complete-protocols.html#cb276-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_design)</span></code></pre></div>
<pre><code>  Wells condition treatment1 treatment2
1   A01         -        HEK          -
2   B01         -        HEK          -
3   C01         -        HEK          -
4   D01         -        HEK          -
5   E01         -     neuron          -
6   F01         -     neuron          -</code></pre>
<p>You can see that the design file is tidy. In contrast the excel file with data is far from tidy. In the excel sheet, two ‘rectangles’ of cells define the data for the firefly &amp; renilla reads.
The data is subset and converted to a vector</p>
<div class="sourceCode" id="cb278"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb278-1"><a href="complete-protocols.html#cb278-1" aria-hidden="true" tabindex="-1"></a>firefly <span class="ot">&lt;-</span> df_raw[<span class="dv">19</span><span class="sc">:</span><span class="dv">26</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">17</span>] <span class="sc">%&gt;%</span> <span class="fu">unlist</span>(<span class="at">use.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb278-2"><a href="complete-protocols.html#cb278-2" aria-hidden="true" tabindex="-1"></a>renilla <span class="ot">&lt;-</span> df_raw[<span class="dv">40</span><span class="sc">:</span><span class="dv">47</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">17</span>] <span class="sc">%&gt;%</span> <span class="fu">unlist</span>(<span class="at">use.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Define a dataframe with wells</p>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="complete-protocols.html#cb279-1" aria-hidden="true" tabindex="-1"></a>column <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="at">each=</span><span class="dv">8</span>)</span>
<span id="cb279-2"><a href="complete-protocols.html#cb279-2" aria-hidden="true" tabindex="-1"></a>row <span class="ot">&lt;-</span> <span class="fu">rep</span>(LETTERS[<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>],<span class="dv">12</span>)</span></code></pre></div>
<p>For convenience, all numbers should consist of 2 digits and so we add a zero that precedes the single digit numbers:</p>
<div class="sourceCode" id="cb280"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb280-1"><a href="complete-protocols.html#cb280-1" aria-hidden="true" tabindex="-1"></a>x0 <span class="ot">&lt;-</span> <span class="fu">str_pad</span>(column, <span class="dv">2</span>, <span class="at">pad =</span> <span class="st">&quot;0&quot;</span>)</span></code></pre></div>
<p>To generate a unique index for each row in the dataframe, we define ‘Wells’, which combines the row and column index:</p>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="complete-protocols.html#cb281-1" aria-hidden="true" tabindex="-1"></a>Wells <span class="ot">&lt;-</span> <span class="fu">paste0</span>(row,x0)</span></code></pre></div>
<p>Next, we create a dataframe that holds the data of individual columns, rows and wells:</p>
<div class="sourceCode" id="cb282"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb282-1"><a href="complete-protocols.html#cb282-1" aria-hidden="true" tabindex="-1"></a>df_plate <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(column,row,Wells)</span>
<span id="cb282-2"><a href="complete-protocols.html#cb282-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_plate)</span></code></pre></div>
<pre><code>  column row Wells
1      1   A   A01
2      1   B   B01
3      1   C   C01
4      1   D   D01
5      1   E   E01
6      1   F   F01</code></pre>
<p>Add to df_plate the vectors with data from firefly and renilla reads</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="complete-protocols.html#cb284-1" aria-hidden="true" tabindex="-1"></a>df_4 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(df_plate,firefly,renilla)</span></code></pre></div>
<p>Merge the design with the data, based on the well ID - left_join() is used to add only data for wells that are listed in the design dataframe</p>
<div class="sourceCode" id="cb285"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb285-1"><a href="complete-protocols.html#cb285-1" aria-hidden="true" tabindex="-1"></a>df_4 <span class="ot">&lt;-</span> <span class="fu">left_join</span>(df_design, df_4, <span class="at">by=</span><span class="st">&#39;Wells&#39;</span>)</span>
<span id="cb285-2"><a href="complete-protocols.html#cb285-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_4)</span></code></pre></div>
<pre><code>  Wells condition treatment1 treatment2 column row firefly renilla
1   A01         -        HEK          -      1   A    2010 2391540
2   B01         -        HEK          -      1   B    3210 2391639
3   C01         -        HEK          -      1   C    1965 2390991
4   D01         -        HEK          -      1   D    2381 2391774
5   E01         -     neuron          -      1   E    1292  269021
6   F01         -     neuron          -      1   F     991  268918</code></pre>
<p>Calculate the relative expression from the firefly/renilla ratio</p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="complete-protocols.html#cb287-1" aria-hidden="true" tabindex="-1"></a>df_4 <span class="ot">&lt;-</span> df_4 <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">expression=</span>firefly<span class="sc">/</span>renilla)</span></code></pre></div>
<p>Take all control conditions and calculate the average value</p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="complete-protocols.html#cb288-1" aria-hidden="true" tabindex="-1"></a>df_norm <span class="ot">&lt;-</span> df_4 <span class="sc">%&gt;%</span> <span class="fu">filter</span>(condition <span class="sc">==</span> <span class="st">&quot;-&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb288-2"><a href="complete-protocols.html#cb288-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(treatment1,treatment2) <span class="sc">%&gt;%</span></span>
<span id="cb288-3"><a href="complete-protocols.html#cb288-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean=</span><span class="fu">mean</span>(expression)) </span></code></pre></div>
<pre><code>`summarise()` has grouped output by &#39;treatment1&#39;. You can override using the
`.groups` argument.</code></pre>
<p>Combine the mean values (needed for normalization) values with the df_4 dataframe</p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="complete-protocols.html#cb290-1" aria-hidden="true" tabindex="-1"></a>df_4 <span class="ot">&lt;-</span> df_4 <span class="sc">%&gt;%</span></span>
<span id="cb290-2"><a href="complete-protocols.html#cb290-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">full_join</span>(df_norm, <span class="at">by=</span><span class="fu">c</span>(<span class="st">&quot;treatment1&quot;</span>,<span class="st">&quot;treatment2&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb290-3"><a href="complete-protocols.html#cb290-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>(condition)</span></code></pre></div>
<p>Calculate the <code>Fold Change</code> by normalizing all measurements against the control (-)</p>
<div class="sourceCode" id="cb291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb291-1"><a href="complete-protocols.html#cb291-1" aria-hidden="true" tabindex="-1"></a>df_4 <span class="ot">&lt;-</span> df_4 <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="st">`</span><span class="at">Fold Change</span><span class="st">`</span> <span class="ot">=</span> expression<span class="sc">/</span>mean)</span></code></pre></div>
<p>The result is a dataframe that holds all the necessary data:</p>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="complete-protocols.html#cb292-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_4)</span></code></pre></div>
<pre><code>  Wells condition treatment1 treatment2 column row firefly renilla   expression
1   A01         -        HEK          -      1   A    2010 2391540 0.0008404626
2   B01         -        HEK          -      1   B    3210 2391639 0.0013421758
3   C01         -        HEK          -      1   C    1965 2390991 0.0008218350
4   D01         -        HEK          -      1   D    2381 2391774 0.0009954954
5   E01         -     neuron          -      1   E    1292  269021 0.0048025991
6   F01         -     neuron          -      1   F     991  268918 0.0036851382
          mean Fold Change
1 0.0009999922   0.8404692
2 0.0009999922   1.3421863
3 0.0009999922   0.8218414
4 0.0009999922   0.9955032
5 0.0040858977   1.1754085
6 0.0040858977   0.9019164</code></pre>
<p>The data is in the right shape now, so let’s save it:</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="complete-protocols.html#cb294-1" aria-hidden="true" tabindex="-1"></a>df_4 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_4.csv&quot;</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Based on the dataframe, we can create a plot with jittered dots that show the data:</p>
<div class="sourceCode" id="cb295"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb295-1"><a href="complete-protocols.html#cb295-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_4, <span class="fu">aes</span>(<span class="at">x=</span>condition, <span class="at">y=</span><span class="st">`</span><span class="at">Fold Change</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb295-2"><a href="complete-protocols.html#cb295-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.2</span>, <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">size=</span><span class="dv">3</span>)</span></code></pre></div>
<p>To splot the graphs based on treatment1 (vertical) and treatment2 (horizontal) we use the <code>facet_grid()</code> function:</p>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="complete-protocols.html#cb296-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">facet_grid</span>(treatment1<span class="sc">~</span>treatment2)</span></code></pre></div>
<p>We add a horizontal line for mean value:</p>
<div class="sourceCode" id="cb297"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb297-1"><a href="complete-protocols.html#cb297-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">stat_summary</span>(<span class="at">fun.min=</span>mean, <span class="at">fun.max=</span>mean, <span class="at">geom=</span><span class="st">&#39;errorbar&#39;</span>, <span class="at">width=</span><span class="fl">0.6</span>, <span class="at">size=</span><span class="fl">0.5</span>)</span></code></pre></div>
<p>Add labels:</p>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="complete-protocols.html#cb298-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb298-2"><a href="complete-protocols.html#cb298-2" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb298-3"><a href="complete-protocols.html#cb298-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Effect of DNA sequences on reporter levels under different conditions&quot;</span>,</span>
<span id="cb298-4"><a href="complete-protocols.html#cb298-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;The expression level was determined by a dual luciferase assay</span><span class="sc">\n</span><span class="st"> and the values were normalized to a control with no DNA sequence (-)&quot;</span>,</span>
<span id="cb298-5"><a href="complete-protocols.html#cb298-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;DNA Sequence&quot;</span>,</span>
<span id="cb298-6"><a href="complete-protocols.html#cb298-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Fold change of the reporter relative to the control (-)&quot;</span>,</span>
<span id="cb298-7"><a href="complete-protocols.html#cb298-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">(based on data from Brandorff et al., DOI: 10.1101/2021.07.08.451595)&quot;</span>,</span>
<span id="cb298-8"><a href="complete-protocols.html#cb298-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 4&quot;</span></span>
<span id="cb298-9"><a href="complete-protocols.html#cb298-9" aria-hidden="true" tabindex="-1"></a>  ) </span></code></pre></div>
<p>Set the theme and font size:</p>
<div class="sourceCode" id="cb299"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb299-1"><a href="complete-protocols.html#cb299-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">14</span>)</span></code></pre></div>
<p>Format the facet labels (strips) and the caption + subtitle</p>
<div class="sourceCode" id="cb300"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb300-1"><a href="complete-protocols.html#cb300-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="st">&quot;grey90&quot;</span>, <span class="at">color=</span><span class="st">&quot;grey50&quot;</span>),</span>
<span id="cb300-2"><a href="complete-protocols.html#cb300-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">color=</span><span class="st">&quot;grey50&quot;</span>),</span>
<span id="cb300-3"><a href="complete-protocols.html#cb300-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>),</span>
<span id="cb300-4"><a href="complete-protocols.html#cb300-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey50&quot;</span>, <span class="at">face =</span> <span class="st">&quot;italic&quot;</span>),</span>
<span id="cb300-5"><a href="complete-protocols.html#cb300-5" aria-hidden="true" tabindex="-1"></a>               <span class="co">#Remove the grid</span></span>
<span id="cb300-6"><a href="complete-protocols.html#cb300-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb300-7"><a href="complete-protocols.html#cb300-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>()</span>
<span id="cb300-8"><a href="complete-protocols.html#cb300-8" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Let’s look at the result:</p>
<div class="sourceCode" id="cb301"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb301-1"><a href="complete-protocols.html#cb301-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-251-1.png" width="672" /></p>
<p>To save the plot as a png file:</p>
<div class="sourceCode" id="cb302"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb302-1"><a href="complete-protocols.html#cb302-1" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_4.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb302-2"><a href="complete-protocols.html#cb302-2" aria-hidden="true" tabindex="-1"></a> p</span>
<span id="cb302-3"><a href="complete-protocols.html#cb302-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-5" class="section level2 hasAnchor" number="4.5">
<h2><span class="header-section-number">4.5</span> Protocol 5 - A map of amino acids<a href="complete-protocols.html#protocol-5" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This data visualization plots the position of amino acids in a given protein. It is inspired by figure 2A of the paper by <a href="https://www.sciencedirect.com/science/article/pii/S0092867420304815">Basu et al. (2020)</a></p>
<p>Let’s load the tidyverse package:</p>
<div class="sourceCode" id="cb304"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb304-1"><a href="complete-protocols.html#cb304-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>First, we define a vector with the 20 amino acids and the order in which we plot them. The amino acids are grouped as hydrophobic (G,A,V,C,P,I,L,M,F,W), hydrophilic (S,T,Y,N,Q), acidic (D,E) and basic (R,H,K).</p>
<div class="sourceCode" id="cb305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb305-1"><a href="complete-protocols.html#cb305-1" aria-hidden="true" tabindex="-1"></a>amino_acid_ordered <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(<span class="st">&quot;GAVCPILMFWSTYNQDERHK&quot;</span>,<span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span> <span class="fu">unlist</span>()</span>
<span id="cb305-2"><a href="complete-protocols.html#cb305-2" aria-hidden="true" tabindex="-1"></a>amino_acid_ordered</span></code></pre></div>
<pre><code> [1] &quot;G&quot; &quot;A&quot; &quot;V&quot; &quot;C&quot; &quot;P&quot; &quot;I&quot; &quot;L&quot; &quot;M&quot; &quot;F&quot; &quot;W&quot; &quot;S&quot; &quot;T&quot; &quot;Y&quot; &quot;N&quot; &quot;Q&quot; &quot;D&quot; &quot;E&quot; &quot;R&quot; &quot;H&quot;
[20] &quot;K&quot;</code></pre>
<p>The protein sequence that we will use is the <em>Homo sapiens</em> Homeobox protein Hox-D13:</p>
<div class="sourceCode" id="cb307"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb307-1"><a href="complete-protocols.html#cb307-1" aria-hidden="true" tabindex="-1"></a>protein <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;MSRAGSWDMDGLRADGGGAGGAPASSSSSSVAAAAASGQCRGFLSAPVFAGTHSGRAAAA</span></span>
<span id="cb307-2"><a href="complete-protocols.html#cb307-2" aria-hidden="true" tabindex="-1"></a><span class="st">AAAAAAAAAAASGFAYPGTSERTGSSSSSSSSAVVAARPEAPPAKECPAPTPAAAAAAPP</span></span>
<span id="cb307-3"><a href="complete-protocols.html#cb307-3" aria-hidden="true" tabindex="-1"></a><span class="st">SAPALGYGYHFGNGYYSCRMSHGVGLQQNALKSSPHASLGGFPVEKYMDVSGLASSSVPA</span></span>
<span id="cb307-4"><a href="complete-protocols.html#cb307-4" aria-hidden="true" tabindex="-1"></a><span class="st">NEVPARAKEVSFYQGYTSPYQHVPGYIDMVSTFGSGEPRHEAYISMEGYQSWTLANGWNS</span></span>
<span id="cb307-5"><a href="complete-protocols.html#cb307-5" aria-hidden="true" tabindex="-1"></a><span class="st">QVYCTKDQPQGSHFWKSSFPGDVALNQPDMCVYRRGRKKRVPYTKLQLKELENEYAINKF</span></span>
<span id="cb307-6"><a href="complete-protocols.html#cb307-6" aria-hidden="true" tabindex="-1"></a><span class="st">INKDKRRRISAATNLSERQVTIWFQNRRVKDKKIVSKLKDTVS&quot;</span>)</span></code></pre></div>
<p>The protein sequence may contain end-of-line characters “” after copy pasting and we need to remove these. The <code>gsub()</code> function can be used:</p>
<div class="sourceCode" id="cb308"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb308-1"><a href="complete-protocols.html#cb308-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gsub</span>(<span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>, <span class="st">&quot;&quot;</span>, protein)</span></code></pre></div>
<pre><code>[1] &quot;MSRAGSWDMDGLRADGGGAGGAPASSSSSSVAAAAASGQCRGFLSAPVFAGTHSGRAAAAAAAAAAAAAAASGFAYPGTSERTGSSSSSSSSAVVAARPEAPPAKECPAPTPAAAAAAPPSAPALGYGYHFGNGYYSCRMSHGVGLQQNALKSSPHASLGGFPVEKYMDVSGLASSSVPANEVPARAKEVSFYQGYTSPYQHVPGYIDMVSTFGSGEPRHEAYISMEGYQSWTLANGWNSQVYCTKDQPQGSHFWKSSFPGDVALNQPDMCVYRRGRKKRVPYTKLQLKELENEYAINKFINKDKRRRISAATNLSERQVTIWFQNRRVKDKKIVSKLKDTVS&quot;</code></pre>
<p>But we can also use <code>str_replace_all()</code> from the {tidyverse} package:</p>
<div class="sourceCode" id="cb310"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb310-1"><a href="complete-protocols.html#cb310-1" aria-hidden="true" tabindex="-1"></a>protein <span class="ot">&lt;-</span> protein <span class="sc">%&gt;%</span> <span class="fu">str_replace_all</span>(<span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>, <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Next, the protein sequence is split into single characters and we assign this vector to <code>aa</code>:</p>
<div class="sourceCode" id="cb311"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb311-1"><a href="complete-protocols.html#cb311-1" aria-hidden="true" tabindex="-1"></a>aa <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(protein, <span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span> <span class="fu">unlist</span>()</span></code></pre></div>
<p>We generate a dataframe with a column with the amino acids and a column that defines their position:</p>
<div class="sourceCode" id="cb312"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb312-1"><a href="complete-protocols.html#cb312-1" aria-hidden="true" tabindex="-1"></a>df_5 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">aa=</span>aa, <span class="at">position=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(aa))</span></code></pre></div>
<p>Now we reorder the data frame to the order of the amino acids that we defined earlier in the vector <code>amino_acid_ordered</code>:</p>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="complete-protocols.html#cb313-1" aria-hidden="true" tabindex="-1"></a>df_5 <span class="ot">&lt;-</span> df_5 <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">aa =</span> <span class="fu">fct_relevel</span>(aa, amino_acid_ordered))</span></code></pre></div>
<p>The basic plot shows a black tile for each amino acid. Note that the y-axis order is defined by the vector <code>amino_acid_ordered</code>, but it needs to be reverted to order the amino acids from top to bottom along the y-axis (which is naturally starts at the bottom it corresponds to the origin).</p>
<p>The data is in the right shape now, so let’s save it:</p>
<div class="sourceCode" id="cb314"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb314-1"><a href="complete-protocols.html#cb314-1" aria-hidden="true" tabindex="-1"></a>df_5 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_5.csv&quot;</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb315"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb315-1"><a href="complete-protocols.html#cb315-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span> <span class="fu">geom_tile</span>(<span class="at">data=</span>df_5, <span class="fu">aes</span>(<span class="at">x=</span>position, <span class="at">y=</span>aa)) <span class="sc">+</span></span>
<span id="cb315-2"><a href="complete-protocols.html#cb315-2" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_y_discrete</span>(<span class="at">limits =</span> <span class="fu">rev</span>(amino_acid_ordered))</span>
<span id="cb315-3"><a href="complete-protocols.html#cb315-3" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-262-1.png" width="672" /></p>
<p>Set the theme to classic, to get rid off the ‘frame’ around the plot and the grid.</p>
<div class="sourceCode" id="cb316"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb316-1"><a href="complete-protocols.html#cb316-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p<span class="sc">+</span><span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">16</span>)</span></code></pre></div>
<p>For each of the four classes of amino acids we can define a box with a color that indicates the class. For example, there are three basic residues that will have a rectangle filled with blue in the background. The amino acids are factors, but we need numbers to define the coordinates for the rectangle. In a plot with a factors (here on the y-axis) their position is defined by a (non visible) natural number. Therefore we can define a box with the function <code>annotate()</code> for the first residue with y-coordinates <code>ymin=0.5</code> and <code>ymax=1.5</code>:</p>
<div class="sourceCode" id="cb317"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb317-1"><a href="complete-protocols.html#cb317-1" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="fl">0.5</span>, <span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymax=</span><span class="fl">1.5</span>, <span class="at">fill=</span><span class="st">&#39;blue&#39;</span>, <span class="at">alpha=</span><span class="fl">0.4</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-264-1.png" width="672" />
In this way, we define four colored rectangles that reflect the different amino acids categories; blue=basic, red=acidic, yellow=hydrophilic, grey=hydrophobic:</p>
<div class="sourceCode" id="cb318"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb318-1"><a href="complete-protocols.html#cb318-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="fl">0.5</span>, <span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymax=</span><span class="fl">3.5</span>, <span class="at">fill=</span><span class="st">&#39;blue&#39;</span>, <span class="at">alpha=</span><span class="fl">0.15</span>)</span>
<span id="cb318-2"><a href="complete-protocols.html#cb318-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="fl">3.5</span>, <span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymax=</span><span class="fl">5.5</span>, <span class="at">fill=</span><span class="st">&#39;red&#39;</span>, <span class="at">alpha=</span><span class="fl">0.15</span>)</span>
<span id="cb318-3"><a href="complete-protocols.html#cb318-3" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="fl">5.5</span>, <span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymax=</span><span class="fl">10.5</span>, <span class="at">fill=</span><span class="st">&#39;yellow&#39;</span>, <span class="at">alpha=</span><span class="fl">0.15</span>)</span>
<span id="cb318-4"><a href="complete-protocols.html#cb318-4" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymin =</span> <span class="fl">10.5</span>, <span class="at">xmax =</span> <span class="cn">Inf</span>, <span class="at">ymax=</span><span class="fl">20.5</span>, <span class="at">fill=</span><span class="st">&#39;black&#39;</span>, <span class="at">alpha=</span><span class="fl">0.15</span>)</span></code></pre></div>
<p>Let’s look at the result:</p>
<div class="sourceCode" id="cb319"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb319-1"><a href="complete-protocols.html#cb319-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-266-1.png" width="672" /></p>
<p>Adjusting the axis labels and adding a title and caption:</p>
<div class="sourceCode" id="cb320"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb320-1"><a href="complete-protocols.html#cb320-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb320-2"><a href="complete-protocols.html#cb320-2" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb320-3"><a href="complete-protocols.html#cb320-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Mapping the amino acid positions of HOXD13&quot;</span>,</span>
<span id="cb320-4"><a href="complete-protocols.html#cb320-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;shows a high abundance of alanines in the IDR&quot;</span>,</span>
<span id="cb320-5"><a href="complete-protocols.html#cb320-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Amino acid&quot;</span>,</span>
<span id="cb320-6"><a href="complete-protocols.html#cb320-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart | based on data from Basu et al, DOI: 10.1016/j.cell.2020.04.018&quot;</span>,</span>
<span id="cb320-7"><a href="complete-protocols.html#cb320-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 5&quot;</span></span>
<span id="cb320-8"><a href="complete-protocols.html#cb320-8" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>And a final tweak of the label style and location:</p>
<div class="sourceCode" id="cb321"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb321-1"><a href="complete-protocols.html#cb321-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(</span>
<span id="cb321-2"><a href="complete-protocols.html#cb321-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">plot.caption =</span> <span class="fu">element_text</span>(</span>
<span id="cb321-3"><a href="complete-protocols.html#cb321-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>,</span>
<span id="cb321-4"><a href="complete-protocols.html#cb321-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">hjust =</span> <span class="dv">1</span></span>
<span id="cb321-5"><a href="complete-protocols.html#cb321-5" aria-hidden="true" tabindex="-1"></a>    ))</span></code></pre></div>
<p>In the original paper, a region of the protein is annotated as an ‘intrinsically disordered region’ abbreviated as IDR. Here, we we use the <code>annotate()</code> function to add a rectangle and a label to the top of the plot:</p>
<div class="sourceCode" id="cb322"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb322-1"><a href="complete-protocols.html#cb322-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="st">&quot;rect&quot;</span>, <span class="at">xmin=</span><span class="dv">0</span>, <span class="at">xmax=</span><span class="dv">118</span>, <span class="at">ymin=</span><span class="dv">21</span>, <span class="at">ymax=</span><span class="dv">22</span>, <span class="at">fill=</span><span class="st">&#39;darkblue&#39;</span>) <span class="sc">+</span></span>
<span id="cb322-2"><a href="complete-protocols.html#cb322-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="at">x=</span><span class="dv">59</span>, <span class="at">y=</span><span class="dv">23</span>, <span class="at">alpha=</span><span class="dv">1</span>, <span class="at">color=</span><span class="st">&#39;darkblue&#39;</span>, <span class="at">size=</span><span class="dv">4</span>,<span class="at">label=</span><span class="st">&#39;IDR&#39;</span>)</span>
<span id="cb322-3"><a href="complete-protocols.html#cb322-3" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-269-1.png" width="672" /></p>
<p>To avoid clipping of the new label by the subtitle of the plot:</p>
<div class="sourceCode" id="cb323"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb323-1"><a href="complete-protocols.html#cb323-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">coord_cartesian</span>(<span class="at">clip =</span> <span class="st">&#39;off&#39;</span>)</span>
<span id="cb323-2"><a href="complete-protocols.html#cb323-2" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-270-1.png" width="672" /></p>
<p>The subtitle is quite close to the IDR label. Let’s give the subtitle a bit more room, by adding a margin at the bottom of the subtitle. This can be done with the <code>theme()</code> function to style the subtitle:</p>
<div class="sourceCode" id="cb324"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb324-1"><a href="complete-protocols.html#cb324-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">margin =</span> <span class="fu">margin</span>(<span class="at">b=</span><span class="dv">20</span>)))</span>
<span id="cb324-2"><a href="complete-protocols.html#cb324-2" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-271-1.png" width="672" /></p>
<p>Finally we can save the plot:</p>
<div class="sourceCode" id="cb325"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb325-1"><a href="complete-protocols.html#cb325-1" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_5.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb325-2"><a href="complete-protocols.html#cb325-2" aria-hidden="true" tabindex="-1"></a> p</span>
<span id="cb325-3"><a href="complete-protocols.html#cb325-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-6" class="section level2 hasAnchor" number="4.6">
<h2><span class="header-section-number">4.6</span> Protocol 6 - Heatmap style visualization of timelapse data<a href="complete-protocols.html#protocol-6" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Lineplots are typically used to plot data from timeseries (also known as longitudinal data). However, in case of many samples/objects, this may result in a cluttered data visualization. To solve this, the same data can be presented as a heatmap, where every row is an object and the response is coded as a color. A downside is that it is less quantitative as it is difficult to ‘read’ the numbers from this kind of visualization. Still, it is a powerful visualization to plot a lot of data and show its dynamics and heterogeneity. The data and visualization is originally published by <a href="https://www.biorxiv.org/content/10.1101/2021.07.27.453948v3">Chavez-Abiega et al. (2021)</a></p>
<p>The multipurpose <code>{tidyverse}</code> package is used for data wrangling and plotting:</p>
<div class="sourceCode" id="cb327"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb327-1"><a href="complete-protocols.html#cb327-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>A CSV file with the data is loaded. Since the file is a couple of Megabytes, we use the faster <code>fread()</code> function from the package <code>data.table</code>:</p>
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb328-1"><a href="complete-protocols.html#cb328-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span></code></pre></div>
<pre><code>
Attaching package: &#39;data.table&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:dplyr&#39;:

    between, first, last</code></pre>
<pre><code>The following object is masked from &#39;package:purrr&#39;:

    transpose</code></pre>
<div class="sourceCode" id="cb332"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb332-1"><a href="complete-protocols.html#cb332-1" aria-hidden="true" tabindex="-1"></a>df_S1P <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="st">&quot;ERK_repsonse_to_S1P.csv&quot;</span>) </span>
<span id="cb332-2"><a href="complete-protocols.html#cb332-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_S1P)</span></code></pre></div>
<pre><code>   Time_in_min Ligand Unique_Object CN_ERK Slide
1:         2.5    S1P           100 0.3941    P1
2:         3.0    S1P           100 0.3921    P1
3:         3.5    S1P           100 0.3900    P1
4:         4.0    S1P           100 0.3880    P1
5:         4.5    S1P           100 0.3859    P1
6:         5.0    S1P           100 0.3839    P1</code></pre>
<p>The column ‘CN_ERK’ has the data on the activity that we will plot over time. Each number in the ‘Unique_Object’ column reflects an individual cell measurement and so we can use that to group the measurements using <code>group_by(Unique_Object)</code>. We subtract the average baseline activity from each trace by subtracting the data acquired at the first 5 timepoints: <code>CN_ERK[1:5]</code>. The data is stored in a new column with normalized ERK activity data ‘ERKn’:</p>
<div class="sourceCode" id="cb334"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb334-1"><a href="complete-protocols.html#cb334-1" aria-hidden="true" tabindex="-1"></a>df_sub <span class="ot">&lt;-</span> df_S1P <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Unique_Object) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(Time_in_min) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">ERKn=</span>CN_ERK<span class="sc">-</span><span class="fu">mean</span>(CN_ERK[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])) <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>()</span></code></pre></div>
<p>Around Timepoint 23 (minutes), the ligand was added. To set this time point to zero, we subtract a value of 23 (minutes) from each point:</p>
<div class="sourceCode" id="cb335"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb335-1"><a href="complete-protocols.html#cb335-1" aria-hidden="true" tabindex="-1"></a>df_sub<span class="sc">$</span>Time_in_min <span class="ot">&lt;-</span> df_sub<span class="sc">$</span>Time_in_min<span class="dv">-23</span></span></code></pre></div>
<p>The column ‘Unique_Object’ that identifies the individual cells contains natural numbers, but these need to be treated as qualitative data. Therefore, we change the data type in this column to a factor with <code>as.factor()</code>:</p>
<div class="sourceCode" id="cb336"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb336-1"><a href="complete-protocols.html#cb336-1" aria-hidden="true" tabindex="-1"></a>df_sub <span class="ot">&lt;-</span> df_sub <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Unique_Object=</span><span class="fu">as.factor</span>(Unique_Object))</span></code></pre></div>
<p>To order objects, we need to order ‘Unique_Object’ according to something. That something can be the maximum value of the Erk activity:</p>
<div class="sourceCode" id="cb337"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb337-1"><a href="complete-protocols.html#cb337-1" aria-hidden="true" tabindex="-1"></a>df_sub <span class="ot">&lt;-</span> df_sub <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Unique_Object =</span> <span class="fu">fct_reorder</span>(Unique_Object, ERKn, max))</span></code></pre></div>
<p>Plot the data in heatmap style. We use theme_void here to focus only on the data and we will deal with styling of the axes and labels later:</p>
<div class="sourceCode" id="cb338"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb338-1"><a href="complete-protocols.html#cb338-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_sub, <span class="fu">aes</span>(<span class="at">x=</span>Time_in_min, <span class="at">y=</span>Unique_Object,<span class="at">fill=</span>ERKn)) <span class="sc">+</span></span>
<span id="cb338-2"><a href="complete-protocols.html#cb338-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>() <span class="sc">+</span> <span class="fu">theme_void</span>()<span class="sc">+</span></span>
<span id="cb338-3"><a href="complete-protocols.html#cb338-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>() </span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-279-1.png" width="672" /></p>
<p>Can we also sort the data based on something else? Definitely, but it requires a bit of understanding of functions. Previously, we used the maximum value. This is defined by the function <code>max</code>, which takes the maxium value from a vector of numbers. Let’s look at an artificial example:</p>
<div class="sourceCode" id="cb339"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb339-1"><a href="complete-protocols.html#cb339-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>,<span class="dv">9</span>)</span>
<span id="cb339-2"><a href="complete-protocols.html#cb339-2" aria-hidden="true" tabindex="-1"></a><span class="fu">max</span>(x)</span></code></pre></div>
<pre><code>[1] 9</code></pre>
<p>Other functions that take a vector as input and return a single value as output can be used. Other existing examples are <code>mean()</code>, <code>sum()</code> and <code>min()</code>. We can also define a function:</p>
<div class="sourceCode" id="cb341"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb341-1"><a href="complete-protocols.html#cb341-1" aria-hidden="true" tabindex="-1"></a>second <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {x[<span class="dv">2</span>]}</span>
<span id="cb341-2"><a href="complete-protocols.html#cb341-2" aria-hidden="true" tabindex="-1"></a><span class="fu">second</span>(x)</span></code></pre></div>
<pre><code>[1] 3</code></pre>
<div class="sourceCode" id="cb343"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb343-1"><a href="complete-protocols.html#cb343-1" aria-hidden="true" tabindex="-1"></a>one_but_last <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {x[<span class="fu">length</span>(x)<span class="sc">-</span><span class="dv">1</span>]}</span>
<span id="cb343-2"><a href="complete-protocols.html#cb343-2" aria-hidden="true" tabindex="-1"></a><span class="fu">one_but_last</span>(x)</span></code></pre></div>
<pre><code>[1] 7</code></pre>
<p>We can use the new function to sort the dataframe:</p>
<div class="sourceCode" id="cb345"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb345-1"><a href="complete-protocols.html#cb345-1" aria-hidden="true" tabindex="-1"></a>df_sub <span class="ot">&lt;-</span> df_sub <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Unique_Object =</span> <span class="fu">fct_reorder</span>(Unique_Object, ERKn, one_but_last)) </span>
<span id="cb345-2"><a href="complete-protocols.html#cb345-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb345-3"><a href="complete-protocols.html#cb345-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_sub, <span class="fu">aes</span>(<span class="at">x=</span>Time_in_min, <span class="at">y=</span>Unique_Object,<span class="at">fill=</span>ERKn)) <span class="sc">+</span></span>
<span id="cb345-4"><a href="complete-protocols.html#cb345-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>() <span class="sc">+</span> <span class="fu">theme_void</span>()<span class="sc">+</span></span>
<span id="cb345-5"><a href="complete-protocols.html#cb345-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>() </span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-282-1.png" width="672" /></p>
<p>If we want to sort on the sum of the top five values we can define a function:</p>
<div class="sourceCode" id="cb346"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb346-1"><a href="complete-protocols.html#cb346-1" aria-hidden="true" tabindex="-1"></a>top_five <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {<span class="fu">sum</span>(<span class="fu">tail</span>(<span class="fu">sort</span>(x),<span class="dv">5</span>))}</span></code></pre></div>
<p>But we can also directly implement the function in the <code>fct_reorder()</code> function:</p>
<div class="sourceCode" id="cb347"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb347-1"><a href="complete-protocols.html#cb347-1" aria-hidden="true" tabindex="-1"></a>df_6 <span class="ot">&lt;-</span> df_sub <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Unique_Object =</span> <span class="fu">fct_reorder</span>(Unique_Object, ERKn, <span class="cf">function</span>(x) {<span class="fu">sum</span>(<span class="fu">tail</span>(<span class="fu">sort</span>(x),<span class="dv">5</span>))}))</span></code></pre></div>
<p>Let’s save this data:</p>
<div class="sourceCode" id="cb348"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb348-1"><a href="complete-protocols.html#cb348-1" aria-hidden="true" tabindex="-1"></a>df_6 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_6.csv&quot;</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb349"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb349-1"><a href="complete-protocols.html#cb349-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span>  <span class="fu">ggplot</span>(df_6, <span class="fu">aes</span>(<span class="at">x=</span>Time_in_min, <span class="at">y=</span>Unique_Object,<span class="at">fill=</span>ERKn)) <span class="sc">+</span> <span class="fu">geom_tile</span>() <span class="sc">+</span></span>
<span id="cb349-2"><a href="complete-protocols.html#cb349-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># scale_x_continuous(breaks=seq(0,60, by=15), labels=seq(0,60, by=15), limits=c(-8,60))  +</span></span>
<span id="cb349-3"><a href="complete-protocols.html#cb349-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>(<span class="at">na.value=</span><span class="st">&quot;black&quot;</span>, <span class="at">limits =</span> <span class="fu">range</span>(<span class="sc">-</span><span class="fl">0.1</span>,<span class="fl">1.5</span>))</span></code></pre></div>
<p>Let’s look at the plot</p>
<div class="sourceCode" id="cb350"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb350-1"><a href="complete-protocols.html#cb350-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-287-1.png" width="672" /></p>
<p>Add labels:</p>
<div class="sourceCode" id="cb351"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb351-1"><a href="complete-protocols.html#cb351-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb351-2"><a href="complete-protocols.html#cb351-2" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb351-3"><a href="complete-protocols.html#cb351-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;ERK activity is increased by S1P&quot;</span>,</span>
<span id="cb351-4"><a href="complete-protocols.html#cb351-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;and varies strongly between cells&quot;</span>,</span>
<span id="cb351-5"><a href="complete-protocols.html#cb351-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Time after addition of S1P (min)&quot;</span>,</span>
<span id="cb351-6"><a href="complete-protocols.html#cb351-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Cells sorted according to response&quot;</span>,</span>
<span id="cb351-7"><a href="complete-protocols.html#cb351-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart | data from Chavez-Abiega, DOI: 10.1101/2021.07.27.453948&quot;</span>,</span>
<span id="cb351-8"><a href="complete-protocols.html#cb351-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 6&quot;</span>,</span>
<span id="cb351-9"><a href="complete-protocols.html#cb351-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill=</span> <span class="st">&quot;ERK C/N ratio&quot;</span></span>
<span id="cb351-10"><a href="complete-protocols.html#cb351-10" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>The <code>theme_void()</code> would be close to what we want as a theme, but I prefer to start from <code>theme_light</code> and remove the redundant features (grids and y-axis labels):</p>
<div class="sourceCode" id="cb352"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb352-1"><a href="complete-protocols.html#cb352-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb352-2"><a href="complete-protocols.html#cb352-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb352-3"><a href="complete-protocols.html#cb352-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>, <span class="at">hjust =</span> <span class="fl">2.0</span>),</span>
<span id="cb352-4"><a href="complete-protocols.html#cb352-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.1</span>, <span class="at">margin =</span> <span class="fu">margin</span>(<span class="at">t=</span><span class="dv">10</span>)),</span>
<span id="cb352-5"><a href="complete-protocols.html#cb352-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.1</span>, <span class="at">margin =</span> <span class="fu">margin</span>(<span class="at">t=</span><span class="dv">2</span>, <span class="at">b=</span><span class="dv">5</span>)),</span>
<span id="cb352-6"><a href="complete-protocols.html#cb352-6" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Remove background</span></span>
<span id="cb352-7"><a href="complete-protocols.html#cb352-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb352-8"><a href="complete-protocols.html#cb352-8" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Remove borders</span></span>
<span id="cb352-9"><a href="complete-protocols.html#cb352-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb352-10"><a href="complete-protocols.html#cb352-10" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Remove grid</span></span>
<span id="cb352-11"><a href="complete-protocols.html#cb352-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb352-12"><a href="complete-protocols.html#cb352-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb352-13"><a href="complete-protocols.html#cb352-13" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Remove text of the y-axis</span></span>
<span id="cb352-14"><a href="complete-protocols.html#cb352-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb352-15"><a href="complete-protocols.html#cb352-15" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Remove ticks on y-axis</span></span>
<span id="cb352-16"><a href="complete-protocols.html#cb352-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb352-17"><a href="complete-protocols.html#cb352-17" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Remove label of y-axis</span></span>
<span id="cb352-18"><a href="complete-protocols.html#cb352-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb352-19"><a href="complete-protocols.html#cb352-19" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Make x-axis ticks more pronounced</span></span>
<span id="cb352-20"><a href="complete-protocols.html#cb352-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks =</span> <span class="fu">element_line</span>(<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb352-21"><a href="complete-protocols.html#cb352-21" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb352-22"><a href="complete-protocols.html#cb352-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb352-23"><a href="complete-protocols.html#cb352-23" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-289-1.png" width="672" /></p>
<p>The caption does not look good in this plot, but it has been optimized to look good in the saved PNG. To get a proper aligned caption in the Rmd you may need to optimize the hjust value in <code>theme((plot.caption = element_text()))</code></p>
<p>To save the plot as PNG:</p>
<div class="sourceCode" id="cb353"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb353-1"><a href="complete-protocols.html#cb353-1" aria-hidden="true" tabindex="-1"></a> <span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_6.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb353-2"><a href="complete-protocols.html#cb353-2" aria-hidden="true" tabindex="-1"></a>   p</span>
<span id="cb353-3"><a href="complete-protocols.html#cb353-3" aria-hidden="true" tabindex="-1"></a> <span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-7" class="section level2 hasAnchor" number="4.7">
<h2><span class="header-section-number">4.7</span> Protocol 7 - Ridgeline plot<a href="complete-protocols.html#protocol-7" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In protocol 6 we have seen how time traces can be visualized with a heatmap style data visualization.
The advantage is that huge amounts of data can be simultaneously visualized. However, it is difficult to relate the color to numbers.
As such the color coded values in a heatmap give a qualitative view of heterogeneity.
In this protocol, we use a method that allows more data than an ordinary lineplot, but keeps the advantage of this plot, i.e. allowing a quantitative comparison.
To do this, an offset between curves is introduced. A package to this directly in ggplot2 is available (<code>{ggridges}</code>) but it may not work well when the values show a negative and positive deviation from baseline values (usually the baseline is zero).
In addition, we have more flexibility (and insight) when we do this ourselves.</p>
<p>Load the <code>{tidyverse}</code> package and the data from a CSV file:</p>
<div class="sourceCode" id="cb355"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb355-1"><a href="complete-protocols.html#cb355-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(tidyverse)</span>
<span id="cb355-2"><a href="complete-protocols.html#cb355-2" aria-hidden="true" tabindex="-1"></a>df1 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;Rac_S1P.csv&quot;</span>)</span>
<span id="cb355-3"><a href="complete-protocols.html#cb355-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df1)</span></code></pre></div>
<pre><code>       Time    Cell.1    Cell.2    Cell.3    Cell.4    Cell.5    Cell.6
1 0.0000000 1.0012160 1.0026460 1.0022090 0.9917870 0.9935569 0.9961453
2 0.1666667 0.9994997 0.9928106 0.9997658 0.9975348 1.0018910 1.0039790
3 0.3333333 0.9908362 0.9964057 0.9905094 0.9946743 0.9961497 0.9953369
4 0.5000000 0.9991967 0.9972504 0.9972806 1.0074250 1.0060510 1.0062390
5 0.6666667 1.0093450 1.0109910 1.0103590 1.0084080 1.0022130 0.9982496
6 0.8333333 0.9941078 0.9940830 0.9990720 1.0181230 1.0110220 1.0139400
     Cell.7    Cell.8    Cell.9   Cell.10   Cell.11   Cell.12   Cell.13
1 0.9964277 1.0006770 0.9999106 1.0043270 1.0086590 1.0074670 0.9875549
2 1.0098780 1.0015050 1.0014380 0.9938114 0.9791933 0.9925425 1.0106440
3 0.9917276 1.0000020 0.9997350 1.0049370 1.0250660 1.0079290 1.0062720
4 1.0039710 1.0046840 0.9976190 0.9971925 1.0008340 0.9975954 1.0050230
5 0.9979655 0.9930890 1.0012990 0.9997435 0.9865567 0.9944269 0.9907017
6 1.0087770 0.9916067 1.0025310 0.9919835 0.9930975 1.0012420 1.0025800
    Cell.14   Cell.15   Cell.16   Cell.17   Cell.18   Cell.19   Cell.20
1 0.9941667 0.9931722 0.9918481 0.9973316 0.9976466 0.9982940 0.9954924
2 1.0024420 0.9980425 0.9974822 0.9976164 0.9881339 1.0122250 1.0132220
3 1.0096800 1.0074450 1.0073690 0.9980090 0.9980346 0.9923310 0.9891400
4 1.0007800 0.9995486 1.0018860 1.0064760 1.0073410 0.9959588 0.9984370
5 0.9930816 1.0018040 1.0014300 1.0006490 1.0088410 1.0013600 1.0040700
6 1.0117190 1.0082190 1.0092750 1.0018410 0.9976196 1.0008680 1.0013100
    Cell.21   Cell.22   Cell.23   Cell.24   Cell.25   Cell.26   Cell.27
1 1.0024980 0.9986917 0.9987728 0.9988987 1.0019450 0.9954534 1.0025060
2 1.0022340 1.0023420 0.9963444 0.9969288 0.9990684 0.9973828 0.9944580
3 1.0003150 1.0014660 0.9985501 0.9983435 0.9944611 0.9987581 0.9935282
4 0.9969226 0.9956030 1.0020300 1.0063250 1.0036700 1.0056470 1.0065540
5 0.9980916 1.0018990 1.0042990 0.9995825 1.0009020 1.0028340 1.0031070
6 0.9993777 1.0041500 0.9984087 1.0016190 1.0007090 1.0022360 1.0007950
    Cell.28   Cell.29   Cell.30   Cell.31   Cell.32
1 0.9957569 0.9852318 1.0007450 0.9927866 0.9871355
2 0.9952635 0.9931840 1.0030060 0.9977890 1.0028110
3 0.9976221 1.0010230 0.9998598 1.0029210 0.9985340
4 1.0068090 1.0139990 0.9968880 1.0026060 1.0039670
5 1.0046560 1.0070850 0.9995161 1.0039100 1.0076550
6 1.0073290 1.0128910 1.0000250 1.0025260 0.9998251</code></pre>
<p>The data is not tidy, so it needs to be re-arranged:</p>
<div class="sourceCode" id="cb357"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb357-1"><a href="complete-protocols.html#cb357-1" aria-hidden="true" tabindex="-1"></a>df_tidy <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(df1, <span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(Time), <span class="at">names_to =</span> <span class="st">&quot;Cell&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;Activity&quot;</span>)</span>
<span id="cb357-2"><a href="complete-protocols.html#cb357-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_tidy)</span></code></pre></div>
<pre><code># A tibble: 6 × 3
   Time Cell   Activity
  &lt;dbl&gt; &lt;chr&gt;     &lt;dbl&gt;
1     0 Cell.1    1.00 
2     0 Cell.2    1.00 
3     0 Cell.3    1.00 
4     0 Cell.4    0.992
5     0 Cell.5    0.994
6     0 Cell.6    0.996</code></pre>
<p>In the next step, we create a new dataframe ‘df_rank’ to order the traces. We group the data by ‘Cell’ and extract the data from a specified time window with <code>filter()</code>. The filtered data is used to integrate the activity by using the function <code>sum()</code>. This summed value is used to generate a rank, ranging from 0 to 1:</p>
<div class="sourceCode" id="cb359"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb359-1"><a href="complete-protocols.html#cb359-1" aria-hidden="true" tabindex="-1"></a>df_rank <span class="ot">&lt;-</span> df_tidy <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Cell) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Time<span class="sc">&gt;=</span><span class="dv">2</span> <span class="sc">&amp;</span> Time <span class="sc">&lt;=</span><span class="dv">10</span>) <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(<span class="at">amplitude=</span><span class="fu">sum</span>(Activity)) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">rank=</span><span class="fu">percent_rank</span>(amplitude))</span>
<span id="cb359-2"><a href="complete-protocols.html#cb359-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_rank)</span></code></pre></div>
<pre><code># A tibble: 6 × 3
  Cell    amplitude  rank
  &lt;chr&gt;       &lt;dbl&gt; &lt;dbl&gt;
1 Cell.1       50.3 0.742
2 Cell.10      50.9 0.806
3 Cell.11      49.8 0.645
4 Cell.12      50.3 0.774
5 Cell.13      49.6 0.548
6 Cell.14      49.4 0.484</code></pre>
<p>We can add the rank information from ‘df_rank’ to the ‘df_tidy’ dataframe:</p>
<div class="sourceCode" id="cb361"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb361-1"><a href="complete-protocols.html#cb361-1" aria-hidden="true" tabindex="-1"></a>df_7 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">full_join</span>(df_tidy, df_rank,<span class="at">by=</span><span class="st">&quot;Cell&quot;</span>))</span></code></pre></div>
<p>This data is saved:</p>
<div class="sourceCode" id="cb362"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb362-1"><a href="complete-protocols.html#cb362-1" aria-hidden="true" tabindex="-1"></a>df_7 <span class="sc">%&gt;%</span> <span class="fu">write.csv</span>(<span class="st">&quot;protocol_7.csv&quot;</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>Let’s make a lineplot of this data and use the rank to shift the data plotted on the y-axis:</p>
<div class="sourceCode" id="cb363"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb363-1"><a href="complete-protocols.html#cb363-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_7, <span class="fu">aes</span>(<span class="at">x=</span>Time, <span class="at">y=</span>Activity<span class="sc">+</span>(rank<span class="sc">*</span><span class="dv">1</span>), <span class="at">group=</span>Cell, <span class="at">height =</span> rank)) <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">size=</span><span class="fl">0.5</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-296-1.png" width="672" /></p>
<p>We can use the rank to shift the plot also in the horizontal direction:</p>
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb364-1"><a href="complete-protocols.html#cb364-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_7, <span class="fu">aes</span>(<span class="at">x=</span>Time<span class="sc">+</span>(rank<span class="sc">*</span><span class="dv">10</span>), <span class="at">y=</span>(Activity<span class="sc">+</span>(rank<span class="sc">*</span><span class="dv">1</span>)), <span class="at">group=</span>Cell)) <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">size=</span><span class="fl">0.5</span>)</span>
<span id="cb364-2"><a href="complete-protocols.html#cb364-2" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-297-1.png" width="672" /></p>
<p>Add labels:</p>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb365-1"><a href="complete-protocols.html#cb365-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb365-2"><a href="complete-protocols.html#cb365-2" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb365-3"><a href="complete-protocols.html#cb365-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Activities measured in single cells over time&quot;</span>,</span>
<span id="cb365-4"><a href="complete-protocols.html#cb365-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;Sorting of the cells is based on integrated activity&quot;</span>,</span>
<span id="cb365-5"><a href="complete-protocols.html#cb365-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;Time [min]&quot;</span>,</span>
<span id="cb365-6"><a href="complete-protocols.html#cb365-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Activity [arbitrary units]&quot;</span>,</span>
<span id="cb365-7"><a href="complete-protocols.html#cb365-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart | data from Reinhard et al; doi: 10.1091/mbc.E17-03-0136&quot;</span>,</span>
<span id="cb365-8"><a href="complete-protocols.html#cb365-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 7&quot;</span></span>
<span id="cb365-9"><a href="complete-protocols.html#cb365-9" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>If we would like to use color, this would be a way to do that:</p>
<div class="sourceCode" id="cb366"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb366-1"><a href="complete-protocols.html#cb366-1" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">color=</span><span class="fu">as.factor</span>(rank))) <span class="sc">+</span></span>
<span id="cb366-2"><a href="complete-protocols.html#cb366-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-299-1.png" width="672" /></p>
<p>Although it looks flashy, we do not really need color here. So we stick to black and white and make some adjustments to the layout by tweaking the theme settings. To remove the grid and show the axis:</p>
<div class="sourceCode" id="cb367"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb367-1"><a href="complete-protocols.html#cb367-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">16</span>)</span>
<span id="cb367-2"><a href="complete-protocols.html#cb367-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb367-3"><a href="complete-protocols.html#cb367-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>),</span>
<span id="cb367-4"><a href="complete-protocols.html#cb367-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb367-5"><a href="complete-protocols.html#cb367-5" aria-hidden="true" tabindex="-1"></a>           <span class="cn">NULL</span>)</span>
<span id="cb367-6"><a href="complete-protocols.html#cb367-6" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-300-1.png" width="672" /></p>
<p>To give it more of a 3D feel we can add a third axis by defining a line:</p>
<div class="sourceCode" id="cb368"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb368-1"><a href="complete-protocols.html#cb368-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;segment&quot;</span>, <span class="at">x=</span><span class="dv">0</span>,<span class="at">y=</span><span class="dv">1</span>,<span class="at">xend=</span><span class="dv">10</span>,<span class="at">yend=</span><span class="dv">2</span>, <span class="at">size=</span><span class="dv">1</span>)</span></code></pre></div>
<p>The next step is to remove the original x-axis, which is a bit too long and also replace that with a line that runs until 20 (minutes):</p>
<div class="sourceCode" id="cb369"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb369-1"><a href="complete-protocols.html#cb369-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.line.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb369-2"><a href="complete-protocols.html#cb369-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.3</span>)</span>
<span id="cb369-3"><a href="complete-protocols.html#cb369-3" aria-hidden="true" tabindex="-1"></a>               ) <span class="sc">+</span></span>
<span id="cb369-4"><a href="complete-protocols.html#cb369-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">&quot;segment&quot;</span>, <span class="at">x=</span><span class="dv">0</span>,<span class="at">y=</span><span class="dv">1</span>,<span class="at">xend=</span><span class="dv">20</span>,<span class="at">yend=</span><span class="dv">1</span>, <span class="at">size=</span><span class="dv">1</span>)</span>
<span id="cb369-5"><a href="complete-protocols.html#cb369-5" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-302-1.png" width="672" /></p>
<p>Finally, we can use <code>scale_x_continuous()</code> to improve two aspects of the x-axis. First, the labels run up to 30, but we can set the scale to 0-20 with <code>breaks = c(0,20)</code>. Second, the default in ggplot is add a bit of margin to the plot, that’s why the x-axis does not touch the y-axis. This can be solved by using <code>expand = c(0, 0)</code>:
We can manually add two</p>
<div class="sourceCode" id="cb370"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb370-1"><a href="complete-protocols.html#cb370-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">10</span>,<span class="dv">20</span>), <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>))</span>
<span id="cb370-2"><a href="complete-protocols.html#cb370-2" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-303-1.png" width="672" /></p>
<p>To save the plot as PNG:</p>
<div class="sourceCode" id="cb371"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb371-1"><a href="complete-protocols.html#cb371-1" aria-hidden="true" tabindex="-1"></a> <span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_7.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb371-2"><a href="complete-protocols.html#cb371-2" aria-hidden="true" tabindex="-1"></a>   p</span>
<span id="cb371-3"><a href="complete-protocols.html#cb371-3" aria-hidden="true" tabindex="-1"></a> <span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-8" class="section level2 hasAnchor" number="4.8">
<h2><span class="header-section-number">4.8</span> Protocol 8 - Plotting data in a 96-wells layout<a href="complete-protocols.html#protocol-8" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This protocol displays data from a 96-wells plate in the same format as the plate. This data visualization of 96-well plate data is also used in the <a href="https://doi.org/10.12688/f1000research.73641.1">plotXpress app</a>.</p>
<p>We start by loading a package that we need:</p>
<div class="sourceCode" id="cb373"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb373-1"><a href="complete-protocols.html#cb373-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>A good example of data wrangling related to 96-wells data is given in <a href="complete-protocols.html#protocol-4">protocol 4</a>. Here we start from the tidy data that was the result of data tiyding in the section <a href="#Data-in-96-wells-format">Data in 96-wells format</a>:</p>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb374-1"><a href="complete-protocols.html#cb374-1" aria-hidden="true" tabindex="-1"></a>df_8 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;df_tidy_wells.csv&quot;</span>)</span>
<span id="cb374-2"><a href="complete-protocols.html#cb374-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_8)</span></code></pre></div>
<pre><code>  X column row Well Intensity
1 1      1   A   A1      2010
2 2      1   B   B1      3210
3 3      1   C   C1      1965
4 4      1   D   D1      2381
5 5      1   E   E1      1292
6 6      1   F   F1       991</code></pre>
<p>We can construct a basic plot where each well is represented by a large dot and the color used to fill the dot represents the intensity. This is much like a heatmap, but it uses a dot instead of a tile:</p>
<div class="sourceCode" id="cb376"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb376-1"><a href="complete-protocols.html#cb376-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>df_8, <span class="fu">aes</span>(<span class="at">x=</span>column, <span class="at">y=</span>row)) <span class="sc">+</span></span>
<span id="cb376-2"><a href="complete-protocols.html#cb376-2" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color=</span>Intensity), <span class="at">size=</span><span class="dv">10</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-307-1.png" width="672" />
There is a couple of things that need to be fixed. First, we invert the row names, to put the row ‘A’ at the top. In addition, the column names are depicted as factors (this can also be achieved by mutating the dataframe):</p>
<div class="sourceCode" id="cb377"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb377-1"><a href="complete-protocols.html#cb377-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>df_8, <span class="fu">aes</span>(<span class="at">x=</span><span class="fu">as.factor</span>(column), <span class="at">y=</span><span class="fu">fct_rev</span>(row))) <span class="sc">+</span></span>
<span id="cb377-2"><a href="complete-protocols.html#cb377-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color=</span>Intensity), <span class="at">size=</span><span class="dv">10</span>)</span>
<span id="cb377-3"><a href="complete-protocols.html#cb377-3" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-308-1.png" width="672" /></p>
<p>Let’s change the color palette that displays intensity to viridis. The numbers in the legend are high and that is why they are difficult to read quickly. So we use the <code>scales</code> package to use the scientific notation:</p>
<div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb378-1"><a href="complete-protocols.html#cb378-1" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_x_discrete</span>(<span class="at">position =</span> <span class="st">&quot;top&quot;</span>) <span class="sc">+</span></span>
<span id="cb378-2"><a href="complete-protocols.html#cb378-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">label =</span> scales<span class="sc">::</span>scientific)</span>
<span id="cb378-3"><a href="complete-protocols.html#cb378-3" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-309-1.png" width="672" /></p>
<p>In my opinion, the scientific notation does not work well, so let’s try to use commas as a thousands separator:</p>
<div class="sourceCode" id="cb379"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb379-1"><a href="complete-protocols.html#cb379-1" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_x_discrete</span>(<span class="at">position =</span> <span class="st">&quot;top&quot;</span>) <span class="sc">+</span></span>
<span id="cb379-2"><a href="complete-protocols.html#cb379-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">label =</span> scales<span class="sc">::</span>comma)</span></code></pre></div>
<pre><code>Scale for &#39;x&#39; is already present. Adding another scale for &#39;x&#39;, which will
replace the existing scale.</code></pre>
<pre><code>Scale for &#39;colour&#39; is already present. Adding another scale for &#39;colour&#39;,
which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb382-1"><a href="complete-protocols.html#cb382-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-310-1.png" width="672" /></p>
<p>This is a bit better, but the numbers need to be right aligned:</p>
<div class="sourceCode" id="cb383"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb383-1"><a href="complete-protocols.html#cb383-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.text.align =</span> <span class="dv">1</span>)</span>
<span id="cb383-2"><a href="complete-protocols.html#cb383-2" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-311-1.png" width="672" />
This looks better, now add the labels (and we skip the labels for the x- and y-axis):</p>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb384-1"><a href="complete-protocols.html#cb384-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span></span>
<span id="cb384-2"><a href="complete-protocols.html#cb384-2" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb384-3"><a href="complete-protocols.html#cb384-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Readings from a 96-wells plate&quot;</span>,</span>
<span id="cb384-4"><a href="complete-protocols.html#cb384-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="cn">NULL</span>,</span>
<span id="cb384-5"><a href="complete-protocols.html#cb384-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="cn">NULL</span>,</span>
<span id="cb384-6"><a href="complete-protocols.html#cb384-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="cn">NULL</span>,</span>
<span id="cb384-7"><a href="complete-protocols.html#cb384-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">&quot;</span><span class="sc">\n</span><span class="st">@joachimgoedhart</span><span class="sc">\n</span><span class="st">based on data from Brandorff et al., DOI: 10.1101/2021.07.08.451595&quot;</span>,</span>
<span id="cb384-8"><a href="complete-protocols.html#cb384-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">color=</span> <span class="st">&#39;Intensity [a.u.]&#39;</span>,</span>
<span id="cb384-9"><a href="complete-protocols.html#cb384-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">tag =</span> <span class="st">&quot;Protocol 8&quot;</span></span>
<span id="cb384-10"><a href="complete-protocols.html#cb384-10" aria-hidden="true" tabindex="-1"></a>  ) </span>
<span id="cb384-11"><a href="complete-protocols.html#cb384-11" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-312-1.png" width="672" /></p>
<p>Set the theme and font size:</p>
<div class="sourceCode" id="cb385"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb385-1"><a href="complete-protocols.html#cb385-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_light</span>(<span class="at">base_size =</span> <span class="dv">14</span>)</span></code></pre></div>
<p>Note that this overrides any previously defined modifications of the <code>theme()</code> function, such as the alignment of the legend labels. So we need to define this again.</p>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb386-1"><a href="complete-protocols.html#cb386-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.text.align =</span> <span class="dv">1</span>)</span></code></pre></div>
<p>We also adjust the theme settings for the other elements:</p>
<div class="sourceCode" id="cb387"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb387-1"><a href="complete-protocols.html#cb387-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>),</span>
<span id="cb387-2"><a href="complete-protocols.html#cb387-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey50&quot;</span>, <span class="at">face =</span> <span class="st">&quot;italic&quot;</span>),</span>
<span id="cb387-3"><a href="complete-protocols.html#cb387-3" aria-hidden="true" tabindex="-1"></a>               <span class="co">#Remove the grid</span></span>
<span id="cb387-4"><a href="complete-protocols.html#cb387-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb387-5"><a href="complete-protocols.html#cb387-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>()</span>
<span id="cb387-6"><a href="complete-protocols.html#cb387-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Let’s look at the result:</p>
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb388-1"><a href="complete-protocols.html#cb388-1" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-316-1.png" width="672" /></p>
<p>To save the plot as a png file:</p>
<div class="sourceCode" id="cb389"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb389-1"><a href="complete-protocols.html#cb389-1" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_8.png&quot;</span>), <span class="at">width =</span> <span class="dv">3000</span>, <span class="at">height =</span> <span class="dv">2000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb389-2"><a href="complete-protocols.html#cb389-2" aria-hidden="true" tabindex="-1"></a> p <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span>
<span id="cb389-3"><a href="complete-protocols.html#cb389-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
</div>
<div id="protocol-9" class="section level2 hasAnchor" number="4.9">
<h2><span class="header-section-number">4.9</span> Protocol 9 - A dose response curve with fit<a href="complete-protocols.html#protocol-9" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this protocol we visualize the data that was used to make a ‘dose reponse curve’. The data consists of some response that is measured at different concentrations of a compound that induces the effect (agonist). This is a typical experiment in pharmacology to understand the relation between the dose and the response. One of the relevant parameters is the ‘Half maximal effective concentration’ abbreviated as EC50. This is the concentration at which 50% of the maximal response is measured and we obtain this value by curve fitting. In this protocol we use data acquired from single cells from different biological replicates. We generate a ‘superplot’ <a href="https://doi.org/10.1083/jcb.202001064">(Lord et al., 2021)</a> to distuinguish technical and sample replicates. More details about the experimental approach are published by <a href="https://doi.org/10.1242/jcs.259685">Chavez-Abiega et al. (2022)</a>.</p>
<p>We start by loading the <code>{tidyverse}</code> package:</p>
<div class="sourceCode" id="cb391"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb391-1"><a href="complete-protocols.html#cb391-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>Read the data:</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb392-1"><a href="complete-protocols.html#cb392-1" aria-hidden="true" tabindex="-1"></a>df_DRC <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;DRC_Histamine.csv&quot;</span>, <span class="at">stringsAsFactors =</span> <span class="cn">TRUE</span>)</span>
<span id="cb392-2"><a href="complete-protocols.html#cb392-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_DRC)</span></code></pre></div>
<pre><code>  X Unique_Unique_Object  Experiment  Aktnac  ERKnac Condition Concen
1 1                   12 20200122 P1  0.0558 -0.2106       PTx      0
2 2                   15 20200122 P1  0.2458 -0.1424       PTx      0
3 3                   16 20200122 P1  0.0412 -0.2493       PTx      0
4 4                   22 20200122 P1 -0.1194 -0.1980       PTx      0
5 5                   23 20200122 P1  0.1838 -0.1540       PTx      0
6 6                   25 20200122 P1  0.3291 -0.7078       PTx      0</code></pre>
<p>The concentration is listed in the column ‘Concen’, the measured response is in the column ‘ERKnac’ and the column ‘Experiment’ identifies the replicates. Let’s rename the columns for easier identification of what they represent:</p>
<div class="sourceCode" id="cb394"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb394-1"><a href="complete-protocols.html#cb394-1" aria-hidden="true" tabindex="-1"></a>df_DRC <span class="ot">&lt;-</span> df_DRC <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">Concentration=</span>Concen, <span class="at">Response=</span>ERKnac)</span>
<span id="cb394-2"><a href="complete-protocols.html#cb394-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_DRC)</span></code></pre></div>
<pre><code>  X Unique_Unique_Object  Experiment  Aktnac Response Condition Concentration
1 1                   12 20200122 P1  0.0558  -0.2106       PTx             0
2 2                   15 20200122 P1  0.2458  -0.1424       PTx             0
3 3                   16 20200122 P1  0.0412  -0.2493       PTx             0
4 4                   22 20200122 P1 -0.1194  -0.1980       PTx             0
5 5                   23 20200122 P1  0.1838  -0.1540       PTx             0
6 6                   25 20200122 P1  0.3291  -0.7078       PTx             0</code></pre>
<p>To change the column ‘Experiment’ from text into a number that represents the replicate, we can convert it using <code>as.numeric()</code>. Since we need these numbers as qualitative data (a label), we convert the numbers to a factor with <code>as.factor()</code>:</p>
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb396-1"><a href="complete-protocols.html#cb396-1" aria-hidden="true" tabindex="-1"></a>df_DRC <span class="ot">&lt;-</span> df_DRC <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Replicate=</span><span class="fu">as.factor</span>(<span class="fu">as.integer</span>(Experiment)))</span>
<span id="cb396-2"><a href="complete-protocols.html#cb396-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df_DRC)</span></code></pre></div>
<pre><code>  X Unique_Unique_Object  Experiment  Aktnac Response Condition Concentration
1 1                   12 20200122 P1  0.0558  -0.2106       PTx             0
2 2                   15 20200122 P1  0.2458  -0.1424       PTx             0
3 3                   16 20200122 P1  0.0412  -0.2493       PTx             0
4 4                   22 20200122 P1 -0.1194  -0.1980       PTx             0
5 5                   23 20200122 P1  0.1838  -0.1540       PTx             0
6 6                   25 20200122 P1  0.3291  -0.7078       PTx             0
  Replicate
1         1
2         1
3         1
4         1
5         1
6         1</code></pre>
<p>The range of concentrations at which the compound is examined spans a few orders of magnitude and therefore a log scale is used to display the concentrations. The minimal response is usually measured at a concentration of 0, but the logarithm of 0 is undefined. Therefore, plotting 0 on a logscale will give an error. The logarithm of 0 can be approached by minus infinity. Therefore, we convert the concentration of 0 to a low value, in this case 0.01:</p>
<div class="sourceCode" id="cb398"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb398-1"><a href="complete-protocols.html#cb398-1" aria-hidden="true" tabindex="-1"></a>df_DRC <span class="ot">&lt;-</span> df_DRC  <span class="sc">%&gt;%</span></span>
<span id="cb398-2"><a href="complete-protocols.html#cb398-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Concentration =</span> <span class="fu">ifelse</span>((Concentration <span class="sc">==</span> <span class="dv">0</span>),</span>
<span id="cb398-3"><a href="complete-protocols.html#cb398-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">yes =</span> <span class="fl">0.01</span>,</span>
<span id="cb398-4"><a href="complete-protocols.html#cb398-4" aria-hidden="true" tabindex="-1"></a>                         <span class="at">no =</span> Concentration)</span>
<span id="cb398-5"><a href="complete-protocols.html#cb398-5" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>Next we take the ‘Response’ and calculate the average per concentration and for each biological replicate and store that information in a new dataframe:</p>
<div class="sourceCode" id="cb399"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb399-1"><a href="complete-protocols.html#cb399-1" aria-hidden="true" tabindex="-1"></a>df_summary <span class="ot">&lt;-</span> df_DRC <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Concentration, Replicate) <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(<span class="at">mean_Response=</span><span class="fu">mean</span>(Response))</span></code></pre></div>
<p>We can define a plot that shows the data with <code>geom_jitter()</code> and the average with a large dot with <code>geom_point()</code>. Each replicate has its own color and the data is plotted on a log scale with <code>scale_x_log10()</code>:</p>
<div class="sourceCode" id="cb400"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb400-1"><a href="complete-protocols.html#cb400-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span>  <span class="fu">ggplot</span>(<span class="at">data =</span> df_DRC, <span class="fu">aes</span>(<span class="at">x =</span> Concentration, <span class="at">y =</span> Response)) <span class="sc">+</span> </span>
<span id="cb400-2"><a href="complete-protocols.html#cb400-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(<span class="at">x =</span> Concentration, <span class="at">y =</span> Response, <span class="at">color=</span>Replicate), <span class="at">width=</span><span class="fl">0.2</span>, <span class="at">size=</span><span class="dv">2</span>, <span class="at">shape=</span><span class="dv">16</span>, <span class="at">alpha=</span><span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb400-3"><a href="complete-protocols.html#cb400-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>df_summary, <span class="fu">aes</span>(<span class="at">x =</span> Concentration, <span class="at">y =</span> mean_Response, <span class="at">fill=</span>Replicate), <span class="at">size=</span><span class="dv">8</span>, <span class="at">shape=</span><span class="dv">21</span>, <span class="at">alpha=</span><span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb400-4"><a href="complete-protocols.html#cb400-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>()</span>
<span id="cb400-5"><a href="complete-protocols.html#cb400-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb400-6"><a href="complete-protocols.html#cb400-6" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-325-1.png" width="672" /></p>
<p>Adjusting the theme and the y-axis scale improves the plot:</p>
<div class="sourceCode" id="cb401"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb401-1"><a href="complete-protocols.html#cb401-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span>  p <span class="sc">+</span></span>
<span id="cb401-2"><a href="complete-protocols.html#cb401-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb401-3"><a href="complete-protocols.html#cb401-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_light</span>(<span class="dv">16</span>)</span>
<span id="cb401-4"><a href="complete-protocols.html#cb401-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb401-5"><a href="complete-protocols.html#cb401-5" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-326-1.png" width="672" /></p>
<p>Add labels:</p>
<div class="sourceCode" id="cb402"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb402-1"><a href="complete-protocols.html#cb402-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">labs</span>(</span>
<span id="cb402-2"><a href="complete-protocols.html#cb402-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;Responses of individual cells to Histamine&quot;</span>,</span>
<span id="cb402-3"><a href="complete-protocols.html#cb402-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="st">&quot;Histamine concentration [µM]&quot;</span>,</span>
<span id="cb402-4"><a href="complete-protocols.html#cb402-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="st">&quot;Response [arbitrary units]&quot;</span>,</span>
<span id="cb402-5"><a href="complete-protocols.html#cb402-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">caption =</span> <span class="st">&quot;@joachimgoedhart</span><span class="sc">\n</span><span class="st">based on data from Chavez-Abiega et al., DOI: 10.1242/jcs.259685&quot;</span>,</span>
<span id="cb402-6"><a href="complete-protocols.html#cb402-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">tag =</span> <span class="st">&quot;Protocol 9&quot;</span></span>
<span id="cb402-7"><a href="complete-protocols.html#cb402-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb402-8"><a href="complete-protocols.html#cb402-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.caption =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;grey80&quot;</span>))</span></code></pre></div>
<p>And to label the different replicates with a colorblind friendly palette, we define a set of colors that were proposed by Paul Tol: <a href="https://personal.sron.nl/~pault/" class="uri">https://personal.sron.nl/~pault/</a></p>
<div class="sourceCode" id="cb403"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb403-1"><a href="complete-protocols.html#cb403-1" aria-hidden="true" tabindex="-1"></a>Tol_bright <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;66CCEE&#39;</span>, <span class="st">&#39;#BBBBBB&#39;</span>, <span class="st">&#39;#CCBB44&#39;</span>,<span class="st">&#39;#AA3377&#39;</span>,<span class="st">&#39;#228833&#39;</span>, <span class="st">&#39;#4477AA&#39;</span>)</span></code></pre></div>
<p>To use these colors we define manual color scales for both ‘fill’ (used for <code>geom_point()</code>) and ‘color’ (used for <code>geom_jitter()</code>):</p>
<div class="sourceCode" id="cb404"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb404-1"><a href="complete-protocols.html#cb404-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> Tol_bright) <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values =</span> Tol_bright)</span>
<span id="cb404-2"><a href="complete-protocols.html#cb404-2" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-329-1.png" width="672" /></p>
<p>The result is a dose response curve in which the replicates are clearly identified with colorblind friendly colors. The EC50 can be determined from fitting a curve. To this end, we use the function <code>nls()</code>, which needs an equation, the data and estimates of the values. It will perform a ‘nonlinear least squares’ optimization to find parameters for the equation that best fit with the data of ‘df_summary’. The parameters are ‘bot’, ‘top’, ‘EC50’ and ‘slope’:</p>
<div class="sourceCode" id="cb405"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb405-1"><a href="complete-protocols.html#cb405-1" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">nls</span>(mean_Response <span class="sc">~</span> top<span class="sc">+</span>(bot<span class="sc">-</span>top)<span class="sc">/</span>(<span class="dv">1</span><span class="sc">+</span>(Concentration<span class="sc">/</span>EC50)<span class="sc">^</span>slope),</span>
<span id="cb405-2"><a href="complete-protocols.html#cb405-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">data =</span> df_summary,</span>
<span id="cb405-3"><a href="complete-protocols.html#cb405-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">start=</span><span class="fu">list</span>(<span class="at">bot=</span><span class="sc">-</span><span class="dv">2</span>, <span class="at">top=</span><span class="dv">8</span>, <span class="at">EC50=</span><span class="dv">1</span>, <span class="at">slope=</span><span class="dv">1</span>))</span></code></pre></div>
<p>The result is stored in the object ‘fit’ and the parameters can be listed:</p>
<div class="sourceCode" id="cb406"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb406-1"><a href="complete-protocols.html#cb406-1" aria-hidden="true" tabindex="-1"></a><span class="fu">coef</span>(fit)</span></code></pre></div>
<pre><code>       bot        top       EC50      slope 
-0.1990747  3.2915665  0.4968078  0.7141988 </code></pre>
<p>To plot the fitted data, we use <code>augment()</code> function that is part of the <code>{broom}</code> package, which we need to load:</p>
<div class="sourceCode" id="cb408"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb408-1"><a href="complete-protocols.html#cb408-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(broom)</span>
<span id="cb408-2"><a href="complete-protocols.html#cb408-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb408-3"><a href="complete-protocols.html#cb408-3" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_line</span>(<span class="at">data =</span> <span class="fu">augment</span>(fit), <span class="fu">aes</span>(<span class="at">x=</span>Concentration,<span class="at">y=</span>.fitted), <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size=</span><span class="dv">1</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-332-1.png" width="672" /></p>
<p>Instead of <code>geom_lin()</code> for plotting the fit, we can also use <code>geom_smooth()</code>:</p>
<div class="sourceCode" id="cb409"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb409-1"><a href="complete-protocols.html#cb409-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">data =</span> <span class="fu">augment</span>(fit), <span class="fu">aes</span>(<span class="at">x=</span>Concentration,<span class="at">y=</span>.fitted), <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size=</span><span class="dv">1</span>)</span>
<span id="cb409-2"><a href="complete-protocols.html#cb409-2" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-333-1.png" width="672" /></p>
<p>An alternative way to do the curve fit is provided through the <code>{drc}</code> package. To demonstrate this, we load the package and add the curve fit to the data:</p>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb410-1"><a href="complete-protocols.html#cb410-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(drc)</span>
<span id="cb410-2"><a href="complete-protocols.html#cb410-2" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">data=</span>df_summary, <span class="fu">aes</span>(<span class="at">x =</span> Concentration, <span class="at">y =</span> mean_Response), <span class="at">color=</span><span class="st">&#39;blue&#39;</span>, <span class="at">method =</span> drm, <span class="at">method.args =</span> <span class="fu">list</span>(<span class="at">fct =</span> <span class="fu">L.4</span>()), <span class="at">se =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="DataViz-protocols_files/figure-html/unnamed-chunk-334-1.png" width="672" /></p>
<p>The <code>{drc}</code> package was used to fit the data in the paper by <a href="https://doi.org/10.1242/jcs.259685">Chavez-Abiega et al. (2022)</a>. But the approach followed above with <code>nls()</code> is easier to adapt for data that requires other models for the fit.</p>
<p>Finally, we can save the plot:</p>
<div class="sourceCode" id="cb411"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb411-1"><a href="complete-protocols.html#cb411-1" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file=</span><span class="fu">paste0</span>(<span class="st">&quot;Protocol_9.png&quot;</span>), <span class="at">width =</span> <span class="dv">4000</span>, <span class="at">height =</span> <span class="dv">3000</span>, <span class="at">units =</span> <span class="st">&quot;px&quot;</span>, <span class="at">res =</span> <span class="dv">400</span>)</span>
<span id="cb411-2"><a href="complete-protocols.html#cb411-2" aria-hidden="true" tabindex="-1"></a> p</span>
<span id="cb411-3"><a href="complete-protocols.html#cb411-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>quartz_off_screen 
                2 </code></pre>
<div id="in-progress" class="section level4 unnumbered hasAnchor">
<h4>In progress<a href="complete-protocols.html#in-progress" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>This chapter is ‘in progress’ and I plan to add more protocols in the future.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="plotting-the-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="questions-and-answers.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": false
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
